commit 294efaa9bd7e3a36345f45326ca2e22e1cccc9b6
Author: Siddharth Wagle <swagle@hortonworks.com>
Date:   Fri Feb 27 11:05:28 2015 -0800

    HBASE-13120. Allow disabling hadoop classpath and native library lookup.

diff --git a/bin/hbase b/bin/hbase
index a7e65d6..b2c8681 100755
--- a/bin/hbase
+++ b/bin/hbase
@@ -199,14 +199,17 @@ function append_path() {
 JAVA_PLATFORM=""
 
 #If avail, add Hadoop to the CLASSPATH and to the JAVA_LIBRARY_PATH
-HADOOP_IN_PATH=$(PATH="${HADOOP_HOME:-${HADOOP_PREFIX}}/bin:$PATH" which hadoop 2>/dev/null)
-if [ -f ${HADOOP_IN_PATH} ]; then
-  HADOOP_JAVA_LIBRARY_PATH=$(HADOOP_CLASSPATH="$CLASSPATH" ${HADOOP_IN_PATH} \
-                             org.apache.hadoop.hbase.util.GetJavaProperty java.library.path 2>/dev/null)
-  if [ -n "$HADOOP_JAVA_LIBRARY_PATH" ]; then
-    JAVA_LIBRARY_PATH=$(append_path "${JAVA_LIBRARY_PATH}" "$HADOOP_JAVA_LIBRARY_PATH")
+# Allow this functionality to be disabled
+if [ "$HBASE_DISABLE_HADOOP_CLASSPATH_LOOKUP" != "true" ] ; then
+  HADOOP_IN_PATH=$(PATH="${HADOOP_HOME:-${HADOOP_PREFIX}}/bin:$PATH" which hadoop 2>/dev/null)
+  if [ -f ${HADOOP_IN_PATH} ]; then
+    HADOOP_JAVA_LIBRARY_PATH=$(HADOOP_CLASSPATH="$CLASSPATH" ${HADOOP_IN_PATH} \
+                               org.apache.hadoop.hbase.util.GetJavaProperty java.library.path 2>/dev/null)
+    if [ -n "$HADOOP_JAVA_LIBRARY_PATH" ]; then
+      JAVA_LIBRARY_PATH=$(append_path "${JAVA_LIBRARY_PATH}" "$HADOOP_JAVA_LIBRARY_PATH")
+    fi
+    CLASSPATH=$(append_path "${CLASSPATH}" `${HADOOP_IN_PATH} classpath 2>/dev/null`)
   fi
-  CLASSPATH=$(append_path "${CLASSPATH}" `${HADOOP_IN_PATH} classpath 2>/dev/null`)
 fi
 
 # Add user-specified CLASSPATH last
