diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupCommands.java hbase-client/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupCommands.java
index 63f7a55..618b0be 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupCommands.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupCommands.java
@@ -194,8 +194,10 @@ public final class BackupCommands {
         request.setBackupType(BackupType.valueOf(args[1].toUpperCase()))
         .setTableList(tables != null?Lists.newArrayList(BackupClientUtil.parseTableNames(tables)): null)
         .setTargetRootDir(args[2]).setWorkers(workers).setBandwidth(bandwidth);
-        backupAdmin.backupTables(request);
+        String backupId = backupAdmin.backupTables(request);
+        System.out.println("Backup session "+ backupId+" finished. Status: SUCCESS");
       } catch (IOException e) {
+        System.out.println("Backup session finished. Status: FAILURE");
         throw e;
       }
     }
