diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java
index ad9241f..f752cd0 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java
@@ -775,7 +775,7 @@ public final class HConstants {
   /**
    * Default value of {@link #HBASE_META_VERSIONS}.
    */
-  public static int DEFAULT_HBASE_META_VERSIONS = 10;
+  public static int DEFAULT_HBASE_META_VERSIONS = 3;
 
   /**
    * Parameter name for number of versions, kept by meta table.
