diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionSplitPolicy.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionSplitPolicy.java
index 9cb2b44..dd7c61a 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionSplitPolicy.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionSplitPolicy.java
@@ -111,7 +111,7 @@ public class TestRegionSplitPolicy {
     regions.add(mockRegion);
     assertFalse(policy.shouldSplit());
     // make sure its just over; verify it'll split
-    Mockito.doReturn((long)(maxSplitSize * 1.025 + 1)).when(mockStore).getSize();
+    Mockito.doReturn((long)(maxSplitSize * 1.25 + 1)).when(mockStore).getSize();
     assertTrue(policy.shouldSplit());
 
     // Finally assert that even if loads of regions, we'll split at max size
