diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
index 49c7f8a..daf2eab 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
@@ -2674,6 +2674,14 @@ public class RpcServer implements RpcServerInterface, ConfigurationObserver {
         }
       }
     }
+    if(param instanceof ClientProtos.CoprocessorServiceRequest) {
+      ClientProtos.CoprocessorServiceRequest r = (ClientProtos.CoprocessorServiceRequest) param;
+      Region region = rsRpcServices.getRegion(r.getRegion());
+      String serviceClass = region.getServiceClassbyName(r.getCall().getServiceName()).getClass().toString();
+      String cpInfo = "endpoint coprocessor= " + serviceClass+ ":" + r.getCall().getMethodName();
+      responseInfo.put("param", cpInfo);
+   }
+
     LOG.warn("(response" + tag + "): " + MAPPER.writeValueAsString(responseInfo));
   }
 
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
index 31b2a15..1513896 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
@@ -7704,6 +7704,15 @@ public class HRegion implements HeapSize, PropagatingConfigurationObserver, Regi
   }
 
   @Override
+  public com.google.protobuf.Service getServiceClassbyName(String serviceName) {
+    if (!coprocessorServiceHandlers.containsKey(serviceName)) {
+      LOG.error("Coprocessor service " + serviceName + "not registered yet");
+      return null;
+    }
+    return coprocessorServiceHandlers.get(serviceName);
+  }
+
+  @Override
   public com.google.protobuf.Message execService(com.google.protobuf.RpcController controller,
       CoprocessorServiceCall call)
   throws IOException {
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/Region.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/Region.java
index 1b106b2..32ae3a0 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/Region.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/Region.java
@@ -603,6 +603,7 @@ public interface Region extends ConfigurationObserver {
       CoprocessorServiceCall call)
   throws IOException;
 
+
   /**
    * Registers a new protocol buffer {@link Service} subclass as a coprocessor endpoint to
    * be available for handling Region#execService(com.google.protobuf.RpcController,
@@ -620,6 +621,14 @@ public interface Region extends ConfigurationObserver {
    */
   boolean registerService(com.google.protobuf.Service instance);
 
+  /**
+   * Return com.google.protobuf.Service instance full class name if service registered
+   * see {@link Region#registerService(com.google.protobuf.Service)}
+   * @param serviceName
+   * @return
+   */
+   com.google.protobuf.Service getServiceClassbyName(String serviceName);
+
   ///////////////////////////////////////////////////////////////////////////
   // RowMutation processor support
 
