diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/HRegionInfo.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/HRegionInfo.java
index c134063..dd55d39 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/HRegionInfo.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/HRegionInfo.java
@@ -744,6 +744,13 @@ public class HRegionInfo implements Comparable<HRegionInfo> {
     return replicaId;
   }
 
+  /*
+   * @return true if this is the primary replica
+   */
+  public boolean isPrimary() {
+    return replicaId == DEFAULT_REPLICA_ID;
+  }
+
   /**
    * @see java.lang.Object#toString()
    */
diff --git a/hbase-server/src/main/resources/hbase-webapps/master/table.jsp b/hbase-server/src/main/resources/hbase-webapps/master/table.jsp
index 45be52b..e04b3bf 100644
--- a/hbase-server/src/main/resources/hbase-webapps/master/table.jsp
+++ b/hbase-server/src/main/resources/hbase-webapps/master/table.jsp
@@ -282,6 +282,7 @@ if ( fqtn != null ) {
 </table>
 <%
   Map<ServerName, Integer> regDistribution = new TreeMap<ServerName, Integer>();
+  Map<ServerName, Integer> primaryRegDistribution = new TreeMap<ServerName, Integer>();
   List<HRegionLocation> regions = r.getAllRegionLocations();
   if(regions != null && regions.size() > 0) { %>
 <%=     tableHeader %>
@@ -304,6 +305,11 @@ if ( fqtn != null ) {
         Integer i = regDistribution.get(addr);
         if (null == i) i = Integer.valueOf(0);
         regDistribution.put(addr, i + 1);
+        if (regionInfo.isPrimary()) {
+          i = primaryRegDistribution.get(addr);
+          if (null == i) i = Integer.valueOf(0);
+          primaryRegDistribution.put(addr, i+1);
+        }
       }
     }
 %>
@@ -341,7 +347,7 @@ if ( fqtn != null ) {
 <% } %>
 </table>
 <h2>Regions by Region Server</h2>
-<table class="table table-striped"><tr><th>Region Server</th><th>Region Count</th></tr>
+<table class="table table-striped"><tr><th>Region Server</th><th>Region Count</th><th>Primary Region Count</th></tr>
 <%
   for (Map.Entry<ServerName, Integer> rdEntry : regDistribution.entrySet()) {
      ServerName addr = rdEntry.getKey();
@@ -350,6 +356,7 @@ if ( fqtn != null ) {
 <tr>
   <td><a href="<%= url %>"><%= addr.getHostname().toString() + ":" + addr.getPort() %></a></td>
   <td><%= rdEntry.getValue()%></td>
+  <td><%= primaryRegDistribution.get(addr)%></td>
 </tr>
 <% } %>
 </table>
