diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
index de3439b..7342749 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
@@ -117,8 +117,12 @@ public class MobFileCache {
       }
       this.scheduleThreadPool.scheduleAtFixedRate(new EvictionThread(this), period, period,
           TimeUnit.SECONDS);
+
+      LOG.info("MobFileCache enabled with cache size: " + mobFileMaxCacheSize +
+          ", evict periods: " +  period + "sec, evict remain ratio: " + evictRemainRatio);
+    } else {
+      LOG.info("MobFileCache disabled");
     }
-    LOG.info("MobFileCache is initialized, and the cache size is " + mobFileMaxCacheSize);
   }
 
   /**
