diff --git a/dev-support/publish_hbase_website.sh b/dev-support/publish_hbase_website.sh
index 072de8d..ecea874 100755
--- a/dev-support/publish_hbase_website.sh
+++ b/dev-support/publish_hbase_website.sh
@@ -76,6 +76,8 @@ if [ ! -d $GIT_DIR -o ! -d $SVN_DIR ]; then
   exit 1
 fi
 
+export MAVEN_OPTS=${MAVEN_OPTS:-"-Xmx3g -XX:MaxPermSize=256m"}
+
 cd $GIT_DIR
 
 # Get the latest
@@ -88,7 +90,7 @@ if [ $INTERACTIVE ]; then
     read -p "Build the site? (y/n)" yn
     case $yn in
         [Yy]* ) 
-    			mvn clean package javadoc:aggregate site post-site site:stage -DskipTests
+            mvn clean package javadoc:aggregate site post-site site:stage -DskipTests -Dcheckstyle.skip
           status=$?
           if [ $status -ne 0 ]; then
             echo "The website does not build. Aborting."
@@ -101,7 +103,7 @@ if [ $INTERACTIVE ]; then
     esac
 else
   echo "Building the site in auto mode."
-  mvn clean package javadoc:aggregate site post-site site:stage -DskipTests
+  mvn clean package javadoc:aggregate site post-site site:stage -DskipTests -Dcheckstyle.skip
   status=$?
   if [ $status != 0 ]; then
     echo "The website does not build. Aborting."
