diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/MultiActionResultTooLarge.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/MultiActionResultTooLarge.java
index d06eea1..fdff554 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/MultiActionResultTooLarge.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/MultiActionResultTooLarge.java
@@ -18,13 +18,17 @@
 
 package org.apache.hadoop.hbase;
 
+import org.apache.hadoop.hbase.classification.InterfaceAudience;
+import org.apache.hadoop.hbase.classification.InterfaceStability;
+
 /**
  * Exception thrown when the result needs to be chunked on the server side.
  * It signals that retries should happen right away and not count against the number of
  * retries because some of the multi was a success.
  */
+@InterfaceAudience.Public
+@InterfaceStability.Evolving
 public class MultiActionResultTooLarge extends RetryImmediatelyException {
-
   public MultiActionResultTooLarge(String s) {
     super(s);
   }
diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/RetryImmediatelyException.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/RetryImmediatelyException.java
index 1b39904..e0b90fd 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/RetryImmediatelyException.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/RetryImmediatelyException.java
@@ -19,7 +19,11 @@
 package org.apache.hadoop.hbase;
 
 import java.io.IOException;
+import org.apache.hadoop.hbase.classification.InterfaceAudience;
+import org.apache.hadoop.hbase.classification.InterfaceStability;
 
+@InterfaceAudience.Public
+@InterfaceStability.Evolving
 public class RetryImmediatelyException extends IOException {
   public RetryImmediatelyException(String s) {
     super(s);
