diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestScannerHeartbeatMessages.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestScannerHeartbeatMessages.java
index 495be30..f86c178 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestScannerHeartbeatMessages.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestScannerHeartbeatMessages.java
@@ -345,7 +345,7 @@ public class TestScannerHeartbeatMessages {
     }
 
     public HeartbeatHRegionServer(Configuration conf, CoordinatedStateManager csm)
-        throws IOException {
+        throws IOException, InterruptedException {
       super(conf, csm);
     }
 
