diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
index d813575..5498d66 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
@@ -367,6 +367,10 @@ public class TestRegionServerMetrics {
       }
       metricsHelper.assertCounter("ScanNext_num_ops", numScanNext, serverSource);
     }
+    try (Admin admin = TEST_UTIL.getHBaseAdmin()) {
+      admin.disableTable(tableName);
+      admin.deleteTable(tableName);
+    }
   }
 
   @Test
@@ -413,5 +417,9 @@ public class TestRegionServerMetrics {
       }
       metricsHelper.assertCounter("ScanNext_num_ops", numScanNext, serverSource);
     }
+    try (Admin admin = TEST_UTIL.getHBaseAdmin()) {
+      admin.disableTable(tableName);
+      admin.deleteTable(tableName);
+    }
   }
 }
