diff --git hbase-shell/src/main/ruby/hbase/admin.rb hbase-shell/src/main/ruby/hbase/admin.rb
index 532f073..d74d229 100644
--- hbase-shell/src/main/ruby/hbase/admin.rb
+++ hbase-shell/src/main/ruby/hbase/admin.rb
@@ -242,6 +242,7 @@ module Hbase
           File.foreach(splits_file) do |line|
             arg[SPLITS].push(line.strip())
           end
+          htd.setValue(SPLITS_FILE, arg[SPLITS_FILE])
         end
 
         if arg.has_key?(SPLITS) 
