diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Put.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Put.java
index 6b06875..f2dff7c 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Put.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Put.java
@@ -117,6 +117,43 @@ public class Put extends Mutation implements HeapSize, Comparable<Row> {
   }
 
   /**
+   * Create a Put operation for the specified row.
+   *
+   * @param row row key
+   * @param copyRow whether to make a local copy of the row specified.
+   *                Set to true if the caller can make sure that the row is immutable
+   *                and will not be changed for the Put duration.
+   */
+  public Put(byte [] row, boolean copyRow) {
+    this(row, HConstants.LATEST_TIMESTAMP, copyRow);
+  }
+
+  /**
+   * Create a Put operation for the specified row, using a given timestamp.
+   *
+   * @param row row key
+   * @param ts timestamp
+   * @param copyRow whether to make a local copy of the row specified.
+   *                Set to false if the caller can make sure that the row is immutable
+   *                and will not be changed for the Put duration.
+   */
+  public Put(byte[] row, long ts, boolean copyRow) {
+    // Check and set timestamp
+    if (ts < 0) {
+      throw new IllegalArgumentException("Timestamp cannot be negative. ts=" + ts);
+    }
+    this.ts = ts;
+
+    // Handle row
+    if (copyRow) {  // make a local copy of row
+      checkRow(row, 0, row.length);
+      this.row = Bytes.copy(row, 0, row.length);
+    } else {  // do not make a local copy of row
+      this.row = row;
+    }
+  }
+
+  /**
    * Copy constructor.  Creates a Put operation cloned from the specified Put.
    * @param putToCopy put to copy
    */
diff --git a/hbase-client/src/test/java/org/apache/hadoop/hbase/client/TestPut.java b/hbase-client/src/test/java/org/apache/hadoop/hbase/client/TestPut.java
index 3a877dc..ec2dbea 100644
--- a/hbase-client/src/test/java/org/apache/hadoop/hbase/client/TestPut.java
+++ b/hbase-client/src/test/java/org/apache/hadoop/hbase/client/TestPut.java
@@ -23,6 +23,8 @@ import org.apache.hadoop.hbase.util.Bytes;
 import org.junit.Test;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+
 public class TestPut {
   @Test
   public void testCopyConstructor() {
@@ -40,4 +42,18 @@ public class TestPut {
     assertNotEquals(origin.getCellList(family), clone.getCellList(family));
 
   }
+
+  // HBASE-14881
+  @Test
+  public void testCopyRowAndNotCopyRow() {
+    byte[] rowKey = Bytes.toBytes("immutable");
+
+    // Do not make a local copy of row
+    Put putNotCopyRow = new Put(rowKey, false);
+    assertTrue(rowKey == putNotCopyRow.getRow());
+
+    // Make a local copy of row
+    Put putCopyRow = new Put(rowKey, 1000L, true);
+    assertTrue(rowKey != putCopyRow.getRow());
+  }
 }
