diff --git a/hbase-it/src/test/java/org/apache/hadoop/hbase/test/IntegrationTestBigLinkedListWithVisibility.java b/hbase-it/src/test/java/org/apache/hadoop/hbase/test/IntegrationTestBigLinkedListWithVisibility.java
index 50c638a..23f3132 100644
--- a/hbase-it/src/test/java/org/apache/hadoop/hbase/test/IntegrationTestBigLinkedListWithVisibility.java
+++ b/hbase-it/src/test/java/org/apache/hadoop/hbase/test/IntegrationTestBigLinkedListWithVisibility.java
@@ -187,8 +187,8 @@ protected void setup(org.apache.hadoop.mapreduce.Mapper.Context context) throws
       protected void instantiateHTable() throws IOException {
         for (int i = 0; i < DEFAULT_TABLES_COUNT; i++) {
           Table table = connection.getTable(getTableName(i));
-          table.setAutoFlushTo(true);
-          //table.setWriteBufferSize(4 * 1024 * 1024);
+          table.setAutoFlushTo(false);
+          table.setWriteBufferSize(4 * 1024 * 1024);
           this.tables[i] = table;
         }
       }
