diff --git a/hbase-shell/src/main/ruby/shell/commands/list.rb b/hbase-shell/src/main/ruby/shell/commands/list.rb
index f7a09878f2..dc3d1687ee 100644
--- a/hbase-shell/src/main/ruby/shell/commands/list.rb
+++ b/hbase-shell/src/main/ruby/shell/commands/list.rb
@@ -22,7 +22,7 @@ module Shell
     class List < Command
       def help
         return <<-EOF
-List all tables in hbase. Optional regular expression parameter could
+List all user tables in hbase. Optional regular expression parameter could
 be used to filter the output. Examples:
 
   hbase> list
