diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFiles.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFiles.java
index c04794b..09dedec 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFiles.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFiles.java
@@ -818,7 +818,7 @@ public class LoadIncrementalHFiles extends Configured implements Tool {
                 Path hfileStagingPath = null;
                 Path hfileOrigPath = new Path(el.getSecond());
                 try {
-                  hfileStagingPath= new Path(secureClient.getStagingPath(bulkToken, el.getFirst()),
+                  hfileStagingPath= new Path(new Path(bulkToken, Bytes.toString(el.getFirst())),
                     hfileOrigPath.getName());
                   if(targetFs.rename(hfileStagingPath, hfileOrigPath)) {
                     LOG.debug("Moved back file " + hfileOrigPath + " from " +
