diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/BalancedQueueRpcExecutor.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/BalancedQueueRpcExecutor.java
index 40ba9fe..9c14028 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/BalancedQueueRpcExecutor.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/BalancedQueueRpcExecutor.java
@@ -22,6 +22,7 @@ import java.util.List;
 import java.util.Random;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadLocalRandom;
 
 import org.apache.hadoop.hbase.classification.InterfaceAudience;
 import org.apache.hadoop.hbase.classification.InterfaceStability;
@@ -113,15 +114,13 @@ public class BalancedQueueRpcExecutor extends RpcExecutor {
    */
   private static class RandomQueueBalancer extends QueueBalancer {
     private int queueSize;
-    private Random random;
 
     public RandomQueueBalancer(int queueSize) {
       this.queueSize = queueSize;
-      this.random = new Random();
     }
 
     public int getNextQueue() {
-      return random.nextInt(queueSize);
+      return ThreadLocalRandom.current().nextInt(queueSize);
     }
   }
 }
