diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKServerTool.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKServerTool.java
index c82e96b..6465aa7 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKServerTool.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKServerTool.java
@@ -58,7 +58,7 @@ public class ZKServerTool {
    */
   public static void main(String args[]) {
     for(ServerName server: readZKNodes(HBaseConfiguration.create())) {
-      System.out.println("Zk host: " + server.getHostname());
+      System.out.println("ZK host: " + server.getHostname());
     }
   }
 }
