diff --git hbase-common/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java hbase-common/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
index 79c8ac3..d02c5e9 100644
--- hbase-common/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
+++ hbase-common/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
@@ -93,7 +93,7 @@ public class VersionInfo {
   static String[] versionReport() {
     return new String[] {
       "HBase " + getVersion(),
-      "Source code repository " + getUrl() + " -r " + getRevision(),
+      "Source code repository " + getUrl() + " revision=" + getRevision(),
       "Compiled by " + getUser() + " on " + getDate(),
       "From source with checksum " + getSrcChecksum()
       };
diff --git hbase-rest/src/main/resources/hbase-webapps/rest/rest.jsp hbase-rest/src/main/resources/hbase-webapps/rest/rest.jsp
index 810569c..43895a8 100644
--- hbase-rest/src/main/resources/hbase-webapps/rest/rest.jsp
+++ hbase-rest/src/main/resources/hbase-webapps/rest/rest.jsp
@@ -86,7 +86,7 @@ String listenPort = conf.get("hbase.rest.port", "8080");
         </tr>
         <tr>
             <td>HBase Version</td>
-            <td><%= VersionInfo.getVersion() %>, r<%= VersionInfo.getRevision() %></td>
+            <td><%= VersionInfo.getVersion() %>, revision=<%= VersionInfo.getRevision() %></td>
             <td>HBase version and revision</td>
         </tr>
         <tr>
diff --git hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
index d75406e..2f0451c 100644
--- hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
+++ hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
@@ -225,7 +225,7 @@ AssignmentManager assignmentManager = master.getAssignmentManager();
                 </tr>
                 <tr>
                     <td>HBase Version</td>
-                    <td><% org.apache.hadoop.hbase.util.VersionInfo.getVersion() %>, r<% org.apache.hadoop.hbase.util.VersionInfo.getRevision() %></td><td>HBase version and revision</td>
+                    <td><% org.apache.hadoop.hbase.util.VersionInfo.getVersion() %>, revision=<% org.apache.hadoop.hbase.util.VersionInfo.getRevision() %></td><td>HBase version and revision</td>
                 </tr>
                 <tr>
                     <td>HBase Compiled</td>
@@ -239,7 +239,7 @@ AssignmentManager assignmentManager = master.getAssignmentManager();
                 </tr>
                 <tr>
                     <td>Hadoop Version</td>
-                    <td><% org.apache.hadoop.util.VersionInfo.getVersion() %>, r<% org.apache.hadoop.util.VersionInfo.getRevision() %></td>
+                    <td><% org.apache.hadoop.util.VersionInfo.getVersion() %>, revision=<% org.apache.hadoop.util.VersionInfo.getRevision() %></td>
                     <td>Hadoop version and revision</td>
                 </tr>
                 <tr>
diff --git hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/regionserver/RSStatusTmpl.jamon hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/regionserver/RSStatusTmpl.jamon
index 68b4b7d..8ee1193 100644
--- hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/regionserver/RSStatusTmpl.jamon
+++ hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/regionserver/RSStatusTmpl.jamon
@@ -132,7 +132,7 @@ org.apache.hadoop.hbase.zookeeper.MasterAddressTracker;
         </tr>
         <tr>
             <td>HBase Version</td>
-            <td><% org.apache.hadoop.hbase.util.VersionInfo.getVersion() %>, r<% org.apache.hadoop.hbase.util.VersionInfo.getRevision() %></td>
+            <td><% org.apache.hadoop.hbase.util.VersionInfo.getVersion() %>, revision=<% org.apache.hadoop.hbase.util.VersionInfo.getRevision() %></td>
             <td>HBase version and revision</td>
         </tr>
         <tr>
@@ -146,6 +146,16 @@ org.apache.hadoop.hbase.zookeeper.MasterAddressTracker;
             <td>HBase source MD5 checksum</td>
         </tr>
         <tr>
+            <td>Hadoop Version</td>
+            <td><% org.apache.hadoop.util.VersionInfo.getVersion() %>, revision=<% org.apache.hadoop.util.VersionInfo.getRevision() %></td>
+            <td>Hadoop version and revision</td>
+        </tr>
+        <tr>
+            <td>Hadoop Compiled</td>
+            <td><% org.apache.hadoop.util.VersionInfo.getDate() %>, <% org.apache.hadoop.util.VersionInfo.getUser() %></td>
+            <td>When Hadoop version was compiled and by whom</td>
+        </tr>
+        <tr>
             <td>Zookeeper Quorum</td>
             <td><% regionServer.getZooKeeper().getQuorum() %></td>
             <td>Addresses of all registered ZK servers</td>
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/monitoring/StateDumpServlet.java hbase-server/src/main/java/org/apache/hadoop/hbase/monitoring/StateDumpServlet.java
index 8b16438..e495e52 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/monitoring/StateDumpServlet.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/monitoring/StateDumpServlet.java
@@ -38,10 +38,10 @@ public abstract class StateDumpServlet extends HttpServlet {
 
   protected void dumpVersionInfo(PrintWriter out) {
     VersionInfo.writeTo(out);
-  
+
     out.println("Hadoop " + org.apache.hadoop.util.VersionInfo.getVersion());
-    out.println("Subversion " + org.apache.hadoop.util.VersionInfo.getUrl() + " -r " +
-        org.apache.hadoop.util.VersionInfo.getRevision());
+    out.println("Source code repository " + org.apache.hadoop.util.VersionInfo.getUrl()
+      + " revision=" + org.apache.hadoop.util.VersionInfo.getRevision());
     out.println("Compiled by " + org.apache.hadoop.util.VersionInfo.getUser() +
         " on " + org.apache.hadoop.util.VersionInfo.getDate());
   }
