diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/client/RegionCoprocessorServiceExec.java hbase-client/src/main/java/org/apache/hadoop/hbase/client/RegionCoprocessorServiceExec.java
index 2d62332..cd4716e 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/client/RegionCoprocessorServiceExec.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/client/RegionCoprocessorServiceExec.java
@@ -101,4 +101,18 @@ public class RegionCoprocessorServiceExec implements Row {
     Row other = (Row) obj;
     return compareTo(other) == 0;
   }
+
+  @Override
+  public String toString() {
+    StringBuilder builder = new StringBuilder();
+    builder.append("region:")
+           .append(Bytes.toStringBinary(region))
+           .append(", startKey:")
+           .append(Bytes.toStringBinary(startKey))
+           .append(", method:")
+           .append(method.getFullName())
+           .append(", request:")
+           .append(request);
+    return builder.toString();
+  }
 }
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/CoordinatedStateManager.java hbase-server/src/main/java/org/apache/hadoop/hbase/CoordinatedStateManager.java
index b7bfa75..bdb202d 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/CoordinatedStateManager.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/CoordinatedStateManager.java
@@ -20,7 +20,7 @@ package org.apache.hadoop.hbase;
 import org.apache.hadoop.hbase.classification.InterfaceAudience;
 
 /**
- * Implementations of this interface will keep and return to clients 
+ * Implementations of this interface will keep and return to clients
  * implementations of classes providing API to execute
  * coordinated operations. This interface is client-side, so it does NOT
  * include methods to retrieve the particular interface implementations.
