diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/tool/Canary.java hbase-server/src/main/java/org/apache/hadoop/hbase/tool/Canary.java
index 8efb73d..c8f0a72 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/tool/Canary.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/tool/Canary.java
@@ -153,14 +153,13 @@ public final class Canary implements Tool {
       } catch (IOException e) {
         LOG.debug("sniffRegion failed", e);
         sink.publishReadFailure(region, e);
-        return null;
-      } finally {
         if (table != null) {
           try {
             table.close();
-          } catch (IOException e) {
+          } catch (IOException ioe) {
           }
         }
+        return null;
       }
 
       byte[] startKey = null;
@@ -204,17 +203,15 @@ public final class Canary implements Tool {
           if (rs != null) {
             rs.close();
           }
-          if (table != null) {
-            try {
-              table.close();
-            } catch (IOException e) {
-            }
-          }
           scan = null;
           get = null;
           startKey = null;
         }
       }
+      try {
+        table.close();
+      } catch (IOException e) {
+      }
       return null;
     }
   }
