diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerStatisticTracker.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerStatisticTracker.java
index 0c7b683..bca442d 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerStatisticTracker.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerStatisticTracker.java
@@ -34,7 +34,7 @@ import java.util.concurrent.ConcurrentHashMap;
 @InterfaceAudience.Private
 public class ServerStatisticTracker {
 
-  private final Map<ServerName, ServerStatistics> stats =
+  private final ConcurrentHashMap<ServerName, ServerStatistics> stats =
       new ConcurrentHashMap<ServerName, ServerStatistics>();
 
   public void updateRegionStats(ServerName server, byte[] region, ClientProtos.RegionLoadStats
@@ -42,14 +42,12 @@ public class ServerStatisticTracker {
     ServerStatistics stat = stats.get(server);
 
     if (stat == null) {
-      // create a stats object and update the stats
-      synchronized (this) {
-        stat = stats.get(server);
-        // we don't have stats for that server yet, so we need to make some
-        if (stat == null) {
-          stat = new ServerStatistics();
-          stats.put(server, stat);
-        }
+      stat = stats.get(server);
+      // We don't have stats for that server yet, so we need to make an entry.
+      // If we race with another thread it's a harmless unnecessary allocation.
+      if (stat == null) {
+        stat = new ServerStatistics();
+        stats.putIfAbsent(server, stat);
       }
     }
     stat.update(region, currentStats);
@@ -71,4 +69,4 @@ public class ServerStatisticTracker {
   ServerStatistics getServerStatsForTesting(ServerName server) {
     return stats.get(server);
   }
-}
\ No newline at end of file
+}
