diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/HTable.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/HTable.java
index 35d9a74..24dc06a 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/HTable.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/HTable.java
@@ -756,7 +756,7 @@ public class HTable implements HTableInterface, RegionLocator {
          try {
            ClientProtos.GetResponse response = getStub().get(controller, request);
            if (!response.hasResult()) return null;
-           return ProtobufUtil.toResult(response.getResult());
+           return ProtobufUtil.toResult(response.getResult(), controller.cellScanner());
          } catch (ServiceException se) {
            throw ProtobufUtil.getRemoteException(se);
          }
diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java
index 907650b..47305ff 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java
@@ -1591,6 +1591,7 @@ public final class ProtobufUtil {
     try {
       GetResponse response = client.get(null, request);
       if (!response.hasResult()) return null;
+      // We pass 'null' RpcController. So Result will be pure RB.
       return toResult(response.getResult());
     } catch (ServiceException se) {
       throw getRemoteException(se);
