diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
index 4e8784a..a173eac 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
@@ -1245,7 +1245,9 @@ public class FSHLog implements WAL {
     void offer(final long sequence, final SyncFuture [] syncFutures, final int syncFutureCount) {
       // Set sequence first because the add to the queue will wake the thread if sleeping.
       this.sequence = sequence;
-      this.syncFutures.addAll(Arrays.asList(syncFutures).subList(0, syncFutureCount));
+      for (int i = 0; i < syncFutureCount; ++i) {
+        this.syncFutures.add(syncFutures[i]);
+      }
     }
 
     /**
