From 3fbc14c5b5d483fc26c3ffa3aabe9ad496f248db Mon Sep 17 00:00:00 2001
From: Nick Dimiduk <ndimiduk@apache.org>
Date: Sat, 16 Jan 2016 15:44:07 -0800
Subject: [PATCH] HBASE-15119 Include git SHA in check_compatibility reports

---
 dev-support/check_compatibility.sh | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dev-support/check_compatibility.sh b/dev-support/check_compatibility.sh
index b6c11e0..95dba00 100755
--- a/dev-support/check_compatibility.sh
+++ b/dev-support/check_compatibility.sh
@@ -188,6 +188,8 @@ for ref in 1 2; do
       echo "Maven could not successfully package ${COMMIT[${ref}]}. Exiting..." >&2
       exit 2
     fi
+    # grab sha for future reference
+    SHA[${ref}]=$(git rev-parse --short HEAD)
     popd > /dev/null
   fi
 
@@ -207,6 +209,7 @@ for ref in 1 2; do
           echo "Maven could not successfully package ${COMMIT[${ref}]}. Exiting..." >&2
           exit 2
         fi
+        SHA[${ref}]=$(git rev-parse --short HEAD)
         popd > /dev/null
       done
 
@@ -259,7 +262,8 @@ fi
 
 # Generate command line arguments for Java ACC.
 JAVA_ACC_COMMAND+=(-l HBase)
-JAVA_ACC_COMMAND+=(-v1 ${COMMIT[1]} -v2 ${COMMIT[2]})
+JAVA_ACC_COMMAND+=(-v1 ${COMMIT[1]}${SHA[1]+"/${SHA[1]}"})
+JAVA_ACC_COMMAND+=(-v2 ${COMMIT[2]}${SHA[2]+"/${SHA[2]}"})
 JAVA_ACC_COMMAND+=(-d1 ${JARS[1]} -d2 ${JARS[2]})
 JAVA_ACC_COMMAND+=(-report-path \
     ${SCRIPT_DIRECTORY}/target/compatibility/report/${COMMIT[1]}_${COMMIT[2]}_compat_report.html)
-- 
2.5.4 (Apple Git-61)

