diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionCoprocessorHost.java hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionCoprocessorHost.java
index a7d0627..1383730 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionCoprocessorHost.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionCoprocessorHost.java
@@ -269,7 +269,7 @@ public class RegionCoprocessorHost
               // ignore
             }
             Configuration ourConf;
-            if (cfgSpec != null) {
+            if (cfgSpec != null && !cfgSpec.trim().equals("|")) {
               cfgSpec = cfgSpec.substring(cfgSpec.indexOf('|') + 1);
               // do an explicit deep copy of the passed configuration
               ourConf = new Configuration(false);
