diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/mapred/TableSplit.java hbase-server/src/main/java/org/apache/hadoop/hbase/mapred/TableSplit.java
index b3c8b82..1d8a55d 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/mapred/TableSplit.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/mapred/TableSplit.java
@@ -119,8 +119,14 @@ public class TableSplit implements InputSplit, Comparable<TableSplit> {
 
   @Override
   public String toString() {
-    return m_regionLocation + ":" +
-      Bytes.toStringBinary(m_startRow) + "," + Bytes.toStringBinary(m_endRow);
+      StringBuilder sb = new StringBuilder();
+      sb.append("HBase table split. ");
+      sb.append("Table name: ").append(m_tableName.getNameAsString());
+      sb.append(", start row: ").append(Bytes.toStringBinary(m_startRow));
+      sb.append(", end row: ").append(Bytes.toStringBinary(m_endRow));
+      sb.append(", region location: ").append(m_regionLocation);
+      sb.append(".");
+      return sb.toString();
   }
 
   @Override
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableSplit.java hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableSplit.java
index 5e3aeef..84afbe1 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableSplit.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableSplit.java
@@ -314,8 +314,14 @@ implements Writable, Comparable<TableSplit> {
    */
   @Override
   public String toString() {
-    return regionLocation + ":" +
-      Bytes.toStringBinary(startRow) + "," + Bytes.toStringBinary(endRow);
+    StringBuilder sb = new StringBuilder();
+    sb.append("HBase table split. ");
+    sb.append("Table name: ").append(tableName.getNameAsString());
+    sb.append(", start row: ").append(Bytes.toStringBinary(startRow));
+    sb.append(", end row: ").append(Bytes.toStringBinary(endRow));
+    sb.append(", region location: ").append(regionLocation);
+    sb.append(".");
+    return sb.toString();
   }
 
   /**
