diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java
index 3d40c70..1c8dd40 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java
@@ -1829,29 +1829,32 @@ public class MetaTableAccessor {
   /**
    * Deletes the specified regions from META.
    * @param connection connection we're using
-   * @param regionsInfo list of regions to be deleted from META
+   * @param regionInfos list of regions to be deleted from META
    * @throws IOException
    */
   public static void deleteRegions(Connection connection,
-                                   List<HRegionInfo> regionsInfo) throws IOException {
-    deleteRegions(connection, regionsInfo, EnvironmentEdgeManager.currentTime());
+                                   List<HRegionInfo> regionInfos) throws IOException {
+    deleteRegions(connection, regionInfos, EnvironmentEdgeManager.currentTime());
   }
   /**
    * Deletes the specified regions from META.
    * @param connection connection we're using
-   * @param regionsInfo list of regions to be deleted from META
+   * @param regionInfos list of regions to be deleted from META
    * @throws IOException
    */
-  public static void deleteRegions(Connection connection,
-                                   List<HRegionInfo> regionsInfo, long ts) throws IOException {
-    List<Delete> deletes = new ArrayList<Delete>(regionsInfo.size());
-    for (HRegionInfo hri: regionsInfo) {
+  public static void deleteRegions(Connection connection, List<HRegionInfo> regionInfos, long ts)
+      throws IOException {
+    List<Delete> deletes = new ArrayList<Delete>(regionInfos.size());
+    for (HRegionInfo hri : regionInfos) {
       Delete e = new Delete(hri.getRegionName());
       e.addFamily(getCatalogFamily(), ts);
       deletes.add(e);
     }
     deleteFromMetaTable(connection, deletes);
-    LOG.info("Deleted " + regionsInfo);
+    LOG.info("Deleted " + regionInfos.size() + " regions from META");
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Deleted regions: " + regionInfos);
+    }
   }
 
   /**
@@ -1903,7 +1906,10 @@ public class MetaTableAccessor {
     //
     // HBASE-13875 uses master timestamp for the mutations. The 20ms sleep is not needed
     addRegionsToMeta(connection, regionInfos, regionReplication, now+1);
-    LOG.info("Overwritten " + regionInfos);
+    LOG.info("Overwritten " + regionInfos.size() + " regions to Meta");
+    if (LOG.isDebugEnabled()) {
+      LOG.info("Overwritten regions: " + regionInfos);
+    }
   }
 
   /**
