diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
index de3439b..512ab82 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
@@ -39,6 +39,7 @@ import org.apache.hadoop.hbase.classification.InterfaceAudience;
 import org.apache.hadoop.hbase.util.IdLock;
 
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.util.StringUtils;
 
 /**
  * The cache for mob files.
@@ -117,8 +118,13 @@ public class MobFileCache {
       }
       this.scheduleThreadPool.scheduleAtFixedRate(new EvictionThread(this), period, period,
           TimeUnit.SECONDS);
+
+      LOG.info("MobFileCache enabled with cache size: " +
+          StringUtils.TraditionalBinaryPrefix.long2String(mobFileMaxCacheSize, "", 1) +
+          ", evict periods: " +  period + "sec, evict remain ratio: " + evictRemainRatio);
+    } else {
+      LOG.info("MobFileCache disabled");
     }
-    LOG.info("MobFileCache is initialized, and the cache size is " + mobFileMaxCacheSize);
   }
 
   /**
