diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/client/RequestController.java hbase-client/src/main/java/org/apache/hadoop/hbase/client/RequestController.java
index 7e9c968..46e730e 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/client/RequestController.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/client/RequestController.java
@@ -95,7 +95,7 @@ public interface RequestController {
   /**
    * @return The number of running task.
    */
-  long getNumberOfTsksInProgress();
+  long getNumberOfTasksInProgress();
 
   /**
    * Waits for the running tasks to complete.
diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/client/SimpleRequestController.java hbase-client/src/main/java/org/apache/hadoop/hbase/client/SimpleRequestController.java
index 473f264..6343af6 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/client/SimpleRequestController.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/client/SimpleRequestController.java
@@ -246,7 +246,7 @@ class SimpleRequestController implements RequestController {
   }
 
   @Override
-  public long getNumberOfTsksInProgress() {
+  public long getNumberOfTasksInProgress() {
     return tasksInProgress.get();
   }
 
