diff --git a/hbase-shell/src/main/ruby/hbase/visibility_labels.rb b/hbase-shell/src/main/ruby/hbase/visibility_labels.rb
index a3f8b1a..b715521 100644
--- a/hbase-shell/src/main/ruby/hbase/visibility_labels.rb
+++ b/hbase-shell/src/main/ruby/hbase/visibility_labels.rb
@@ -45,7 +45,8 @@ module Hbase
       end

       begin
-        response = VisibilityClient.addLabels(@config, labels.to_java(:string))
+        connection = @admin.getConnection()
+        response = VisibilityClient.addLabels(connection, labels.to_java(:string))
         if response.nil?
           raise(ArgumentError, "DISABLED: Visibility labels feature is not available")
         end
@@ -70,7 +71,8 @@ module Hbase
       end

       begin
-        response = VisibilityClient.setAuths(@config, auths.to_java(:string), user)
+        connection = @admin.getConnection()
+        response = VisibilityClient.setAuths(connection, auths.to_java(:string), user)
         if response.nil?
           raise(ArgumentError, "DISABLED: Visibility labels feature is not available")
         end
@@ -90,7 +92,8 @@ module Hbase
     def get_auths(user)
       lables_table_available?
       begin
-        response = VisibilityClient.getAuths(@config, user)
+        connection = @admin.getConnection()
+        response = VisibilityClient.getAuths(connection, user)
         if response.nil?
           raise(ArgumentError, "DISABLED: Visibility labels feature is not available")
         end
