diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/DelegatingRetryingCallable.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/DelegatingRetryingCallable.java
index 984a867..cdb23bd 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/DelegatingRetryingCallable.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/DelegatingRetryingCallable.java
@@ -21,6 +21,13 @@ import java.io.IOException;
 
 import org.apache.hadoop.hbase.classification.InterfaceAudience;
 
+/**
+ * Helper callable for internal use whe you just want to override a single method of a {@link
+ * RetryingCallable}. By default, this just delegates all {@link RetryingCallable} methods to the
+ * specified delegate.
+ * @param <T> Result class from calls to the delegate {@link RetryingCallable}
+ * @param <D> Type of the delegate class
+ */
 @InterfaceAudience.Private
 public class DelegatingRetryingCallable<T, D extends RetryingCallable<T>> implements
     RetryingCallable<T> {
diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RetryingCallable.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RetryingCallable.java
index ea65fcf..7a32175 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RetryingCallable.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RetryingCallable.java
@@ -26,7 +26,7 @@ import org.apache.hadoop.hbase.classification.InterfaceAudience;
 /**
  * A Callable&lt;T&gt; that will be retried.  If {@link #call(int)} invocation throws exceptions,
  * we will call {@link #throwable(Throwable, boolean)} with whatever the exception was.
- * @param <T>
+ * @param <T> result class from executing <tt>this</tt>
  */
 @InterfaceAudience.Private
 public interface RetryingCallable<T> {
