diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/ColumnRangeFilter.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/ColumnRangeFilter.java
index 9963af6..d8ea094 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/ColumnRangeFilter.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/ColumnRangeFilter.java
@@ -133,7 +133,7 @@ public class ColumnRangeFilter extends FilterBase {
     }
 
     if (!this.minColumnInclusive && cmpMin == 0) {
-      return ReturnCode.SKIP;
+      return ReturnCode.NEXT_COL;
     }
 
     if (this.maxColumn == null) {
diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/FamilyFilter.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/FamilyFilter.java
index e289026..e79a4d5 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/FamilyFilter.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/filter/FamilyFilter.java
@@ -65,7 +65,7 @@ public class FamilyFilter extends CompareFilter {
     if (familyLength > 0) {
       if (doCompare(this.compareOp, this.comparator, v.getFamilyArray(),
           v.getFamilyOffset(), familyLength)) {
-        return ReturnCode.SKIP;
+        return ReturnCode.NEXT_ROW;
       }
     }
     return ReturnCode.INCLUDE;
