diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/SizeCachedNoTagsKeyValue.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/SizeCachedNoTagsKeyValue.java
index 23ed0c6..619196d 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/SizeCachedNoTagsKeyValue.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/SizeCachedNoTagsKeyValue.java
@@ -19,7 +19,11 @@
  */
 package org.apache.hadoop.hbase;
 
+import java.io.IOException;
+import java.io.OutputStream;
+
 import org.apache.hadoop.hbase.classification.InterfaceAudience;
+import org.apache.hadoop.hbase.util.Bytes;
 
 /**
  * This class is an extension to ContentSizeCachedKeyValue where there are no tags in Cell.
@@ -42,4 +46,11 @@ public class SizeCachedNoTagsKeyValue extends SizeCachedKeyValue {
   public long heapSizeWithoutTags() {
     return super.heapSize();
   }
+
+  @Override
+  public int write(OutputStream out, boolean withTags) throws IOException {
+    writeInt(out, this.length);
+    out.write(this.bytes, this.offset, this.length);
+    return this.length + Bytes.SIZEOF_INT;
+  }
 }
