diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java.rej b/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java.rej
deleted file mode 100644
index 7cd2dad..0000000
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java.rej
+++ /dev/null
@@ -1,16 +0,0 @@
---- hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java
-+++ hbase-client/src/main/java/org/apache/hadoop/hbase/protobuf/ProtobufUtil.java
-@@ -124,12 +126,12 @@
- import org.apache.hadoop.hbase.protobuf.generated.RegionServerStatusProtos.RegionServerReportRequest;
- import org.apache.hadoop.hbase.protobuf.generated.RegionServerStatusProtos.RegionServerStartupRequest;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos;
-+import org.apache.hadoop.hbase.protobuf.generated.WALProtos.BulkLoadDescriptor;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.CompactionDescriptor;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.FlushDescriptor;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.FlushDescriptor.FlushAction;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.RegionEventDescriptor;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.RegionEventDescriptor.EventType;
--import org.apache.hadoop.hbase.protobuf.generated.WALProtos.BulkLoadDescriptor;
- import org.apache.hadoop.hbase.protobuf.generated.WALProtos.StoreDescriptor;
- import org.apache.hadoop.hbase.quotas.QuotaScope;
- import org.apache.hadoop.hbase.quotas.QuotaType;
