diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/InternalScan.java hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/InternalScan.java
index 9baac9b..143f800 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/InternalScan.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/InternalScan.java
@@ -18,12 +18,15 @@
  */
 package org.apache.hadoop.hbase.regionserver;
 
+import java.io.IOException;
+
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
 import org.apache.hadoop.hbase.classification.InterfaceAudience;
 import org.apache.hadoop.hbase.client.Get;
 import org.apache.hadoop.hbase.client.Scan;
 
 /**
- * Special internal-only scanner, currently used for increment operations to
+ * Special scanner, currently used for increment operations to
  * allow additional server-side arguments for Scan operations.
  * <p>
  * Rather than adding new options/parameters to the public Scan API, this new
@@ -33,8 +36,8 @@ import org.apache.hadoop.hbase.client.Scan;
  * {@link #checkOnlyMemStore()} or to only read from StoreFiles with
  * {@link #checkOnlyStoreFiles()}.
  */
-@InterfaceAudience.Private
-class InternalScan extends Scan {
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
+public class InternalScan extends Scan {
   private boolean memOnly = false;
   private boolean filesOnly = false;
 
@@ -46,6 +49,16 @@ class InternalScan extends Scan {
   }
 
   /**
+   * @param scan - original scan object
+   * @throws IOException 
+   */
+  public InternalScan(Scan scan) 
+      throws IOException 
+  {
+    super(scan);
+  }
+  
+  /**
    * StoreFiles will not be scanned. Only MemStore will be scanned.
    */
   public void checkOnlyMemStore() {
