diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/master/NamespaceJanitor.java hbase-server/src/main/java/org/apache/hadoop/hbase/master/NamespaceJanitor.java
deleted file mode 100644
index e69de29..0000000
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java
index dc593ac..a74a463 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java
@@ -707,6 +707,8 @@ public class HStore implements Store {
       }
 
       if (verifyBulkLoads) {
+        long verificationStartTime = EnvironmentEdgeManager.currentTimeMillis();
+        LOG.info("Full verification started for bulk load hfile: " + srcPath.toString());
         Cell prevKV = null;
         HFileScanner scanner = reader.getScanner(false, false, false);
         scanner.seekTo();
@@ -732,6 +734,9 @@ public class HStore implements Store {
           }
           prevKV = kv;
         } while (scanner.next());
+      LOG.info("Full verification complete for bulk load hfile: " + srcPath.toString()
+         + " took " + (EnvironmentEdgeManager.currentTimeMillis() - verificationStartTime)
+         + " ms");
       }
     } finally {
       if (reader != null) reader.close();
