Index: C:/Documents and Settings/nomi/workspace/kernel/src/org/apache/axis2/receivers/RawXMLINOutMessageReceiver.java
===================================================================
--- C:/Documents and Settings/nomi/workspace/kernel/src/org/apache/axis2/receivers/RawXMLINOutMessageReceiver.java	(revision 436938)
+++ C:/Documents and Settings/nomi/workspace/kernel/src/org/apache/axis2/receivers/RawXMLINOutMessageReceiver.java	(working copy)
@@ -99,8 +99,12 @@
                     result = (OMElement) method.invoke(obj, args);
 
                 AxisService service = msgContext.getAxisService();
-                result.declareNamespace(service.getTargetNamespace(),
-                        service.getTargetNamespacePrefix());
+
+                // handle empty body
+                if(result != null)
+                	result.declareNamespace(service.getTargetNamespace(),
+                			service.getTargetNamespacePrefix());
+                
                 OMElement bodyContent;
 
                 SOAPFactory fac = getSOAPFactory(msgContext);
