Index: src/org/apache/axis2/deployment/DeploymentClassLoader.java
===================================================================
--- src/org/apache/axis2/deployment/DeploymentClassLoader.java	(revision 775133)
+++ src/org/apache/axis2/deployment/DeploymentClassLoader.java	(working copy)
@@ -260,4 +260,17 @@
         }
         return null;
     }
+
+    protected synchronized Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+
+        Class c = findLoadedClass(name);
+        if (c == null) {
+            try {
+                c = findClass(name);
+            } catch (Exception e) {
+                c = super.loadClass(name, resolve);
+            }
+        }
+        return c;
+    }
 }
