Index: modules/saaj/src/org/apache/axis2/om/impl/dom/ParentNode.java
===================================================================
--- modules/saaj/src/org/apache/axis2/om/impl/dom/ParentNode.java	(revision 358707)
+++ modules/saaj/src/org/apache/axis2/om/impl/dom/ParentNode.java	(working copy)
@@ -116,7 +116,10 @@
 	///	
 	
 	public NodeList getChildNodes() {
-		return new NodeListImpl(this, null,null);
+        if(!this.done){
+            this.build();
+        }
+        return new NodeListImpl(this, null,null);
 	}
 	
 	public Node getFirstChild() {
