Index: AxisConfiguration.java
===================================================================
--- AxisConfiguration.java	(revision 483469)
+++ AxisConfiguration.java	(working copy)
@@ -680,11 +680,11 @@
         return this.systemClassLoader;
     }
 
-    public TransportInDescription getTransportIn(QName name) throws AxisFault {
+    public TransportInDescription getTransportIn(QName name) {
         return (TransportInDescription) transportsIn.get(name);
     }
 
-    public TransportOutDescription getTransportOut(QName name) throws AxisFault {
+    public TransportOutDescription getTransportOut(QName name) {
         return (TransportOutDescription) transportsOut.get(name);
     }
 
