Index: src/main/webapp/axis2-web/listGroupService.jsp
===================================================================
--- src/main/webapp/axis2-web/listGroupService.jsp	(revision 1169767)
+++ src/main/webapp/axis2-web/listGroupService.jsp	(working copy)
@@ -52,7 +52,7 @@
 <h2><font color="blue"><a href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a>
 </font></h2>
 <font color="blue">Service EPR :</font><font color="black"><%=prefix + axisService.getName()%></font>
-<h4>Service Description : <font color="black"><%=axisService.getServiceDescription()%></font></h4>
+<h4>Service Description : <font color="black"><%=axisService.getDocumentation()%></font></h4>
 <i><font color="blue">Service Status : <%=axisService.isActive() ? "Active" : "InActive"%></font></i><br/>
 <%
   Collection engagedModules = axisService.getEngagedModules();
Index: src/main/webapp/axis2-web/listSingleService.jsp
===================================================================
--- src/main/webapp/axis2-web/listSingleService.jsp	(revision 1169767)
+++ src/main/webapp/axis2-web/listSingleService.jsp	(working copy)
@@ -68,7 +68,7 @@
     }
 
 
-    String serviceDescription = axisService.getServiceDescription();
+    String serviceDescription = axisService.getDocumentation();
     if (serviceDescription == null || "".equals(serviceDescription)) {
         serviceDescription = "No description available for this service";
     }
Index: src/main/webapp/axis2-web/listServices.jsp
===================================================================
--- src/main/webapp/axis2-web/listServices.jsp	(revision 1169767)
+++ src/main/webapp/axis2-web/listServices.jsp	(working copy)
@@ -77,7 +77,7 @@
     }
 
 
-    String serviceDescription = axisService.getServiceDescription();
+    String serviceDescription = axisService.getDocumentation();
     if (serviceDescription == null || "".equals(serviceDescription)) {
         serviceDescription = "No description available for this service";
     }
Index: src/main/webapp/axis2-web/listFaultyService.jsp
===================================================================
--- src/main/webapp/axis2-web/listFaultyService.jsp	(revision 1169767)
+++ src/main/webapp/axis2-web/listFaultyService.jsp	(working copy)
@@ -55,7 +55,7 @@
           String  serviceName = axisService.getName();
             %><h2><font color="blue"><a href="<%=prifix + axisService.getName()%>?wsdl"><%=serviceName%></a></font></h2>
            <font color="blue">Service EPR : </font><font color="black"><%=prifix + axisService.getName()%></font><br>
-           <h4>Service Description : <font color="black"><%=axisService.getServiceDescription()%></font></h4>
+           <h4>Service Description : <font color="black"><%=axisService.getDocumentation()%></font></h4>
            <i><font color="blue">Service Status : <%=axisService.isActive()?"Active":"InActive"%></font></i><br>
            <%
             if (opItr.hasNext()) {
Index: src/main/webapp/axis2-web/listService.jsp
===================================================================
--- src/main/webapp/axis2-web/listService.jsp	(revision 1169767)
+++ src/main/webapp/axis2-web/listService.jsp	(working copy)
@@ -70,7 +70,7 @@
     }
 
 
-    String serviceDescription = axisService.getServiceDescription();
+    String serviceDescription = axisService.getDocumentation();
     if (serviceDescription == null || "".equals(serviceDescription)) {
         serviceDescription = "No description available for this service";
     }
