Index: ListingAgent.java
===================================================================
--- ListingAgent.java	(revision 127)
+++ ListingAgent.java	(working copy)
@@ -27,6 +27,7 @@
 import org.apache.axis2.deployment.DeploymentConstants;
 import org.apache.axis2.description.AxisDescription;
 import org.apache.axis2.description.AxisService;
+import org.apache.axis2.description.Parameter;
 import org.apache.axis2.description.PolicyInclude;
 import org.apache.axis2.description.TransportInDescription;
 import org.apache.axis2.transport.TransportListener;
@@ -418,7 +419,7 @@
      * This class is just to add tarnsport at the runtime if user send requet using
      * diffrent schemes , simly to handle http/https seperetaly
      */
-    private class HTTPSListener implements TransportListener {
+    public class HTTPSListener implements TransportListener {
 
         private int port;
         private String schema;
@@ -432,6 +433,11 @@
         public void init(ConfigurationContext axisConf,
                          TransportInDescription transprtIn) throws AxisFault {
             this.axisConf = axisConf;
+            
+            Parameter param = transprtIn.getParameter(PARAM_PORT);
+            if (param != null) {
+                this.port = Integer.parseInt((String) param.getValue());
+            }
         }
 
         public void start() throws AxisFault {
@@ -442,7 +448,7 @@
 
         public EndpointReference[] getEPRsForService(String serviceName, String ip)
                 throws AxisFault {
-            return new EndpointReference[]{new EndpointReference(schema + "://" + ip + ":" + port +
+            return new EndpointReference[]{new EndpointReference("https" + "://" + ip + ":" + port +
                                                                  "/" + axisConf.getServiceContextPath() + "/" +
                                                                  serviceName)};  //To change body of implemented methods use File | Settings | File Templates.
         }
