Index: modules/metadata/src/org/apache/axis2/jaxws/description/validator/EndpointDescriptionValidator.java
===================================================================
--- modules/metadata/src/org/apache/axis2/jaxws/description/validator/EndpointDescriptionValidator.java	(revision 1023120)
+++ modules/metadata/src/org/apache/axis2/jaxws/description/validator/EndpointDescriptionValidator.java	(working copy)
@@ -138,7 +138,9 @@
         // Validate that the WSDL value is valid
         else if (!SOAPBinding.SOAP11HTTP_BINDING.equals(wsdlBindingType)
                 && !SOAPBinding.SOAP12HTTP_BINDING.equals(wsdlBindingType)
-                && !javax.xml.ws.http.HTTPBinding.HTTP_BINDING.equals(wsdlBindingType)) {
+                && !javax.xml.ws.http.HTTPBinding.HTTP_BINDING.equals(wsdlBindingType)
+                && !MDQConstants.SOAP11JMS_BINDING.equals(wsdlBindingType)
+                && !MDQConstants.SOAP12JMS_BINDING.equals(wsdlBindingType)) {
             addValidationFailure(this, "Invalid wsdl binding value specified: " 
                                  + DescriptionUtils.mapBindingTypeAnnotationToWsdl(wsdlBindingType));
             isBindingValid = false;
@@ -156,6 +158,12 @@
                  && bindingType.equals(HTTPBinding.HTTP_BINDING)) {
             isBindingValid = true;
         }
+        else if (wsdlBindingType.equals(MDQConstants.SOAP11JMS_BINDING)&& bindingType.startsWith(MDQConstants.SOAP11JMS_BINDING)) {
+            isBindingValid = true;
+        }
+        else if (wsdlBindingType.equals(MDQConstants.SOAP12JMS_BINDING)&& bindingType.startsWith(MDQConstants.SOAP12JMS_BINDING)) {
+            isBindingValid = true;
+        }
         // The HTTP binding is not valid on a Java Bean SEI-based endpoint; only on a Provider based one.
         else if (wsdlBindingType.equals(HTTPBinding.HTTP_BINDING) &&
                 endpointDesc.isEndpointBased()) {
Index: modules/metadata/src/org/apache/axis2/jaxws/description/builder/MDQConstants.java
===================================================================
--- modules/metadata/src/org/apache/axis2/jaxws/description/builder/MDQConstants.java	(revision 1023120)
+++ modules/metadata/src/org/apache/axis2/jaxws/description/builder/MDQConstants.java	(working copy)
@@ -83,10 +83,10 @@
     //Represent SOAP/JMS Bindings
     //REVIEW: SOAP-JMS may be using the same NS for SOAP11 and SOAP12, 
     //  if so we could remove some duplicate values below
-    public static final String SOAP11JMS_BINDING = "http://www.example.org/2006/06/soap/bindings/JMS/";
-    public static final String SOAP12JMS_BINDING = "http://www.example.org/2006/06/soap/bindings/JMS/";
-    public static final String SOAP11JMS_MTOM_BINDING = "http://http://www.example.org/2006/06/soap/bindings/JMS/?mtom=true";
-    public static final String SOAP12JMS_MTOM_BINDING = "http://http://www.example.org/2006/06/soap/bindings/JMS/?mtom=true";
+    public static final String SOAP11JMS_BINDING = "http://www.w3.org/2010/soapjms/";
+    public static final String SOAP12JMS_BINDING = SOAP11JMS_BINDING;
+    public static final String SOAP11JMS_MTOM_BINDING = "http://www.w3.org/2010/soapjms/?mtom=true";
+    public static final String SOAP12JMS_MTOM_BINDING = SOAP11JMS_MTOM_BINDING;
     public static final String SOAP_HTTP_BINDING ="SOAP_HTTP_BINDING";
     
     public static final String USE_LEGACY_WEB_METHOD_RULES_SUN = "com.sun.xml.ws.model.RuntimeModeler.legacyWebMethod";
Index: modules/jaxws/src/org/apache/axis2/jaxws/message/Protocol.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/message/Protocol.java	(revision 1023120)
+++ modules/jaxws/src/org/apache/axis2/jaxws/message/Protocol.java	(working copy)
@@ -61,8 +61,8 @@
         protocolMappings.put(SOAPBinding.SOAP12HTTP_BINDING, Protocol.soap12);
         protocolMappings.put(SOAPBinding.SOAP12HTTP_MTOM_BINDING, Protocol.soap12);
         protocolMappings.put(HTTPBinding.HTTP_BINDING, Protocol.rest);
-        protocolMappings.put(MDQConstants.SOAP12JMS_BINDING, Protocol.soap12);
-        protocolMappings.put(MDQConstants.SOAP12JMS_MTOM_BINDING, Protocol.soap12);
+        // protocolMappings.put(MDQConstants.SOAP12JMS_BINDING, Protocol.soap12);
+        // protocolMappings.put(MDQConstants.SOAP12JMS_MTOM_BINDING, Protocol.soap12);
         
         // Add each of the URLs with a "/" at the end for flexibility
         Map<String, Protocol> updates = new HashMap<String, Protocol>();
