Index: modules/kernel/src/org/apache/axis2/engine/ListenerManager.java
===================================================================
--- modules/kernel/src/org/apache/axis2/engine/ListenerManager.java	(revision 2455)
+++ modules/kernel/src/org/apache/axis2/engine/ListenerManager.java	(revision 2456)
@@ -90,11 +90,7 @@
                         (TransportListener) startedTransports.get(exposeTransport.get(0));
 
                 EndpointReference[] eprsForService;
-                if (opName == null) {
-                    eprsForService = listener.getEPRsForService(serviceName, null);
-                } else {
-                    eprsForService = listener.getEPRsForService(serviceName + "/" + opName, null);
-                }
+                eprsForService = listener.getEPRsForService(serviceName, null);
                 return eprsForService != null ? eprsForService[0] : null;
             }
 
@@ -103,11 +99,7 @@
                     configctx.getAxisConfiguration().getTransportIn(new QName(transportName));
             TransportListener listener = trsIN.getReceiver();
             EndpointReference[] eprsForService;
-            if (opName == null) {
-                eprsForService = listener.getEPRsForService(serviceName, null);
-            } else {
-                eprsForService = listener.getEPRsForService(serviceName + "/" + opName, null);
-            }
+            eprsForService = listener.getEPRsForService(serviceName, null);
             return eprsForService != null ? eprsForService[0] : null;
         }
     }
