Index: C:/Vladimirov/workspace/Axis2_1.4/modules/kernel/src/org/apache/axis2/engine/ListenerManager.java
===================================================================
--- C:/Vladimirov/workspace/Axis2_1.4/modules/kernel/src/org/apache/axis2/engine/ListenerManager.java	(revision 669456)
+++ C:/Vladimirov/workspace/Axis2_1.4/modules/kernel/src/org/apache/axis2/engine/ListenerManager.java	(working copy)
@@ -50,6 +50,7 @@
     private ConfigurationContext configctx;
     private HashMap startedTransports = new HashMap();
     private boolean stopped ;
+    private Thread shutdownHook;
 
     public void init(ConfigurationContext configCtx) {
         configCtx.setTransportManager(this);
@@ -132,7 +133,9 @@
                 log.info(e.getMessage(), e);
             }
         }
-        Runtime.getRuntime().addShutdownHook(new ListenerManagerShutdownThread(this));
+
+        shutdownHook = new ListenerManagerShutdownThread(this);
+        Runtime.getRuntime().addShutdownHook(shutdownHook);
     }
 
     public synchronized void startSystem(ConfigurationContext configurationContext) {
@@ -188,6 +191,10 @@
                 serviceLifeCycle.shutDown(configctx, axisService);
             }
         }
+
+        if (shutdownHook != null)
+            Runtime.getRuntime().removeShutdownHook(shutdownHook);
+
         stopped = true;
     }
 
