Index: C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/Parameter.java
===================================================================
--- C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/Parameter.java	(revision 502673)
+++ C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/Parameter.java	(working copy)
@@ -309,9 +309,7 @@
         {
             log.trace(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
         }
-        //System.out.println(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
 
-
         // serialization version ID
         long suid = in.readLong();
 
@@ -393,12 +391,5 @@
         // done
         //---------------------------------------------------------
 
-        // trace point
-        if (log.isTraceEnabled())
-        {
-            log.trace(myClassName+":readExternal():  END");
-        }
-        //System.out.println(myClassName+":readExternal():  END");
-
     }
 }
Index: C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/ParameterIncludeImpl.java
===================================================================
--- C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/ParameterIncludeImpl.java	(revision 502673)
+++ C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/description/ParameterIncludeImpl.java	(working copy)
@@ -246,9 +246,7 @@
         {
             log.trace(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
         }
-        //System.out.println(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
 
-
         // serialization version ID
         long suid = in.readLong();
 
@@ -285,7 +283,6 @@
                 {
                     log.trace(myClassName+":readExternal():  WARNING: parameters doesnot have a defined HashMap ");
                 }
-                //System.out.println(myClassName+":readExternal():  WARNING: parameters doesnot have a defined HashMap ");
             }
         }
 
@@ -293,13 +290,6 @@
         // done
         //---------------------------------------------------------
 
-        // trace point
-        if (log.isTraceEnabled())
-        {
-            log.trace(myClassName+":readExternal():  END");
-        }
-        //System.out.println(myClassName+":readExternal():  END");
-
     }
 
 }
Index: C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/ServiceGroupContext.java
===================================================================
--- C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/ServiceGroupContext.java	(revision 502673)
+++ C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/ServiceGroupContext.java	(working copy)
@@ -387,9 +387,7 @@
         {
             log.trace(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
         }
-        //System.out.println(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
 
-
         // serialization version ID
         long suid = in.readLong();
 
Index: C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/SessionContext.java
===================================================================
--- C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/SessionContext.java	(revision 502673)
+++ C:/WS2/apache/axis2/r502673.2002b/modules/kernel/src/org/apache/axis2/context/SessionContext.java	(working copy)
@@ -267,9 +267,7 @@
         {
             log.trace(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
         }
-        //System.out.println(myClassName+":readExternal():  BEGIN  bytes available in stream ["+in.available()+"]  ");
 
-
         // serialization version ID
         long suid = in.readLong();
 
@@ -333,13 +331,6 @@
         // done
         //---------------------------------------------------------
 
-        // trace point
-        if (log.isTraceEnabled())
-        {
-            log.trace(myClassName+":readExternal():  END");
-        }
-        //System.out.println(myClassName+":readExternal():  END");
-
     }
 
 
