Index: modules/kernel/src/org/apache/axis2/engine/AddressingBasedDispatcher.java
===================================================================
--- modules/kernel/src/org/apache/axis2/engine/AddressingBasedDispatcher.java	(revision 520774)
+++ modules/kernel/src/org/apache/axis2/engine/AddressingBasedDispatcher.java	(working copy)
@@ -48,7 +48,7 @@
     public static final String NAME = "AddressingBasedDispatcher";
     private static final Log log = LogFactory.getLog(AddressingBasedDispatcher.class);
 
-    // TODO this logic needed to be improved, as the Dispatching is almost guaranteed to fail
+    //TODO This logic needs to be improved, as the dispatching is almost guaranteed to fail.
     public AxisOperation findOperation(AxisService service, MessageContext messageContext)
             throws AxisFault {
         if (LoggingControl.debugLoggingAllowed && log.isDebugEnabled()) {
@@ -100,8 +100,8 @@
 
                 AxisService axisService = registry.getService(serviceName.getLocalPart());
 
-                // If the axisService is not null we get the binding that the request came to add
-                // add it as a property to the messageContext
+                // If the axisService is not null, the binding that the request came to add is obtained.
+                // Add endpoints as a property to the messageContext
                 if (axisService != null) {
                     Map endpoints = axisService.getEndpoints();
                     if (endpoints != null) {
@@ -131,8 +131,8 @@
      * @param msgctx
      * @throws org.apache.axis2.AxisFault
      */
-    public InvocationResponse invoke(MessageContext msgctx) throws AxisFault {
-        // first check we can dispatch using the relates to
+    public InvocationResponse invoke(MessageContext msgctx) throws AxisFault { 
+     // First check if dispatches using the relatesTo are possible.
         if (msgctx.getRelatesTo() != null) {
             String relatesTo = msgctx.getRelatesTo().getValue();
 
@@ -147,7 +147,7 @@
 
                 if (operationContext != null) {
                     if(operationContext.isComplete()){
-                        // If the dispatch happens because of the RelatesTo and the mep is complete
+                        // If the dispatch happens because of the RelatesTo and the MEP is complete
                         // we should throw a more descriptive fault.
                         throw new AxisFault(Messages.getMessage("duplicaterelatesto",relatesTo));
                     }
