Index: C:/Vladimirov/workspace/Axis2/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java
===================================================================
--- C:/Vladimirov/workspace/Axis2/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java	(revision 668142)
+++ C:/Vladimirov/workspace/Axis2/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java	(working copy)
@@ -433,16 +433,7 @@
             }
             axisConfiguration = configContext.getAxisConfiguration();
 
-            ListenerManager listenerManager = new ListenerManager();
-            listenerManager.init(configContext);
-            TransportInDescription transportInDescription = new TransportInDescription(
-                    Constants.TRANSPORT_HTTP);
-            transportInDescription.setReceiver(this);
-            listenerManager.addListener(transportInDescription, true);
-            listenerManager.start();
-            ListenerManager.defaultConfigurationContext = configContext;
-            agent = new ListingAgent(configContext);
-
+            initTransport();
             initParams();
 
         } catch (Exception e) {
@@ -450,6 +441,18 @@
         }
     }
 
+    protected void initTransport() throws Exception {
+        ListenerManager listenerManager = new ListenerManager();
+        listenerManager.init(configContext);
+        TransportInDescription transportInDescription = new TransportInDescription(
+                Constants.TRANSPORT_HTTP);
+        transportInDescription.setReceiver(this);
+        listenerManager.addListener(transportInDescription, true);
+        listenerManager.start();
+        ListenerManager.defaultConfigurationContext = configContext;
+        agent = new ListingAgent(configContext);
+    }
+
     /**
      * distroy the ConfigurationContext
      */
