Index: src/org/apache/axis2/clustering/tribes/TribesClusteringAgent.java
===================================================================
--- src/org/apache/axis2/clustering/tribes/TribesClusteringAgent.java	(revision 5953)
+++ src/org/apache/axis2/clustering/tribes/TribesClusteringAgent.java	(working copy)
@@ -262,11 +262,14 @@
                 OMElement propEle = (OMElement) iter.next();
                 OMAttribute nameAttrib = propEle.getAttribute(new QName("name"));
                 if(nameAttrib != null){
+                    String attribName = nameAttrib.getAttributeValue();
+                    attribName = replaceProperty(attribName, memberInfo);
+
                     OMAttribute valueAttrib = propEle.getAttribute(new QName("value"));
                     if  (valueAttrib != null) {
                         String attribVal = valueAttrib.getAttributeValue();
                         attribVal = replaceProperty(attribVal, memberInfo);
-                        memberInfo.setProperty(nameAttrib.getAttributeValue(), attribVal);
+                        memberInfo.setProperty(attribName, attribVal);
                     }
                 }
             }
@@ -298,6 +301,9 @@
             String sysProp = text.substring(indexOfStartingChars + 2,
                                             indexOfClosingBrace);
             String propValue = props.getProperty(sysProp);
+            if (propValue == null) {
+                propValue = System.getProperty(sysProp);
+            }
             if (propValue != null) {
                 text = text.substring(0, indexOfStartingChars) + propValue +
                        text.substring(indexOfClosingBrace + 1);
