diff -urw ../origsrc/jms/JMSConstants.java org/apache/axis2/transport/jms/JMSConstants.java
--- ../origsrc/jms/JMSConstants.java	2008-12-03 16:11:38.562500000 +0100
+++ org/apache/axis2/transport/jms/JMSConstants.java	2008-12-03 14:11:27.828125000 +0100
@@ -70,6 +70,16 @@
 
     /**
      * The Parameter name indicating the JMS connection factory username (useful for WebsphereMQ CLIENT connections)
+     */
+    public static final String CONFAC_USER = "transport.jms.ConnectionFactoryUser";
+
+    /**
+     * The Parameter name indicating the JMS connection factory password (useful for WebsphereMQ CLIENT connections)
+     */
+    public static final String CONFAC_PASS = "transport.jms.ConnectionFactoryPass";
+
+    /**
+     * The Parameter name indicating the JMS connection factory username (useful for WebsphereMQ CLIENT connections)
      * n.b. This is not the actual username, it is the JNDI name of the variable that will hold the username
      */
     public static final String CONFAC_JNDI_NAME_USER = "transport.jms.ConnectionFactoryJNDIUser";
diff -urw ../origsrc/jms/JMSOutTransportInfo.java org/apache/axis2/transport/jms/JMSOutTransportInfo.java
--- ../origsrc/jms/JMSOutTransportInfo.java	2008-12-03 16:11:38.609375000 +0100
+++ org/apache/axis2/transport/jms/JMSOutTransportInfo.java	2008-12-03 14:14:12.203125000 +0100
@@ -125,6 +125,11 @@
      * @return the connection factory username (or null if one is not in the JNDI tree)
      */
     private String getConnectionFactoryUser(Context context, Hashtable props) {
+    	String conFacUser = (String) props.get(JMSConstants.CONFAC_USER);
+    	if (conFacUser != null) {
+    		return conFacUser;
+    	}
+    	
         try {
 
             String conFacJndiUser = (String) props.get(JMSConstants.CONFAC_JNDI_NAME_USER);
@@ -147,6 +152,10 @@
      * @return the connection factory password (or null if one is not in the JNDI tree)
      */
     private String getConnectionFactoryPass(Context context, Hashtable props) {
+    	String conFacPass = (String) props.get(JMSConstants.CONFAC_PASS);
+    	if (conFacPass != null) {
+    		return conFacPass;
+    	}
         try {
 
             String conFacJndiPass = (String) props.get(JMSConstants.CONFAC_JNDI_NAME_PASS);
