Index: modules/kernel/src/org/apache/axis2/receivers/AbstractMessageReceiver.java
===================================================================
--- modules/kernel/src/org/apache/axis2/receivers/AbstractMessageReceiver.java	(revision 507856)
+++ modules/kernel/src/org/apache/axis2/receivers/AbstractMessageReceiver.java	(working copy)
@@ -107,9 +107,9 @@
             ClassLoader classLoader = service.getClassLoader();
 
             // allow alternative definition of makeNewServiceObject
-            if (service.getParameter(Constants.SERVICE_OBJECT_SUPPLIER) != null) {
-                Parameter serviceObjectParam =
+            Parameter serviceObjectParam =
                         service.getParameter(Constants.SERVICE_OBJECT_SUPPLIER);
+            if (serviceObjectParam != null) {                
                 Class serviceObjectMaker = Loader.loadClass(classLoader, ((String)
                         serviceObjectParam.getValue()).trim());
 
