Index: modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java	(revision 606826)
+++ modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java	(working copy)
@@ -18,6 +18,7 @@
  */
 package org.apache.axis2.jaxws.message.impl;
 
+import org.apache.axiom.om.OMDataSource;
 import org.apache.axiom.om.OMDataSourceExt;
 import org.apache.axiom.om.OMElement;
 import org.apache.axiom.om.OMException;
@@ -22,6 +23,7 @@
 import org.apache.axiom.om.OMElement;
 import org.apache.axiom.om.OMException;
 import org.apache.axiom.om.OMOutputFormat;
+import org.apache.axiom.om.OMSourcedElement;
 import org.apache.axiom.om.ds.ByteArrayDataSource;
 import org.apache.axiom.om.impl.MTOMXMLStreamWriter;
 import org.apache.axiom.om.impl.builder.StAXOMBuilder;
@@ -138,6 +140,19 @@
             throw ExceptionFactory.makeWebServiceException(
                     Messages.getMessage("BlockImplErr1", this.getClass().getName()));
         }
+        
+        // If an object has been already created in previous component, set the object to this Block.
+        // https://issues.apache.org/jira/browse/AXIS2-3417
+        if (busObject == null && omElement instanceof OMSourcedElement) {
+            OMSourcedElement omse = (OMSourcedElement) omElement;
+            OMDataSource omDataSource = omse.getDataSource();
+            if (omDataSource.getClass() == this.getClass()
+            		&& omDataSource instanceof OMDataSourceExt) {
+            	OMDataSourceExt omDataSourceExt = (OMDataSourceExt) omDataSource;
+                busObject = omDataSourceExt.getObject();
+            }
+        }
+
         if (busObject != null) {
             busObject = _getBOFromBO(busObject, busContext, consume);
         } else {
