Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ResponseWrapperAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ResponseWrapperAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ResponseWrapperAnnot.java	(revision 0)
@@ -0,0 +1,109 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class ResponseWrapperAnnot implements javax.xml.ws.ResponseWrapper{
+
+	private String 	localName;
+	private String 	targetNamespace;
+	private String 	className;			
+
+	
+	/**
+     * A ResponseWrapperAnnot cannot be instantiated.
+     */
+	private  ResponseWrapperAnnot(){
+		
+	}
+	
+	private  ResponseWrapperAnnot(
+			String localName,
+			String targetNamespace,
+			String className)
+	{
+		this.localName = localName;
+		this.targetNamespace = targetNamespace;
+		this.className = className;
+	}
+
+    public static ResponseWrapperAnnot createResponseWrapperAnnotImpl() {
+        return new ResponseWrapperAnnot();
+    }
+
+    public static ResponseWrapperAnnot createResponseWrapperAnnotImpl( 
+    			String localName,
+    			String targetNamespace,
+    			String className
+    		) 
+    {
+        return new ResponseWrapperAnnot( localName, 
+        								targetNamespace, 
+        								className);
+    }
+	
+	
+	/**
+	 * @return Returns the name.
+	 */
+	public String localName() {
+		return localName;
+	}
+	
+	/**
+	 * @return Returns the targetNamespace.
+	 */
+	public String targetNamespace() {
+		return targetNamespace;
+	}
+	
+	/**
+	 * @return Returns the wsdlLocation.
+	 */
+	public String className() {
+		return className;
+	}
+
+	/**
+	 * @param name The name to set.
+	 */
+	public void setLocalName(String localName) {
+		this.localName = localName;
+	}
+
+	/**
+	 * @param targetNamespace The targetNamespace to set.
+	 */
+	public void setTargetNamespace(String targetNamespace) {
+		this.targetNamespace = targetNamespace;
+	}
+
+	/**
+	 * @param wsdlLocation The wsdlLocation to set.
+	 */
+	public void setClassName(String className) {
+		this.className = className;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java	(revision 0)
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+import java.security.Principal;
+
+import javax.xml.ws.handler.MessageContext;
+
+public class WebServiceContextAnnot implements javax.xml.ws.WebServiceContext{
+	
+	private MessageContext messageContext;
+	private Principal userPrincipal;
+	private boolean isUserInRole = false;
+
+	/**
+     * A WebServiceContextAnnot cannot be instantiated.
+     */
+	private  WebServiceContextAnnot(){
+		
+	}
+	
+	private  WebServiceContextAnnot(
+				MessageContext messageContext,
+				Principal userPrincipal,
+				boolean isUserInRole)
+	{
+		this.messageContext = messageContext;
+		this.userPrincipal = userPrincipal;
+		this.isUserInRole = isUserInRole;
+	}
+
+    public static WebServiceContextAnnot createWebServiceConextAnnot() {
+        return new WebServiceContextAnnot();
+    }
+
+    public static WebServiceContextAnnot createWebServiceContextAnnot( 
+			MessageContext messageContext,
+			Principal userPrincipal,
+			boolean isUserInRole
+    		) 
+    {
+        return new WebServiceContextAnnot( 	messageContext,
+        									userPrincipal,
+        									isUserInRole);
+    }
+
+	/**
+	 * @return Returns the isUserInRole.
+	 */
+	public boolean isUserInRole() {
+		return isUserInRole;
+	}
+
+	/**
+	 * @return Returns the messageContext.
+	 */
+	public MessageContext getMessageContext() {
+		return messageContext;
+	}
+
+	/**
+	 * @return Returns the userPrincipal.
+	 */
+	public Principal getUserPrincipal() {
+		return userPrincipal;
+	}
+
+	/**
+	 * @param isUserInRole The isUserInRole to set.
+	 */
+	public void setUserInRole(boolean isUserInRole) {
+		this.isUserInRole = isUserInRole;
+	}
+
+	/**
+	 * @param messageContext The messageContext to set.
+	 */
+	public void setMessageContext(MessageContext messageContext) {
+		this.messageContext = messageContext;
+	}
+
+	/**
+	 * @param userPrincipal The userPrincipal to set.
+	 */
+	public void setUserPrincipal(Principal userPrincipal) {
+		this.userPrincipal = userPrincipal;
+	}
+
+	/**
+	 * @param role The role to check.
+	 * @return Returns boolean indicating whether user is in Role
+	 */
+	public boolean isUserInRole(String role) {
+		return false;
+	}
+
+
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+	
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ServiceModeAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ServiceModeAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ServiceModeAnnot.java	(revision 0)
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+import javax.xml.ws.Service;
+ 
+public class ServiceModeAnnot implements javax.xml.ws.ServiceMode{
+	
+	private Service.Mode value = Service.Mode.PAYLOAD;
+
+	/**
+     * A ServiceModeAnnot cannot be instantiated.
+     */
+	private  ServiceModeAnnot(){
+		
+	}
+	
+	private  ServiceModeAnnot( Service.Mode value)
+	{
+		this.value = value;
+	}
+
+    public static ServiceModeAnnot createWebServiceAnnotImpl() {
+        return new ServiceModeAnnot();
+    }
+
+    public static ServiceModeAnnot createWebServiceAnnotImpl( Service.Mode value) 
+    {
+        return new ServiceModeAnnot( value );
+    }
+
+	public Service.Mode value(){
+		return this.value;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+	
+	//Setters
+	public void setValue(Service.Mode value){
+		this.value = value;
+	}
+	
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/MethodDescriptionComposite.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/MethodDescriptionComposite.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/MethodDescriptionComposite.java	(revision 0)
@@ -0,0 +1,157 @@
+/**
+ * 
+ */
+package org.apache.axis2.jaxws.description.builder;
+
+import java.util.List;
+
+import org.apache.axis2.jaxws.description.ParameterDescription;
+
+
+public class MethodDescriptionComposite {
+	
+	/**
+	 * 
+	 */
+	public MethodDescriptionComposite(String className) {
+		super();
+		// TODO Auto-generated constructor stub
+	}
+	
+	public MethodDescriptionComposite () {
+		
+	}
+	
+	public MethodDescriptionComposite (	
+			String methodName,
+			WebMethodAnnot webMethodAnnot,
+			WebResultAnnot webResultAnnot,
+			boolean oneWayAnnotated,
+			HandlerChainAnnot handlerChainAnnot,
+			SoapBindingAnnot soapBindingAnnot,
+			WebServiceRefAnnot webServiceRefAnnot,
+			WebEndpointAnnot webEndpointAnnot,
+			RequestWrapperAnnot requestWrapperAnnot,
+			ResponseWrapperAnnot responseWrapperAnnot,
+			WebServiceContextAnnot webServiceContextAnnot
+	) {
+		
+		this.methodName 			= methodName;
+		this.webMethodAnnot 		= webMethodAnnot;
+		this.webResultAnnot 		= webResultAnnot;
+		this.oneWayAnnotated 		= oneWayAnnotated;
+		this.handlerChainAnnot 		= handlerChainAnnot;
+		this.soapBindingAnnot 		= soapBindingAnnot;
+		this.webServiceRefAnnot 	= webServiceRefAnnot;
+		this.webEndpointAnnot 		= webEndpointAnnot;
+		this.requestWrapperAnnot 	= requestWrapperAnnot;
+		this.responseWrapperAnnot 	= responseWrapperAnnot;
+		this.webServiceContextAnnot = webServiceContextAnnot;
+	}
+	
+	//Method reflective information
+	private String methodName;	//a public method name in this class
+
+	//'WebMethod' annotations for this method		
+	WebMethodAnnot webMethodAnnot;
+	
+	//'WebResult' annotations for this method
+	WebResultAnnot webResultAnnot;
+
+	//'WebServiceContext' annotation
+	private WebServiceContextAnnot webServiceContextAnnot;
+	
+	//'OneWay' annotations for this method
+	boolean oneWayAnnotated;
+	
+	//'HandlerChain' annotations for this method
+	private HandlerChainAnnot handlerChainAnnot;
+	
+    //'SoapBinding' annotations for this method
+	private SoapBindingAnnot soapBindingAnnot;
+
+	//'WebServiceRef' annotations for this web service
+	private WebServiceRefAnnot webServiceRefAnnot;
+	
+	//Annotations for 'WebEndpoint'
+	private WebEndpointAnnot webEndpointAnnot;
+	
+	//Annotations for 'RequestWrapper'
+	//	TODO EDIT CHECK: Should only be on the methods of an SEI	
+	private RequestWrapperAnnot requestWrapperAnnot;
+
+	//Annotations for 'ResponseWrapper'
+	//	TODO EDIT CHECK: Should only be on the methods of an SEI	
+	private ResponseWrapperAnnot responseWrapperAnnot;
+
+	private List<ParameterDescriptionComposite> parameterDescriptions;
+
+	/**
+	 * @return Returns the webEndpointAnnot.
+	 */
+	public WebEndpointAnnot getWebEndpointAnnot() {
+		return webEndpointAnnot;
+	}
+
+	/**
+	 * @return Returns the requestWrapperAnnot.
+	 */
+	public RequestWrapperAnnot getRequestWrapperAnnot() {
+		return requestWrapperAnnot;
+	}
+
+	/**
+	 * @return Returns the responseWrapperAnnot.
+	 */
+	public ResponseWrapperAnnot getResponseWrapperAnnot() {
+		return responseWrapperAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceContextAnnot.
+	 */
+	public WebServiceContextAnnot getWebServiceContextAnnot() {
+		return webServiceContextAnnot;
+	}
+
+	/**
+	 * @param webEndpointAnnotImpl The webEndpointAnnotImpl to set.
+	 */
+	public void setWebEndpointAnnot(WebEndpointAnnot webEndpointAnnot) {
+		this.webEndpointAnnot = webEndpointAnnot;
+	}
+
+	/**
+	 * @param requestWrapperAnnot The requestWrapperAnnot to set.
+	 */
+	public void setRequestWrapperAnnot(
+			RequestWrapperAnnot requestWrapperAnnot) {
+		this.requestWrapperAnnot = requestWrapperAnnot;
+	}
+
+	/**
+	 * @param responseWrapperAnnot The responseWrapperAnnot to set.
+	 */
+	public void setResponseWrapperAnnot(
+			ResponseWrapperAnnot responseWrapperAnnot) {
+		this.responseWrapperAnnot = responseWrapperAnnot;
+	}
+	
+	/**
+	 * @param webServiceContextAnnot The webServiceContextAnnot to set.
+	 */
+	private void setWebServiceContextAnnot(WebServiceContextAnnot webServiceContextAnnot) {
+		this.webServiceContextAnnot = webServiceContextAnnot;
+	}
+
+
+	/**
+	 *  @param parameterDescription The parameterDescription to add to the set.
+	 */
+	public void addParameterDescriptionComposite(ParameterDescriptionComposite parameterDescription) {
+		parameterDescriptions.add(parameterDescription);
+	}
+	
+
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ParameterDescriptionComposite.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ParameterDescriptionComposite.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/ParameterDescriptionComposite.java	(revision 0)
@@ -0,0 +1,100 @@
+/**
+ * 
+ */
+package org.apache.axis2.jaxws.description.builder;
+
+public class ParameterDescriptionComposite {
+	
+	/**
+	 * 
+	 */
+	public ParameterDescriptionComposite(String className) {
+		super();
+		// TODO Auto-generated constructor stub
+	}
+	
+	public ParameterDescriptionComposite () {
+		
+	}
+	
+	public ParameterDescriptionComposite (	
+			Class parameterType,
+			WebParamAnnot webParamAnnot,
+			WebServiceRefAnnot webServiceRefAnnot,
+			WebServiceContextAnnot webServiceContextAnnot) {
+
+		this.parameterType 			= parameterType;
+		this.webParamAnnot 			= webParamAnnot;
+		this.webServiceRefAnnot 	= webServiceRefAnnot;
+		this.webServiceContextAnnot = webServiceContextAnnot;
+	}
+	
+	//Parameter reflective information
+	private Class parameterType;	
+	
+	//'WebParam' annotations for this parameter
+	private WebParamAnnot	webParamAnnot;
+	
+    //'WebServiceRef' annotations for this parameter
+	private WebServiceRefAnnot webServiceRefAnnot;
+
+    //'WebServiceContext' annotations for this parameter
+	private WebServiceContextAnnot webServiceContextAnnot;
+
+	/**
+	 * @return Returns the parameterType.
+	 */
+	private Class getParameterType() {
+		return parameterType;
+	}
+
+	/**
+	 * @return Returns the webParamAnnot.
+	 */
+	private WebParamAnnot getWebParamAnnot() {
+		return webParamAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceRefAnnot.
+	 */
+	private WebServiceRefAnnot getWebServiceRefAnnot() {
+		return webServiceRefAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceContextAnnot.
+	 */
+	public WebServiceContextAnnot getWebServiceContextAnnot() {
+		return webServiceContextAnnot;
+	}
+
+	/**
+	 * @param parameterType The parameterType to set.
+	 */
+	private void setParameterType(Class parameterType) {
+		this.parameterType = parameterType;
+	}
+
+	/**
+	 * @param webParamAnnot The webParamAnnot to set.
+	 */
+	private void setWebParamAnnot(WebParamAnnot webParamAnnot) {
+		this.webParamAnnot = webParamAnnot;
+	}
+
+	/**
+	 * @param webServiceRefAnnot The webServiceRefAnnot to set.
+	 */
+	private void setWebServiceRefAnnot(WebServiceRefAnnot webServiceRefAnnot) {
+		this.webServiceRefAnnot = webServiceRefAnnot;
+	}
+
+	/**
+	 * @param webServiceContextAnnot The webServiceContextAnnot to set.
+	 */
+	private void setWebServiceContextAnnot(WebServiceContextAnnot webServiceContextAnnot) {
+		this.webServiceContextAnnot = webServiceContextAnnot;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceProviderAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceProviderAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceProviderAnnot.java	(revision 0)
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebServiceProviderAnnot implements javax.xml.ws.WebServiceProvider{
+	
+	private String wsdlLocation = "";
+	private String serviceName = "";
+	private String portName = "";
+	private String targetNamespace = "";
+
+	/**
+     * A WebServiceProviderAnnot cannot be instantiated.
+     */
+	private  WebServiceProviderAnnot(){
+		
+	}
+	
+	private  WebServiceProviderAnnot(
+			String wsdlLocation,
+			String serviceName,
+			String portName,
+			String targetNamespace)
+	{
+		this.targetNamespace = targetNamespace;
+		this.serviceName = serviceName;
+		this.wsdlLocation = wsdlLocation;
+		this.portName = portName;
+	}
+
+    public static WebServiceProviderAnnot createWebServiceAnnotImpl() {
+        return new WebServiceProviderAnnot();
+    }
+
+    public static WebServiceProviderAnnot createWebServiceAnnotImpl( 
+    			String name,
+    			String targetNamespace,
+    			String serviceName,
+    			String wsdlLocation,
+    			String endpointInterface,
+    			String portName
+    		) 
+    {
+        return new WebServiceProviderAnnot( wsdlLocation,
+        										serviceName,
+        										portName,
+        										targetNamespace); 
+    }
+
+	public String targetNamespace(){
+		return this.targetNamespace;
+	}
+	
+	public String serviceName(){
+		return this.serviceName;
+	}
+	
+	public String wsdlLocation(){
+		return this.wsdlLocation;
+	}
+	
+	public String portName(){
+		return this.portName;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+	
+	//Setters
+	public void setTargetNamespace(String targetNamespace){
+		this.targetNamespace = targetNamespace;
+	}
+	
+	public void setServiceName(String serviceName){
+		this.serviceName = serviceName;
+	}
+	
+	public void setWsdlLocation(String wsdlLocation){
+		this.wsdlLocation = wsdlLocation;
+	}
+	
+	public void setPortName(String portName){
+		this.portName = portName;
+	}
+	
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebFaultAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebFaultAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebFaultAnnot.java	(revision 0)
@@ -0,0 +1,55 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebFaultAnnot implements javax.xml.ws.WebFault{
+
+	/**
+     * A WebFaultAnnot cannot be instantiated.
+     */
+	private  WebFaultAnnot(){
+		
+	}
+	
+    public static WebFaultAnnot createWebFaultAnnotImpl() {
+        return new WebFaultAnnot();
+    }
+    
+    /**
+     * Get the 'name'
+     * @return String 
+     */
+	public String name() {
+		return new String("return");
+	}
+	
+	public String targetNamespace() {
+		return new String("");
+	}
+	
+	public String faultBean() {
+		return new String("");
+	}
+				
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebResultAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebResultAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebResultAnnot.java	(revision 0)
@@ -0,0 +1,59 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebResultAnnot implements javax.jws.WebResult{
+
+	/**
+     * A WebResultAnnot cannot be instantiated.
+     */
+	private  WebResultAnnot(){
+		
+	}
+	
+    public static WebResultAnnot createWebResultAnnotImpl() {
+        return new WebResultAnnot();
+    }
+    
+    /**
+     * Get the 'name'
+     * @return String 
+     */
+	public String name() {
+		return new String("return");
+	}
+	
+	public String targetNamespace() {
+		return new String("");
+	}
+	
+	public boolean header() {
+		return false;
+	}
+	
+	public String partName() {
+		return new String("");
+	}
+				
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/BindingTypeAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/BindingTypeAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/BindingTypeAnnot.java	(revision 0)
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class BindingTypeAnnot implements javax.xml.ws.BindingType{
+	
+	private String value = "";
+
+	/**
+     * A BindingTypeAnnot cannot be instantiated.
+     */
+	private  BindingTypeAnnot(){
+		
+	}
+	
+	private  BindingTypeAnnot( String value)
+	{
+		this.value = value;
+	}
+
+    public static BindingTypeAnnot createBindingTypeAnnotImpl() {
+        return new BindingTypeAnnot();
+    }
+
+    public static BindingTypeAnnot createBindingTypeAnnotImpl( String value) 
+    {
+        return new BindingTypeAnnot( value );
+    }
+
+	/**
+	 * @return Returns the value.
+	 */
+	public String value() {
+		return value;
+	}
+
+	/**
+	 * @param value The value to set.
+	 */
+	public void setValue(String value) {
+		this.value = value;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java	(revision 0)
@@ -0,0 +1,272 @@
+/**
+ * 
+ */
+package org.apache.axis2.jaxws.description.builder;
+
+import java.util.List;
+
+import javax.wsdl.Definition;
+
+
+public class DescriptionBuilderComposite {
+	/*
+	 * This structure contains the full reflected class, as well as, the
+	 * possible annotations found for this class...the class description 
+	 * must be complete enough for full validation between class info and annotations
+	 * The annotations will be added to the corresponding class members.
+	 */
+
+	/**
+	 * 
+	 */
+	public DescriptionBuilderComposite(String className) {
+		super();
+		// TODO Auto-generated constructor stub
+	}
+	
+	public DescriptionBuilderComposite () {
+		
+	}
+
+	//Class type within the module
+	private enum ModuleClassType { SERVICEIMPL, SEI, SERVICE, SUPER, PROVIDER, FAULT }
+	
+	//Note: a WSDL is not necessary
+	private Definition	wsdlDefinition;
+
+	// Class-level annotations
+//	TODO: Need getters and setters
+
+	//Annotations for 'WebService'
+	private WebServiceAnnot webServiceAnnot;
+
+	//	TODO EDIT CHECK: WebService and WebServiceProvider are mutually exclusive	
+	//Annotations for 'WebServiceProvider'
+	private WebServiceProviderAnnot webServiceProviderAnnot;
+	
+	//Annotations for 'ServiceMode'
+	//	TODO EDIT CHECK: Should only be set if this is a provider class	
+	private ServiceModeAnnot serviceModeAnnot;
+
+	//Annotations for 'WebServiceClient'
+	private WebServiceClientAnnot webServiceClientAnnot;
+
+	//JAXWS Class level annotations
+	//Annotations for 'WebFault'
+	private WebFaultAnnot webFaultAnnot;
+	
+	//'HandlerChain' annotations for this web service
+	private HandlerChainAnnot handlerChainAnnot;
+	
+    //'SoapBinding' annotations for this web service
+	private SoapBindingAnnot soapBindingAnnot;
+	
+    //'WebServiceRef' annotations for this web service
+	private WebServiceRefAnnot webServiceRefAnnot;
+	
+    //'BindingType' annotations for this web service
+	private BindingTypeAnnot bindingTypeAnnot;
+
+	//'WebServiceContext' annotation
+	private WebServiceContextAnnot webServiceContextAnnot;
+	
+	// Class information
+	private String className;
+	private String[] classModifiers; //public, abstract, final, strictfp...should make this ENUM
+
+	// Methods
+	// Need getters and setters
+	private List<MethodDescriptionComposite> methodDescriptions;		
+
+	public WebServiceAnnot getWebServiceAnnot() {
+		return this.webServiceAnnot;
+	}
+
+	/**
+	 * @return Returns the classModifiers.
+	 */
+	public String[] getClassModifiers() {
+		return classModifiers;
+	}
+
+	/**
+	 * @return Returns the className.
+	 */
+	public String getClassName() {
+		return className;
+	}
+
+	/**
+	 * @return Returns the handlerChainAnnotImpl.
+	 */
+	public HandlerChainAnnot getHandlerChainAnnot() {
+		return handlerChainAnnot;
+	}
+
+	/**
+	 * @return Returns the serviceModeAnnot.
+	 */
+	public ServiceModeAnnot getServiceModeAnnot() {
+		return serviceModeAnnot;
+	}
+
+	/**
+	 * @return Returns the soapBindingAnnot.
+	 */
+	public SoapBindingAnnot getSoapBindingAnnot() {
+		return soapBindingAnnot;
+	}
+
+	/**
+	 * @return Returns the webFaultAnnot.
+	 */
+	public WebFaultAnnot getWebFaultAnnot() {
+		return webFaultAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceClientAnnot.
+	 */
+	public WebServiceClientAnnot getWebServiceClientAnnot() {
+		return webServiceClientAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceProviderAnnot.
+	 */
+	public WebServiceProviderAnnot getWebServiceProviderAnnot() {
+		return webServiceProviderAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceRefAnnot.
+	 */
+	public WebServiceRefAnnot getWebServiceRefAnnot() {
+		return webServiceRefAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceRefAnnot.
+	 */
+	public BindingTypeAnnot getBindingTypeAnnot() {
+		return bindingTypeAnnot;
+	}
+
+	/**
+	 * @return Returns the webServiceContextAnnot.
+	 */
+	public WebServiceContextAnnot getWebServiceContextAnnot() {
+		return webServiceContextAnnot;
+	}
+
+	/**
+	 * @return Returns the wsdlDefinition.
+	 */
+	public Definition getWsdlDefinition() {
+		return wsdlDefinition;
+	}
+	
+	//++++++++
+	//Setters
+	//++++++++
+	public void setWebServiceAnnot(WebServiceAnnot webServiceAnnot) {
+		this.webServiceAnnot = webServiceAnnot;
+	}
+
+	/**
+	 * @param classModifiers The classModifiers to set.
+	 */
+	public void setClassModifiers(String[] classModifiers) {
+		this.classModifiers = classModifiers;
+	}
+
+	/**
+	 * @param className The className to set.
+	 */
+	public void setClassName(String className) {
+		this.className = className;
+	}
+
+	/**
+	 * @param handlerChainAnnot The handlerChainAnnot to set.
+	 */
+	public void setHandlerChainAnnot(HandlerChainAnnot handlerChainAnnot) {
+		this.handlerChainAnnot = handlerChainAnnot;
+	}
+
+	/**
+	 * @param serviceModeAnnot The serviceModeAnnot to set.
+	 */
+	public void setServiceModeAnnot(ServiceModeAnnot serviceModeAnnot) {
+		this.serviceModeAnnot = serviceModeAnnot;
+	}
+
+	/**
+	 * @param soapBindingAnnot The soapBindingAnnot to set.
+	 */
+	public void setSoapBindingAnnot(SoapBindingAnnot soapBindingAnnot) {
+		this.soapBindingAnnot = soapBindingAnnot;
+	}
+
+	/**
+	 * @param webFaultAnnot The webFaultAnnot to set.
+	 */
+	public void setWebFaultAnnot(WebFaultAnnot webFaultAnnot) {
+		this.webFaultAnnot = webFaultAnnot;
+	}
+
+	/**
+	 * @param webServiceClientAnnot The webServiceClientAnnot to set.
+	 */
+	public void setWebServiceClientAnnot(
+			WebServiceClientAnnot webServiceClientAnnot) {
+		this.webServiceClientAnnot = webServiceClientAnnot;
+	}
+
+	/**
+	 * @param webServiceProviderAnnot The webServiceProviderAnnot to set.
+	 */
+	public void setWebServiceProviderAnnot(
+			WebServiceProviderAnnot webServiceProviderAnnot) {
+		this.webServiceProviderAnnot = webServiceProviderAnnot;
+	}
+
+	/**
+	 * @param webServiceRefAnnot The webServiceRefAnnot to set.
+	 */
+	public void setWebServiceRefAnnot(
+			WebServiceRefAnnot webServiceRefAnnot) {
+		this.webServiceRefAnnot = webServiceRefAnnot;
+	}
+
+	/**
+	 * @param wsdlDefinition The wsdlDefinition to set.
+	 */
+	public void setWsdlDefinition(Definition wsdlDefinition) {
+		this.wsdlDefinition = wsdlDefinition;
+	}
+
+	/**
+	 * @param BindingTypeAnnot The BindingTypeAnnot to set.
+	 */
+	public void setBindingTypeAnnot(
+			BindingTypeAnnot bindingTypeAnnot) {
+		this.bindingTypeAnnot = bindingTypeAnnot;
+	}
+
+	/**
+	 * @param webServiceContextAnnot The webServiceContextAnnot to set.
+	 */
+	public void setWebServiceContextAnnot(
+			WebServiceContextAnnot webServiceContextAnnot) {
+		this.webServiceContextAnnot = webServiceContextAnnot;
+	}
+
+	/**
+	 *  @param methodDescription The methodDescription to add to the set.
+	 */
+	public void addMethodDescriptionComposite(MethodDescriptionComposite methodDescription) {
+		methodDescriptions.add(methodDescription);
+	}
+	
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebMethodAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebMethodAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebMethodAnnot.java	(revision 0)
@@ -0,0 +1,51 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebMethodAnnot implements javax.jws.WebMethod{
+
+	/**
+     * A WebMethodAnnot cannot be instantiated.
+     */
+	private  WebMethodAnnot(){
+		
+	}
+	
+    public static WebMethodAnnot createWebMethodAnnotImpl() {
+        return new WebMethodAnnot();
+    }
+    
+	public String operationName() {
+		return new String("");
+	}
+	
+	public String action() {
+		return new String("");
+	}
+	
+	public String exclude() {
+		return new String("");
+	}
+				
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebEndpointAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebEndpointAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebEndpointAnnot.java	(revision 0)
@@ -0,0 +1,65 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebEndpointAnnot implements javax.xml.ws.WebEndpoint{
+
+	private String 	name;
+
+	/**
+     * A WebEndpointAnnot cannot be instantiated.
+     */
+	private	WebEndpointAnnot( String name )
+	{
+		this.name = name;
+	}
+	
+	private WebEndpointAnnot()
+	{
+		
+	}
+
+    public static WebEndpointAnnot createWebEndpointAnnotImpl() {
+        return new WebEndpointAnnot();
+    }
+
+    public static WebEndpointAnnot createWebEndpointAnnotImpl( String name )
+    {
+        return new WebEndpointAnnot( name );
+    }
+
+	/**
+	 * @return Returns the name.
+	 */
+	public String name() {
+		return name;
+	}
+	
+	/**
+	 * @param name The name to set.
+	 */
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/HandlerChainAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/HandlerChainAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/HandlerChainAnnot.java	(revision 0)
@@ -0,0 +1,57 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.net.URL;
+
+import javax.jws.HandlerChain;
+import javax.xml.namespace.QName;
+
+import org.apache.axis2.description.AxisService;
+import org.apache.axis2.jaxws.ExceptionFactory;
+
+import java.lang.annotation.Annotation;
+
+public class HandlerChainAnnot implements javax.jws.HandlerChain{
+
+	/**
+     * A WebServiceAnnot cannot be instantiated.
+     */
+	private  HandlerChainAnnot(){
+		
+	}
+	
+    public static HandlerChainAnnot createHandlerChainAnnotImpl() {
+        return new HandlerChainAnnot();
+    }
+
+    public String file(){
+		return new String("");
+	}
+		
+	public String name(){
+		return new String("");
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/RequestWrapperAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/RequestWrapperAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/RequestWrapperAnnot.java	(revision 0)
@@ -0,0 +1,109 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class RequestWrapperAnnot implements javax.xml.ws.RequestWrapper{
+
+	private String 	localName;
+	private String 	targetNamespace;
+	private String 	className;			
+
+	
+	/**
+     * A RequestWrapperAnnot cannot be instantiated.
+     */
+	private  RequestWrapperAnnot(){
+		
+	}
+	
+	private  RequestWrapperAnnot(
+			String localName,
+			String targetNamespace,
+			String className)
+	{
+		this.localName = localName;
+		this.targetNamespace = targetNamespace;
+		this.className = className;
+	}
+
+    public static RequestWrapperAnnot createRequestWrapperAnnotImpl() {
+        return new RequestWrapperAnnot();
+    }
+
+    public static RequestWrapperAnnot createRequestWrapperAnnotImpl( 
+    			String localName,
+    			String targetNamespace,
+    			String className
+    		) 
+    {
+        return new RequestWrapperAnnot( localName, 
+        								targetNamespace, 
+        								className);
+    }
+	
+	
+	/**
+	 * @return Returns the name.
+	 */
+	public String localName() {
+		return localName;
+	}
+	
+	/**
+	 * @return Returns the targetNamespace.
+	 */
+	public String targetNamespace() {
+		return targetNamespace;
+	}
+	
+	/**
+	 * @return Returns the wsdlLocation.
+	 */
+	public String className() {
+		return className;
+	}
+
+	/**
+	 * @param name The name to set.
+	 */
+	public void setLocalName(String localName) {
+		this.localName = localName;
+	}
+
+	/**
+	 * @param targetNamespace The targetNamespace to set.
+	 */
+	public void setTargetNamespace(String targetNamespace) {
+		this.targetNamespace = targetNamespace;
+	}
+
+	/**
+	 * @param wsdlLocation The wsdlLocation to set.
+	 */
+	public void setClassName(String className) {
+		this.className = className;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceRefAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceRefAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceRefAnnot.java	(revision 0)
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebServiceRefAnnot implements javax.xml.ws.WebServiceRef{
+	
+	private String name = "";
+	private String wsdlLocation = "";
+	private Class type;
+	private Class value;
+	private String mappedName = "";
+	
+	/**
+     * A WebServiceRefAnnot cannot be instantiated.
+     */
+	private  WebServiceRefAnnot(){
+		
+	}
+	
+	private  WebServiceRefAnnot(
+			String 	name,
+			String 	wsdlLocation,
+			Class	type,
+			Class	value,
+			String mappedName)
+	{
+		this.name = name;
+		this.wsdlLocation = wsdlLocation;
+		this.type = type;
+		this.value = value;
+		this.mappedName = mappedName;
+	}
+
+    public static WebServiceRefAnnot createWebServiceRefAnnotImpl() {
+        return new WebServiceRefAnnot();
+    }
+
+    public static WebServiceRefAnnot createWebServiceRefAnnotImpl( 
+    			String 	name,
+    			String 	wsdlLocation,
+    			Class	type,
+    			Class	value,
+    			String	mappedName
+    		) 
+    {
+        return new WebServiceRefAnnot( 	name, 
+        									wsdlLocation, 
+        									type, 
+        									value, 
+        									mappedName);
+    }
+
+	
+	/**
+	 * @return Returns the mappedName.
+	 */
+	public String mappedName() {
+		return mappedName;
+	}
+
+	/**
+	 * @return Returns the name.
+	 */
+	public String name() {
+		return name;
+	}
+
+	/**
+	 * @return Returns the type.
+	 */
+	public Class type() {
+		return type;
+	}
+
+	/**
+	 * @return Returns the value.
+	 */
+	public Class value() {
+		return value;
+	}
+
+	/**
+	 * @return Returns the wsdlLocation.
+	 */
+	public String wsdlLocation() {
+		return wsdlLocation;
+	}
+
+	/**
+	 * @param mappedName The mappedName to set.
+	 */
+	public void setMappedName(String mappedName) {
+		this.mappedName = mappedName;
+	}
+
+	/**
+	 * @param name The name to set.
+	 */
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	/**
+	 * @param type The type to set.
+	 */
+	public void setType(Class type) {
+		this.type = type;
+	}
+
+	/**
+	 * @param value The value to set.
+	 */
+	public void setValue(Class value) {
+		this.value = value;
+	}
+
+	/**
+	 * @param wsdlLocation The wsdlLocation to set.
+	 */
+	public void setWsdlLocation(String wsdlLocation) {
+		this.wsdlLocation = wsdlLocation;
+	}
+
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/SoapBindingAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/SoapBindingAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/SoapBindingAnnot.java	(revision 0)
@@ -0,0 +1,52 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class SoapBindingAnnot implements javax.jws.soap.SOAPBinding{
+
+	/**
+     * A SoapBindingAnnot cannot be instantiated.
+     */
+	private  SoapBindingAnnot(){
+		
+	}
+	
+    public static SoapBindingAnnot createSoapBindingAnnotImpl() {
+        return new SoapBindingAnnot();
+    }
+
+
+    public Style style(){
+		return Style.DOCUMENT;
+	}
+		
+    public Use use(){
+		return Use.LITERAL;
+	}
+	
+    public ParameterStyle parameterStyle(){
+		return ParameterStyle.WRAPPED;
+	}
+			
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceClientAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceClientAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceClientAnnot.java	(revision 0)
@@ -0,0 +1,109 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebServiceClientAnnot implements javax.xml.ws.WebServiceClient{
+
+	private String 	name;
+	private String 	targetNamespace;
+	private String 	wsdlLocation;			
+
+	
+	/**
+     * A WebServiceClientAnnot cannot be instantiated.
+     */
+	private  WebServiceClientAnnot(){
+		
+	}
+	
+	private  WebServiceClientAnnot(
+			String name,
+			String targetNamespace,
+			String wsdlLocation)
+	{
+		this.name = name;
+		this.targetNamespace = targetNamespace;
+		this.wsdlLocation = wsdlLocation;
+	}
+
+    public static WebServiceClientAnnot createWebServiceClientAnnotImpl() {
+        return new WebServiceClientAnnot();
+    }
+
+    public static WebServiceClientAnnot createWebServiceClientAnnotImpl( 
+    			String name,
+    			String targetNamespace,
+    			String wsdlLocation
+    		) 
+    {
+        return new WebServiceClientAnnot( name, 
+        								targetNamespace, 
+        								wsdlLocation);
+    }
+	
+	
+	/**
+	 * @return Returns the name.
+	 */
+	public String name() {
+		return name;
+	}
+	
+	/**
+	 * @return Returns the targetNamespace.
+	 */
+	public String targetNamespace() {
+		return targetNamespace;
+	}
+	
+	/**
+	 * @return Returns the wsdlLocation.
+	 */
+	public String wsdlLocation() {
+		return wsdlLocation;
+	}
+
+	/**
+	 * @param name The name to set.
+	 */
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	/**
+	 * @param targetNamespace The targetNamespace to set.
+	 */
+	public void setTargetNamespace(String targetNamespace) {
+		this.targetNamespace = targetNamespace;
+	}
+
+	/**
+	 * @param wsdlLocation The wsdlLocation to set.
+	 */
+	public void setWsdlLocation(String wsdlLocation) {
+		this.wsdlLocation = wsdlLocation;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebParamAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebParamAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebParamAnnot.java	(revision 0)
@@ -0,0 +1,109 @@
+/* Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+import javax.jws.WebParam.Mode;
+
+public class WebParamAnnot implements javax.jws.WebParam{
+
+	private String 	name;
+	private String 	targetNamespace;
+	private Mode 	mode;
+	private boolean header;
+	private String 	partName;
+
+	/**
+     * A WebParamAnnot cannot be instantiated.
+     */
+	private  WebParamAnnot(){
+		
+	}
+	
+    public static WebParamAnnot createWebParamAnnotImpl() {
+        return new WebParamAnnot();
+    }
+    
+    /**
+     * Get the 'name'
+     * @return String 
+     */
+	public String name() {
+		return new String("return");
+	}
+	
+	public String targetNamespace() {
+		return new String("");
+	}
+	
+	public Mode mode() {
+		return Mode.IN;
+	}
+	
+	public boolean header () {
+		return false;
+	}
+	
+	public String partName() {
+		return new String("");
+	}
+	
+	/**
+	 * @param name The name to set.
+	 */
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	/**
+	 * @param targetNamespace The targetNamespace to set.
+	 */
+	public void setTargetNamespace(String targetNamespace) {
+		this.targetNamespace = targetNamespace;
+	}
+
+	/**
+	 * @param mode The mode to set.
+	 */
+	public void setMode(Mode mode) {
+		this.mode = mode;
+	}
+
+	/**
+	 * @param header The header to set.
+	 */
+	public void setHeader(boolean header) {
+		this.header = header;
+	}
+
+	/**
+	 * @param partName The partName to set.
+	 */
+	public void setPartName(String partName) {
+		this.partName = partName;
+	}
+
+
+	
+				
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+
+}
Index: C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceAnnot.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceAnnot.java	(revision 0)
+++ C:/wasx/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/builder/WebServiceAnnot.java	(revision 0)
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.jaxws.description.builder;
+
+import java.lang.annotation.Annotation;
+
+public class WebServiceAnnot implements javax.jws.WebService{
+	
+	private String name = "";
+	private String targetNamespace = "";
+	private String serviceName = "";
+	private String wsdlLocation = "";
+	private String endpointInterface = "";
+	private String portName = "";
+
+	/**
+     * A WebServiceAnnot cannot be instantiated.
+     */
+	private  WebServiceAnnot(){
+		
+	}
+	
+	private  WebServiceAnnot(
+			String name,
+			String targetNamespace,
+			String serviceName,
+			String wsdlLocation,
+			String endpointInterface,
+			String portName)
+	{
+		this.name = name;
+		this.targetNamespace = targetNamespace;
+		this.serviceName = serviceName;
+		this.wsdlLocation = wsdlLocation;
+		this.endpointInterface = endpointInterface;
+		this.portName = portName;
+	}
+
+    public static WebServiceAnnot createWebServiceAnnotImpl() {
+        return new WebServiceAnnot();
+    }
+
+    public static WebServiceAnnot createWebServiceAnnotImpl( 
+    			String name,
+    			String targetNamespace,
+    			String serviceName,
+    			String wsdlLocation,
+    			String endpointInterface,
+    			String portName
+    		) 
+    {
+        return new WebServiceAnnot( name, 
+        								targetNamespace, 
+        								serviceName, 
+        								wsdlLocation, 
+        								endpointInterface, 
+        								portName);
+    }
+
+	public String name(){
+		return this.name;
+	}
+	
+	public String targetNamespace(){
+		return this.targetNamespace;
+	}
+	
+	public String serviceName(){
+		return this.serviceName;
+	}
+	
+	public String wsdlLocation(){
+		return this.wsdlLocation;
+	}
+	
+	public String endpointInterface(){
+		return this.endpointInterface;
+	}
+	
+	public String portName(){
+		return this.portName;
+	}
+	
+	//hmm, should we really do this
+	public Class<Annotation> annotationType(){
+		return Annotation.class;
+	}
+	
+	//Setters
+	public void setName(String name){
+		this.name = name;
+	}
+	
+	public void setTargetNamespace(String targetNamespace){
+		this.targetNamespace = targetNamespace;
+	}
+	
+	public void setServiceName(String serviceName){
+		this.serviceName = serviceName;
+	}
+	
+	public void setWsdlLocation(String wsdlLocation){
+		this.wsdlLocation = wsdlLocation;
+	}
+	
+	public void setEndpointInterface(String endpointInterface){
+		this.endpointInterface = endpointInterface;
+	}
+	
+	public void setPortName(String portName){
+		this.portName = portName;
+	}
+	
+}
