Index: modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java
===================================================================
--- modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java	(revision 904087)
+++ modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java	(working copy)
@@ -452,7 +452,7 @@
                     epr = new EndpointReference(anonymous);
                 }
             }
-            else if (!isFinalAddressingNamespace && epr.hasNoneAddress()) {
+            else if ((!isFinalAddressingNamespace || messageContext.isServerSide()) && epr.hasNoneAddress()) {
                 return; //Omit the header.
             }
             else if (epr.isWSAddressingAnonymous())
Index: modules/integration/test/org/apache/axis2/addressing/MultiHopRedirectService1.java
===================================================================
--- modules/integration/test/org/apache/axis2/addressing/MultiHopRedirectService1.java	(revision 904087)
+++ modules/integration/test/org/apache/axis2/addressing/MultiHopRedirectService1.java	(working copy)
@@ -32,6 +32,7 @@
         currOutMc.setTo(new EndpointReference("http://127.0.0.1:" + (UtilServer.TESTING_PORT) +
                 "/axis2/services/MultiHopRedirectService2/echoRedirect"));
         currOutMc.setReplyTo(currInMc.getReplyTo());
+        currOutMc.setServerSide(false);
         return ome;
     }
 }
