Index: C:/apache/os/axis2/modules/metadata/test/org/apache/axis2/jaxws/description/impl/DescriptionFactoryImplTests.java
===================================================================
--- C:/apache/os/axis2/modules/metadata/test/org/apache/axis2/jaxws/description/impl/DescriptionFactoryImplTests.java	(revision 588288)
+++ C:/apache/os/axis2/modules/metadata/test/org/apache/axis2/jaxws/description/impl/DescriptionFactoryImplTests.java	(working copy)
@@ -19,10 +19,13 @@
 
 package org.apache.axis2.jaxws.description.impl;
 
+import java.io.File;
+import java.io.InputStream;
 import java.lang.annotation.ElementType;
 import java.lang.reflect.Field;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -178,6 +181,43 @@
         }
     }
     
+    public void testHandlerChainSource() {
+    	JavaClassToDBCConverter converter = new JavaClassToDBCConverter(AnnotatedService.class);
+        HashMap<String, DescriptionBuilderComposite> dbcMap = converter.produceDBC();
+        DescriptionBuilderComposite dbc = dbcMap.get(AnnotatedService.class.getName());
+        assertNotNull(dbc);
+        InputStream is = getXMLFileStream();
+        assertNotNull(is);
+        dbc.setHandlerChainSource(is);
+        List<ServiceDescription> sdList = DescriptionFactoryImpl.createServiceDescriptionFromDBCMap(dbcMap);
+        assertNotNull(sdList);
+        assertTrue(sdList.size() > 0);
+        ServiceDescription sd = sdList.get(0);
+        assertNotNull(sd.getEndpointDescriptions_AsCollection());
+        Collection<EndpointDescription> edColl = sd.getEndpointDescriptions_AsCollection();
+        assertNotNull(edColl);
+        assertTrue(edColl.size() > 0);
+        EndpointDescription ed = edColl.iterator().next();
+        assertNotNull(ed);
+        assertNotNull(ed.getHandlerChain());
+    }
+    
+    private InputStream getXMLFileStream() {
+    	InputStream is = null;
+    	String configLoc = null;
+        try {
+            String sep = "/";
+            configLoc = sep + "test-resources" + sep + "test-handler.xml";
+            String baseDir = new File(System.getProperty("basedir",".")).getCanonicalPath();
+            is = new File(baseDir + configLoc).toURL().openStream();
+        }
+        catch(Exception e) {
+            e.printStackTrace();
+        }
+    	return is;
+    }
+    
+    
     private void resetClientConfigFactory() throws Exception {
         Field field = DescriptionFactoryImpl.class.getDeclaredField("clientConfigFactory");
         field.setAccessible(true);
Index: C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/i18n/resource.properties
===================================================================
--- C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/i18n/resource.properties	(revision 588288)
+++ C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/i18n/resource.properties	(working copy)
@@ -183,3 +183,4 @@
 onewayAsync=The async listener property was set.  A one way invocation cannot be performed using an async response channel.
 soapBindingUseEncoded=The SEI {0} contains method {1} with an unsupported SOAPBinding annotation value. SOAPBinding.Use field cannot be set to ENCODED.
 soapBindingStyle=The SEI {0} contains method {1} with annotation SOAPBinding.Style set to RPC.
+handlerSourceFail = Internal error. Multiple sources were supplied for the handler chain configuration information.
Index: C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java
===================================================================
--- C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java	(revision 588288)
+++ C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/builder/DescriptionBuilderComposite.java	(working copy)
@@ -28,6 +28,8 @@
 
 
 import javax.wsdl.Definition;
+
+import java.io.InputStream;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -98,6 +100,11 @@
 
     private WsdlGenerator wsdlGenerator;
     private ClassLoader classLoader;
+    
+    // represents a stream to the XML handler chain configuration
+    // either this or the HandlerChainAnnot may be present, but 
+    // not both, they may both be null
+    private InputStream handlerChainSource;
 
     // Methods
     public WebServiceAnnot getWebServiceAnnot() {
@@ -455,6 +462,14 @@
 
         this.classLoader = classLoader;
     }
+    
+    public void setHandlerChainSource(InputStream handlerChainSource) {
+    	this.handlerChainSource = handlerChainSource;
+    }
+    
+    public InputStream getHandlerChainSource() {
+    	return handlerChainSource;
+    }
 
     /**
      * Convenience method for unit testing. We will print all of the
Index: C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/impl/EndpointDescriptionImpl.java
===================================================================
--- C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/impl/EndpointDescriptionImpl.java	(revision 588288)
+++ C:/apache/os/axis2/modules/metadata/src/org/apache/axis2/jaxws/description/impl/EndpointDescriptionImpl.java	(working copy)
@@ -145,6 +145,7 @@
     //ANNOTATION: @HandlerChain
     private HandlerChain handlerChainAnnotation;
     private HandlerChainsType handlerChainsType;
+    private InputStream handlerChainSource;
 
     // Information common to both WebService and WebServiceProvider annotations
     private String annotation_WsdlLocation;
@@ -275,12 +276,20 @@
         this.parentServiceDescription = parent;
         this.serviceImplName = serviceImplName;
         this.implOrSEIClass = null;
+        
 
         composite = getServiceDescriptionImpl().getDescriptionBuilderComposite();
         if (composite == null) {
             throw ExceptionFactory.makeWebServiceException(
                     "EndpointDescription.EndpointDescription: parents DBC is null");
         }
+        
+        if(composite.getHandlerChainAnnot() != null && composite.getHandlerChainSource() != null) {
+        	throw ExceptionFactory.makeWebServiceException(
+            Messages.getMessage("handlerSourceFail"));
+        }
+        
+        handlerChainSource = composite.getHandlerChainSource();
 
         //Set the base level of annotation that we are processing...currently
         // a 'WebService' or a 'WebServiceProvider'
@@ -1346,6 +1355,13 @@
                     handlerChainsType = DescriptionUtils.loadHandlerChains(is);
                 }
             }
+            else if(handlerChainSource != null) {
+            	
+            	if(log.isDebugEnabled()) {
+            		log.debug("Loading handlers from provided source");
+            	}
+            	handlerChainsType = DescriptionUtils.loadHandlerChains(handlerChainSource);
+            }
         }
         return handlerChainsType;
     }
Index: C:/apache/os/axis2/modules/metadata/test-resources/test-handler.xml
===================================================================
--- C:/apache/os/axis2/modules/metadata/test-resources/test-handler.xml	(revision 0)
+++ C:/apache/os/axis2/modules/metadata/test-resources/test-handler.xml	(revision 0)
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jws:handler-chains xmlns:jws="http://java.sun.com/xml/ns/javaee">
+	<jws:handler-chain name="demo">
+		<jws:handler>
+			<jws:handler-class>
+				handler.Test
+			</jws:handler-class>
+		</jws:handler>
+	</jws:handler-chain>	
+</jws:handler-chains>
\ No newline at end of file
