Index: C:/WebService/Eclipse/workspace/axis2_012207/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java
===================================================================
--- C:/WebService/Eclipse/workspace/axis2_012207/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java	(revision 498796)
+++ C:/WebService/Eclipse/workspace/axis2_012207/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java	(working copy)
@@ -112,6 +112,7 @@
                 getTransportHeaders(req));
         msgContext.setIncomingTransportName(Constants.TRANSPORT_HTTP);
         msgContext.setProperty(HTTPConstants.MC_HTTP_SERVLETREQUEST, req);
+        msgContext.setProperty(HTTPConstants.MC_HTTP_SERVLETRESPONSE, resp);
         return msgContext;
     }
 
