Index: modules/kernel/src/org/apache/axis2/description/AxisDescription.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/AxisDescription.java	(revision 774182)
+++ modules/kernel/src/org/apache/axis2/description/AxisDescription.java	(working copy)
@@ -26,6 +26,7 @@
 import org.apache.axiom.om.OMText;
 import org.apache.axis2.AxisFault;
 import org.apache.axis2.engine.AxisConfiguration;
+import org.apache.axis2.engine.AxisEvent;
 import org.apache.axis2.i18n.Messages;
 import org.apache.axis2.modules.Module;
 import org.apache.axis2.util.JavaUtils;
@@ -431,6 +432,17 @@
      */
     public void engageModule(AxisModule axisModule) throws AxisFault {
         engageModule(axisModule, this);
+
+        /**
+         * if a module engaged to a servivce or an operation
+         * notify with a serviceUpdate 
+         */
+        AxisConfiguration config = getAxisConfiguration();
+        if (this instanceof AxisService) {
+            config.notifyObservers(AxisEvent.MODULE_ENGAGED , (AxisService)this);
+        } else if (this instanceof AxisOperation) {
+            config.notifyObservers(AxisEvent.MODULE_ENGAGED ,((AxisOperation)this).getAxisService());
+        }
     }
 
     /**
@@ -511,6 +523,19 @@
             onDisengage(module);
             engagedModules.remove(Utils.getModuleName(module.getName(), module
                     .getVersion()));
+            /**
+             * if a Disengaged module belogs to an AxisService or an Operation
+             * notify with a serviceUpdate
+             */
+            if(this instanceof AxisService) {
+                getAxisConfiguration().notifyObservers(AxisEvent.MODULE_DISENGAGED ,
+                        (AxisService)this);
+            }
+
+            if(this instanceof AxisOperation) {
+                getAxisConfiguration().notifyObservers(AxisEvent.MODULE_DISENGAGED ,
+                        ((AxisOperation)this).getAxisService());
+            }
         }
     }
 
Index: modules/kernel/src/org/apache/axis2/engine/AxisEvent.java
===================================================================
--- modules/kernel/src/org/apache/axis2/engine/AxisEvent.java	(revision 774182)
+++ modules/kernel/src/org/apache/axis2/engine/AxisEvent.java	(working copy)
@@ -32,6 +32,8 @@
     public static final int SERVICE_START = 3;
     public static final int MODULE_DEPLOY = 4;
     public static final int MODULE_REMOVE = 5;
+    public static final int MODULE_ENGAGED = 6;
+    public static final int MODULE_DISENGAGED = 7;
     private int EVENT_TYPE;
 
     public AxisEvent(int EVENT_TYPE) {
