Index: HTTPConstants.java
===================================================================
--- HTTPConstants.java	(revision 765513)
+++ HTTPConstants.java	(working copy)
@@ -22,6 +22,8 @@
 
 import java.io.UnsupportedEncodingException;
 
+import org.apache.commons.httpclient.HttpState;
+
 /**
  * HTTP protocol and message context constants.
  */
@@ -438,6 +440,12 @@
     public static final String CACHED_HTTP_CLIENT = "CACHED_HTTP_CLIENT";
 
     /**
+     * The name of the property that sets a HTTP state to be cached. 
+     * The property value should be of type {@link HttpState}.  
+     */
+    public static final String CACHED_HTTP_STATE = "CACHED_HTTP_STATE";    
+
+    /**
      * @deprecated please use MULTITHREAD_HTTP_CONNECTION_MANAGER
      */
     public static final String MUTTITHREAD_HTTP_CONNECTION_MANAGER = "MUTTITHREAD_HTTP_CONNECTION_MANAGER";
