Index: modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 93931)
+++ modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -49,6 +49,7 @@
     public static ConfigurationContext CONFIG_CONTEXT;
     private ConfigurationContext confContext;
     private MessageContext inMessageContext;
+    private TransportInDescription tIn;
     /** Whether the call is blocking or non-blocking */
     private boolean nonBlocking = false;
     /** If the call is non-blocking the in message context will be stored in this property */
@@ -117,7 +118,7 @@
                                inMessageContext.getProperty(MessageContext.REMOTE_ADDR));
         }
 
-        TransportInDescription tIn = confContext.getAxisConfiguration().getTransportIn(
+        tIn = confContext.getAxisConfiguration().getTransportIn(
                 Constants.TRANSPORT_LOCAL);
         TransportOutDescription tOut = confContext.getAxisConfiguration().getTransportOut(
                 Constants.TRANSPORT_LOCAL);
@@ -184,7 +185,13 @@
         }
     }
 
-    public void init(ConfigurationContext axisConf, TransportInDescription transprtIn) throws AxisFault {
+    public void init(ConfigurationContext configurationContext, TransportInDescription transportInDescription) throws AxisFault {
+        this.confContext = configurationContext;
+        this.tIn = transportInDescription;
+        String scheme = transportInDescription.getName();
+        if (!"local".equals(scheme)) {
+            throw new AxisFault(LocalTransportReceiver.class.getName() + " can only be used for local");
+        }
     }
 
     public void start() throws AxisFault {
