Index: modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java
===================================================================
--- modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java	(revision 507867)
+++ modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java	(working copy)
@@ -213,19 +213,18 @@
                     new RequestURIOperationDispatcher();
             requestURIOperationDispatcher.invoke(msgContext);
 
-            if (msgContext.getAxisOperation() == null) {
+            AxisOperation axisOperation = msgContext.getAxisOperation();
+            if (axisOperation == null) {
                 HTTPLocationBasedDispatcher httpLocationBasedDispatcher =
                         new HTTPLocationBasedDispatcher();
                 httpLocationBasedDispatcher.invoke(msgContext);
-            }
 
-            if (msgContext.getAxisOperation() == null) {
                 SOAPActionBasedDispatcher soapActionBasedDispatcher =
                         new SOAPActionBasedDispatcher();
                 soapActionBasedDispatcher.invoke(msgContext);
             }
-            AxisOperation axisOperation;
-            if ((axisOperation = msgContext.getAxisOperation()) != null) {
+
+            if (axisOperation != null) {
                 AxisEndpoint axisEndpoint =
                         (AxisEndpoint) msgContext.getProperty(WSDL2Constants.ENDPOINT_LOCAL_NAME);
                 if (axisEndpoint != null) {
@@ -237,7 +236,7 @@
             }
 
             // check for the dispatching result
-            if (msgContext.getAxisOperation() == null) {
+            if (axisOperation == null) {
                 throw new AxisFault("I can not find a service for this request to be serviced." +
                         " Check the WSDL and the request URI");
             }
