Index: kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java
===================================================================
--- kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java	(revision 480099)
+++ kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java	(working copy)
@@ -43,6 +43,7 @@
         String FLATTEN_FILES_OPTION = "f";
         String UNWRAP_PARAMETERS = "uw";
         String BACKWORD_COMPATIBILITY_OPTION = "b";
+        String SUPPRESS_PREFIXES_OPTION = "sp";
         String SOURCE_FOLDER_NAME_OPTION = "S";
         String RESOURCE_FOLDER_OPTION = "R";
 
@@ -68,9 +69,10 @@
         String SERVER_SIDE_INTERFACE_OPTION_LONG = "serverside-interface";
         String EXTERNAL_MAPPING_OPTION_LONG = "external-mapping";
         String WSDL_VERSION_OPTION_LONG = "wsdl-version";
-        String FLATTEN_FILES_OPTION_LONG = "flatten-filess";
+        String FLATTEN_FILES_OPTION_LONG = "flatten-files";
         String UNWRAP_PARAMETERS_LONG = "unwrap-params";
         String BACKWORD_COMPATIBILITY_OPTION_LONG = "backword-compatible";
+        String SUPPRESS_PREFIXES_OPTION_LONG = "suppress-prefixes";
         String SOURCE_FOLDER_NAME_OPTION_LONG = "source-folder";
         String RESOURCE_FOLDER_OPTION_LONG = "resource-folder";
 
Index: adb-codegen/src/org/apache/axis2/schema/ExtensionUtility.java
===================================================================
--- adb-codegen/src/org/apache/axis2/schema/ExtensionUtility.java	(revision 480099)
+++ adb-codegen/src/org/apache/axis2/schema/ExtensionUtility.java	(working copy)
@@ -426,6 +426,10 @@
         if (configuration.isBackwordCompatibilityMode()) {
             options.setBackwordCompatibilityMode(true);
         }
+
+        if (configuration.isSuppressPrefixesMode()) {
+            options.setSuppressPrefixesMode(true);
+        }
     }
 
 }
Index: adb-codegen/src/org/apache/axis2/schema/CompilerOptions.java
===================================================================
--- adb-codegen/src/org/apache/axis2/schema/CompilerOptions.java	(revision 480099)
+++ adb-codegen/src/org/apache/axis2/schema/CompilerOptions.java	(working copy)
@@ -168,4 +168,17 @@
 	public void setBackwordCompatibilityMode(boolean backwordCompatibilityMode) {
 		this.backwordCompatibilityMode = backwordCompatibilityMode;
 	}
+
+    /**
+     * Should we suppress namespace prefixes
+     */
+    private boolean suppressPrefixesMode = false;
+
+    public boolean isSuppressPrefixesMode() {
+        return suppressPrefixesMode;
+    }
+
+    public void setSuppressPrefixesMode(boolean suppressPrefixesMode) {
+        this.suppressPrefixesMode = suppressPrefixesMode;
+    }
 }
Index: adb-codegen/src/org/apache/axis2/schema/writer/JavaBeanWriter.java
===================================================================
--- adb-codegen/src/org/apache/axis2/schema/writer/JavaBeanWriter.java	(revision 480099)
+++ adb-codegen/src/org/apache/axis2/schema/writer/JavaBeanWriter.java	(working copy)
@@ -79,6 +79,8 @@
 
     private boolean isHelperMode = false;
 
+    private boolean isSuppressPrefixesMode = false;
+
     /**
      * package for the mapping class
      */
@@ -146,6 +148,9 @@
             this.ns2packageNameMap = options.getNs2PackageMap();
             //set helper mode
             this.isHelperMode = options.isHelperMode();
+            // set suppress prefixes mode
+            this.isSuppressPrefixesMode = options.isSuppressPrefixesMode();
+
             //set mapper class package if present
             if (options.isMapperClassPackagePresent()) {
                 this.mappingClassPackage = options.getMapperClassPackage();
@@ -442,7 +447,7 @@
         XSLTUtils.addAttribute(model, "package", packageName, rootElt);
         XSLTUtils
                 .addAttribute(model, "nsuri", qName.getNamespaceURI(), rootElt);
-        XSLTUtils.addAttribute(model, "nsprefix", getPrefixForURI(qName
+        XSLTUtils.addAttribute(model, "nsprefix", isSuppressPrefixesMode ? "" : getPrefixForURI(qName
                 .getNamespaceURI(), qName.getPrefix()), rootElt);
 
         if (!wrapClasses) {
Index: codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java
===================================================================
--- codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java	(revision 480099)
+++ codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java	(working copy)
@@ -205,6 +205,14 @@
 			config.setBackwordCompatibilityMode(true);
 		}
 
+        commandLineOption = loadOption(
+                WSDL2JavaConstants.SUPPRESS_PREFIXES_OPTION,
+                WSDL2JavaConstants.SUPPRESS_PREFIXES_OPTION_LONG,
+                optionMap);
+        if (commandLineOption != null) {
+            config.setSuppressPrefixesMode(true);
+        }
+
         // loop through the map and find parameters having the extra prefix.
         //put them in the property map
         Iterator keyIterator = optionMap.keySet().iterator();
Index: codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java
===================================================================
--- codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java	(revision 480099)
+++ codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java	(working copy)
@@ -469,4 +469,17 @@
 	public void setBackwordCompatibilityMode(boolean backwordCompatibilityMode) {
 		this.backwordCompatibilityMode = backwordCompatibilityMode;
 	}
+
+    /**
+     * Should we suppress namespace prefixes
+     */
+    private boolean suppressPrefixesMode = false;
+
+    public boolean isSuppressPrefixesMode() {
+        return suppressPrefixesMode;
+    }
+
+    public void setSuppressPrefixesMode(boolean suppressPrefixesMode) {
+        this.suppressPrefixesMode = suppressPrefixesMode;
+    }
 }
Index: codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java
===================================================================
--- codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java	(revision 480099)
+++ codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java	(working copy)
@@ -53,6 +53,7 @@
                     (WSDL2JavaConstants.FLATTEN_FILES_OPTION).equalsIgnoreCase(optionType) ||
                     (WSDL2JavaConstants.UNWRAP_PARAMETERS).equalsIgnoreCase(optionType) ||
                     (WSDL2JavaConstants.BACKWORD_COMPATIBILITY_OPTION).equalsIgnoreCase(optionType) ||
+                    (WSDL2JavaConstants.SUPPRESS_PREFIXES_OPTION).equalsIgnoreCase(optionType) ||
                     (WSDL2JavaConstants.SOURCE_FOLDER_NAME_OPTION).equalsIgnoreCase(optionType) ||
                     (WSDL2JavaConstants.RESOURCE_FOLDER_OPTION).equalsIgnoreCase(optionType) ||
 
@@ -78,7 +79,8 @@
                     (WSDL2JavaConstants.REPOSITORY_PATH_OPTION_LONG).equalsIgnoreCase(optionType)||
                     (WSDL2JavaConstants.SOURCE_FOLDER_NAME_OPTION_LONG).equalsIgnoreCase(optionType)||
                     (WSDL2JavaConstants.RESOURCE_FOLDER_OPTION_LONG).equalsIgnoreCase(optionType)||
-                    (WSDL2JavaConstants.BACKWORD_COMPATIBILITY_OPTION_LONG).equalsIgnoreCase(optionType)
+                    (WSDL2JavaConstants.BACKWORD_COMPATIBILITY_OPTION_LONG).equalsIgnoreCase(optionType) ||
+                    (WSDL2JavaConstants.SUPPRESS_PREFIXES_OPTION_LONG).equalsIgnoreCase(optionType)
             );
 
         }
