Index: MTOM10AssertionBuilder.java
===================================================================
--- MTOM10AssertionBuilder.java	(revision 736025)
+++ MTOM10AssertionBuilder.java	(working copy)
@@ -38,6 +38,8 @@
 import javax.xml.namespace.QName;
 
 /**
+ * This builder is responsible for the creation of a MTOM assertion object </br>
+ * (compliant with the WS-MTOMPolicy verion 1.0).
  * The builder will be picked by the
  * "org.apache.neethi.AssertionBuilderFactory".
  */
