Index: modules/jaxws/src/org/apache/axis2/jaxws/server/EndpointController.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/server/EndpointController.java	(revision 543191)
+++ modules/jaxws/src/org/apache/axis2/jaxws/server/EndpointController.java	(working copy)
@@ -213,8 +213,9 @@
       */
     private EndpointDispatcher getEndpointDispatcher(Class serviceImplClass, Object serviceInstance)
             throws Exception {
-        return EndpointDispatcherFactory
-                .createEndpointDispatcher(serviceImplClass, serviceInstance);
+        EndpointDispatcherFactory factory = 
+            (EndpointDispatcherFactory)FactoryRegistry.getFactory(EndpointDispatcherFactory.class);        
+        return factory.createEndpointDispatcher(serviceImplClass, serviceInstance);       
     }
 
     /*
Index: modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/factory/EndpointDispatcherFactory.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/factory/EndpointDispatcherFactory.java	(revision 543191)
+++ modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/factory/EndpointDispatcherFactory.java	(working copy)
@@ -28,11 +28,10 @@
 
     public EndpointDispatcherFactory() {
         super();
-
     }
 
-    public static EndpointDispatcher createEndpointDispatcher(Class serviceImplClass,
-                                                              Object serviceInstance) {
+    public EndpointDispatcher createEndpointDispatcher(Class serviceImplClass,
+                                                       Object serviceInstance) {
         //		 TODO:  This check should be based on the EndpointDescription processing of annotations
         //        It is left this way for now because some tests have an @WebService annotation on
         //        Provider-based endpoints as a pre-existing workaround.
Index: modules/jaxws/src/org/apache/axis2/jaxws/registry/FactoryRegistry.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/registry/FactoryRegistry.java	(revision 543191)
+++ modules/jaxws/src/org/apache/axis2/jaxws/registry/FactoryRegistry.java	(working copy)
@@ -34,6 +34,7 @@
 import org.apache.axis2.jaxws.message.impl.MessageFactoryImpl;
 import org.apache.axis2.jaxws.message.impl.XMLPartFactoryImpl;
 import org.apache.axis2.jaxws.message.util.impl.SAAJConverterFactoryImpl;
+import org.apache.axis2.jaxws.server.dispatcher.factory.EndpointDispatcherFactory;
 import org.apache.axis2.jaxws.server.endpoint.lifecycle.factory.EndpointLifecycleManagerFactory;
 
 import java.util.Hashtable;
@@ -56,6 +57,7 @@
         table.put(SAAJConverterFactory.class, new SAAJConverterFactoryImpl());
         table.put(EndpointLifecycleManagerFactory.class, new EndpointLifecycleManagerFactory());
         table.put(ClassFinderFactory.class, new ClassFinderFactory());
+        table.put(EndpointDispatcherFactory.class, new EndpointDispatcherFactory());
     }
 
     /** FactoryRegistry is currently a static singleton */
