Index: modules/core/src/org/apache/axis2/transport/http/HTTPWorker.java
===================================================================
--- modules/core/src/org/apache/axis2/transport/http/HTTPWorker.java	(revision 414844)
+++ modules/core/src/org/apache/axis2/transport/http/HTTPWorker.java	(working copy)
@@ -26,8 +26,6 @@
 import org.apache.axis2.context.MessageContext;
 import org.apache.axis2.context.OperationContext;
 import org.apache.axis2.description.AxisService;
-import org.apache.axis2.description.Parameter;
-import org.apache.axis2.description.TransportOutDescription;
 import org.apache.axis2.transport.http.server.HttpUtils;
 import org.apache.axis2.transport.http.server.OutputBuffer;
 import org.apache.axis2.transport.http.server.Worker;
@@ -64,26 +62,6 @@
         String method = request.getRequestLine().getMethod();
         String soapAction = HttpUtils.getSoapAction(request);
         
-        // Adjust version and content chunking based on the config
-        boolean chunked = false;
-        TransportOutDescription transportOut = msgContext.getTransportOut();
-        if (transportOut != null) {
-            Parameter p = transportOut.getParameter(HTTPConstants.PROTOCOL_VERSION);
-            if (p != null) {
-                if (HTTPConstants.HEADER_PROTOCOL_10.equals(p.getValue())) {
-                    ver = HttpVersion.HTTP_1_0;
-                }
-            }
-            if (ver.greaterEquals(HttpVersion.HTTP_1_1)) {
-                p = transportOut.getParameter(HTTPConstants.HEADER_TRANSFER_ENCODING);
-                if (p != null) {
-                    if (HTTPConstants.HEADER_TRANSFER_ENCODING_CHUNKED.equals(p.getValue())) {
-                        chunked = true;
-                    }
-                }
-            }
-        }
-        
         if (method.equals(HTTPConstants.HEADER_GET)) {
             if (uri.equals("/favicon.ico")) {
                 response.setStatusLine(new StatusLine(ver, 301, "Redirect"));
@@ -102,7 +80,6 @@
                         String res = HTTPTransportReceiver.printServiceHTML(serviceName, configurationContext);
                         StringEntity entity = new StringEntity(res);
                         entity.setContentType("text/html");
-                        entity.setChunked(chunked);
                         response.setEntity(entity);
                         return;
                     }
@@ -122,7 +99,6 @@
                         
                     });
                     entity.setContentType("text/xml");
-                    entity.setChunked(chunked);
                     response.setEntity(entity);
                     return;
                 }
@@ -141,7 +117,6 @@
                         
                     });
                     entity.setContentType("text/xml");
-                    entity.setChunked(chunked);
                     response.setEntity(entity);
                     return;
                 }
@@ -159,7 +134,6 @@
                         
                     });
                     entity.setContentType("text/xml");
-                    entity.setChunked(chunked);
                     response.setEntity(entity);
                     return;
                 }
@@ -187,7 +161,6 @@
                             
                         });
                         entity.setContentType("text/xml");
-                        entity.setChunked(chunked);
                         response.setEntity(entity);
                         return;
                     } else {
@@ -212,14 +185,12 @@
                     HTTPTransportReceiver.getGetRequestParameters(uri));
             
             if (processed) {
-                outbuffer.setChunked(chunked);
                 response.setEntity(outbuffer);
             } else {
                 response.setStatusLine(new StatusLine(ver, 200, "OK"));
                 String s = HTTPTransportReceiver.getServicesHTML(configurationContext);
                 StringEntity entity = new StringEntity(s);
                 entity.setContentType("text/html");
-                entity.setChunked(chunked);
                 response.setEntity(entity);
             }
             
@@ -242,8 +213,6 @@
                     contenttype, 
                     soapAction, 
                     uri);
-            
-            outbuffer.setChunked(chunked);
             response.setEntity(outbuffer);
             
         } else {
Index: modules/core/src/org/apache/axis2/transport/http/RESTSender.java
===================================================================
--- modules/core/src/org/apache/axis2/transport/http/RESTSender.java	(revision 414844)
+++ modules/core/src/org/apache/axis2/transport/http/RESTSender.java	(working copy)
@@ -25,7 +25,6 @@
 import org.apache.axis2.context.MessageContext;
 import org.apache.axis2.i18n.Messages;
 import org.apache.commons.httpclient.Header;
-import org.apache.commons.httpclient.HostConfiguration;
 import org.apache.commons.httpclient.HttpClient;
 import org.apache.commons.httpclient.HttpStatus;
 import org.apache.commons.httpclient.HttpVersion;
