Index: src/main/webapp/axis2-web/upload.jsp
===================================================================
--- src/main/webapp/axis2-web/upload.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/upload.jsp	(working copy)
@@ -18,7 +18,6 @@
   --%>
 
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-
 <jsp:include page="include/adminheader.jsp"/>
 <script type="text/javascript">
 <!--
@@ -42,23 +41,30 @@
 <p>Simple as that!</p>
 
 <% if ("success".equals(request.getAttribute("status"))) { %>
-<font color="green">File <%= request.getAttribute("filename") %> successfully uploaded </font><br/><br/>
+<p style="color:green">
+File <%= request.getAttribute("filename") %> successfully uploaded 
+</p>
 <%
 } else if ("failure".equals(request.getAttribute("status"))) {
 %>
-<font color="red">The following error occurred <br/> <%= request.getAttribute("cause") %></font><br/>
+<p style="color:red">
+The following error occurred
+</p>
+<p style="color:red">
+<%= request.getAttribute("cause") %>
+</p>
 <% } %>
 
 <form method="post" name="Axis2upload" id="Axis2upload" action="axis2-admin/upload"
       enctype="multipart/form-data">
-  <table>
+  <table summary="main content table">
     <tr>
       <td>Service archive : </td>
-      <td><input id="serviceUpload" type="file" name="filename" size="50"/></td>
+      <td><input id="serviceUpload" type="file" name="filename" size="50"></td>
     </tr>
     <tr>
       <td>&nbsp;</td>
-      <td><input name="upload" type="button" onclick="javascript:checkFileUpload();" value=" Upload "/></td>
+      <td><input name="upload" type="button" onclick="javascript:checkFileUpload();" value=" Upload "></td>
     </tr>
   </table>
 </form>
@@ -66,4 +72,4 @@
 <p>Hot deployment of new service archives is <%=request.getAttribute("hotDeployment") %></p>
 <p>Hot update of existing service archives is <%=request.getAttribute("hotUpdate") %></p>
 </blockquote>
-<jsp:include page="include/adminfooter.inc"/>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/InActivateService.jsp
===================================================================
--- src/main/webapp/axis2-web/InActivateService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/InActivateService.jsp	(working copy)
@@ -23,10 +23,10 @@
                  java.util.HashMap,
                  java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Deactivate Service</h1>
 <form method="get" name="serviceInActivate" action="axis2-admin/deactivateService">
-  <table width="100%"  border="0">
+  <table summary="main content table" style="width: 100%"  border="0">
 <tr>
   <td colspan="2" >
      <p>Only the services that are active are listed below. Note that although you can activate a service from this page, once system is restarted the service will be active again</p>
@@ -51,16 +51,16 @@
 if (count > 0) {
 %>
 
-    <td width="20%"> Select Service : </td>
-    <td width="80%">
+    <td style="width: 20%"> Select Service : </td>
+    <td style="width: 80%">
        <select name="axisService" class="selectBoxes">
 <%=html%>
 		</select>
 	</td>
   </tr>
   <tr>
-    <td width="20%">Deactivate Service </td>
-    <td width="80%"><input type="checkbox" name="turnoff">
+    <td style="width: 20%">Deactivate Service </td>
+    <td style="width: 80%"><input type="checkbox" name="turnoff">
     </td>
   </tr>
   <tr>
@@ -80,4 +80,4 @@
 
 </table>
 </form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/viewServiceGroupContext.jsp
===================================================================
--- src/main/webapp/axis2-web/viewServiceGroupContext.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/viewServiceGroupContext.jsp	(working copy)
@@ -21,8 +21,8 @@
 <%@ page import="org.apache.axis2.context.ServiceGroupContext"%>
 <%@ page import="java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
-<h1>Runing Context hierachy</h1>
+<jsp:include page="include/adminheader.jsp"/>
+<h1>Running Context hierarchy</h1>
 <%
     String type = (String) request.getSession().getAttribute("TYPE");
     ConfigurationContext configCtx = (ConfigurationContext) request.getSession().getAttribute("ConfigurationContext");
@@ -32,7 +32,7 @@
             if ("VIEW".equals(type)) {
               Iterator propertyNames = sgContext.getPropertyNames();
 %>
-             <h4>Persistance properties</h4><ul>
+             <h4>Persistence properties</h4><ul>
              <%
                  while (propertyNames.hasNext()) {
                      String key = (String) propertyNames.next();
@@ -52,4 +52,4 @@
 %> <h4>No service group context found</h4><%
     }
 %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/ViewServiceHandlers.jsp
===================================================================
--- src/main/webapp/axis2-web/ViewServiceHandlers.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ViewServiceHandlers.jsp	(working copy)
@@ -27,7 +27,7 @@
                  java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
 <html>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>View Operation Specific Chains</h1>
    <%
             AxisService axisService = (AxisService)request.getSession().
@@ -144,4 +144,4 @@
                  }
 
               %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/listGroupService.jsp
===================================================================
--- src/main/webapp/axis2-web/listGroupService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listGroupService.jsp	(working copy)
@@ -29,7 +29,7 @@
 <head>
   <jsp:include page="include/httpbase.jsp"/>
   <title>List Services</title>
-  <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css" />
+  <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
 </head>
 
 <body>
@@ -49,17 +49,28 @@
         serviceName = axisService.getName();
 %><hr>
 
-<h2><font color="blue"><a href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a>
-</font></h2>
-<font color="blue">Service EPR :</font><font color="black"><%=prefix + axisService.getName()%></font>
-<h4>Service Description : <font color="black"><%=axisService.getDocumentation()%></font></h4>
-<i><font color="blue">Service Status : <%=axisService.isActive() ? "Active" : "InActive"%></font></i><br/>
+<h2><a style="color:blue" href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></h2>
+<p>
+<span style="color:blue">Service EPR :</span>
+<span style="color:black"><%=prefix + axisService.getName()%></span>
+</p>
+
+<h4>Service Description : <span style="color:black"><%=axisService.getDocumentation()%></span></h4>
+
+<p style="color:blue; font-style:italic">
+Service Status : <%=axisService.isActive() ? "Active" : "InActive"%>
+</p>
+
 <%
   Collection engagedModules = axisService.getEngagedModules();
   String moduleName;
   if (engagedModules.size() > 0) {
 %>
-<i>Engaged Modules for the Axis Service</i><ul>
+<p style="font-style:italic">
+Engaged Modules for the Axis Service
+</p>
+
+<ul>
   <%
     for (Iterator iteratorm = engagedModules.iterator(); iteratorm.hasNext();) {
       AxisModule axisOperation = (AxisModule) iteratorm.next();
@@ -71,9 +82,18 @@
 <%
   }
   if (operations.hasNext()) {
-%><br><i>Available operations</i><%
+%>
+
+<p style="font-style:italic">
+Available operations
+</p>
+<%
 } else {
-%><i> There are no operations specified</i><%
+%>
+<p style="font-style:italic">
+There are no operations specified
+</p>
+<%
   }
 %><ul><%
   operations = axisService.getOperations();
@@ -84,12 +104,17 @@
     engagedModules = axisOperation.getEngagedModules();
     if (engagedModules.size() > 0) {
   %>
-  <br><i>Engaged Modules for the Operation</i><ul>
+  
+<p style="font-style:italic">
+Engaged Modules for the Operation
+</p>
+<ul>
   <%
     for (Iterator iterator2 = engagedModules.iterator(); iterator2.hasNext();) {
       AxisModule moduleDecription = (AxisModule) iterator2.next();
       moduleName = moduleDecription.getName();
-  %><li><%=moduleName%></li><br><%
+  %><li><%=moduleName%></li>
+  <%
   }
 %></ul><%
     }
Index: src/main/webapp/axis2-web/errorModule.jsp
===================================================================
--- src/main/webapp/axis2-web/errorModule.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/errorModule.jsp	(working copy)
@@ -20,11 +20,12 @@
 <%@ page import="org.apache.axis2.Constants,
                  java.util.Hashtable"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
   <%
       Hashtable errornessModules =(Hashtable)request.getSession().getAttribute(Constants.ERROR_MODULE_MAP);
       String moduleName = request.getParameter("moduleName");
   %> <h3>The Module has deployment faults</h3><%
-  %><font color="red" ><%=(String)errornessModules.get(moduleName) %></font><%
+  %><span style="color:red"><%=(String)errornessModules.get(moduleName) %></span><%
+  
   %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/SelectService.jsp
===================================================================
--- src/main/webapp/axis2-web/SelectService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/SelectService.jsp	(working copy)
@@ -23,7 +23,7 @@
                  java.util.HashMap,
                  java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 
   <%
       String action ="";
@@ -51,9 +51,9 @@
 <h1><%=heading%></h1>
 <p><%=disc%></p>
 <form method="get" name="selectServiceForm" action="axis2-admin/<%=action%>">
-<table border="0" width="50%" cellspacing="1" cellpadding="1">
+<table summary="main content table" border="0" style="width: 50%" cellspacing="1" cellpadding="1">
          <tr>
-        <td width="35%">Select a Service :</td><td width="65%">
+        <td style="width: 35%">Select a Service :</td><td style="width:65%">
             <select name="axisService">
             <%
                 HashMap services = (HashMap)request.getSession().getAttribute(Constants.SERVICE_MAP);
@@ -61,11 +61,12 @@
                 for (Iterator iterator = serviceCol.iterator(); iterator.hasNext();) {
                     AxisService axisService = (AxisService)iterator.next();
                     String serviceName = axisService.getName();
-            %> <option  align="left" value="<%=serviceName%>"><%=serviceName%></option>
+            %> <option value="<%=serviceName%>"><%=serviceName%></option>
              <%
                 }
                 request.getSession().setAttribute(Constants.SERVICE_MAP,null);
              %>
+             </select>
            </td>
         </tr>
 	<tr><td colspan="2">&nbsp;</td></tr>
@@ -76,4 +77,4 @@
          </tr>
       </table>
    </form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/engagingtoaservice.jsp
===================================================================
--- src/main/webapp/axis2-web/engagingtoaservice.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/engagingtoaservice.jsp	(working copy)
@@ -25,7 +25,7 @@
                                                                              java.util.Iterator"
          %>
 <%@ page import="org.apache.axis2.util.Utils" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <%
     String status = (String) request.getSession().getAttribute(Constants.ENGAGE_STATUS);
 %>
@@ -40,7 +40,7 @@
     </ol>
 
 <form method="get" name="selectModuleForm" action="axis2-admin/engageToService">
-    <table border="0" width="100%" cellspacing="1" cellpadding="1">
+    <table summary="main content table" border="0" width="100%" cellspacing="1" cellpadding="1">
         <tr>
             <td>
                 &nbsp;
@@ -63,13 +63,14 @@
                             String modulename = axisOperation.getName();
 
                     %>
-                    <option align="left" value="<%=modulename%>"><%=modulename%>
+                    <option value="<%=modulename%>"><%=modulename%>
                     </option>
                     <%
 
                        }
 
                     %>
+                 </select>
             </td>
         </tr>
         <tr>
@@ -93,7 +94,7 @@
                             String serviceName = axisService.getName();
 
                     %>
-                    <option align="left" value="<%=serviceName%>"><%=serviceName%>
+                    <option value="<%=serviceName%>"><%=serviceName%>
                     </option>
                     <%
 
@@ -101,6 +102,7 @@
                        request.getSession().setAttribute(Constants.SERVICE_MAP,null);
 
                     %>
+                 </select>
             </td>
         </tr>
         <tr>
@@ -128,17 +130,21 @@
         </tr>
         <tr>
             <td>
-                <textarea cols="50"  <%
-                    if (status == null) {
+                <textarea cols="50" rows="5" <%
+                    //if (status == null) {
                 %>
                           style="display:none"
                         <%
-                            } %>
-                        ><%=Utils.sanitizeWebOutput(status)%>
+                            //} %>
+                        >
+                       
+                <%--
+                Following call causes exception so I commented out.
+                	<%=Utils.sanitizeWebOutput(status)%>
+                --%>
                 </textarea>
             </td>
         </tr>
     </table>
 </form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
-
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/Error/GenError.html
===================================================================
--- src/main/webapp/axis2-web/Error/GenError.html	(revision 1179398)
+++ src/main/webapp/axis2-web/Error/GenError.html	(working copy)
@@ -20,13 +20,13 @@
 <html>
  <head><title>Error In Admin Application</title></head>
  <body>
- <table width="100%">
+ <table summary="header table" width="100%">
   <tr>
-    <td align="left"><img src="../images/asf-logo.gif"/></td>
-    <td align="right"><img src="../images/axis_l.jpg"/></td>
+    <td align="left"><img src="../images/asf-logo.gif" alt="Apache Software Foundation Logo"/></td>
+    <td align="right"><img src="../images/axis_l.jpg" alt="AXIS Logo"/></td>
     </tr>
 </table>
-<table width="100%">
+<table summary="main content table" width="100%">
   <tr>
     <td align="center">Server Encountered an Error!</td>
   
@@ -35,7 +35,7 @@
 </table>
 
 <hr/>
-<table width="100%">
+<table summary="footer table" width="100%">
  <tr>
    <td align="center">All rights reserved by Apache Software Foundation</td>
    </tr>
Index: src/main/webapp/axis2-web/Error/error500.jsp
===================================================================
--- src/main/webapp/axis2-web/Error/error500.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/Error/error500.jsp	(working copy)
@@ -18,41 +18,41 @@
   --%>
 
 <%@ page contentType="text/html;charset=iso-8859-1" language="java" %>
-<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
   <head>
     <jsp:include page="../include/httpbase.jsp"/>
     <title>Axis2 :: Internal server error</title>
-    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css"/>
-    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
+    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
+    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   </head>
 
   <body>
-    <table width="100%">
+    <table summary="header table" style="width: 100%">
       <tr>
-        <td align="left"><img src="axis2-web/images/asf-logo.gif" alt=""/></td>
-        <td align="right"><img src="axis2-web/images/axis_l.jpg" alt=""/></td>
+        <td align="left"><img src="axis2-web/images/asf-logo.gif" alt="Apache Software Foundation Logo"></td>
+        <td align="right"><img src="axis2-web/images/axis_l.jpg" alt="AXIS Logo"></td>
       </tr>
     </table>
-    <table width="100%">
+    <table summary="main content table" style="width: 100%">
       <tr>
         <td>
           <h1>Internal server error</h1>
-          <br/>
-          <br/>
-          <br/>
-          <br/>
+          <br>
+          <br>
+          <br>
+          <br>
         </td>
       </tr>
-      <tr><td align="center"><a href="axis2-web/index.jsp">home</a></td></tr>
+      <tr><td style="align: center"><a href="axis2-web/index.jsp">home</a></td></tr>
     </table>
-    <hr/>
-    <table width="100%">
+    <hr>
+    <table summary="footer table" style="width: 100%">
       <tr><td>
-        <table width="950px">
-          <tr><td><hr size="1" noshade=""/></td></tr>
+        <table summary="embedded footer table">
+          <tr><td><hr></td></tr>
           <tr>
-            <td align="center">Copyright &#169; 1999-2006, The Apache Software Foundation<br/>Licensed under the <a
+            <td style="align: center">Copyright &#169; 1999-2006, The Apache Software Foundation<br>Licensed under the <a
               href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a>.</td>
           </tr>
         </table>
Index: src/main/webapp/axis2-web/Error/error404.jsp
===================================================================
--- src/main/webapp/axis2-web/Error/error404.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/Error/error404.jsp	(working copy)
@@ -18,40 +18,40 @@
   --%>
 
 <%@ page contentType="text/html;charset=iso-8859-1" language="java" %>
-<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
   <head>
     <jsp:include page="../include/httpbase.jsp"/>
     <title>Axis2 :: Resource not found!</title>
-    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css"/>
-    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
+    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
+    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   </head>
 
   <body>
-    <table width="100%">
+    <table summary="header table" style="width: 100%">
       <tr>
-        <td align="left"><img src="axis2-web/images/asf-logo.gif" alt=""/></td>
-        <td align="right"><img src="axis2-web/images/axis_l.jpg" alt=""/></td>
+        <td align="left"><img src="axis2-web/images/asf-logo.gif" alt="Apache Software Foundation Logo"></td>
+        <td align="right"><img src="axis2-web/images/axis_l.jpg" alt="AXIS Logo"></td>
       </tr>
     </table>
-    <table width="100%">
+    <table summary="main content table" style="width: 100%">
       <tr>
         <td>
           <h1>Requested resource not found!</h1>
-          <br/>
-          <br/>
-          <br/>
-          <br/>
+          <br>
+          <br>
+          <br>
+          <br>
         </td>
       </tr>
       <tr><td align="center"><a href="axis2-web/index.jsp">home</a></td></tr>
     </table>
-    <table width="100%">
+    <table summary="footer table" style="width: 100%">
       <tr><td>
-        <table width="950px">
-          <tr><td><hr size="1" noshade=""/></td></tr>
+        <table summary="embedded footer table">
+          <tr><td><hr size="1" noshade="noshade"></td></tr>
           <tr>
-            <td align="center">Copyright &#169; 1999-2006, The Apache Software Foundation<br/>Licensed under the <a
+            <td align="center">Copyright &#169; 1999-2006, The Apache Software Foundation<br>Licensed under the <a
               href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a>.</td>
           </tr>
         </table>
Index: src/main/webapp/axis2-web/Error/AuthError.html
===================================================================
--- src/main/webapp/axis2-web/Error/AuthError.html	(revision 1179398)
+++ src/main/webapp/axis2-web/Error/AuthError.html	(working copy)
@@ -20,13 +20,13 @@
 <html>
  <head><title>Error In Admin Application</title></head>
  <body>
- <table width="100%">
+ <table summary="header table" width="100%">
   <tr>
-    <td align="left"><img src="../images/asf-logo.gif"/></td>
-    <td align="right"><img src="../images/axis_l.jpg"/></td>
+    <td align="left"><img src="../images/asf-logo.gif" alt="Apache Software Founation Logo"/></td>
+    <td align="right"><img src="../images/axis_l.jpg" alt="AXIS Logo"/></td>
     </tr>
 </table>
-<table width="100%">
+<table summary="main content table" width="100%">
   <tr>
     <td align="center">Failure in Authentication! Please check your username and password</td>
   
@@ -35,7 +35,7 @@
 </table>
 
 <hr/>
-<table width="100%">
+<table summary="footer table" width="100%">
  <tr>
    <td align="center">All rights reserved by Apache Software Foundation</td>
    </tr>
Index: src/main/webapp/axis2-web/ActivateService.jsp
===================================================================
--- src/main/webapp/axis2-web/ActivateService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ActivateService.jsp	(working copy)
@@ -23,10 +23,10 @@
                  java.util.HashMap,
                  java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Turn On Service</h1>
 <form method="get" name="serviceActivate" action="axis2-admin/activateService">
-  <table width="100%"  border="0">
+  <table summary="main content table" width="100%"  border="0">
 <tr>
   <td colspan="2" >
      <p>The services that are inactive are listed below. Although you can activate the services from this page, once system is restarted the services will be inactive again</p>
@@ -79,4 +79,4 @@
 
 </table>
 </form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/viewServiceContext.jsp
===================================================================
--- src/main/webapp/axis2-web/viewServiceContext.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/viewServiceContext.jsp	(working copy)
@@ -47,4 +47,4 @@
         }
     }
 %>
-<jsp:include page="include/adminfooter.inc"/>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/EngageToServiceGroup.jsp
===================================================================
--- src/main/webapp/axis2-web/EngageToServiceGroup.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/EngageToServiceGroup.jsp	(working copy)
@@ -24,7 +24,7 @@
 <%@ page import="java.util.HashMap"%>
 <%@ page import="java.util.Iterator"%>
 <%@ page import="org.apache.axis2.util.Utils" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <%
     String status = (String)request.getSession().getAttribute(Constants.ENGAGE_STATUS);
     HashMap moduels = (HashMap)request.getSession().getAttribute(Constants.MODULE_MAP);
@@ -52,7 +52,7 @@
 		<%} else {
 %>
 <form method="get" name="selectModuleForm" action="axis2-admin/engageToServiceGroup">
-    <table border="0" width="100%" cellspacing="1" cellpadding="1">
+    <table summary="main content table" border="0" width="100%" cellspacing="1" cellpadding="1">
         <tr>
             <td>Select a Module :</td>
         </tr>
@@ -63,10 +63,11 @@
                 for (Iterator iterator = moduleCol.iterator(); iterator.hasNext();) {
                     AxisModule axisOperation = (AxisModule) iterator.next();
                     String modulename = axisOperation.getName();
-            %> <option  align="left" value="<%=modulename%>"><%=modulename%></option>
+            %> <option  value="<%=modulename%>"><%=modulename%></option>
                     <%
                 }
              %>
+                </select>
             </td>
         </tr>
         <tr>
@@ -86,7 +87,7 @@
                         while(servicesGroups.hasNext()){
                             AxisServiceGroup axisServiceGroup = (AxisServiceGroup) servicesGroups.next();
                             String serviceName = axisServiceGroup.getServiceGroupName();
-                    %> <option  align="left" value="<%=serviceName%>"><%=serviceName%></option>
+                    %> <option value="<%=serviceName%>"><%=serviceName%></option>
                     <%
                         }
 
@@ -134,5 +135,4 @@
 		}
 	}
 %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
-
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/engagingtoanoperation.jsp
===================================================================
--- src/main/webapp/axis2-web/engagingtoanoperation.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/engagingtoanoperation.jsp	(working copy)
@@ -24,7 +24,7 @@
                                                                              java.util.Collection,
                                                                              java.util.HashMap,
                                                                              java.util.Iterator"%>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <%
     String status = (String)request.getSession().getAttribute(Constants.ENGAGE_STATUS);
 %>
@@ -37,7 +37,7 @@
         </ol>
 
 <form method="get" name="selectModuleForm" action="axis2-admin/listOperations">
-<table border="0" width="100%" cellspacing="1" cellpadding="1">
+<table summary="main content table" border="0" width="100%" cellspacing="1" cellpadding="1">
     <tr>
         <td>
             &nbsp;
@@ -57,10 +57,11 @@
                 for (Iterator iterator = moduleCol.iterator(); iterator.hasNext();) {
                     AxisModule axisOperation = (AxisModule) iterator.next();
                     String modulename = axisOperation.getName();
-            %> <option  align="left" value="<%=modulename%>"><%=modulename%></option>
+            %> <option  value="<%=modulename%>"><%=modulename%></option>
              <%
                 }
              %>
+             </select>
            </td>
         </tr>
         <tr>
@@ -80,11 +81,12 @@
                 while (operations.hasNext()) {
                     AxisOperation axisOperationtion = (AxisOperation)operations.next();
                     String opname = axisOperationtion.getName().getLocalPart();
-            %> <option  align="left" value="<%=opname%>"><%=opname%></option>
+            %> <option value="<%=opname%>"><%=opname%></option>
              <%
                 }
                 request.getSession().setAttribute(Constants.OPERATION_MAP,null);
              %>
+             </select>
            </td>
         </tr>
         <tr>
@@ -123,4 +125,4 @@
            </tr>
       </table>
    </form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/viewphases.jsp
===================================================================
--- src/main/webapp/axis2-web/viewphases.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/viewphases.jsp	(working copy)
@@ -20,13 +20,13 @@
                  org.apache.axis2.engine.Phase"%>
  <%@ page import="java.util.List"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Available Phases</h1>
      <%
          PhasesInfo phases = (PhasesInfo)request.getSession().getAttribute(Constants.PHASE_LIST);
          request.getSession().setAttribute(Constants.PHASE_LIST,null);
          List tempList = phases.getGlobalInflow();
-     %><h2><font color="blue">System Pre-defined Phases</font></h2>
+     %><h2 style="color:blue">System Pre-defined Phases</h2>
      <b>InFlow Up to Dispatcher</b>
      <blockquote>
          <%
@@ -67,7 +67,7 @@
          %>
          </blockquote>
          <br>
-         <h2><font color="blue">User Defined Phases</font></h2>
+         <h2 style="color:blue">User Defined Phases</h2>
          <b>Inflow after Dispatcher</b>
          <blockquote>
          <%
@@ -108,4 +108,4 @@
              }
          %>
          </blockquote>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/ServiceParaEdit.jsp
===================================================================
--- src/main/webapp/axis2-web/ServiceParaEdit.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ServiceParaEdit.jsp	(working copy)
@@ -24,14 +24,14 @@
                  java.util.ArrayList,
                  java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Edit Service Parameters</h1>
   <form method="get" name="editServicepara" action="axis2-admin/editServicepara">
    <%
             AxisService axisService = (AxisService)request.getSession().
                     getAttribute(Constants.SERVICE);
              if(axisService != null ){
-   %>     <table width="100%">
+   %>     <table summary="main content table" width="100%">
 
             <tr>
                  <td colspan="2" ><b>
@@ -118,5 +118,5 @@
              }
 
        %>
-       <form>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+       </form>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/disengage.jsp
===================================================================
--- src/main/webapp/axis2-web/disengage.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/disengage.jsp	(working copy)
@@ -19,4 +19,4 @@
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
 <jsp:include page="include/adminheader.jsp"/>
 <%=request.getSession().getAttribute("status")%>
-<jsp:include page="include/adminfooter.inc" />
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/admin.jsp
===================================================================
--- src/main/webapp/axis2-web/admin.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/admin.jsp	(working copy)
@@ -18,7 +18,7 @@
   --%>
 
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Welcome to Axis2 Web Admin Module !!</h1>
 <p>You are now logged into the Axis2 administration console from inside
 the console you will be able to</p>
@@ -28,4 +28,4 @@
         <li>to upload new services into Axis2 [Service hot-deployment].</li>
     </ul>
 
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/Login.jsp
===================================================================
--- src/main/webapp/axis2-web/Login.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/Login.jsp	(working copy)
@@ -18,6 +18,7 @@
   --%>
 
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
   <head>
     <jsp:include page="include/httpbase.jsp"/>
@@ -28,7 +29,7 @@
   <body onload="javascript:document.LoginForm.userName.focus();">
     <jsp:include page="include/header.inc"/>
     <jsp:include page="include/link-footer.jsp"/>
-    <table class="FULL_BLANK">
+    <table summary="main content table" class="FULL_BLANK">
       <tr>
         <td valign="top">
         	<h2>Welcome :</h2>
@@ -38,26 +39,26 @@
  	       		<li>Change any parameters at run time.</li>
  	       		<li>Upload new services into Axis2 [Service hot-deployment].</li>
  	   		</ul>
-	        <font color="orange">Warning: Please note that configuration changes done through the administration console
-    	    will be lost when the server is restarted.</font>
+	        <span style="color:orange">Warning: Please note that configuration changes done through the administration console
+    	    will be lost when the server is restarted.</span>
     	</td>
         <td valign="top" align="left">
           <form method="post" name="LoginForm" action="axis2-admin/login">
-            <table class="LOG_IN_FORM">
+            <table summary="login table" class="LOG_IN_FORM">
               <tr>
-                <td align="center" colspan="2" bgcolor="#b0b0b0" color="#FFFFFF"><font color="#FFFFFF">Login</font></td>
+                <td align="center" colspan="2" style="bgcolor:#b0b0b0"><span style="color:#FFFFFF">Login</span></td>
               </tr>
               <tr>
                  <td align="center" colspan="2">&nbsp;</td>
                </tr>
               <tr>
                 <td align="right">Username:</td>
-                <td><input align="left" type="text" name="userName" tabindex="1">
+                <td><input style="align:left" type="text" name="userName" tabindex="1">
                 </td>
               </tr>
               <tr>
                 <td align="right">Password : </td>
-                <td><input align="left" type="password" name="password" tabindex="2">
+                <td><input style="align:left" type="password" name="password" tabindex="2">
                 </td>
               </tr>
               <tr>
@@ -73,12 +74,12 @@
               </tr>
               <tr>
                 <td align="center" colspan="2">
-                  <font color="red">&nbsp;<% if (request.getAttribute("errorMessage") != null) {%><%= request.getAttribute("errorMessage")%><% } %>&nbsp;</font>
+                  <span style="color:red">&nbsp;<% if (request.getAttribute("errorMessage") != null) {%><%= request.getAttribute("errorMessage")%><% } %>&nbsp;</span>
                 </td>
               </tr>
             </table>
           </form>
-          <br/><br/><br/><br/><br/><br/>
+          <br><br><br><br><br><br>
         </td>
       </tr>
     </table>
Index: src/main/webapp/axis2-web/listSingleService.jsp
===================================================================
--- src/main/webapp/axis2-web/listSingleService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listSingleService.jsp	(working copy)
@@ -37,8 +37,8 @@
     if (Constants.IS_FAULTY.equals(isFault)) {
         Hashtable errornessservices = (Hashtable) request.getSession().getAttribute(Constants.ERROR_SERVICE_MAP);
 %>
-<h3>This Axis service has deployment faults.</h3><%
-%><font color="red"><%=(String) errornessservices.get(servicName) %></font>
+<h3>This Axis service has deployment faults.</h3>
+<p style="color:red"><%=(String) errornessservices.get(servicName) %></p>
 <%
 
 } else {
@@ -49,8 +49,12 @@
         Iterator opItr = axisService.getOperations();
         //operationsList = operations.values();
         String serviceName = axisService.getName();
-%><h2><font color="blue"><a href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></font></h2>
-<font color="blue">Service EPR : </font><font color="black"><%=prefix + axisService.getName()%></font><br>
+%>
+<h2><a style="color:blue" href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></h2>
+<p>
+<span style="color:blue">Service EPR : </span>
+<span style="color:black"><%=prefix + axisService.getName()%></span>
+</p>
 <%
     boolean disableREST = false;
     AxisConfiguration axisConfiguration = axisService.getAxisConfiguration();
@@ -73,9 +77,10 @@
         serviceDescription = "No description available for this service";
     }
 %>
-<h4>Service Description : <font color="black"><%=serviceDescription%></font></h4>
-
-<i><font color="blue">Service Status : <%=axisService.isActive() ? "Active" : "InActive"%></font></i><br>
+<h4>Service Description : <span style="color:black"><%=serviceDescription%></span></h4>
+<p style="color:blue; font-style:italic">
+Service Status : <%=axisService.isActive() ? "Active" : "InActive"%>
+</p>
 <%
     if (opItr.hasNext()) {
 %><i>Available operations</i><%
@@ -94,10 +99,10 @@
 <%
 } else {
 %>
-<h3><font color="red">No services found in this location</font></h3>
+<h3 style="color:red">No services found in this location</h3>
 <%
         }
 
     }
 %>
-<jsp:include page="include/adminfooter.inc"/>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/HappyAxis.jsp
===================================================================
--- src/main/webapp/axis2-web/HappyAxis.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/HappyAxis.jsp	(working copy)
@@ -16,7 +16,7 @@
   ~ specific language governing permissions and limitations
   ~ under the License.
   --%>
-
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
 <%@ page import="org.apache.axis2.AxisFault,
                  org.apache.axis2.Constants,
@@ -132,9 +132,9 @@
             } else {
                 String location = getLocation(out, clazz);
                 if (location == null) {
-                    out.write("Found " + axisOperation + " (" + classname + ")<br/>");
+                    out.write("Found " + axisOperation + " (" + classname + ")<br>");
                 } else {
-                    out.write("Found " + axisOperation + " (" + classname + ") <br/> &nbsp;&nbsp;at " + location + "<br/>");
+                    out.write("Found " + axisOperation + " (" + classname + ") <br> &nbsp;&nbsp;at " + location + "<br/>");
                 }
                 return 0;
             }
@@ -349,6 +349,8 @@
             result.serialize(XMLOutputFactory.newInstance().createXMLStreamWriter(writer));
             writer.flush();
             value = writer.toString();
+            value = value.replaceAll("<", "&lt;");
+            value = value.replaceAll(">", "&gt;");
             return true;
         } catch (AxisFault axisFault) {
             System.out.println(value);
@@ -370,7 +372,7 @@
 
 <h2>Examining webapp configuration</h2>
 
-<blockquote>
+
 
 <h4>Essential Components</h4>
 
@@ -416,50 +418,50 @@
     //is everything we need here
     if (needed == 0) {
         //yes, be happy
-        out.write("<p><font color='green'><strong>The core axis2 libraries are present.</strong></font></p>");
+        out.write("<p style=\"color:green; font-style:bold\">The core axis2 libraries are present.</p>");
     } else {
         //no, be very unhappy
         response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
-        out.write("<font color='red'><i>"
+        out.write("<p style=\"color:red; font-style:italic\">"
                 + needed
                 + " core axis2 librar"
                 + (needed == 1 ? "y is" : "ies are")
-                + " missing</i></font>");
+                + " missing</p>");
     }
     //now look at wanted stuff
 %>
 <p>
-    <B><I>Note:</I></B> Even if everything this page probes for is present,
+    <span style="font-style:bold italic">Note:</span> Even if everything this page probes for is present,
     there is no guarantee your Axis Service will work, because there are many configuration options
-    that we do not check for. These tests are <i>necessary</i> but not <i>sufficient</i>
+    that we do not check for. These tests are <span style="font-style:italic">necessary</span> but not <span style="font-style:italic">sufficient</span>
 </p>
-</blockquote>
+
 <h2>Examining Version Service</h2>
 <%
     boolean serviceStatus = invokeTheService();
     if (serviceStatus) {
 %>
-<blockquote>
-    <font color="green"><strong>
+<div>
+    <p style="color:green; font-style:bold">
         Found Axis2 default Version service and Axis2 is working
-        properly.</strong></font>
+        properly.
+    </p>
+        
     <p>Now you can drop a service archive in axis2/WEB-INF/services.
         Following output was produced while invoking Axis2 version service
-        </p>
-        <p><%= value%></p>
-</blockquote>
+    </p>
+    <p><%= value%></p>
+</div>
 
 <%
 } else {
 %>
-<p>
-    <font color="brown"> There was a problem in Axis2 version service , may be
+<p style="color:brown">
+    There was a problem in Axis2 version service , may be
         the service not available or some thing has gone wrong. But this does
         not mean system is not working !
         Try to upload some other service and check to see whether it is
         working.
-        <br>
-    </font>
 </p>
 
 <%
@@ -467,7 +469,7 @@
 %>
 <h2>Examining Application Server</h2>
 <blockquote>
-<table>
+<table summary="main content table">
     <tr><td>Servlet version</td><td><%=getServletVersion()%></td></tr>
     <tr><td>Platform</td>
         <td><%=getServletConfig().getServletContext().getServerInfo()%></td>
@@ -485,17 +487,16 @@
     } catch (SecurityException se) {
     }
     if (e != null) {
-        out.write("<pre>");
-        out.write("<table cellpadding='5px' cellspacing='0px' style='border: .5px blue solid;'>");
+        out.write("<table summary=\"main content table\" cellpadding=\"5\" cellspacing=\"0\" style=\"border: .5 blue solid;\">");
         for (; e.hasMoreElements();) {
             out.write("<tr>");
             String key = (String) e.nextElement();
             out.write("<th style='border: .5px #A3BBFF solid;'>" + key + "</th>");
             out.write("<td style='border: .5px #A3BBFF solid;'>" + getFormatedSystemProperty(System.getProperty(key)) + "&nbsp;</td>");
-            out.write("<tr>");
+            out.write("</tr>");
         }
         out.write("</table>");
-        out.write("</pre><p>");
+        out.write("<p>");
     } else {
         out.write("System properties are not accessible<p>");
     }
Index: src/main/webapp/axis2-web/listServices.jsp
===================================================================
--- src/main/webapp/axis2-web/listServices.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listServices.jsp	(working copy)
@@ -29,15 +29,17 @@
 <%@ page import="java.util.Hashtable" %>
 <%@ page import="java.util.Iterator" %>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
+<head>
 <jsp:include page="include/httpbase.jsp"/>
-<head><title>List Services</title>
-    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css"/>
+<title>List Services</title>
+    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
 </head>
 
 <body>
-<jsp:include page="include/header.inc" />
-<jsp:include page="include/link-footer.jsp" />
+<jsp:include page="include/header.inc"/>
+<jsp:include page="include/link-footer.jsp"/>
 <h1>Available services</h1>
 <% String prefix = request.getAttribute("frontendHostUrl") + (String)request.getSession().getAttribute(Constants.SERVICE_PATH) + "/";
 %>
@@ -57,7 +59,7 @@
             opItr = axisService.getOperations();
             //operationsList = operations.values();
             serviceName = axisService.getName();
-%><h2><font color="blue"><a href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></font></h2>
+%><h2><a style="color:blue" href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></h2>
 <%
     boolean disableREST = false;
     AxisConfiguration axisConfiguration = axisService.getAxisConfiguration();
@@ -82,7 +84,7 @@
         serviceDescription = "No description available for this service";
     }
 %>
-<h5>Service Description : <font color="black"><%=serviceDescription%></font></h5>
+<h5>Service Description : <span style="color:black;"><%=serviceDescription%></span></h5>
 <h5>Service EPR : <%=prefix + axisService.getName()%></h5>
 <h5>Service Status : <%=axisService.isActive() ? "Active" : "InActive"%></h5><br>
 <%
@@ -110,13 +112,14 @@
 %>
 <hr>
 
-<h3><font color="blue">Faulty Services</font></h3>
+<h3 style="color:blue">Faulty Services</h3>
 <%
     Enumeration faultyservices = errornessservice.keys();
     while (faultyservices.hasMoreElements()) {
         String faultyserviceName = (String) faultyservices.nextElement();
-%><h3><font color="blue"><a href="services/ListFaultyServices?serviceName=<%=faultyserviceName%>">
-    <%=faultyserviceName%></a></font></h3>
+%>
+<h3><a style="color:blue" href="services/ListFaultyServices?serviceName=<%=faultyserviceName%>">
+    <%=faultyserviceName%></a></h3>
 <%
             }
         }
@@ -126,6 +129,6 @@
 %> No services listed! Try hitting refresh. <%
     }
 %>
-<jsp:include page="include/footer.inc" />
+<jsp:include page="include/footer.inc"/>
 </body>
 </html>
Index: src/main/webapp/axis2-web/ViewGlobalHandlers.jsp
===================================================================
--- src/main/webapp/axis2-web/ViewGlobalHandlers.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ViewGlobalHandlers.jsp	(working copy)
@@ -23,7 +23,7 @@
                  org.apache.axis2.engine.Phase,
                  java.util.List"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>View Global Execution Chains</h1>
    <%
             AxisConfiguration axisConfig = (AxisConfiguration)request.getSession().
@@ -134,4 +134,4 @@
 
              }
               %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/include/link-footer.inc
===================================================================
--- src/main/webapp/axis2-web/include/link-footer.inc	(revision 1179398)
+++ src/main/webapp/axis2-web/include/link-footer.inc	(working copy)
@@ -16,7 +16,7 @@
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
-<table width="100%">
+<table summary="back home table" width="100%">
  <tr>
    <td align="center"><a href="index.jsp">Back Home</a></td>
    </tr>
Index: src/main/webapp/axis2-web/include/adminheader.inc
===================================================================
--- src/main/webapp/axis2-web/include/adminheader.inc	(revision 1179398)
+++ src/main/webapp/axis2-web/include/adminheader.inc	(working copy)
@@ -16,13 +16,13 @@
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
-<table width="100%">
+<table summary="header table"width="100%">
   <tr>
-    <td align="left"><img src="axis2-web/images/asf-logo.gif"/><h1>Axis 2</h1></td>
-    <td align="right"><img src="axis2-web/images/axis.jpg"/></td>
-    </tr>
-    <tr>
+    <td align="left"><img src="axis2-web/images/asf-logo.gif" alt="Apache Software Foundation Logo"><h1>Axis 2</h1></td>
+    <td align="right"><img src="axis2-web/images/axis.jpg" alt="AXIS Logo"></td>
+  </tr>
+  <tr>
     <td>.</td>
-   <td align="right"><a href="logout" target="_parent" >Log out</a></td>
-   </tr>
+    <td align="right"><a href="logout" target="_parent" >Log out</a></td>
+  </tr>
 </table>
Index: src/main/webapp/axis2-web/include/adminheader.jsp
===================================================================
--- src/main/webapp/axis2-web/include/adminheader.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/include/adminheader.jsp	(working copy)
@@ -56,24 +56,24 @@
 ~ under the License.
 --%>
 <%@ page contentType="text/html;charset=iso-8859-1" language="java" %>
-<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
   <head>
     <jsp:include page="../include/httpbase.jsp"/>
     <title>Axis2 :: Administration Page</title>
-    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css"/>
-    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
+    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
+    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   </head>
 
   <body>
     <jsp:include page="header.inc"/>
-    <table class="FULL_BLANK">
+    <table summary="left frame table" class="FULL_BLANK">
       <tr>
-        <td valign="top" width="20%">
+        <td valign="top" style="width:20%">
           <jsp:include page="../LeftFrame.jsp"/>
         </td>
-        <td valign="top" align="left" width="80%">
-        <table width="100%">
+        <td valign="top" align="left" style="width:80%">
+        <table summary="logout table" width="100%">
           <tr>
             <td align="right" colspan="2"><a href="#" onclick="javascript:history.back();">Back</a>&nbsp;&nbsp;|&nbsp;&nbsp;<a
               href="axis2-admin/logout">Log out</a></td>
Index: src/main/webapp/axis2-web/include/link-footer.jsp
===================================================================
--- src/main/webapp/axis2-web/include/link-footer.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/include/link-footer.jsp	(working copy)
@@ -57,9 +57,9 @@
 ~ specific language governing permissions and limitations
 ~ under the License.
 --%>
-<table width="100%">
+<table summary="back home table"width="100%">
 	<tr><td>
-		<table width="950px">
+		<table summary="embedded back home table">
 			<tr>
 				<td align="right"><a href="axis2-web/index.jsp">Back Home</a>&nbsp;&nbsp;|&nbsp;&nbsp; <a href="#" onclick="javascript:window.location.reload(true);return false;">Refresh</a></td>
 			</tr>
Index: src/main/webapp/axis2-web/include/adminfooter.jsp
===================================================================
--- src/main/webapp/axis2-web/include/adminfooter.jsp	(revision 0)
+++ src/main/webapp/axis2-web/include/adminfooter.jsp	(revision 0)
@@ -0,0 +1,25 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<%@ page contentType="text/html;charset=iso-8859-1" language="java" %>
+</td>
+</tr>
+</table>
+<jsp:include page="footer.inc"/>
+</body>
+</html>
\ No newline at end of file
Index: src/main/webapp/axis2-web/include/adminfooter.inc
===================================================================
--- src/main/webapp/axis2-web/include/adminfooter.inc	(revision 1179398)
+++ src/main/webapp/axis2-web/include/adminfooter.inc	(working copy)
@@ -16,8 +16,13 @@
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
-</td></tr>
+<table summary="header table" width="100%">
+  <tr>
+    <td align="left"><img src="axis2-web/images/asf-logo.gif" alt="Apache Software Foundation Logo"><h1>Axis 2</h1></td>
+    <td align="right"><img src="axis2-web/images/axis.jpg" alt="AXIS Logo"></td>
+  </tr>
+  <tr>
+    <td>.</td>
+    <td align="right"><a href="logout" target="_parent" >Log out</a></td>
+  </tr>
 </table>
-<jsp:include page="footer.inc"></jsp:include>
-</body>
-</html>
\ No newline at end of file
Index: src/main/webapp/axis2-web/include/httpbase.jsp
===================================================================
--- src/main/webapp/axis2-web/include/httpbase.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/include/httpbase.jsp	(working copy)
@@ -68,4 +68,4 @@
         return curentUrl;
     }
 %>
-<base href="<%= calculateHttpBase(request)%>"/>
\ No newline at end of file
+<base href="<%= calculateHttpBase(request)%>">
\ No newline at end of file
Index: src/main/webapp/axis2-web/include/footer.inc
===================================================================
--- src/main/webapp/axis2-web/include/footer.inc	(revision 1179398)
+++ src/main/webapp/axis2-web/include/footer.inc	(working copy)
@@ -18,18 +18,24 @@
   -->
 </td>
 </tr>
-<tr height="20px">
+<tr>
 	<td>
-	<table width="100%">
-		<tr><td>
-			<table width="950px">
-				<tr><td><hr size="1" noshade="" /></td></tr>
+	<table summary="footer table"width="100%">
+		<tr>
+			<td>
+			<table summary="footer table" width="100%">
+				<tr>
+					<td>
+					<hr>
+					</td>
+				</tr>
 				<tr>
-					<td align="center">Copyright &#169; 1999-2006, The Apache Software Foundation<br />Licensed under the <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a>.</td>
+					<td align="center">
+					Copyright &#169; 1999-2006, The Apache Software Foundation<br />Licensed under the <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a>.
+					</td>
 				</tr>
 			</table>
-		</td>
-		<td>&nbsp;</td>
+			</td>
 		</tr>
 	</table>
 	</td>
Index: src/main/webapp/axis2-web/include/header.inc
===================================================================
--- src/main/webapp/axis2-web/include/header.inc	(revision 1179398)
+++ src/main/webapp/axis2-web/include/header.inc	(working copy)
@@ -16,14 +16,21 @@
   ~ specific language governing permissions and limitations
   ~ under the License.
   -->
-<table class="FULL_BLANK"><tr height="100%"><td valign="top">
-<table width="100%">
-<tr><td>
-<table width="950px">
+<table summary="header table" class="FULL_BLANK">
   <tr>
-    <td align="left"><img src="axis2-web/images/asf-logo.gif"/></td>
-    <td align="right"><img src="axis2-web/images/axis_l.jpg"/></td>
-    </tr>
-</table>
-</td>
-<td>&nbsp;</td></tr></table>
+    <td>
+	<table summary="header table" width="100%">
+	  <tr>
+	  	<td>
+		<table summary="embedded header table">
+  		  <tr>
+    		<td align="left"><img src="axis2-web/images/asf-logo.gif" alt="Apache Software Foundation Logo"></td>
+    	 	<td align="right"><img src="axis2-web/images/axis_l.jpg" alt="AXIS Logo"></td>
+          </tr>
+        </table>
+		</td>
+		<td>
+		&nbsp;
+		</td>
+	  </tr>
+	</table>
Index: src/main/webapp/axis2-web/globalModules.jsp
===================================================================
--- src/main/webapp/axis2-web/globalModules.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/globalModules.jsp	(working copy)
@@ -20,7 +20,7 @@
                  java.util.Collection,
                  java.util.Iterator"%>
  <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Globally Engaged Modules</h1>
      <%
          boolean foundModules = false;
@@ -48,8 +48,8 @@
 		}
          } else{
      %>
-    <h2><font color="blue">There are no modules engaged globally</font></h2> 
+    <h2 style="color:blue">There are no modules engaged globally</h2> 
                  <%
          }
                  %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/index.jsp
===================================================================
--- src/main/webapp/axis2-web/index.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/index.jsp	(working copy)
@@ -18,38 +18,39 @@
   --%>
 
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
 <html>
   <head>
     <jsp:include page="include/httpbase.jsp"/>
     <title>Axis 2 - Home</title>
-    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css"/>
+    <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
   </head>
 
   <body>
-    <jsp:include page="include/header.inc" />
-    <br/>
+    <jsp:include page="include/header.inc"/>
+    <br>
 
     <h1>Welcome!</h1>
     Welcome to the new generation of Axis. If you can see this page you have
     successfully deployed the Axis2 Web Application. However, to ensure that Axis2
     is properly working, we encourage you to click on the validate link.
-    <br/>
+    <br>
     <ul>
       <li><a href="services/listServices">Services</a>
-        <br/>
+        <br>
         View the list of all the available services deployed in this server.
       </li>
       <li><a href="axis2-web/HappyAxis.jsp">Validate</a>
-        <br/>
+        <br>
         Check the system to see whether all the required libraries are in place
         and view the system information.
       </li>
       <%--<li><a href="interop.jsp">Do an interop test</a>
-      <br/>
+      <br>
       You can run Interoperability tests here
       </li>--%>
       <li><a href="axis2-admin/">Administration</a>
-        <br/>
+        <br>
         Console for administering this Axis2 installation.
       </li>
     </ul>
Index: src/main/webapp/axis2-web/listModules.jsp
===================================================================
--- src/main/webapp/axis2-web/listModules.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listModules.jsp	(working copy)
@@ -23,7 +23,7 @@
  <%@ page import="java.util.Hashtable"%>
  <%@ page import="java.util.Iterator"%>
  <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
   <h1>Available Modules</h1>
      <%
          boolean foundModules = false;
@@ -61,7 +61,7 @@
       %>
       <%if(errornesModules.size()>0){
           %>
-      <h3><font color="red">Faulty Modules</font></h3>
+      <h3 style="color:red">Faulty Modules</h3>
              <%
              Enumeration faultyModules = errornesModules.keys();
              while (faultyModules.hasMoreElements()) {
@@ -74,8 +74,8 @@
       }
              if(! foundModules) {
                  %>
-                 <h2><font color="blue">There are no modules deployed in the system.</font></h2>
+                 <h2 style="color:blue">There are no modules deployed in the system.</h2>
                  <%
              }
      %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/engagingglobally.jsp
===================================================================
--- src/main/webapp/axis2-web/engagingglobally.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/engagingglobally.jsp	(working copy)
@@ -23,7 +23,6 @@
                  java.util.HashMap,
                  java.util.Iterator" %>
 <%@ page import="org.apache.axis2.util.Utils" %>
-<html>
 <jsp:include page="include/adminheader.jsp"/>
 <%
     String status = (String) request.getSession().getAttribute(Constants.ENGAGE_STATUS);
@@ -35,10 +34,10 @@
     phase needs to be engaged globally.</p>
 
 <form method="get" name="selectModuleForm" action="axis2-admin/engagingglobally">
-    <table border="0" width="100%" cellspacing="1" cellpadding="1">
+    <table summary="main content table" border="0" style="width:100%" cellspacing="1" cellpadding="1">
         <tr>
-            <td width="15%">Select a Module :</td>
-            <td width="75%" align="left">
+            <td style="width: 15%">Select a Module :</td>
+            <td style="width: 75%" align="left">
                 <select name="modules">
                     <%
                         HashMap modules = (HashMap) request.getSession().getAttribute(Constants.MODULE_MAP);
@@ -48,7 +47,7 @@
                             AxisModule axisOperation = (AxisModule) iterator.next();
                             String modulename = axisOperation.getName();
                     %>
-                    <option align="left" value="<%=modulename%>"><%=modulename%>
+                    <option value="<%=modulename%>"><%=modulename%>
                     </option>
                     <%
                         }
@@ -69,4 +68,4 @@
 <p style="color:blue"><%=Utils.sanitizeWebOutput(status)%></p>
 <%
     } %>
-<jsp:include page="include/adminfooter.inc"/>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/LeftFrame.jsp
===================================================================
--- src/main/webapp/axis2-web/LeftFrame.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/LeftFrame.jsp	(working copy)
@@ -38,7 +38,7 @@
 <%@ page import="java.util.Hashtable" %>
 <%@ page import="java.util.Iterator" %>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<table width="100%" style="border-right:1px solid #CCCCCC;">
+<table summary="main content table" width="100%" style="border-right:1px solid #CCCCCC;">
     <tr>
      <td colspan="2" >
        <b>Tools </b>
@@ -54,7 +54,7 @@
     </tr>
 <tr>
      <td colspan="2" >
-      <b><nobr>System Components&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</nobr></b>
+      <span style="font-style:bold; white-space:nowrap">System Components&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span>
      </td>
   </tr>
   <tr>
Index: src/main/webapp/axis2-web/ListServiceGroup.jsp
===================================================================
--- src/main/webapp/axis2-web/ListServiceGroup.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ListServiceGroup.jsp	(working copy)
@@ -36,14 +36,15 @@
         Collection modules = axisServiceGroup.getEngagedModules();
         Iterator axisServiceIter = axisServiceGroup.getServices();
 %>
-<h2><%=groupName%></h2><ul>
+<h2><%=groupName%></h2>
+<ul>
     <%
         while (axisServiceIter.hasNext()){
             AxisService axisService = (AxisService) axisServiceIter.next();
             String serviceName = axisService.getName();
     %>
-    <li><font color="blue"><a href="axis2-admin/ListSingleService?serviceName=<%=serviceName%>">
-        <%=serviceName%></a></font></li>
+    <li><a style="color:blue" href="axis2-admin/ListSingleService?serviceName=<%=serviceName%>">
+        <%=serviceName%></a></li>
     <%
         }
     %>
@@ -64,4 +65,4 @@
         }
     }
 %>
-<jsp:include page="include/adminfooter.inc"/>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/listFaultyService.jsp
===================================================================
--- src/main/webapp/axis2-web/listFaultyService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listFaultyService.jsp	(working copy)
@@ -30,8 +30,8 @@
     <link href="axis2-web/css/axis-style.css" rel="stylesheet" type="text/css">
   </head>
   <body>
-  <jsp:include page="include/header.inc"></jsp:include>
-    <jsp:include page="include/link-footer.jsp"></jsp:include>
+  <jsp:include page="include/header.inc"/>
+    <jsp:include page="include/link-footer.jsp"/>
   <%
         String prifix = request.getAttribute("frontendHostUrl") + (String)request.getSession().getAttribute(Constants.SERVICE_PATH) +"services/";
     %>
@@ -42,7 +42,7 @@
                 Hashtable errornessservices =(Hashtable)request.getSession().getAttribute(Constants.ERROR_SERVICE_MAP);
                 %>
                     <h3>This Web axisService has deployment faults</h3><%
-                     %><font color="red" ><%=(String)errornessservices.get(servicName) %></font>
+                     %><p style="color:red"><%=(String)errornessservices.get(servicName) %></p>
                 <%
 
                     }else {
@@ -53,15 +53,21 @@
            Iterator opItr = axisService.getOperations();
             //operationsList = operations.values();
           String  serviceName = axisService.getName();
-            %><h2><font color="blue"><a href="<%=prifix + axisService.getName()%>?wsdl"><%=serviceName%></a></font></h2>
-           <font color="blue">Service EPR : </font><font color="black"><%=prifix + axisService.getName()%></font><br>
-           <h4>Service Description : <font color="black"><%=axisService.getDocumentation()%></font></h4>
-           <i><font color="blue">Service Status : <%=axisService.isActive()?"Active":"InActive"%></font></i><br>
+            %><h2><a style="color:blue" href="<%=prifix + axisService.getName()%>?wsdl"><%=serviceName%></a></h2>
+            <p>
+            <span style="color:blue">Service EPR :</span>
+            <span style="color:black"><%=prifix + axisService.getName()%></span>
+            </p>
+           
+           <h4>Service Description : <span style="color:blue"><%=axisService.getDocumentation()%></span></h4>
+           <p>
+           <span style="color:blue; font-style:italic">Service Status : <%=axisService.isActive()?"Active":"InActive"%></span>
+           </p>
            <%
             if (opItr.hasNext()) {
-                %><i>Available operations</i><%
+                %><span style="font-style:italic">Available operations</span><%
             } else {
-                %><i> There are no Operations specified</i><%
+                %><span style="font-style:italic"> There are no Operations specified</span><%
             }
                opItr = axisService.getOperations();
            %><ul><%
@@ -74,12 +80,12 @@
            <%
                     } else{
                            %>
-                <h3><font color="red" >No services found in this location.</font></h3>
+                <h3 style="color:red">No services found in this location.</h3>
  <%
                     }
 
             }
         %>
-<jsp:include page="include/footer.inc"></jsp:include>
+<jsp:include page="include/footer.inc"/>
         </body>
 </html>
Index: src/main/webapp/axis2-web/error.jsp
===================================================================
--- src/main/webapp/axis2-web/error.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/error.jsp	(working copy)
@@ -19,8 +19,8 @@
 
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
 <%@page isErrorPage="true" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
-  <table>
+<jsp:include page="include/adminheader.jsp"/>
+  <table summary="main content table">
   <tr>
   <td align="center">An error has occured!. Please check the details below</td>
   </tr>
@@ -28,4 +28,4 @@
   <td align="center"><%=exception.getMessage()%></td>
   </tr>
   </table>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
Index: src/main/webapp/axis2-web/ViewContexts.jsp
===================================================================
--- src/main/webapp/axis2-web/ViewContexts.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/ViewContexts.jsp	(working copy)
@@ -23,7 +23,7 @@
 <%@ page import="org.apache.axis2.context.ServiceGroupContext"%>
 <%@ page import="java.util.Iterator"%>
 <%@ page contentType="text/html;charset=UTF-8" language="java" %>
-<jsp:include page="include/adminheader.jsp"></jsp:include>
+<jsp:include page="include/adminheader.jsp"/>
 <h1>Running Context Hierarchy</h1>
 <%
     ConfigurationContext configContext = (ConfigurationContext) request.getSession().getAttribute(
@@ -39,9 +39,11 @@
         String groupContextID = serviceGroupIds[i];
         ServiceGroupContext groupContext = configContext.getServiceGroupContext(groupContextID);
         %>
-           <li><%=groupContextID%><font color="blue"><a href="axis2-admin/viewServiceGroupConetxt?TYPE=VIEW&ID=<%=groupContextID%>">
-                    View</a></font>  <font color="red"><a href="axis2-admin/viewServiceGroupConetxt?TYPE=DELETE&ID=<%=groupContextID%>">
-                    Remove</a> </font></li>
+           <li><%=groupContextID%><a style="color:blue" href="axis2-admin/viewServiceGroupConetxt?TYPE=VIEW&ID=<%=groupContextID%>">
+                    View</a>  <a style="color:red" href="axis2-admin/viewServiceGroupConetxt?TYPE=DELETE&ID=<%=groupContextID%>">
+                    Remove</a> </li>
+                    
+                    <li>
         <%
         Iterator serviceContextItr = groupContext.getServiceContexts();
             %><ul><%
@@ -49,11 +51,12 @@
             ServiceContext serviceContext = (ServiceContext)serviceContextItr.next();
              String serviceConID = serviceContext.getAxisService().getName();
         %>
-            <li><%=serviceConID%><font color="blue"><a href="axis2-admin/viewServiceContext?TYPE=VIEW&ID=<%=serviceConID%>&PID=<%=groupContextID%>">
-                    View</a></font></li>
+            <li><%=serviceConID%><a style="color:blue" href="axis2-admin/viewServiceContext?TYPE=VIEW&ID=<%=serviceConID%>&PID=<%=groupContextID%>">
+                    View</a></li>
         <%
         }
-                %></ul><hr><%
+                %></ul><hr>
+                </li><%
     }
     %>  </ul>
         <%
@@ -61,4 +64,4 @@
 	<p>No running contexts were found on the system.</p>
             <%}
 %>
-<jsp:include page="include/adminfooter.inc"></jsp:include>
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/listService.jsp
===================================================================
--- src/main/webapp/axis2-web/listService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/listService.jsp	(working copy)
@@ -48,7 +48,7 @@
             AxisService axisService = (AxisService) iterator.next();
             operations = axisService.getOperations();
             serviceName = axisService.getName();
-%><h2><font color="blue"><a href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></font></h2>
+%><h2><a style="color:blue" href="<%=prefix + axisService.getName()%>?wsdl"><%=serviceName%></a></h2>
 <%
     // do we need to enable REST in the main servlet so that it handles both REST and SOAP messages
     boolean disableREST = false;
@@ -75,9 +75,9 @@
         serviceDescription = "No description available for this service";
     }
 %>
-<p>Service Description : <%=serviceDescription%><br/>
-Service EPR : <%=prefix + axisService.getName()%><br/>
-Service Status : <%=axisService.isActive() ? "Active" : "InActive"%>
+<p>Service Description : <%=serviceDescription%><br>
+Service EPR : <%=prefix + axisService.getName()%><br>
+Service Status : <%=axisService.isActive() ? "Active" : "InActive"%><br>
 Actions : <a href="axis2-admin/deleteService?serviceName=<%=serviceName%>">Remove Service</a></p><br>
 <%
     Collection engagedModules = axisService.getEngagedModules();
@@ -85,7 +85,7 @@
     boolean modules_present = false;
     if (engagedModules.size() > 0) {
 %>
-<i>Engaged modules for the service</i>
+<span style="font-style: italic">Engaged modules for the service</span>
 <%
     for (Iterator iteratorm = engagedModules.iterator(); iteratorm.hasNext();) {
         AxisModule axisOperation = (AxisModule) iteratorm.next();
@@ -95,8 +95,8 @@
 %>
 <ul>
     <% }
-    %><li><%=moduleName%> :: <a href="axis2-admin/disengageModule?type=service&serviceName=<%=serviceName%>&module=<%=moduleName%>">Disengage</a></li>
-    <br>
+    %><li><%=moduleName%> :: <a href="axis2-admin/disengageModule?type=service&amp;serviceName=<%=serviceName%>&amp;module=<%=moduleName%>">Disengage</a></li>
+
     <%
         }
         if (modules_present) {%>
@@ -105,11 +105,16 @@
         }
     }
     if (operations.hasNext()) {
-%><br><i>Available operations</i><%
+%>
+<br>
+<span style="font-style: italic">Available operations</span>
+<%
 } else {
-%><i> There are no Operations specified</i><%
+%><span style="font-style: italic"> There are no Operations specified</span>
+<%
     }
-%><ul><%
+%>
+<ul><%
     operations = axisService.getOperations();
     while (operations.hasNext()) {
         AxisOperation axisOperation = (AxisOperation) operations.next();
@@ -119,18 +124,27 @@
         engagedModules = axisOperation.getEngagedModules();
         if (engagedModules.size() > 0) {
     %>
-    <br><i>Engaged Modules for the Operation</i><ul>
+    <li>
+    
+    <span style="font-style: italic">Engaged Modules for the Operation</span>
+    <ul>
     <%
         for (Iterator iterator2 = engagedModules.iterator(); iterator2.hasNext();) {
             AxisModule moduleDecription = (AxisModule) iterator2.next();
             moduleName = moduleDecription.getName();
-    %><li><%=moduleName%> :: <a href="axis2-admin/disengageModule?type=operation&serviceName=<%=serviceName%>&operation=<%=axisOperation.getName().getLocalPart()%>&module=<%=moduleName%>">Disengage</a></li><br><%
+    %>
+    <li>
+    <%=moduleName%> :: <a href="axis2-admin/disengageModule?type=operation&amp;serviceName=<%=serviceName%>&amp;operation=<%=axisOperation.getName().getLocalPart()%>&amp;module=<%=moduleName%>">Disengage</a>
+    </li>
+    <%
     }
-%></ul><%
+	%>
+</ul></li><%
         }
 
     }
-%></ul>
+%>
+</ul>
 <%
             status = true;
         }
@@ -139,13 +153,13 @@
         if (errornessservice.size() > 0) {
             request.getSession().setAttribute(Constants.IS_FAULTY, Constants.IS_FAULTY);
 %>
-<h3><font color="red">Faulty Services</font></h3>
+<h3 style="color:red">Faulty Services</h3>
 <%
     Enumeration faultyservices = errornessservice.keys();
     while (faultyservices.hasMoreElements()) {
         String faultyserviceName = (String) faultyservices.nextElement();
-%><h3><font color="blue"><a href="services/ListFaultyServices?serviceName=<%=faultyserviceName%>">
-    <%=faultyserviceName%></a></font></h3>
+%><h3><a styel="color:blue" href="services/ListFaultyServices?serviceName=<%=faultyserviceName%>">
+    <%=faultyserviceName%></a></h3>
 <%
             }
         }
@@ -155,4 +169,4 @@
 %> No services listed! Try hitting refresh. <%
     }
 %>
-<jsp:include page="include/adminfooter.inc" />
+<jsp:include page="include/adminfooter.jsp"/>
Index: src/main/webapp/axis2-web/deleteService.jsp
===================================================================
--- src/main/webapp/axis2-web/deleteService.jsp	(revision 1179398)
+++ src/main/webapp/axis2-web/deleteService.jsp	(working copy)
@@ -20,4 +20,4 @@
 <jsp:include page="include/adminheader.jsp"/>
 <p><%=request.getSession().getAttribute("status")%></p>
 <p>Click <a href="services/listServices">here</a> to go to the service list page</p>
-<jsp:include page="include/adminfooter.inc" />
\ No newline at end of file
+<jsp:include page="include/adminfooter.jsp"/>
\ No newline at end of file
