C:\svn\axis2\java\modules\core>svn diff
Index: src/org/apache/axis2/transport/http/AxisServlet.java
===================================================================
--- src/org/apache/axis2/transport/http/AxisServlet.java        (revision 357874)
+++ src/org/apache/axis2/transport/http/AxisServlet.java        (working copy)
@@ -183,21 +183,34 @@
      */
     public void init(ServletConfig config) throws ServletException {
         try {
+            configContext = initConfigContext(config);
+            lister = new ListingAgent(configContext);
+            context.setAttribute(CONFIGURATION_CONTEXT, configContext);
+        } catch (Exception e) {
+            throw new ServletException(e);
+        }
+    }
+
+    /**
+     * Initialize the Axis configuration context
+     *
+     * @param config Servlet configuration
+     * @throws ServletException
+     */
+    protected ConfigurationContext initConfigContext(ServletConfig config) throws ServletException {
+        try {
             ServletContext context = config.getServletContext();
             String repoDir = context.getRealPath("/WEB-INF");
             ConfigurationContextFactory erfac = new ConfigurationContextFactory();
-
-            configContext = erfac.buildConfigurationContext(repoDir);
+            ConfigurationContext configContext = erfac.buildConfigurationContext(repoDir);
             configContext.setProperty(Constants.CONTAINER_MANAGED, Constants.VALUE_TRUE);
             configContext.setRootDir(new File(context.getRealPath("/WEB-INF")));
-            lister = new ListingAgent(configContext);
-            context.setAttribute(CONFIGURATION_CONTEXT, configContext);
-
+            return configContext;
         } catch (Exception e) {
             throw new ServletException(e);
         }
-    }
-
+    }
+
     private HashMap getHTTPParameters(HttpServletRequest httpServletRequest) {
         HashMap map = new HashMap();
         Enumeration enu = httpServletRequest.getParameterNames();
