Index: C:/wasx/axis2/modules/kernel/src/org/apache/axis2/AxisFault.java
===================================================================
--- C:/wasx/axis2/modules/kernel/src/org/apache/axis2/AxisFault.java	(revision 556683)
+++ C:/wasx/axis2/modules/kernel/src/org/apache/axis2/AxisFault.java	(working copy)
@@ -295,7 +295,20 @@
         setFaultCode(faultCode);
     }
 
+    
     /**
+     * @param message
+     * @param faultMessageContext
+     * @param cause
+     */
+    public AxisFault(String message, MessageContext faultMessageContext, Throwable cause) {
+        super(message, cause);
+
+        this.faultMessageContext = faultMessageContext;
+    }
+
+
+    /**
      * @param messageText - this will appear as the Text in the Reason information item of SOAP Fault
      * @param faultCode   - this will appear as the Value in the Code information item of SOAP Fault
      * @param cause       - this will appear under the Detail information item of SOAP Fault
