Index: modules/codegen/src/org/apache/axis2/wsdl/util/XSLTIncludeResolver.java
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/util/XSLTIncludeResolver.java	(revision 551017)
+++ modules/codegen/src/org/apache/axis2/wsdl/util/XSLTIncludeResolver.java	(working copy)
@@ -76,6 +76,8 @@
                 supporterTemplateStream = getClass().getResourceAsStream(templateName);
                 return new StreamSource(supporterTemplateStream);
             }
+        } else if ((href != null) && (!href.equals("externalTemplate"))){
+            return getSourceFromTemplateName(href);
         }
         //if nothing could be found return an empty source
         return getEmptySource();
