Index: modules/kernel/src/org/apache/axis2/addressing/EndpointReference.java
===================================================================
--- modules/kernel/src/org/apache/axis2/addressing/EndpointReference.java	(revision 608073)
+++ modules/kernel/src/org/apache/axis2/addressing/EndpointReference.java	(working copy)
@@ -175,7 +175,7 @@
     /**
      * hasAnonymousAddress
      *
-     * @return true if address is 'Anonymous URI'
+     * @return true if address is 'Anonymous URI' or a 'MakeConnection' anonymous URI
      */
     public boolean hasAnonymousAddress() {
         boolean result = (AddressingConstants.Final.WSA_ANONYMOUS_URL.equals(address) ||
@@ -186,8 +186,7 @@
                 //transports back channel, as opposed to asynchronously. No other
                 //semantics normally associated with WS-Addressing anonymous values should
                 //be assumed, by it's presence here.
-                (address != null && address.startsWith(
-                        "http://docs.oasis-open.org/ws-rx/wsmc/200702/anonymous?id=")));
+                isMakeConnectionAddress());
         if (log.isTraceEnabled()) {
             log.trace("hasAnonymousAddress: " + address + " is Anonymous: " + result);
         }
@@ -195,6 +194,19 @@
     }
 
     /**
+     * isMakeConnectionAddress
+     *
+     * @return true if address is a MakeConnection URI
+     */	
+	public boolean isMakeConnectionAddress(){
+		boolean result = (address != null && address.startsWith("http://docs.oasis-open.org/ws-rx/wsmc/200702/anonymous?id="));
+        if (log.isTraceEnabled()) {
+            log.trace("isMakeConnectionAddress: " + address + " is Anonymous: " + result);
+        }
+        return result;		
+	}
+
+    /**
      * hasNoneAddress
      *
      * @return true if the address is the 'None URI' from the final addressing spec.
