Index: C:/WebServices/Eclipse/workspace/axis2_sandbox/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLSpineImpl.java
===================================================================
--- C:/WebServices/Eclipse/workspace/axis2_sandbox/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLSpineImpl.java	(revision 428035)
+++ C:/WebServices/Eclipse/workspace/axis2_sandbox/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLSpineImpl.java	(working copy)
@@ -34,6 +34,7 @@
 import org.apache.axiom.soap.SOAPFactory;
 import org.apache.axiom.soap.SOAPHeader;
 import org.apache.axiom.soap.impl.llom.soap11.SOAP11Factory;
+import org.apache.axiom.soap.impl.llom.soap12.SOAP12Factory;
 import org.apache.axis2.jaxws.ExceptionFactory;
 import org.apache.axis2.jaxws.message.Block;
 import org.apache.axis2.jaxws.message.MessageException;
@@ -172,7 +173,7 @@
 		if (protocol == Protocol.soap11) {
 			soapFactory = new SOAP11Factory();
 		} else if (protocol == Protocol.soap12) {
-			soapFactory = new SOAP11Factory();
+			soapFactory = new SOAP12Factory();
 		} else {
 			// TODO NLS
 			throw ExceptionFactory.makeMessageInternalException("unsupported protocol", null);
