diff --git a/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java b/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java
index bb27eae..f3fa695 100644
--- a/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java
+++ b/modules/kernel/src/org/apache/axis2/transport/http/AxisServlet.java
@@ -93,6 +93,7 @@ public class AxisServlet extends HttpServlet implements TransportListener {
     private boolean closeReader = true;
 
     private static final int BUFFER_SIZE = 1024 * 8;
+    private static boolean initialized = false;
 
     /**
      * Implementaion of POST interface
@@ -483,7 +484,8 @@ public class AxisServlet extends HttpServlet implements TransportListener {
      * @throws ServletException
      */
     public void init() throws ServletException {
-        if (this.servletConfig != null) {
+        if (this.servletConfig != null && !initialized) {
+			initialized = true;
             init(this.servletConfig);
         }
     }
