Index: modules/kernel/src/org/apache/axis2/transport/http/RESTSender.java
===================================================================
--- modules/kernel/src/org/apache/axis2/transport/http/RESTSender.java	(revision 437980)
+++ modules/kernel/src/org/apache/axis2/transport/http/RESTSender.java	(working copy)
@@ -145,7 +145,11 @@
         String param = getParam(msgContext);
         GetMethod getMethod = new GetMethod();
 
-        getMethod.setPath(url.getFile() + "?" + param);
+        if (param != null && param.length() > 0) {
+            getMethod.setPath(url.getFile() + "?" + param);
+        } else {
+            getMethod.setPath(url.getFile());
+        }
 
         // Serialization as "application/x-www-form-urlencoded"
         String charEncoding =
