Index: modules/kernel/test/org/apache/axis2/xmlfilter/FilterConfigLoadDemo.java
===================================================================
--- modules/kernel/test/org/apache/axis2/xmlfilter/FilterConfigLoadDemo.java	(revision 0)
+++ modules/kernel/test/org/apache/axis2/xmlfilter/FilterConfigLoadDemo.java	(revision 0)
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.axis2.xmlfilter;
+
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.Iterator;
+
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+import org.apache.axiom.om.util.StAXUtils;
+import org.apache.axis2.AxisFault;
+import org.apache.axis2.context.ConfigurationContext;
+import org.apache.axis2.context.ConfigurationContextFactory;
+import org.apache.axis2.description.FilterDescription;
+import org.apache.axis2.engine.AxisConfiguration;
+
+public class FilterConfigLoadDemo {
+
+    /**
+     * @param args
+     */
+    public static void main(String[] args) {
+        try {
+//            String filelocation = System.getProperty("basedir") + "/test-resources/deployment/filter/test"; // for maven
+            String filelocation = "modules/kernel/test-resources/deployment/filter/test"; // for stand alone
+            String filename = filelocation + "/axis2.xml";
+            
+            ConfigurationContext configContext = ConfigurationContextFactory.createConfigurationContextFromFileSystem(filelocation, filename);
+            AxisConfiguration axisConfiguration = configContext.getAxisConfiguration();
+
+            // check FilterDescriptions
+            Iterator filterOrderIte = axisConfiguration.getFilterOrder().iterator();
+            while(filterOrderIte.hasNext()){
+                FilterDescription filterDesc = (FilterDescription) filterOrderIte.next();
+            }
+            
+            // check FilterRegistry
+            InputStream in = new ByteArrayInputStream(new byte[1]);
+            XMLStreamReader xmlReader = StAXUtils.createXMLStreamReader(in);
+            FilterRegistry filterRegistry = axisConfiguration.getFilterRegistry();
+            xmlReader = filterRegistry.enableFilter(xmlReader, null);
+            
+        } catch (AxisFault e) {
+            e.printStackTrace();
+        } catch (XMLStreamException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+}
Index: modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter1.java
===================================================================
--- modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter1.java	(revision 0)
+++ modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter1.java	(revision 0)
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.xmlfilter;
+
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.stream.util.StreamReaderDelegate;
+
+public class TestFilter1 extends StreamReaderDelegate implements Filter {
+
+    Object context = null;
+    
+    public XMLStreamReader init(XMLStreamReader reader, Object context) {
+        
+        super.setParent(reader);
+        this.context = context;
+        
+        return this;
+    }
+
+    public int next() throws XMLStreamException {
+        int eventType = super.next();
+        
+        return eventType;
+    }
+}
Index: modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter2.java
===================================================================
--- modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter2.java	(revision 0)
+++ modules/kernel/test/org/apache/axis2/xmlfilter/TestFilter2.java	(revision 0)
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.xmlfilter;
+
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.stream.util.StreamReaderDelegate;
+
+public class TestFilter2 extends StreamReaderDelegate implements Filter {
+
+    Object context = null;
+    
+    public XMLStreamReader init(XMLStreamReader reader, Object context) {
+        
+        super.setParent(reader);
+        this.context = context;
+        
+        return this;
+    }
+
+    public int next() throws XMLStreamException {
+        int eventType = super.next();
+        
+        return eventType;
+    }
+}
Index: modules/kernel/test/org/apache/axis2/xmlfilter/FilterDemo.java
===================================================================
--- modules/kernel/test/org/apache/axis2/xmlfilter/FilterDemo.java	(revision 0)
+++ modules/kernel/test/org/apache/axis2/xmlfilter/FilterDemo.java	(revision 0)
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.xmlfilter;
+
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+import org.apache.axiom.om.util.StAXUtils;
+import org.apache.axiom.soap.SOAP11Constants;
+import org.apache.axiom.soap.SOAPEnvelope;
+import org.apache.axiom.soap.impl.builder.StAXSOAPModelBuilder;
+import org.apache.axis2.AxisFault;
+import org.apache.axis2.context.ConfigurationContext;
+import org.apache.axis2.context.ConfigurationContextFactory;
+import org.apache.axis2.context.MessageContext;
+
+public class FilterDemo {
+    
+    public static void main(String[] args) {
+        
+        try {
+            MessageContext messageContext = new MessageContext();
+            
+            // load axis2.xml
+            String filelocation = "modules/kernel/test-resources/deployment/filter/test";
+            String filename = filelocation + "/axis2.xml";
+            ConfigurationContext configContext = ConfigurationContextFactory.createConfigurationContextFromFileSystem(filelocation, filename);
+            messageContext.setConfigurationContext(configContext);
+            
+            //prepare byte array
+            String xmlfile = "modules/kernel/test-resources/deployment/filter/demo/soapsample.xml";
+            ByteArrayInputStream bais = getByteArray(xmlfile);
+
+            // set myactor to MessageContext
+            messageContext.setProperty("myactor", "webservice1");
+            
+            // create XML StAX parser
+            XMLStreamReader xmlReader = StAXUtils.createXMLStreamReader(bais);
+            
+            // set filter
+            xmlReader = messageContext
+                          .getConfigurationContext()
+                          .getAxisConfiguration()
+                          .getFilterRegistry().enableFilter(xmlReader, messageContext);
+            
+            // create Builder
+            StAXSOAPModelBuilder builder = new StAXSOAPModelBuilder(xmlReader, SOAP11Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+            
+            // parse soap message
+            SOAPEnvelope soapEnvelope = builder.getSOAPEnvelope();
+            soapEnvelope.build();
+            
+            System.out.println(soapEnvelope);
+        } catch (XMLStreamException e) {
+            e.printStackTrace();
+        } catch (AxisFault e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static ByteArrayInputStream getByteArray(String xmlfile) {
+        try {
+            System.out.println("Reading " + xmlfile);
+            ByteArrayInputStream bais;
+            ByteArrayOutputStream tmpOS;
+
+            byte[] buffer = new byte[1024];
+            FileInputStream in = new FileInputStream(xmlfile);
+            tmpOS = new ByteArrayOutputStream();
+            int length;
+            while ((length = in.read(buffer)) != -1) {
+                tmpOS.write(buffer, 0, length);
+            }
+            in.close();
+            bais = new ByteArrayInputStream(tmpOS.toByteArray());
+            return bais;
+        } catch (FileNotFoundException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+        return null;
+    }
+}
Index: modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java
===================================================================
--- modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java	(revision 466220)
+++ modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java	(working copy)
@@ -33,6 +33,7 @@
 import org.apache.axis2.transport.TransportSender;
 import org.apache.axis2.util.TargetResolver;
 import org.apache.axis2.util.Loader;
+import org.apache.axis2.xmlfilter.FilterRegistry;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
@@ -91,7 +92,12 @@
             Iterator trs_Reivers = config_element.getChildrenWithName(new QName(TAG_TRANSPORT_RECEIVER));
 
             processTransportReceivers(trs_Reivers);
-
+            
+            // Processing filter order
+            Iterator filterOrder_ite = config_element.getChildrenWithName(new QName(TAG_FILTER_ORDER));
+            
+            processFilterOrder(filterOrder_ite);
+            
             // Process Observers
             Iterator obs_ittr = config_element.getChildrenWithName(new QName(TAG_LISTENER));
 
@@ -424,4 +430,52 @@
         Class phaseClass = Loader.loadClass(axisConfig.getSystemClassLoader(), className);
         return (Phase) phaseClass.newInstance();
     }
+    
+    /*
+     * <filterOrder>
+     *     <filter name="" class="" />
+     *     <filter name="" class="">
+     *         <parameter name="" locked="">xxx</parameter>
+     *     </filter>
+     * </filterOrder>
+     */
+    private void processFilterOrder(Iterator filterOrder_ite) throws DeploymentException {
+        while (filterOrder_ite.hasNext()) {
+            
+            // create filter description
+            OMElement filterOrderElem = (OMElement) filterOrder_ite.next();
+            Iterator filter_ite = filterOrderElem.getChildrenWithName(new QName(TAG_FILTER));
+            while(filter_ite.hasNext()) {
+                OMElement filterElem = (OMElement) filter_ite.next();
+
+                // get filter name
+                OMAttribute filterName = filterElem.getAttribute(new QName(ATTRIBUTE_NAME));
+                if(filterName == null)
+                    continue;
+                
+                // get filter class name
+                OMAttribute filterClassName = filterElem.getAttribute(new QName(TAG_CLASS_NAME));
+                if(filterClassName == null)
+                    continue;
+                
+                String name = filterName.getAttributeValue();
+                String className = filterClassName.getAttributeValue();
+                FilterDescription filterDescription = new FilterDescription(new QName(name));
+                filterDescription.setClassname(className);
+                
+                try {
+                    Iterator itr = filterElem.getChildrenWithName(new QName(TAG_PARAMETER));
+                    processParameters(itr, filterDescription, axisConfig);
+                    axisConfig.addFilterDescription(filterDescription);
+                } catch (AxisFault axisFault) {
+                    throw new DeploymentException(axisFault);
+                }
+            }
+        }
+        
+        // register filter
+        FilterRegistry filterRegistry = axisConfig.getFilterRegistry();
+        ArrayList filterOrder = axisConfig.getFilterOrder();
+        filterRegistry.registerFilter(filterOrder);
+    }
 }
Index: modules/kernel/src/org/apache/axis2/deployment/DeploymentConstants.java
===================================================================
--- modules/kernel/src/org/apache/axis2/deployment/DeploymentConstants.java	(revision 466220)
+++ modules/kernel/src/org/apache/axis2/deployment/DeploymentConstants.java	(working copy)
@@ -48,6 +48,8 @@
     String TAG_TARGET_RESOLVER = "targetResolver";
     String TAG_TRANSPORT_SENDER = "transportSender";
     String TAG_TRANSPORT_RECEIVER = "transportReceiver";
+    String TAG_FILTER_ORDER = "filterOrder";
+    String TAG_FILTER = "filter";
     String TAG_SERVICE_GROUP = "serviceGroup";
     String TAG_SERVICE = "service";
     String TAG_REFERENCE = "ref";
Index: modules/kernel/src/org/apache/axis2/description/FilterDescription.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/FilterDescription.java	(revision 0)
+++ modules/kernel/src/org/apache/axis2/description/FilterDescription.java	(revision 0)
@@ -0,0 +1,57 @@
+package org.apache.axis2.description;
+
+import java.util.ArrayList;
+
+import javax.xml.namespace.QName;
+
+import org.apache.axiom.om.OMElement;
+import org.apache.axis2.AxisFault;
+
+public class FilterDescription implements ParameterInclude {
+    
+    protected QName name;
+    protected String classname;
+    protected final ParameterInclude paramInclude;
+
+    public FilterDescription(QName name) {
+        paramInclude = new ParameterIncludeImpl();
+        this.name = name;
+    }
+    
+    public QName getName() {
+        return name;
+    }
+
+    public String getClassname() {
+        return classname;
+    }
+
+    public void setClassname(String classname) {
+        this.classname = classname;
+    }
+
+    public void addParameter(Parameter param) throws AxisFault {
+        paramInclude.addParameter(param);
+    }
+
+    public void removeParameter(Parameter param) throws AxisFault {
+        paramInclude.removeParameter(param);
+    }
+
+    public void deserializeParameters(OMElement parameterElement) throws AxisFault {
+        this.paramInclude.deserializeParameters(parameterElement);
+    }
+
+    public Parameter getParameter(String name) {
+        return paramInclude.getParameter(name);
+    }
+
+    public ArrayList getParameters() {
+        return paramInclude.getParameters();
+    }
+
+    public boolean isParameterLocked(String parameterName) {
+        return paramInclude.isParameterLocked(parameterName);
+    }
+
+}
Index: modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java
===================================================================
--- modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java	(revision 466220)
+++ modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java	(working copy)
@@ -24,6 +24,7 @@
 import org.apache.axis2.phaseresolver.PhaseResolver;
 import org.apache.axis2.util.TargetResolver;
 import org.apache.axis2.util.Utils;
+import org.apache.axis2.xmlfilter.FilterRegistry;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.ws.java2wsdl.Java2WSDLConstants;
@@ -54,6 +55,9 @@
     private final HashMap transportsIn = new HashMap();
 
     private final HashMap transportsOut = new HashMap();
+    
+    private final ArrayList filterOrder = new ArrayList();
+    private final FilterRegistry filterRegistry = new FilterRegistry();
 
     private final HashMap policySupportedModules = new HashMap();
 
@@ -879,4 +883,31 @@
     public boolean isAssertionLocal(QName name) {
         return this.localPolicyAssertions.contains(name);
     }
+    
+    public void addFilterDescription(FilterDescription filterDesc) throws AxisFault {
+        filterOrder.add(filterDesc);
+    }
+    
+    public ArrayList getFilterOrder() {
+        return filterOrder;
+    }
+    
+    public FilterRegistry getFilterRegistry() {
+        return filterRegistry;
+    }
+    
+    public FilterDescription getFilterDescription(String classname) {
+        if(classname==null)
+            return null;
+        FilterDescription ret = null;
+        Iterator ite = filterOrder.iterator();
+        while(ite.hasNext()) {
+            FilterDescription filterDescription = (FilterDescription) ite.next();
+            if(classname.equals(filterDescription.getClassname())){
+                ret = filterDescription;
+                break;
+            }
+        }
+        return ret;
+    }
 }
Index: modules/kernel/src/org/apache/axis2/xmlfilter/Filter.java
===================================================================
--- modules/kernel/src/org/apache/axis2/xmlfilter/Filter.java	(revision 0)
+++ modules/kernel/src/org/apache/axis2/xmlfilter/Filter.java	(revision 0)
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.xmlfilter;
+
+import javax.xml.stream.XMLStreamReader;
+
+public interface Filter extends XMLStreamReader {
+
+    public XMLStreamReader init(XMLStreamReader reader, Object context);
+
+}
Index: modules/kernel/src/org/apache/axis2/xmlfilter/FilterRegistry.java
===================================================================
--- modules/kernel/src/org/apache/axis2/xmlfilter/FilterRegistry.java	(revision 0)
+++ modules/kernel/src/org/apache/axis2/xmlfilter/FilterRegistry.java	(revision 0)
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.axis2.xmlfilter;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+
+import javax.xml.stream.XMLStreamReader;
+
+import org.apache.axis2.description.FilterDescription;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class FilterRegistry {
+
+    private static final Log log = LogFactory.getLog(FilterRegistry.class);
+    private static boolean doDebug = log.isDebugEnabled();
+
+    boolean enable = false;
+    ArrayList registeredFilter;
+
+    public boolean isEnable() {
+        if (doDebug) log.debug("FilterRegistry.enable() : " + enable);
+        return enable;
+    }
+
+    public XMLStreamReader enableFilter(XMLStreamReader reader, Object context) {
+        if(enable) {
+            return insertFilter(reader, context);
+        } else
+            return reader;
+    }
+    
+    public XMLStreamReader insertFilter(XMLStreamReader reader, Object context) {
+        XMLStreamReader ret = reader;
+        Iterator ite = registeredFilter.iterator();
+        while (ite.hasNext()) {
+            Class cls = (Class) ite.next();
+            try {
+                Filter filter = (Filter) cls.newInstance();
+                ret = filter.init(ret,  context);
+                if (doDebug) log.debug("FilterRegistry.insertFilter() : " + filter);
+            } catch (IllegalAccessException e) {
+                e.printStackTrace();
+            } catch (InstantiationException e) {
+                e.printStackTrace();
+            }
+        }
+        return ret;
+    }
+
+    public void registerFilter(ArrayList filterOrder) {
+        ClassLoader classloader = null;
+
+        classloader = Thread.currentThread().getContextClassLoader();
+        int size = filterOrder.size();
+        for (int i = size - 1; i >= 0; i--) {
+            FilterDescription filterDesc = (FilterDescription) filterOrder.get(i);
+            registerFilter(classloader, filterDesc.getClassname());
+        }
+    }
+    
+    public void registerFilter(ClassLoader classloader, String className) {
+        try {
+            Class cls = classloader.loadClass(className);
+            if (cls != null) {
+                if (registeredFilter == null)
+                    registeredFilter = new ArrayList();
+                registeredFilter.add(cls);
+                enable = true;
+                if (doDebug) log.debug("FilterRegistry.registerFilter() success to register filter : " + className);
+            }
+        } catch (ClassNotFoundException e) {
+            if (doDebug) log.debug("FilterRegistry.registerFilter(), failed to get Class object of classname=" + className);
+            // class not found
+            e.printStackTrace();
+        }
+    }
+
+
+}
Index: modules/kernel/test-resources/deployment/filter/test/axis2.xml
===================================================================
--- modules/kernel/test-resources/deployment/filter/test/axis2.xml	(revision 0)
+++ modules/kernel/test-resources/deployment/filter/test/axis2.xml	(revision 0)
@@ -0,0 +1,282 @@
+<axisconfig name="AxisJava2.0">
+    <!-- ================================================= -->
+    <!-- Parameters -->
+    <!-- ================================================= -->
+    <parameter name="hotdeployment" locked="false">true</parameter>
+    <parameter name="hotupdate" locked="false">false</parameter>
+    <parameter name="enableMTOM" locked="false">false</parameter>
+    <parameter name="enableSwA" locked="false">false</parameter>
+
+    <!--Uncomment if you want to enable file caching for attachments -->
+    <!--parameter name="cacheAttachments" locked="false">true</parameter>
+    <parameter name="attachmentDIR" locked="false"></parameter>
+    <parameter name="sizeThreshold" locked="false">4000</parameter-->
+
+    <!--This will give out the timout of the configuration contexts, in seconds-->
+    <parameter name="ConfigContextTimeoutInterval" locked="false">30</parameter>
+
+    <!--During a fault, stacktrace can be sent with the fault message. The following flag will control -->
+    <!--that behaviour.-->
+    <parameter name="sendStacktraceDetailsWithFaults" locked="false">true</parameter>
+
+    <!--If there aren't any information available to find out the fault reason, we set the message of the expcetion-->
+    <!--as the faultreason/Reason. But when a fault is thrown from a service or some where, it will be -->
+    <!--wrapped by different levels. Due to this the initial exception message can be lost. If this flag-->
+    <!--is set then, Axis2 tries to get the first exception and set its message as the faultreason/Reason.-->
+    <parameter name="DrillDownToRootCauseForFaultReason" locked="false">false</parameter>
+
+    <parameter name="userName" locked="false">admin</parameter>
+    <parameter name="password" locked="false">axis2</parameter>
+
+    <!--Following params will set the proper context paths for invocations. All the endpoints will have a commons context-->
+    <!--root which can configured using the following contextRoot parameter-->
+    <!--<parameter name="contextRoot" locked="false">axis2</parameter>-->
+
+    <!--Our HTTP endpoints can handle both REST and SOAP. Following parameters can be used to distingiush those endpoints-->
+    <!--In case of a servlet, if you change this you have to manually change the settings of your servlet container to map this -->
+    <!--context path to proper Axis2 servlets-->
+    <!--<parameter name="servicePath" locked="false">services</parameter>-->
+    <!--<parameter name="restPath" locked="false">rest</parameter>-->
+
+
+    <!--Set the flag to true if you want to enable transport level session mangment-->
+    <parameter name="manageTransportSession" locked="false">false</parameter>
+
+    <!--Following two parameters will be used to handle REST in Axis2. The default settings will make Axis2 to have two-->
+    <!--different endpoints, one for REST (AxisRESTServlet) one for SOAP message handling (AxisServlet). But following-->
+    <!--parameters help to tweak the message handling of two main servlets. -->
+
+    <!-- If the enableRESTInAxis2MainServlet is true, then Axis2MainServlet will handle both SOAP and REST messages -->
+    <parameter name="enableRESTInAxis2MainServlet" locked="true">false</parameter>
+
+    <!-- Following parameter will completely disable REST handling in both the servlets-->
+    <parameter name="disableREST" locked="true">false</parameter>
+
+    <!-- This will disable the separate servlet we have for REST handling. -->
+    <parameter name="disableSeparateEndpointForREST" locked="true">false</parameter>
+
+    <!-- If you have a frontend host which exposes this webservice using a different public URL  -->
+    <!-- use this parameter to override autodetected url -->
+    <!--<parameter name="httpFrontendHostUrl" locked="false">https://someotherhost/context</parameter>-->
+
+
+    <!--    The way of adding listener to the system-->
+    <!--    <listener class="org.apache.axis2.ObserverIMPL">-->
+    <!--        <parameter name="RSS_URL" locked="false">http://127.0.0.1/rss</parameter>-->
+    <!--    </listener>-->
+
+    <!-- ================================================= -->
+    <!-- Message Receivers -->
+    <!-- ================================================= -->
+    <!--This is the Deafult Message Receiver for the system , if you want to have MessageReceivers for -->
+    <!--all the other MEP implement it and add the correct entry to here , so that you can refer from-->
+    <!--any operation -->
+    <!--Note : You can ovride this for particular service by adding the same element with your requirement-->
+    <messageReceivers>
+        <messageReceiver mep="http://www.w3.org/2004/08/wsdl/in-only"
+                         class="org.apache.axis2.receivers.RawXMLINOnlyMessageReceiver"/>
+        <messageReceiver mep="http://www.w3.org/2004/08/wsdl/in-out"
+                         class="org.apache.axis2.receivers.RawXMLINOutMessageReceiver"/>
+    </messageReceivers>
+    <!-- ================================================= -->
+    <!-- Transport Ins -->
+    <!-- ================================================= -->
+    <transportReceiver name="http"
+                       class="org.apache.axis2.transport.http.SimpleHTTPServer">
+        <parameter name="port" locked="false">6060</parameter>
+        <!-- Here is the complete list of supported parameters (see example settings further below):
+            port: the port to listen on (default 6060)
+            hostname:  if non-null, url prefix used in reply-to endpoint references                                 (default null)
+            originServer:  value of http Server header in outgoing messages                                         (default "Simple-Server/1.1")
+            requestTimeout:  value in millis of time that requests can wait for data                                (default 20000)
+            requestTcpNoDelay:  true to maximize performance and minimize latency                                   (default true)
+                                false to minimize bandwidth consumption by combining segments
+            requestCoreThreadPoolSize:  number of threads available for request processing (unless queue fills up)  (default 25)
+            requestMaxThreadPoolSize:  number of threads available for request processing if queue fills us         (default 150)
+                                       note that default queue never fills up:  see HttpFactory
+            threadKeepAliveTime:  time to keep threads in excess of core size alive while inactive                  (default 180)
+                                  note that no such threads can exist with default unbounded request queue
+            threadKeepAliveTimeUnit:  TimeUnit of value in threadKeepAliveTime (default SECONDS)                    (default SECONDS)
+        -->
+        <!-- <parameter name="hostname"                  locked="false">http://www.myApp.com/ws</parameter> -->
+        <!-- <parameter name="originServer"              locked="false">My-Server/1.1</parameter>           -->
+        <!-- <parameter name="requestTimeout"            locked="false">10000</parameter>                   -->
+        <!-- <parameter name="requestTcpNoDelay"         locked="false">false</parameter>                   -->
+        <!-- <parameter name="requestCoreThreadPoolSize" locked="false">50</parameter>                      -->
+        <!-- <parameter name="RequestMaxThreadPoolSize"  locked="false">100</parameter>                     -->
+        <!-- <parameter name="threadKeepAliveTime"       locked="false">240000</parameter>                  -->
+        <!-- <parameter name="threadKeepAliveTimeUnit"   locked="false">MILLISECONDS</parameter>            -->
+    </transportReceiver>
+    
+    <!--Uncomment this and configure as appropriate for JMS transport support, after setting up your JMS environment (e.g. ActiveMQ)
+    <transportReceiver name="jms" class="org.apache.axis2.transport.jms.JMSListener">
+        <parameter name="myTopicConnectionFactory" locked="false">        	        	
+        	<parameter name="java.naming.factory.initial" locked="false">org.apache.activemq.jndi.ActiveMQInitialContextFactory</parameter>
+        	<parameter name="java.naming.provider.url" locked="false">tcp://localhost:61616</parameter>        	
+        	<parameter name="transport.jms.ConnectionFactoryJNDIName" locked="false">TopicConnectionFactory</parameter>
+        </parameter>
+
+        <parameter name="myQueueConnectionFactory" locked="false">        	        	
+        	<parameter name="java.naming.factory.initial" locked="false">org.apache.activemq.jndi.ActiveMQInitialContextFactory</parameter>
+        	<parameter name="java.naming.provider.url" locked="false">tcp://localhost:61616</parameter>        	
+        	<parameter name="transport.jms.ConnectionFactoryJNDIName" locked="false">QueueConnectionFactory</parameter>
+        </parameter>
+
+        <parameter name="default" locked="false">        	        	
+        	<parameter name="java.naming.factory.initial" locked="false">org.apache.activemq.jndi.ActiveMQInitialContextFactory</parameter>
+        	<parameter name="java.naming.provider.url" locked="false">tcp://localhost:61616</parameter>        	
+        	<parameter name="transport.jms.ConnectionFactoryJNDIName" locked="false">QueueConnectionFactory</parameter>
+        </parameter>
+    </transportReceiver>-->
+
+    <!--Uncomment if you want to have SMTP transport support-->
+    <!--<transportReceiver name="mail" class="org.apache.axis2.transport.mail.SimpleMailListener">-->
+    <!--<parameter name="transport.mail.pop3.host" locked="false">127.0.0.1</parameter>-->
+    <!--<parameter name="transport.mail.pop3.user" locked="false">axis2</parameter>-->
+    <!--<parameter name="transport.mail.pop3.password" locked="false">axis2</parameter>-->
+    <!--<parameter name="transport.mail.pop3.port" locked="false">110</parameter>-->
+    <!--<parameter name="transport.mail.replyToAddress" locked="false">axis2@127.0.0.1</parameter>-->
+    <!--</transportReceiver>-->
+
+    <transportReceiver name="tcp"
+                       class="org.apache.axis2.transport.tcp.TCPServer">
+        <parameter name="port" locked="false">6060</parameter>
+        <!--If you want to give your own host address for EPR generation-->
+        <!--uncommet following paramter , and set as you required.-->
+        <!--<parameter name="hostname" locked="false">tcp://myApp.com/ws</parameter>-->
+    </transportReceiver>
+
+    <!-- ================================================= -->
+    <!-- Transport Outs -->
+    <!-- ================================================= -->
+
+    <transportSender name="tcp"
+                     class="org.apache.axis2.transport.tcp.TCPTransportSender"/>
+    <transportSender name="local"
+                     class="org.apache.axis2.transport.local.LocalTransportSender"/>
+    <transportSender name="http"
+                     class="org.apache.axis2.transport.http.CommonsHTTPTransportSender">
+        <parameter name="PROTOCOL" locked="false">HTTP/1.1</parameter>
+        <parameter name="Transfer-Encoding" locked="false">chunked</parameter>
+    </transportSender>
+    <transportSender name="https"
+                     class="org.apache.axis2.transport.http.CommonsHTTPTransportSender">
+        <parameter name="PROTOCOL" locked="false">HTTP/1.1</parameter>
+        <parameter name="Transfer-Encoding" locked="false">chunked</parameter>
+    </transportSender>
+    <transportSender name="jms"
+                     class="org.apache.axis2.transport.jms.JMSSender"/>
+
+    <!-- Uncomment this one with the appropriate papameters to enable the SMTP transport Receiver
+   <transportSender name="mailto" class="org.apache.axis2.transport.mail.MailTransportSender">
+       <parameter name="transport.mail.smtp.host" locked="false">127.0.0.1</parameter>
+       <parameter name="transport.mail.smtp.user" locked="false">axis2</parameter>
+       <parameter name="transport.mail.smtp.password" locked="false">axis2</parameter>
+       <parameter name="transport.mail.smtp.port" locked="false">25</parameter>
+   </transportSender>
+   -->
+
+    <!-- ================================================= -->
+    <!-- StAX filter                                       -->
+    <!-- ================================================= -->
+    <filterOrder>
+        <filter name="TestFilter1" class="org.apache.axis2.xmlfilter.TestFilter1" />
+        <filter name="TestFilter2" class="org.apache.axis2.xmlfilter.TestFilter2">
+            <parameter name="default_length" locked="false">1024</parameter>
+        </filter>
+    </filterOrder>
+
+    <!-- ================================================= -->
+    <!-- Global Modules  -->
+    <!-- ================================================= -->
+    <!-- Comment this to disable Addressing -->
+    <!-- <module ref="addressing"/> -->
+
+    <!--Configuring module , providing parameters for modules whether they refer or not-->
+    <!--<moduleConfig name="addressing">-->
+    <!--<parameter name="addressingPara" locked="false">N/A</parameter>-->
+    <!--</moduleConfig>-->
+
+    <!-- ================================================= -->
+    <!-- Phases  -->
+    <!-- ================================================= -->
+    <phaseOrder type="InFlow">
+        <!--  System pre defined phases       -->
+        <phase name="Transport">
+            <handler name="RequestURIBasedDispatcher"
+                     class="org.apache.axis2.engine.RequestURIBasedDispatcher">
+                <order phase="Transport"/>
+            </handler>
+            <handler name="SOAPActionBasedDispatcher"
+                     class="org.apache.axis2.engine.SOAPActionBasedDispatcher">
+                <order phase="Transport"/>
+            </handler>
+        </phase>
+        <phase name="Security"/>
+        <phase name="PreDispatch"/>
+        <phase name="Dispatch" class="org.apache.axis2.engine.DispatchPhase">
+            <handler name="AddressingBasedDispatcher"
+                     class="org.apache.axis2.engine.AddressingBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+
+            <handler name="SOAPMessageBodyBasedDispatcher"
+                     class="org.apache.axis2.engine.SOAPMessageBodyBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+            <handler name="InstanceDispatcher"
+                     class="org.apache.axis2.engine.InstanceDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+        </phase>
+        <!--  System pre defined phases       -->
+        <!--   After Postdispatch phase module author or or service author can add any phase he want      -->
+        <phase name="OperationInPhase"/>
+    </phaseOrder>
+    <phaseOrder type="OutFlow">
+        <!--      user can add his own phases to this area  -->
+        <phase name="OperationOutPhase"/>
+        <!--system predefined phase-->
+        <!--these phase will run irrespective of the service-->
+        <phase name="PolicyDetermination"/>
+        <phase name="MessageOut"/>
+        <phase name="Security"/>
+    </phaseOrder>
+    <phaseOrder type="InFaultFlow">
+        <phase name="PreDispatch"/>
+        <phase name="Dispatch" class="org.apache.axis2.engine.DispatchPhase">
+            <handler name="RequestURIBasedDispatcher"
+                     class="org.apache.axis2.engine.RequestURIBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+
+            <handler name="SOAPActionBasedDispatcher"
+                     class="org.apache.axis2.engine.SOAPActionBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+
+            <handler name="AddressingBasedDispatcher"
+                     class="org.apache.axis2.engine.AddressingBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+
+            <handler name="SOAPMessageBodyBasedDispatcher"
+                     class="org.apache.axis2.engine.SOAPMessageBodyBasedDispatcher">
+                <order phase="Dispatch"/>
+            </handler>
+       