Index: C:/wasx/axis2/modules/jaxws/src/javax/jws/WebMethod.java
===================================================================
--- C:/wasx/axis2/modules/jaxws/src/javax/jws/WebMethod.java	(revision 443462)
+++ C:/wasx/axis2/modules/jaxws/src/javax/jws/WebMethod.java	(working copy)
@@ -13,5 +13,5 @@
 public @interface WebMethod {
 	String operationName() default "";
 	String action() default "";
-	String exclude() default "";
+	boolean exclude() default false;
 }
