Index: ConfigurationContext.java
===================================================================
--- ConfigurationContext.java	(revision 608741)
+++ ConfigurationContext.java	(working copy)
@@ -877,4 +877,13 @@
     public ConfigurationContext getRootContext() {
         return this;
     }
+
+    public boolean isAnyOperationContextRegistered(){
+        if(operationContextMap.isEmpty()){
+            return false;
+        }
+        else{
+            return true;
+        }
+    }
 }
\ No newline at end of file
