Index: C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java
===================================================================
--- C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java	(revision 469549)
+++ C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/AxisConfigBuilder.java	(working copy)
@@ -38,7 +38,12 @@
 
 import javax.xml.namespace.QName;
 import javax.xml.stream.XMLStreamException;
+
 import java.io.InputStream;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Iterator;
@@ -329,9 +334,17 @@
                 OMAttribute trsClas = transport.getAttribute(new QName(TAG_CLASS_NAME));
                 if (trsClas != null) {
                     try {
-                        String clasName = trsClas.getAttributeValue();
+                        final String clasName = trsClas.getAttributeValue();
                         Class receiverClass;
-                        receiverClass = Loader.loadClass(clasName); 
+                        try {
+                             receiverClass = (Class) org.apache.axis2.java.security.AccessController.doPrivileged(new PrivilegedExceptionAction() {
+                        		public Object run() throws ClassNotFoundException {
+                        			return Loader.loadClass(clasName);	
+                        		}
+                        	});  
+                        } catch (PrivilegedActionException e) {
+                            throw (ClassNotFoundException)e.getException();
+                        }
 
                         TransportListener receiver =
                                 (TransportListener) receiverClass.newInstance();
Index: C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/DescriptionBuilder.java
===================================================================
--- C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/DescriptionBuilder.java	(revision 469549)
+++ C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/deployment/DescriptionBuilder.java	(working copy)
@@ -46,6 +46,10 @@
 import javax.xml.stream.XMLStreamException;
 import javax.xml.stream.XMLStreamReader;
 import java.io.InputStream;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Iterator;
@@ -127,8 +131,17 @@
                 TAG_MESSAGE_RECEIVER));
         while (msgReceivers.hasNext()) {
             OMElement msgReceiver = (OMElement) msgReceivers.next();
-            MessageReceiver receiver = loadMessageReceiver(Thread
-                    .currentThread().getContextClassLoader(), msgReceiver);
+            final OMElement tempMsgReceiver = msgReceiver;
+            MessageReceiver receiver = null;
+            try { 
+                receiver = (MessageReceiver) org.apache.axis2.java.security.AccessController.doPrivileged(new PrivilegedExceptionAction() {
+                    public Object run() throws org.apache.axis2.deployment.DeploymentException {
+                        return loadMessageReceiver(Thread.currentThread().getContextClassLoader(), tempMsgReceiver);
+                    }
+                });
+            } catch (PrivilegedActionException e) {
+                throw (DeploymentException)e.getException();
+            }              
             OMAttribute mepAtt = msgReceiver.getAttribute(new QName(TAG_MEP));
             mr_mep.put(mepAtt.getAttributeValue(), receiver);
         }
Index: C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/description/AxisService.java
===================================================================
--- C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/description/AxisService.java	(revision 469549)
+++ C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/description/AxisService.java	(working copy)
@@ -62,6 +62,8 @@
 import java.io.OutputStream;
 import java.net.SocketException;
 import java.net.URL;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.*;
 
 /**
@@ -226,7 +228,11 @@
         moduleRefs = new ArrayList();
         engagedModules = new ArrayList();
         schemaList = new ArrayList();
-        serviceClassLoader = Thread.currentThread().getContextClassLoader();
+        serviceClassLoader = (ClassLoader) org.apache.axis2.java.security.AccessController.doPrivileged(new PrivilegedAction() {
+            public Object run() {
+                return Thread.currentThread().getContextClassLoader();
+            }
+        });
     }
 
     /**
Index: C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java
===================================================================
--- C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java	(revision 469549)
+++ C:/eclipse_workspace/axis2_os_1031/axis2/modules/kernel/src/org/apache/axis2/engine/AxisConfiguration.java	(working copy)
@@ -30,6 +30,8 @@
 
 import javax.xml.namespace.QName;
 import java.net.URL;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.*;
 
 /**
@@ -127,9 +129,13 @@
         faultyModules = new Hashtable();
         observersList = new ArrayList();
         inPhasesUptoAndIncludingPostDispatch = new ArrayList();
-        systemClassLoader = Thread.currentThread().getContextClassLoader();
-        serviceClassLoader = Thread.currentThread().getContextClassLoader();
-        moduleClassLoader = Thread.currentThread().getContextClassLoader();
+        systemClassLoader = (ClassLoader) org.apache.axis2.java.security.AccessController.doPrivileged(new PrivilegedAction() {
+            public Object run() {
+                return Thread.currentThread().getContextClassLoader();      
+            }
+        });
+        serviceClassLoader = systemClassLoader; 
+        moduleClassLoader = systemClassLoader;
         this.phasesinfo = new PhasesInfo();
         targetResolvers = new ArrayList();
     }
Index: C:/eclipse_workspace/axis2_os_1031/axis2/modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/ProviderDispatcher.java
===================================================================
--- C:/eclipse_workspace/axis2_os_1031/axis2/modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/ProviderDispatcher.java	(revision 469549)
+++ C:/eclipse_workspace/axis2_os_1031/axis2/modules/jaxws/src/org/apache/axis2/jaxws/server/dispatcher/ProviderDispatcher.java	(working copy)
@@ -19,6 +19,8 @@
 
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 
 import javax.activation.DataSource;
 import javax.xml.bind.JAXBContext;
@@ -150,7 +152,11 @@
         // Invoke the actual Provider.invoke() method
         Object responseParamValue = null;
         try {
-            responseParamValue = providerInstance.invoke(input);
+            responseParamValue = (Object) org.apache.axis2.java.security.AccessController.doPrivileged(new PrivilegedAction() {
+            	public Object run() {
+            		return providerInstance.invoke(input);
+            	}
+            });
         } catch (Exception e) {
             e.printStackTrace();
             throw ExceptionFactory.makeWebServiceException(e);
