Index: DeploymentEngine.java
===================================================================
--- DeploymentEngine.java	(Revision 389424)
+++ DeploymentEngine.java	(Arbeitskopie)
@@ -301,18 +301,32 @@
         }
     }
 
+    private boolean isSourceControlDir(File file) {
+        if (file.isDirectory()) {
+            String name = file.getName();
+            if (name.equalsIgnoreCase("CVS") || name.equalsIgnoreCase(".svn")) {
+                return true;
+            }
+        }
+        return false;        
+    }
+    
     /**
      * @param file
      */
     public void addWSToDeploy(ArchiveFileData file) {
-        wsToDeploy.add(file);
+        if (!isSourceControlDir(file.getFile())) {
+            wsToDeploy.add(file);
+        }
     }
 
     /**
      * @param file
      */
     public void addWSToUndeploy(WSInfo file) {
-        wsToUnDeploy.add(file);
+        if (!isSourceControlDir(new File(file.getFileName()))) {
+            wsToUnDeploy.add(file);
+        }
     }
 
     /**
