Index: modules/samples/json/src/sample/json/client/JsonClient.java
===================================================================
--- modules/samples/json/src/sample/json/client/JsonClient.java	(revision 1358584)
+++ modules/samples/json/src/sample/json/client/JsonClient.java	(working copy)
@@ -40,7 +40,7 @@
             "\"My_Street\",\"building\":\"My_Building\",\"flat\":\"My_Flat\",\"zipCode\":\"My_ZipCode\"}}}}";
 
         JsonClient jsonClient = new JsonClient();
-        jsonClient.post(sampleString);
+        jsonClient.post(echoUser);
     }
 
     public boolean post(String message) throws UnsupportedEncodingException {
