Index: modules/kernel/src/org/apache/axis2/description/WSDL11ToAllAxisServicesBuilder.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/WSDL11ToAllAxisServicesBuilder.java	(revision 432613)
+++ modules/kernel/src/org/apache/axis2/description/WSDL11ToAllAxisServicesBuilder.java	(working copy)
@@ -40,6 +40,12 @@
  * Extends the WSDL11ToAxisServiceBuilder class to provide functionality to return
  * multiple AxisService objects; one for each port on each service in the WSDL 1.1 file.
  *
+ * DISCLAIMER: This is intended as a TEMPORARY solution ONLY.   
+ * A more permanent solution will involve changes to the description hierarchy to handle 
+ * services with multiple ports/endpoints.  When the permanent
+ * solution is available this class may no longer be supported.
+ *
+
  */
 public class WSDL11ToAllAxisServicesBuilder extends WSDL11ToAxisServiceBuilder {
     protected static final Log log =
Index: modules/kernel/src/org/apache/axis2/description/WSDL20ToAllAxisServicesBuilder.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/WSDL20ToAllAxisServicesBuilder.java	(revision 432613)
+++ modules/kernel/src/org/apache/axis2/description/WSDL20ToAllAxisServicesBuilder.java	(working copy)
@@ -39,6 +39,11 @@
  * Extends the WSDL20ToAxisServiceBuilder class to provide functionality to return
  * multiple AxisService objects; one for each endpoint on each service in the WSDL 2.0 file.
  *
+ * DISCLAIMER: This is intended as a TEMPORARY solution ONLY.   
+ * A more permanent solution will involve changes to the description hierarchy to handle 
+ * services with multiple ports/endpoints.  When the permanent
+ * solution is available this class may no longer be supported.
+
  */
 public class WSDL20ToAllAxisServicesBuilder extends WSDL20ToAxisServiceBuilder {
     protected static final Log log =
