Index: modules/core/src/org/apache/axis2/transport/jms/JMSSender.java
===================================================================
--- modules/core/src/org/apache/axis2/transport/jms/JMSSender.java	(revision 420918)
+++ modules/core/src/org/apache/axis2/transport/jms/JMSSender.java	(working copy)
@@ -210,6 +210,11 @@
             !AddressingConstants.Final.WSA_ANONYMOUS_URL
                 .equals(msgContext.getTo().getAddress())) {
             endpointAddress = msgContext.getTo().getAddress();
+            
+            if (endpointAddress.equals(AddressingConstants.Final.WSA_NONE_URI)) {
+                //Don't send the message.
+                return;
+            }
         }
         boolean waitForResponse = false;
 
