Index: D:/socWorkspaces/ApacheAxis2/axis2/src/org/apache/axis2/jaxws/marshaller/impl/alt/DocLitBareMethodMarshaller.java
===================================================================
--- D:/socWorkspaces/ApacheAxis2/axis2/src/org/apache/axis2/jaxws/marshaller/impl/alt/DocLitBareMethodMarshaller.java	(revision 652729)
+++ D:/socWorkspaces/ApacheAxis2/axis2/src/org/apache/axis2/jaxws/marshaller/impl/alt/DocLitBareMethodMarshaller.java	(working copy)
@@ -81,6 +81,12 @@
                     MethodMarshallerUtils.getMarshalDesc(endpointDesc);
             TreeSet<String> packages = marshalDesc.getPackages();
 
+            // Remember this unmarshal information so that we can speed up processing
+            // the next time.
+            MethodMarshallerUtils.registerUnmarshalInfo(message.getMessageContext(),
+                                                        packages,
+                                                        marshalDesc.getPackagesKey());
+
             // Get the return value.
             Class returnType = operationDesc.getResultActualType();
             Object returnValue = null;
@@ -162,6 +168,12 @@
                     MethodMarshallerUtils.getMarshalDesc(endpointDesc);
             TreeSet<String> packages = marshalDesc.getPackages();
 
+            // Remember this unmarshal information so that we can speed up processing
+            // the next time.
+            MethodMarshallerUtils.registerUnmarshalInfo(message.getMessageContext(),
+                                                        packages,
+                                                        marshalDesc.getPackagesKey());
+
             // Unmarshal the ParamValues from the message
             List<PDElement> pvList = MethodMarshallerUtils.getPDElements(pds,
                                                                          message,
