Index: modules/transport/http/src/org/apache/axis2/transport/http/HTTPTransportReceiver.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/HTTPTransportReceiver.java	(revision 1167006)
+++ modules/transport/http/src/org/apache/axis2/transport/http/HTTPTransportReceiver.java	(working copy)
@@ -26,12 +26,7 @@
 import org.apache.axis2.description.AxisService;
 import org.apache.axis2.engine.AxisConfiguration;
 
-import java.util.Collection;
-import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.Hashtable;
-import java.util.Iterator;
-import java.util.Map;
+import java.util.*;
 
 /**
  * Class HTTPTransportReceiver
@@ -97,9 +92,13 @@
 
         if ((services != null) && !services.isEmpty()) {
             status = true;
+            Collection serviceCollection = new ArrayList();
+            // sort services by alphabetical order to show in services page
+            TreeSet<String> allServices = new TreeSet<String>(services.keySet());
+            for(String serviceName : allServices) {
+                serviceCollection.add(services.get(serviceName));
+            }
 
-            Collection serviceCollection = services.values();
-
             temp += "<h2>" + "Deployed services" + "</h2>";
 
             for (Iterator it = serviceCollection.iterator(); it.hasNext();) {
