Index: C:/development/kingfisher/workspace/blank-project-kf/config/XmlbeansDatabindingTemplate.xsl
===================================================================
--- C:/development/kingfisher/workspace/blank-project-kf/config/XmlbeansDatabindingTemplate.xsl	(revision 5398)
+++ C:/development/kingfisher/workspace/blank-project-kf/config/XmlbeansDatabindingTemplate.xsl	(working copy)
@@ -41,7 +41,7 @@
             private  org.apache.axiom.om.OMElement  toOM(<xsl:value-of select="@type"/> param, boolean optimizeContent)
             throws org.apache.axis2.AxisFault{
             org.apache.axiom.om.impl.builder.StAXOMBuilder builder = new org.apache.axiom.om.impl.builder.StAXOMBuilder
-            (org.apache.axiom.om.OMAbstractFactory.getOMFactory(),new org.apache.axis2.util.StreamWrapper(param.newXMLStreamReader())) ;
+            (org.apache.axiom.om.OMAbstractFactory.getOMFactory(),new org.apache.axis2.util.StreamWrapper(param.newXMLStreamReader(new org.apache.xmlbeans.XmlOptions().setSaveAggressiveNamespaces()))) ;
             org.apache.axiom.om.OMElement documentElement = builder.getDocumentElement();
 
             <xsl:if test="$base64">
