Index: src/main/java/org/apache/axis2/transport/jms/JMSSender.java
===================================================================
--- src/main/java/org/apache/axis2/transport/jms/JMSSender.java	(revision 89984)
+++ src/main/java/org/apache/axis2/transport/jms/JMSSender.java	(working copy)
@@ -88,7 +88,11 @@
             if (jmsConnectionFactoryName != null) {
                 return connFacManager.getJMSConnectionFactory(jmsConnectionFactoryName);
             } else {
-                return connFacManager.getJMSConnectionFactory(props);
+                JMSConnectionFactory fac = connFacManager.getJMSConnectionFactory(props);
+                if (fac == null) {
+                    fac = connFacManager.getJMSConnectionFactory(JMSConstants.DEFAULT_CONFAC_NAME);
+                }
+                return  fac;
             }
         } else {
             return null;
