Index: src/org/apache/axis2/transport/http/ListingAgent.java
===================================================================
--- src/org/apache/axis2/transport/http/ListingAgent.java	(revision 540688)
+++ src/org/apache/axis2/transport/http/ListingAgent.java	(working copy)
@@ -192,8 +192,16 @@
             throws IOException, ServletException {
 
         String filePart = req.getRequestURL().toString();
-        String serviceName = filePart.substring(filePart.lastIndexOf("/") + 1,
-                                                filePart.length());
+
+        // strip first two path components of the uri to obtain the service name
+        // e.g.   GET /axis2/services/path/to/service  ->  path/to/service
+        String uri = req.getRequestURI();
+        for (int i=0; i<3; i++) {
+            int pos = uri.indexOf("/");
+            if (pos >=0 && pos < uri.length()-1) uri = uri.substring(pos+1); else break;
+        }
+        serviceName = uri;
+
         HashMap services = configContext.getAxisConfiguration().getServices();
         String query = req.getQueryString();
         int wsdl2 = query.indexOf("wsdl2");
