Index: PolicyEngine.java
===================================================================
--- PolicyEngine.java	(revision 542790)
+++ PolicyEngine.java	(working copy)
@@ -21,6 +21,8 @@
 import org.apache.axiom.om.OMNamespace;
 import org.apache.axiom.om.impl.llom.factory.OMXMLBuilderFactory;
 import org.apache.neethi.builders.AssertionBuilder;
+import org.apache.neethi.util.PolicyReaderThread;
+
 import javax.xml.namespace.QName;
 import javax.xml.stream.XMLInputFactory;
 import java.io.InputStream;
@@ -146,7 +148,7 @@
         return reference;
     }
 
-    private static Policy getPolicyOperator(OMElement element) {
+    public static Policy getPolicyOperator(OMElement element) {
         return (Policy) processOperationElement(element, new Policy());
     }
 
@@ -213,8 +215,14 @@
 
                 } else if (Constants.ELEM_POLICY_REF.equals(childElement
                         .getLocalName())) {
-                    operator
-                            .addPolicyComponent(getPolicyReference(childElement));
+                    
+                	//extract the policy reference object
+                	PolicyReference pReference = getPolicyReference(childElement);
+  	  
+                	//pull the policy content in a thred so no blocking occurs
+                	PolicyReaderThread thread = new PolicyReaderThread(pReference,operator);
+                	thread.start();
+                	
                 }
 
             } else {
