Index: src/org/apache/axis2/policy/model/MTOM10Assertion.java
===================================================================
--- src/org/apache/axis2/policy/model/MTOM10Assertion.java	(revision 703931)
+++ src/org/apache/axis2/policy/model/MTOM10Assertion.java	(working copy)
@@ -1,5 +1,7 @@
 package org.apache.axis2.policy.model;
 
+import java.util.logging.Logger;
+
 import javax.xml.namespace.QName;
 import javax.xml.stream.XMLStreamException;
 import javax.xml.stream.XMLStreamWriter;
@@ -10,6 +12,8 @@
 /** Assertion to pick up the QName <wsoma:OptimizedMimeSerialization xmlns:wsoma="http://schemas.xmlsoap.org/ws/2004/09/policy/optimizedmimeserialization"/> */
 public class MTOM10Assertion extends MTOMAssertion {
 
+	public static final Logger log = Logger.getLogger(MTOM10Assertion.class.toString());
+	
     public final static String NS = "http://schemas.xmlsoap.org/ws/2004/09/policy/optimizedmimeserialization";
 
     public final static String MTOM_SERIALIZATION_CONFIG_LN = "OptimizedMimeSerialization";
@@ -39,7 +43,7 @@
         writer.writeStartElement(PREFIX, MTOM_SERIALIZATION_CONFIG_LN, NS);
 
         if (optional)
-            writer.writeAttribute("Optional", "true");
+            writer.writeAttribute(Constants.ATTR_WSP, null, "Optional", "true");
 
         writer.writeNamespace(PREFIX, NS);
         writer.writeEndElement();
Index: src/org/apache/axis2/policy/model/MTOM11Assertion.java
===================================================================
--- src/org/apache/axis2/policy/model/MTOM11Assertion.java	(revision 703931)
+++ src/org/apache/axis2/policy/model/MTOM11Assertion.java	(working copy)
@@ -1,5 +1,7 @@
 package org.apache.axis2.policy.model;
 
+import java.util.logging.Logger;
+
 import javax.xml.namespace.QName;
 import javax.xml.stream.XMLStreamException;
 import javax.xml.stream.XMLStreamWriter;
@@ -8,6 +10,8 @@
 import org.apache.neethi.PolicyComponent;
 
 public class MTOM11Assertion extends MTOMAssertion {
+	
+	public static final Logger log = Logger.getLogger(MTOM11Assertion.class.toString());
     
     public final static String NS = "http://www.w3.org/2007/08/soap12-mtom-policy";
 
@@ -34,7 +38,7 @@
         writer.writeStartElement(PREFIX, MTOM_LN, NS);
 
         if (optional)
-            writer.writeAttribute("Optional", "true");
+            writer.writeAttribute(Constants.ATTR_WSP, null, "Optional", "true");
 
         writer.writeNamespace(PREFIX, NS);
         writer.writeEndElement();
Index: src/org/apache/axis2/policy/model/MTOMAssertion.java
===================================================================
--- src/org/apache/axis2/policy/model/MTOMAssertion.java	(revision 703931)
+++ src/org/apache/axis2/policy/model/MTOMAssertion.java	(working copy)
@@ -19,10 +19,14 @@
 
 package org.apache.axis2.policy.model;
 
+import java.util.logging.Logger;
+
 import org.apache.neethi.Assertion;
 
 public abstract class MTOMAssertion implements Assertion {
 
+	public static Logger log = Logger.getLogger(MTOMAssertion.class.toString());
+	
     protected boolean optional = false;
     
     public boolean isOptional() {
