Index: modules/transport/http-hc3/src/main/java/org/apache/axis2/transport/http/impl/httpclient3/HTTPProxyConfigurator.java
===================================================================
--- modules/transport/http-hc3/src/main/java/org/apache/axis2/transport/http/impl/httpclient3/HTTPProxyConfigurator.java	(revision 1791784)
+++ modules/transport/http-hc3/src/main/java/org/apache/axis2/transport/http/impl/httpclient3/HTTPProxyConfigurator.java	(working copy)
@@ -135,7 +135,7 @@
         }
 
         String port = System.getProperty(HTTPTransportConstants.HTTP_PROXY_PORT);
-        if (port != null) {
+        if (port != null && !port.isEmpty()) {
             proxyPort = Integer.parseInt(port);
         }
 
Index: modules/transport/http/src/org/apache/axis2/transport/http/impl/httpclient4/HTTPProxyConfigurator.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/impl/httpclient4/HTTPProxyConfigurator.java	(revision 1791784)
+++ modules/transport/http/src/org/apache/axis2/transport/http/impl/httpclient4/HTTPProxyConfigurator.java	(working copy)
@@ -141,7 +141,7 @@
         }
 
         String port = System.getProperty(HTTPTransportConstants.HTTP_PROXY_PORT);
-        if (port != null) {
+		if (port != null && !port.isEmpty()) {
             proxyPort = Integer.parseInt(port);
         }
 
