Index: src/org/apache/axis2/description/AxisService.java
===================================================================
--- src/org/apache/axis2/description/AxisService.java	(revision 544863)
+++ src/org/apache/axis2/description/AxisService.java	(working copy)
@@ -900,12 +900,16 @@
     }
 
     private void printUserWSDL(OutputStream out) throws AxisFault {
+        printUserWSDL(out, null);
+    }
+
+    private void printUserWSDL(OutputStream out, String requestIP) throws AxisFault {
         Parameter wsld4jdefinition = getParameter(WSDLConstants.WSDL_4_J_DEFINITION);
         if (wsld4jdefinition != null) {
             try {
                 Definition definition = (Definition) wsld4jdefinition.getValue();
                 if (isModifyUserWSDLPortAddress()) {
-                    setPortAddress(definition);
+                    setPortAddress(definition, requestIP);
                 }
                 WSDLWriter writer = WSDLFactory.newInstance().newWSDLWriter();
                 writer.writeWSDL(definition, out);
@@ -924,7 +928,7 @@
      */
     public void printWSDL(OutputStream out, String requestIP) throws AxisFault {
         if (isUseUserWSDL()) {
-            printUserWSDL(out);
+            printUserWSDL(out, requestIP);
         } else {
             String[] eprArray = getEPRs(requestIP);
             getWSDL(out, eprArray);
