Index: C:/eclipse/workspace_axis2/axis2_latest/modules/core/src/org/apache/axis2/transport/http/util/RESTUtil.java
===================================================================
--- C:/eclipse/workspace_axis2/axis2_latest/modules/core/src/org/apache/axis2/transport/http/util/RESTUtil.java	(revision 411211)
+++ C:/eclipse/workspace_axis2/axis2_latest/modules/core/src/org/apache/axis2/transport/http/util/RESTUtil.java	(working copy)
@@ -181,11 +181,8 @@
                 // If charset is not specified
                 XMLStreamReader xmlreader;
                 if (TransportUtils.getCharSetEncoding(contentType) == null) {
-                    xmlreader = XMLInputFactory.
-                            newInstance().
-                            createXMLStreamReader(inputStream,
-                                    MessageContext.DEFAULT_CHAR_SET_ENCODING);
-
+                	xmlreader = StAXUtils.createXMLStreamReader(inputStream, MessageContext.DEFAULT_CHAR_SET_ENCODING);
+                	
                     // Set the encoding scheme in the message context
                     msgCtxt.setProperty(MessageContext.CHARACTER_SET_ENCODING,
                             MessageContext.DEFAULT_CHAR_SET_ENCODING);
