Index: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageReceiver.java
===================================================================
--- modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageReceiver.java	(revision 1145504)
+++ modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageReceiver.java	(working copy)
@@ -156,7 +156,12 @@
         // set the JMS Message ID as the Message ID of the MessageContext
         try {
             msgContext.setMessageID(message.getJMSMessageID());
-            msgContext.setProperty(JMSConstants.JMS_COORELATION_ID, message.getJMSMessageID());
+            String jmsCorrelationID = message.getJMSCorrelationID();
+            if (jmsCorrelationID != null && jmsCorrelationID.length() > 0) {
+                msgContext.setProperty(JMSConstants.JMS_COORELATION_ID, jmsCorrelationID);
+            } else {
+                msgContext.setProperty(JMSConstants.JMS_COORELATION_ID, message.getJMSMessageID());
+            }
         } catch (JMSException ignore) {}
 
         String soapAction = JMSUtils.getProperty(message, BaseConstants.SOAPACTION);
Index: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java
===================================================================
--- modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java	(revision 1145504)
+++ modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java	(working copy)
@@ -226,7 +226,12 @@
             } catch (JMSException ignore) {}
 
             try {
-                correlationId = message.getJMSMessageID();
+                String jmsCorrelationID = message.getJMSCorrelationID();
+                if (jmsCorrelationID != null && jmsCorrelationID.length() > 0) {
+                    correlationId = jmsCorrelationID;
+                } else {
+                    correlationId = message.getJMSMessageID();
+                }
             } catch(JMSException ignore) {}
 
             // We assume here that the response uses the same message property to
