Index: modules/kernel/src/org/apache/axis2/addressing/AddressingConstants.java
===================================================================
--- modules/kernel/src/org/apache/axis2/addressing/AddressingConstants.java	(revision 702134)
+++ modules/kernel/src/org/apache/axis2/addressing/AddressingConstants.java	(working copy)
@@ -48,6 +48,7 @@
     static final String DISABLE_ADDRESSING_FOR_IN_MESSAGES = "disableAddressingForInMessages";
     static final String ADDR_VALIDATE_ACTION = "addressing.validateAction";
     static final String ADDR_VALIDATE_INVOCATION_PATTERN = "addressing.validateInvocationPattern";
+    static final String ENDPOINT_REFERENCE = "EndpointReference";
 
     // ====================== WSDL Binding Constants ========================
     static final String USING_ADDRESSING = "UsingAddressing";
@@ -132,6 +133,26 @@
     static final String FAULT_ADDRESSING_DESTINATION_UNREACHABLE =
             "DestinationUnreachable";
     
+    // ==================== WS-AddressingAndIdentity ========================
+    
+    static final String ADDRESSING_IDENTITY_PREFIX = "wsid";
+    static final String ADDRESSING_IDENTITY_NS = 
+        "http://schemas.xmlsoap.org/ws/2006/02/addressingidentity";
+    static final String XML_SIG_PREFIX = "ds";
+    static final String XML_SIG_NS = "http://www.w3.org/2000/09/xmldsig#";
+    
+    static final String IDENTITY = "Identity";
+    static final String IDENTITY_KEY_INFO = "KeyInfo";
+    static final String IDENTITY_X509_DATA = "X509Data";
+    static final String IDENTITY_X509_CERT = "X509Certificate";
+    
+    static final String IDENTITY_PARAMETER = "ServiceIdentityData";
+    
+    static final QName QNAME_IDENTITY = new QName(ADDRESSING_IDENTITY_NS,IDENTITY,ADDRESSING_IDENTITY_PREFIX);
+    static final QName QNAME_IDENTITY_KEY_INFO = new QName(XML_SIG_NS,IDENTITY_KEY_INFO,XML_SIG_PREFIX);
+    static final QName QNAME_IDENTITY_X509_DATA = new QName(XML_SIG_NS,IDENTITY_X509_DATA,XML_SIG_PREFIX);
+    static final QName QNAME_IDENTITY_X509_CERT = new QName(XML_SIG_NS,IDENTITY_X509_CERT,XML_SIG_PREFIX);
+    
     interface Final {
 
         // ====================== Addressing 1.0 Final Version Constants ====================
@@ -191,7 +212,8 @@
         static final QName WSAW_USING_ADDRESSING =
                 new QName(WSAW_NAMESPACE, USING_ADDRESSING);
         static final QName WSAW_ANONYMOUS = new QName(WSAW_NAMESPACE, USING_ADDRESSING);
-        static final QName WSA_ENDPOINT_REFERENCE = new QName(WSA_NAMESPACE,"EndpointReference");
+        static final QName WSA_ENDPOINT_REFERENCE = new QName(WSA_NAMESPACE, ENDPOINT_REFERENCE, WSA_DEFAULT_PREFIX);
+        static final QName WSA_ADDRESS = new QName(WSA_NAMESPACE, EPR_ADDRESS,WSA_DEFAULT_PREFIX);
 
         final QName QNAME_WSA_TO = new QName(WSA_NAMESPACE, WSA_TO);
         final QName QNAME_WSA_FROM = new QName(WSA_NAMESPACE, WSA_FROM);
Index: modules/kernel/src/org/apache/axis2/description/AxisService2WSDL11.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/AxisService2WSDL11.java	(revision 702134)
+++ modules/kernel/src/org/apache/axis2/description/AxisService2WSDL11.java	(working copy)
@@ -655,6 +655,7 @@
 					WSDLSerializationUtil.addExtensionElement(fac, port,
 							SOAP_ADDRESS, LOCATION, (endpointURL == null) ? ""
 									: endpointURL, soap);
+					generateEPRElement(fac, port, endpointURL);
 					addPolicyAsExtElement(axisEndpoint, port);
 					generateSoap11Binding(fac, definition, axisEndpoint
 							.getBinding());
@@ -697,6 +698,7 @@
 					WSDLSerializationUtil.addExtensionElement(fac, port,
 							SOAP_ADDRESS, LOCATION, (endpointURL == null) ? ""
 									: endpointURL, soap12);
+					generateEPRElement(fac, port, endpointURL);
 					addPolicyAsExtElement(axisEndpoint, port);
 					generateSoap12Binding(fac, definition, axisEndpoint
 							.getBinding());
@@ -1270,4 +1272,60 @@
 		    return endpointURL;
 		}
 	}
+	
+	/**
+	 * Generate the Identity element according to the WS-AddressingAndIdentity if the 
+	 * AddressingConstants.IDENTITY_PARAMETER parameter is set. 
+	 * http://schemas.xmlsoap.org/ws/2006/02/addressingidentity/ 
+	 */
+	
+	private void generateIdentityElement(OMFactory fac,OMElement epr) {
+	    
+	    Parameter parameter = axisService.getParameter(AddressingConstants.IDENTITY_PARAMETER);
+	    
+	    // If the parameter is not set, return
+	    if (parameter == null) {
+	        return;
+	    }
+	    
+	    // Create the Identity element
+	    OMElement identity = fac.createOMElement(AddressingConstants.QNAME_IDENTITY);
+	    OMElement keyInfo = fac.createOMElement(AddressingConstants.QNAME_IDENTITY_KEY_INFO);
+	    OMElement x509Data = fac.createOMElement(AddressingConstants.QNAME_IDENTITY_X509_DATA);
+	    OMElement x509cert = fac.createOMElement(AddressingConstants.QNAME_IDENTITY_X509_CERT);
+	    x509cert.setText((String)parameter.getValue());
+	    
+	    x509Data.addChild(x509cert);
+	    keyInfo.addChild(x509Data);
+	    identity.addChild(keyInfo);
+	    
+	    epr.addChild(identity);
+	    
+	}
+	
+	
+	/*
+	 * Generate the EndpointReference element
+	 * <wsa:EndpointReference>
+         *    <wsa:Address>
+         *        http://some.service.epr/
+         *     </wsa:Address>
+         * </wsa:EndpointReference>
+	 * 
+	 */
+	private void generateEPRElement(OMFactory fac, OMElement port, String endpointURL){
+	    
+	    OMElement wsaEpr = fac.createOMElement(AddressingConstants.Final.WSA_ENDPOINT_REFERENCE);
+	    
+	    OMElement address = fac.createOMElement(AddressingConstants.Final.WSA_ADDRESS);
+	    address.setText((endpointURL == null) ? "": endpointURL);
+	    
+	    wsaEpr.addChild(address);
+	    
+	    // This will generate the identity element if the service parameter is set
+	    generateIdentityElement(fac, wsaEpr);
+	    
+	    port.addChild(wsaEpr);   
+	    
+	}
 }

