Index: modules/kernel/src/org/apache/axis2/builder/BuilderUtil.java
===================================================================
--- modules/kernel/src/org/apache/axis2/builder/BuilderUtil.java	(revision 1149578)
+++ modules/kernel/src/org/apache/axis2/builder/BuilderUtil.java	(working copy)
@@ -507,7 +507,9 @@
     protected static Attachments createAttachmentsMap(MessageContext msgContext,
                                                       InputStream inStream,
                                                       String contentTypeString) {
+
         boolean fileCacheForAttachments = isAttachmentsCacheEnabled(msgContext);
+        boolean attachmentStreamingEnabled = isAttachmentStreamingEnabled(msgContext);
 
         String attachmentRepoDir = null;
         String attachmentSizeThreshold = null;
@@ -570,7 +572,8 @@
                                  fileCacheForAttachments,
                                  attachmentRepoDir,
                                  attachmentSizeThreshold,
-                                 contentLength);
+                                 contentLength,
+                                 attachmentStreamingEnabled);
     }
 
     public static boolean isAttachmentsCacheEnabled(MessageContext msgContext) {
@@ -591,13 +594,33 @@
         return fileCacheForAttachments;
     }
 
+    public static boolean isAttachmentStreamingEnabled(MessageContext msgContext) {
+        Object streamAttachmentProperty = msgContext.getProperty(
+                Constants.Configuration.STREAM_ATTACHMENTS);
+        String streamAttachmentString;
+        boolean attachmentStreamingEnabled;
+
+        if (streamAttachmentProperty != null && streamAttachmentProperty instanceof String) {
+            streamAttachmentString = (String) streamAttachmentProperty;
+        } else {
+            Parameter streamAttachmentParam =
+                    msgContext.getParameter(Constants.Configuration.STREAM_ATTACHMENTS);
+            streamAttachmentString = (streamAttachmentParam != null) ?
+                    (String) streamAttachmentParam.getValue() : null;
+        }
+        attachmentStreamingEnabled = (Constants.VALUE_TRUE.equals(streamAttachmentString));
+        return attachmentStreamingEnabled;
+    }
+
     public static Attachments createAttachments(MessageContext msgContext,
                                                 InputStream inStream,
                                                 String contentTypeString,
                                                 boolean fileCacheForAttachments,
                                                 String attachmentRepoDir,
                                                 String attachmentSizeThreshold,
-                                                int contentLength) {
+                                                int contentLength,
+                                                boolean attachmentStreamingEnabled) {
+
         LifecycleManager manager = null;
         try {
             AxisConfiguration configuration = msgContext.getRootContext().getAxisConfiguration();
@@ -619,9 +642,27 @@
                                fileCacheForAttachments,
                                attachmentRepoDir,
                                attachmentSizeThreshold,
-                               contentLength);
+                               contentLength,
+                               attachmentStreamingEnabled);
     }
 
+    public static Attachments createAttachments(MessageContext msgContext,
+                                                InputStream inStream,
+                                                String contentTypeString,
+                                                boolean fileCacheForAttachments,
+                                                String attachmentRepoDir,
+                                                String attachmentSizeThreshold,
+                                                int contentLength) {
+        return createAttachments(msgContext,
+                                 inStream,
+                                 contentTypeString,
+                                 fileCacheForAttachments,
+                                 attachmentRepoDir,
+                                 attachmentSizeThreshold,
+                                 contentLength,
+                                 false);
+    }
+
     /**
      * Utility method to get a StAXBuilder
      *
Index: modules/kernel/src/org/apache/axis2/Constants.java
===================================================================
--- modules/kernel/src/org/apache/axis2/Constants.java	(revision 1149578)
+++ modules/kernel/src/org/apache/axis2/Constants.java	(working copy)
@@ -340,6 +340,7 @@
         //Attachment configurations
         public static final String ENABLE_MTOM = "enableMTOM";
         public static final String MTOM_THRESHOLD = "mtomThreshold";
+        public static final String STREAM_ATTACHMENTS = "streamAttachments";
         public static final String CACHE_ATTACHMENTS = "cacheAttachments";
         public static final String ATTACHMENT_TEMP_DIR = "attachmentDIR";
         public static final String FILE_SIZE_THRESHOLD = "sizeThreshold";
