Index: E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/MessageImpl.java
===================================================================
--- E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/MessageImpl.java	(revision 682724)
+++ E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/MessageImpl.java	(working copy)
@@ -437,7 +437,13 @@
      * @see org.apache.axis2.jaxws.message.Message#getDataHandler(java.lang.String)
      */
     public DataHandler getDataHandler(String cid) {
+        // if null DH was specified explicitly, just return
+        if(cid == null){
+             return (DataHandler) null;
+        }
+
         String bcid = getBlobCID(cid);
+
         return attachments.getDataHandler(bcid);
     }
     
Index: E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/Attachment.java
===================================================================
--- E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/Attachment.java	(revision 682724)
+++ E:/eclipse322/workspace/axis2/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/Attachment.java	(working copy)
@@ -95,7 +95,15 @@
      */
     public DataHandler getDataHandler() {
         if (dh == null) {
-            dh = createDataHandler(sigValue, sigClass, aDesc.getMimeTypes(), getContentID());
+        	//if null DH was specified explicitly, return it, don't create something else.
+            if (sigValue == null) {
+                // Create a contentID and null DataHandler
+                getContentID(); // Force setting of content id
+                dh = (DataHandler) null;
+            } else {
+                // Normal case: create a DataHandler from the sigValue object
+                dh = createDataHandler(sigValue, sigClass, aDesc.getMimeTypes(), getContentID());
+            }
         }
         return dh;
     }
@@ -123,6 +131,11 @@
         DataHandler dh = null;
         if (cls.isAssignableFrom(DataHandler.class)) {
             dh = (DataHandler) value;
+            if(dh == null)  
+            {
+                return dh; //return if DataHandler is null
+            }
+
             try {
                 Object content = dh.getContent();
                 // If the content is a Source, convert to a String due to 
