Index: modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 93931)
+++ modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -198,7 +198,8 @@
     }
 
     public EndpointReference[] getEPRsForService(String serviceName, String ip) throws AxisFault {
-        return new EndpointReference[0];
+        EndpointReference [] epr = {new EndpointReference(Constants.TRANSPORT_LOCAL + "://localhost/services/"  + serviceName)};
+        return epr;
     }
 
     public SessionContext getSessionContext(MessageContext messageContext) {
