--- axis2server.sh.orig	2007-04-05 09:15:59.711495300 -0700
+++ axis2server.sh	2007-04-05 09:16:32.992958300 -0700
@@ -29,5 +29,5 @@
 # Get the context and from that find the location of setenv.sh
 . `dirname $0`/setenv.sh
 
-java -classpath $AXIS2_CLASSPATH org.apache.axis2.transport.SimpleAxis2Server \
--repo $AXIS2_HOME/repository -conf $AXIS2_HOME/conf/axis2.xml $*
+java -classpath "$AXIS2_CLASSPATH" org.apache.axis2.transport.SimpleAxis2Server \
+-repo "$AXIS2_HOME"/repository -conf "$AXIS2_HOME"/conf/axis2.xml $*
