Index: src/org/apache/axis2/saaj/SOAPElementImpl.java
===================================================================
--- src/org/apache/axis2/saaj/SOAPElementImpl.java	(revision 495207)
+++ src/org/apache/axis2/saaj/SOAPElementImpl.java	(working copy)
@@ -510,6 +510,8 @@
         if (!encodingStyle.equals(SOAPConstants.URI_NS_SOAP_ENCODING)) {
             throw new IllegalArgumentException("Invalid Encoding style : " + encodingStyle);
         }
+        
+        //TODO setting charset encoding...???
         ((DocumentImpl) getOwnerDocument()).setCharsetEncoding(encodingStyle);
     }
 
Index: src/org/apache/axis2/saaj/SOAPFaultImpl.java
===================================================================
--- src/org/apache/axis2/saaj/SOAPFaultImpl.java	(revision 495207)
+++ src/org/apache/axis2/saaj/SOAPFaultImpl.java	(working copy)
@@ -19,6 +19,7 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.List;
 import java.util.Locale;
 
 import javax.wsdl.extensions.soap12.SOAP12Body;
@@ -61,10 +62,12 @@
 import org.apache.axiom.soap.impl.dom.soap11.SOAP11FaultRoleImpl;
 import org.apache.axiom.soap.impl.dom.soap11.SOAP11FaultTextImpl;
 import org.apache.axiom.soap.impl.dom.soap11.SOAP11Factory;
+import org.apache.axiom.soap.impl.dom.soap11.SOAP11HeaderBlockImpl;
 import org.apache.axiom.soap.impl.dom.soap12.SOAP12FaultDetailImpl;
 import org.apache.axiom.soap.impl.dom.soap12.SOAP12FaultDetailImpl;
 import org.apache.axiom.soap.impl.dom.soap12.SOAP12FaultImpl;
 import org.apache.axiom.soap.impl.dom.soap12.SOAP12FaultTextImpl;
+import org.apache.axiom.soap.impl.dom.soap12.SOAP12HeaderBlockImpl;
 
 public class SOAPFaultImpl extends SOAPBodyElementImpl implements SOAPFault {
 
@@ -315,24 +318,55 @@
         return this.faultReasonLocale;
     }
 
+    /**
+     * Appends or replaces a Reason Text item containing the specified text message and an xml:lang
+     * derived from locale. If a Reason Tex