Index: MessageContext.java
===================================================================
--- MessageContext.java	(revision 703931)
+++ MessageContext.java	(working copy)
@@ -842,6 +842,10 @@
      * @return Parameter <code>Parameter</code>
      */
     public Parameter getParameter(String key) {
+        if (axisMessage != null) {
+    		return axisMessage.getParameter(key);
+    	}
+        
         if (axisOperation != null) {
             return axisOperation.getParameter(key);
         }
