Index: modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java
===================================================================
--- modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java	(revision 998425)
+++ modules/addressing/src/org/apache/axis2/handlers/addressing/AddressingOutHandler.java	(working copy)
@@ -64,13 +64,21 @@
     private static final String MODULE_NAME = "addressing";
     
     public InvocationResponse invoke(MessageContext msgContext) throws AxisFault {
-        //determine whether outbound addressing has been disabled or not.
-        // Get default value from module.xml or axis2.xml files
-        Parameter param = msgContext.getModuleParameter(
-                DISABLE_ADDRESSING_FOR_OUT_MESSAGES, MODULE_NAME, handlerDesc);
-        boolean disableAddressing =
-            msgContext.isPropertyTrue(DISABLE_ADDRESSING_FOR_OUT_MESSAGES,
+        Parameter param = null;
+        boolean disableAddressing = false;
+        
+        Object o = msgContext.getProperty(DISABLE_ADDRESSING_FOR_OUT_MESSAGES);
+        if (o == null || !(o instanceof Boolean)) {
+            //determine whether outbound addressing has been disabled or not.
+            // Get default value from module.xml or axis2.xml files
+            param = msgContext.getModuleParameter(DISABLE_ADDRESSING_FOR_OUT_MESSAGES, MODULE_NAME, handlerDesc);
+            disableAddressing =
+                msgContext.isPropertyTrue(DISABLE_ADDRESSING_FOR_OUT_MESSAGES,
                     JavaUtils.isTrueExplicitly(Utils.getParameterValue(param)));
+        } else {
+           Boolean bool = (Boolean)o;
+           disableAddressing = bool.booleanValue();
+        }
 
         if (disableAddressing) {
             if (LoggingControl.debugLoggingAllowed && log.isTraceEnabled()) {
