diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/filter/TestRegexComparator.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/filter/TestRegexComparator.java
index 9dbe432..84e5e94 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/filter/TestRegexComparator.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/filter/TestRegexComparator.java
@@ -22,14 +22,13 @@ import static org.junit.Assert.*;
 
 import java.util.regex.Pattern;
 
+import org.apache.hadoop.hbase.SmallTests;
 import org.apache.hadoop.hbase.filter.RegexStringComparator.EngineType;
-import org.apache.hadoop.hbase.testclassification.FilterTests;
-import org.apache.hadoop.hbase.testclassification.SmallTests;
 import org.apache.hadoop.hbase.util.Bytes;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 
-@Category({FilterTests.class, SmallTests.class})
+@Category(SmallTests.class)
 public class TestRegexComparator {
 
   @Test
