commit 60d7f798aaef5808d3726615e1beab56e56668d1 (HEAD -> refs/heads/master)
Author: Andrew Wang <andrew.wang@cloudera.com>
Date:   Thu Oct 8 17:53:05 2015 -0700

    Stop accessing the src folder from within unit tests

diff --git a/hbase-server/src/test/data/0000000000000016310 b/hbase-server/src/test/data/0000000000000016310
deleted file mode 100644
index 8e58c98..0000000
Binary files a/hbase-server/src/test/data/0000000000000016310 and /dev/null differ
diff --git a/hbase-server/src/test/data/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c b/hbase-server/src/test/data/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c
deleted file mode 100644
index d16afff..0000000
Binary files a/hbase-server/src/test/data/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c and /dev/null differ
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/io/TestReference.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/io/TestReference.java
index 80295ff..0d18209 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/io/TestReference.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/io/TestReference.java
@@ -43,8 +43,8 @@ public class TestReference {
   @Test
   public void testParsingWritableReference() throws IOException {
     // Read a Reference written w/ 0.94 out of the test data dir.
-    final String datafile = System.getProperty("project.build.testSourceDirectory", "src/test") +
-      File.separator + "data" + File.separator +
+    final String datafile = System.getProperty("test.build.classes",
+        "target/test-classes") + File.separator +
       "a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c";
     FileSystem fs = FileSystem.get(HTU.getConfiguration());
     Reference.read(fs, new Path(datafile));
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRecoveredEdits.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRecoveredEdits.java
index c339bb1..f2acbdc 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRecoveredEdits.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRecoveredEdits.java
@@ -116,9 +116,9 @@ public class TestRecoveredEdits {
     Path regionDir = region.getRegionDir(hbaseRootDir, hri);
     Path recoveredEditsDir = WALSplitter.getRegionDirRecoveredEditsDir(regionDir);
     // This is a little fragile getting this path to a file of 10M of edits.
-    Path recoveredEditsFile = new Path(new Path(
-      System.getProperty("project.build.testSourceDirectory", "src" + Path.SEPARATOR + "test"),
-      "data"), "0000000000000016310");
+    Path recoveredEditsFile = new Path(
+      System.getProperty("test.build.classes", "target/test-classes"),
+        "0000000000000016310");
     // Copy this file under the region's recovered.edits dir so it is replayed on reopen.
     Path destination = new Path(recoveredEditsDir, recoveredEditsFile.getName());
     fs.copyToLocalFile(recoveredEditsFile, destination);
diff --git a/hbase-server/src/test/resources/0000000000000016310 b/hbase-server/src/test/resources/0000000000000016310
new file mode 100644
index 0000000..8e58c98
Binary files /dev/null and b/hbase-server/src/test/resources/0000000000000016310 differ
diff --git a/hbase-server/src/test/resources/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c b/hbase-server/src/test/resources/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c
new file mode 100644
index 0000000..d16afff
Binary files /dev/null and b/hbase-server/src/test/resources/a6a6562b777440fd9c34885428f5cb61.21e75333ada3d5bafb34bb918f29576c differ
diff --git a/pom.xml b/pom.xml
index c9fc6a0..cb28c84 100644
--- a/pom.xml
+++ b/pom.xml
@@ -575,6 +575,7 @@
             <redirectTestOutputToFile>${test.output.tofile}</redirectTestOutputToFile>
             <systemPropertyVariables>
               <jacoco-agent.destfile>target/jacoco.exec</jacoco-agent.destfile>
+              <test.build.classes>${test.build.classes}</test.build.classes>
             </systemPropertyVariables>
 
             <excludes>
@@ -1305,6 +1306,8 @@
     <hbase.skip-jacoco>true</hbase.skip-jacoco>
     <jacoco.version>0.7.5.201505241946</jacoco.version>
     <extra.enforcer.version>1.0-beta-3</extra.enforcer.version>
+    <!-- Location of test resources -->
+    <test.build.classes>${project.build.directory}/test-classes</test.build.classes>
   </properties>
   <!-- Sorted by groups of dependencies then groupId and artifactId -->
   <dependencyManagement>
