diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java
index a51a4ac..696a6c9 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java
@@ -411,7 +411,7 @@ class ConnectionImplementation implements ClusterConnection, Closeable {
           //the queue is full, a new thread will be started
           this.metaLookupPool = getThreadPool(
              conf.getInt("hbase.hconnection.meta.lookup.threads.max", 128),
-             conf.getInt("hbase.hconnection.meta.lookup.threads.max.core", 10),
+             conf.getInt("hbase.hconnection.meta.lookup.threads.core", 10),
              "-metaLookup-shared-", new LinkedBlockingQueue<Runnable>());
         }
       }
