diff --git a/hbase-hadoop1-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java b/hbase-hadoop1-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
index d2d48c8..cfb3dee 100644
--- a/hbase-hadoop1-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
+++ b/hbase-hadoop1-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
@@ -32,6 +32,7 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
   private final MetricMutableCounterLong authorizationSuccesses;
   private final MetricMutableCounterLong authorizationFailures;
   private final MetricMutableCounterLong authenticationSuccesses;
+  private final MetricMutableCounterLong authenticationFallbacks;
   private final MetricMutableCounterLong authenticationFailures;
   private final MetricMutableCounterLong sentBytes;
   private final MetricMutableCounterLong receivedBytes;
@@ -79,6 +80,8 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
 
     this.authenticationSuccesses = this.getMetricsRegistry().newCounter(
         AUTHENTICATION_SUCCESSES_NAME, AUTHENTICATION_SUCCESSES_DESC, 0l);
+    this.authenticationFallbacks = this.getMetricsRegistry().newCounter(
+        AUTHENTICATION_FALLBACKS_NAME, AUTHENTICATION_FALLBACKS_DESC, 0L);
     this.authenticationFailures = this.getMetricsRegistry().newCounter(AUTHENTICATION_FAILURES_NAME,
         AUTHENTICATION_FAILURES_DESC, 0l);
     this.sentBytes = this.getMetricsRegistry().newCounter(SENT_BYTES_NAME,
@@ -118,6 +121,11 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
   }
 
   @Override
+  public void authenticationFallback() {
+    authenticationFallbacks.incr();
+  }
+
+  @Override
   public void exception() {
     exceptions.incr();
   }
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
index f5df475..f637fe4 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/RpcServer.java
@@ -1325,7 +1325,7 @@ public class RpcServer implements RpcServerInterface {
       } else {
         authorizedUgi = UserGroupInformation.createRemoteUser(authorizedId);
       }
-      authorizedUgi.setAuthenticationMethod(authMethod.authenticationMethod.getAuthMethod());
+      authorizedUgi.setAuthenticationMethod(authMethod.authenticationMethod);
       return authorizedUgi;
     }
 
