diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MetricsRegionServerWrapperImpl.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MetricsRegionServerWrapperImpl.java
index e908be6..2e480ad 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MetricsRegionServerWrapperImpl.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MetricsRegionServerWrapperImpl.java
@@ -569,6 +569,7 @@ class MetricsRegionServerWrapperImpl
 
     @Override
     synchronized public void run() {
+      try {
       initBlockCache();
       initMobFileCache();
       cacheStats = blockCache.getStats();
@@ -718,6 +719,9 @@ class MetricsRegionServerWrapperImpl
       mobFileCacheEvictedCount = mobFileCache.getEvictedFileCount();
       mobFileCacheCount = mobFileCache.getCacheSize();
       blockedRequestsCount = tempBlockedRequestsCount;
+      } catch (Throwable e) {
+        LOG.warn("Caught exception! Will suppress and retry.", e);
+      }
     }
   }
 
