commit 00f73de746f5d2df5852735c6ed226fd81c11d68
Author: thiruvel <thiruvel@yahoo-inc.com>
Date:   Sat Jun 25 19:57:14 2016 -0700

    get regionLocation for required regions for MR jobs

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableInputFormatBase.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableInputFormatBase.java
index 53148dc..9ebb3c1 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableInputFormatBase.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableInputFormatBase.java
@@ -275,16 +275,7 @@ extends InputFormat<ImmutableBytesWritable, Result> {
         if (!includeRegionInSplit(keys.getFirst()[i], keys.getSecond()[i])) {
           continue;
         }
-        HRegionLocation location = getRegionLocator().getRegionLocation(keys.getFirst()[i], false);
-        // The below InetSocketAddress creation does a name resolution.
-        InetSocketAddress isa = new InetSocketAddress(location.getHostname(), location.getPort());
-        if (isa.isUnresolved()) {
-          LOG.warn("Failed resolve " + isa);
-        }
-        InetAddress regionAddress = isa.getAddress();
-        String regionLocation;
-        regionLocation = reverseDNS(regionAddress);
-  
+
         byte[] startRow = scan.getStartRow();
         byte[] stopRow = scan.getStopRow();
         // determine if the given start an stop key fall into the region
@@ -299,7 +290,17 @@ extends InputFormat<ImmutableBytesWritable, Result> {
             Bytes.compareTo(keys.getSecond()[i], stopRow) <= 0) &&
             keys.getSecond()[i].length > 0 ?
               keys.getSecond()[i] : stopRow;
-  
+
+          HRegionLocation location = getRegionLocator().getRegionLocation(keys.getFirst()[i], false);
+          // The below InetSocketAddress creation does a name resolution.
+          InetSocketAddress isa = new InetSocketAddress(location.getHostname(), location.getPort());
+          if (isa.isUnresolved()) {
+            LOG.warn("Failed resolve " + isa);
+          }
+          InetAddress regionAddress = isa.getAddress();
+          String regionLocation;
+          regionLocation = reverseDNS(regionAddress);
+
           byte[] regionName = location.getRegionInfo().getRegionName();
           String encodedRegionName = location.getRegionInfo().getEncodedName();
           long regionSize = sizeCalculator.getRegionSize(regionName);
