diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/mob/compactions/PartitionedMobCompactor.java hbase-server/src/main/java/org/apache/hadoop/hbase/mob/compactions/PartitionedMobCompactor.java
index 6c8080c..33f5e71 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/mob/compactions/PartitionedMobCompactor.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/mob/compactions/PartitionedMobCompactor.java
@@ -176,10 +176,11 @@ public class PartitionedMobCompactor extends MobCompactor {
         id.setDate(MobFileName.getDateFromName(fileName));
         CompactionPartition compactionPartition = filesToCompact.get(id);
         if (compactionPartition == null) {
-          compactionPartition = new CompactionPartition(
-              new CompactionPartitionId(id.getStartKey(), id.getDate()));
+          CompactionPartitionId newId = new CompactionPartitionId(id.getStartKey(), id.getDate());
+          compactionPartition = new CompactionPartition(newId);
+
           compactionPartition.addFile(file);
-          filesToCompact.put(id, compactionPartition);
+          filesToCompact.put(newId, compactionPartition);
         } else {
           compactionPartition.addFile(file);
         }
