diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
index de3439b..fd62340 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCache.java
@@ -117,8 +117,12 @@ public MobFileCache(Configuration conf) {
       }
       this.scheduleThreadPool.scheduleAtFixedRate(new EvictionThread(this), period, period,
           TimeUnit.SECONDS);
+
+      LOG.info("MobFileCache enabled with cacheSize=" + mobFileMaxCacheSize +
+          ", evictPeriods=" +  period + "sec, evictRemainRatio=" + evictRemainRatio);
+    } else {
+      LOG.info("MobFileCache disabled");
     }
-    LOG.info("MobFileCache is initialized, and the cache size is " + mobFileMaxCacheSize);
   }
 
   /**
