diff --git src/main/ruby/hbase/admin.rb src/main/ruby/hbase/admin.rb
index 85367bf..762b053 100644
--- src/main/ruby/hbase/admin.rb
+++ src/main/ruby/hbase/admin.rb
@@ -198,6 +198,7 @@ module Hbase
             File.foreach(arg[SPLITS_FILE]) do |line|
               arg[SPLITS].push(line.strip())
             end
+            htd.setValue(SPLITS_FILE, arg[SPLITS_FILE])
           end
 
           splits = Java::byte[][arg[SPLITS].size].new
