diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/AsyncRpcChannelImpl.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/AsyncRpcChannelImpl.java
index 2b9000a..f59e303 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/AsyncRpcChannelImpl.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/AsyncRpcChannelImpl.java
@@ -661,7 +661,7 @@ public class AsyncRpcChannelImpl implements AsyncRpcChannel {
           } else {
             String msg = "Couldn't setup connection for "
                 + UserGroupInformation.getLoginUser().getUserName() + " to " + serverPrincipal;
-            LOG.warn(msg);
+            LOG.warn(msg, ex);
             throw (IOException) new IOException(msg).initCause(ex);
           }
         } else {
diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/RpcClientImpl.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/RpcClientImpl.java
index 00eea7a..abae403 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/RpcClientImpl.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/RpcClientImpl.java
@@ -672,7 +672,7 @@ public class RpcClientImpl extends AbstractRpcClient {
               String msg = "Couldn't setup connection for " +
                 UserGroupInformation.getLoginUser().getUserName() +
                 " to " + serverPrincipal;
-              LOG.warn(msg);
+              LOG.warn(msg, ex);
               throw (IOException) new IOException(msg).initCause(ex);
             }
           } else {
