diff --git hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
index 951a95e..729d6ee 100644
--- hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
+++ hbase-server/src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
@@ -37,6 +37,7 @@ org.apache.hadoop.hbase.HTableDescriptor;
 org.apache.hadoop.hbase.NamespaceDescriptor;
 org.apache.hadoop.hbase.ServerLoad;
 org.apache.hadoop.hbase.ServerName;
+org.apache.hadoop.hbase.snapshot.SnapshotInfo;
 org.apache.hadoop.hbase.TableName;
 org.apache.hadoop.hbase.client.Admin;
 org.apache.hadoop.hbase.master.AssignmentManager;
@@ -458,6 +459,10 @@ AssignmentManager assignmentManager = master.getAssignmentManager();
    try (Admin admin = master.getConnection().getAdmin()) {
      snapshots = master.isInitialized() ? admin.listSnapshots() : null;
    }
+   long totalStoreFileSize = 0;
+   long totalSharedStoreFileSize = 0;
+   long totalArchivedStoreFileSize = 0;
+   long totalMobStoreFileSize = 0;
 </%java>
 <%if (snapshots != null && snapshots.size() > 0)%>
 <table class="table table-striped">
@@ -469,6 +474,11 @@ AssignmentManager assignmentManager = master.getAssignmentManager();
     <%for SnapshotDescription snapshotDesc : snapshots%>
     <%java>
         TableName snapshotTable = TableName.valueOf(snapshotDesc.getTable());
+        SnapshotInfo.SnapshotStats stats = SnapshotInfo.getSnapshotStats(master.getConfiguration(), snapshotDesc);
+        totalStoreFileSize += stats.getStoreFilesSize();
+        totalSharedStoreFileSize += stats.getSharedStoreFilesSize();
+        totalArchivedStoreFileSize += stats.getArchivedStoreFileSize();
+        totalMobStoreFileSize += stats.getMobStoreFilesSize();
     </%java>
     <tr>
         <td><a href="snapshot.jsp?name=<% snapshotDesc.getName() %>"><% snapshotDesc.getName() %></a> </td>
@@ -478,6 +488,10 @@ AssignmentManager assignmentManager = master.getAssignmentManager();
     </tr>
     </%for>
     <p><% snapshots.size() %> snapshot(s) in set.</p>
+    <p>Total Store Files' size: <% totalStoreFileSize %> </p>
+    <p>Total Shared Store Files' size: <% totalSharedStoreFileSize %> </p>
+    <p>Total Archived Store Files' size: <% totalArchivedStoreFileSize %> </p>
+    <p>Total Mob Store Files' size: <% totalMobStoreFileSize %> </p>
 </table>
 </%if>
 </%def>
