Index: src/main/java/org/apache/axis2/transport/mail/MailConstants.java
===================================================================
--- src/main/java/org/apache/axis2/transport/mail/MailConstants.java	(revision 1038039)
+++ src/main/java/org/apache/axis2/transport/mail/MailConstants.java	(working copy)
@@ -25,7 +25,7 @@
 
     public static final String TRANSPORT_NAME = "mailto";
     public static final String TRANSPORT_PREFIX = "mailto:";
-    
+
     public static final String TEXT_PLAIN = "text/plain";
     public static final String APPLICATION_BINARY = "application/binary";
 
@@ -48,9 +48,9 @@
     public static final long   DEFAULT_RECONNECT_TIMEOUT = 30000;
 
     public static final String TRANSPORT_MAIL_ADDRESS  = "transport.mail.Address";
-    
+
     public static final String TRANSPORT_MAIL_DEBUG = "transport.mail.Debug";
-    
+
     /**
      * Key for the mail store protocol parameter.
      * The mail store protocol identifier is used in calls to {@link Session#getStore()}.
@@ -95,5 +95,8 @@
     /** @see org.apache.axis2.transport.mail.WSMimeMessage */
     public static final String MAIL_HEADER_X_MESSAGE_ID= "X-Message-ID";
     public static final String TRANSPORT_MAIL_CUSTOM_HEADERS     = "transport.mail.custom.headers";
-    
+
+    public static final String MAIL_SENT_DATE = "sent-date";
+    public static final String MAIL_RECEIVED_DATE = "received-date";
+
 }
Index: src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java
===================================================================
--- src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java	(revision 1038039)
+++ src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java	(working copy)
@@ -494,6 +494,15 @@
         // save original mail message id message context MessageID
         msgContext.setMessageID(outInfo.getRequestMessageID());
 
+        //Set the Sent date and received date.
+        if(message.getSentDate() != null) {
+            Calendar sentDate = Calendar.getInstance();
+            sentDate.setTime(message.getSentDate());
+            msgContext.setProperty(MailConstants.MAIL_SENT_DATE,sentDate);
+        }
+
+        msgContext.setProperty(MailConstants.MAIL_RECEIVED_DATE,Calendar.getInstance());
+
         // set the message payload to the message context
         InputStream in = messagePart.getInputStream();
         try {
