Index: src/org/apache/axis2/engine/ListenerManager.java
===================================================================
--- src/org/apache/axis2/engine/ListenerManager.java	(revision 762636)
+++ src/org/apache/axis2/engine/ListenerManager.java	(working copy)
@@ -56,9 +56,25 @@
     // We're stopped at first.
     private boolean stopped = true;
 
+    // need to preserve the default behavior of requiring a shutdown hook
+    private boolean shutdownHookRequired = true;
+
     public void init(ConfigurationContext configCtx) {
         configCtx.setTransportManager(this);
         this.configctx = configCtx;
+
+        // initialize all the transport listeners
+        for (Object o : configctx.getAxisConfiguration().getTransportsIn().values()) {
+            try {
+                TransportInDescription transportIn = (TransportInDescription)o;
+                TransportListener listener = transportIn.getReceiver();
+                if (listener != null && startedTransports.get(transportIn.getName()) == null) {
+                    listener.init(configctx, transportIn);
+                }
+            } catch (Exception e) {
+                log.info(e.getMessage(), e);
+            }
+        }
     }
 
     public ConfigurationContext getConfigctx() {
@@ -122,7 +138,6 @@
                 TransportInDescription transportIn = (TransportInDescription)o;
                 TransportListener listener = transportIn.getReceiver();
                 if (listener != null && startedTransports.get(transportIn.getName()) == null) {
-                    listener.init(configctx, transportIn);
                     listener.start();
                     startedTransports.put(transportIn.getName(), listener);
                 }
@@ -131,7 +146,7 @@
             }
         }
 
-        if (shutdownHookThread == null) {
+        if (shutdownHookThread == null && isShutdownHookRequired()) {
             shutdownHookThread = new ListenerManagerShutdownThread(this);
             Runtime.getRuntime().addShutdownHook(shutdownHookThread);
         }
@@ -214,7 +229,7 @@
             if (!started) {
                 transportListener.init(configctx, trsIn);
                 transportListener.start();
-                if (shutdownHookThread == null) {
+                if (shutdownHookThread == null && isShutdownHookRequired()) {
                     shutdownHookThread = new ListenerManagerShutdownThread(this);
                     Runtime.getRuntime().addShutdownHook(shutdownHookThread);
                 }
@@ -244,6 +259,14 @@
         defaultConfigurationContext = null;
     }
 
+    public boolean isShutdownHookRequired() {
+        return shutdownHookRequired;
+    }
+
+    public void setShutdownHookRequired(boolean shutdownHookRequired) {
+        this.shutdownHookRequired = shutdownHookRequired;
+    }
+
     static class ListenerManagerShutdownThread extends Thread {
         ListenerManager listenerManager;
 
