Index: C:/workspace/live/axis2new/modules/jaxws/test/org/apache/axis2/jaxws/framework/JAXWSTest.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/test/org/apache/axis2/jaxws/framework/JAXWSTest.java	(revision 476039)
+++ C:/workspace/live/axis2new/modules/jaxws/test/org/apache/axis2/jaxws/framework/JAXWSTest.java	(working copy)
@@ -72,7 +72,7 @@
     
     static {
         // Enable debug
-        //BasicConfigurator.configure();
+        BasicConfigurator.configure();
     }
     
     /**
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLPartBase.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLPartBase.java	(revision 476039)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/XMLPartBase.java	(working copy)
@@ -42,7 +42,7 @@
 import org.apache.axis2.jaxws.message.XMLPart;
 import org.apache.axis2.jaxws.message.factory.BlockFactory;
 import org.apache.axis2.jaxws.message.factory.SOAPEnvelopeBlockFactory;
-import org.apache.axis2.jaxws.message.util.MessageUtils;
+import org.apache.axis2.jaxws.util.JavaUtils;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
@@ -585,8 +585,8 @@
 		if (consume) {
 			this.consumed = true;
 			if (log.isDebugEnabled()) {
-				log.debug("The follow stack trace indicates where the Block is consumed");
-				log.debug(MessageUtils.stackToString(new RuntimeException()));
+				log.debug("Debug Monitoring When Block is Consumed");
+				log.debug(JavaUtils.stackToString());
 			}
 		} else {
 			consumed = false;
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java	(revision 476039)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/impl/BlockImpl.java	(working copy)
@@ -36,8 +36,8 @@
 import org.apache.axis2.jaxws.message.MessageException;
 import org.apache.axis2.jaxws.message.XMLPart;
 import org.apache.axis2.jaxws.message.factory.BlockFactory;
-import org.apache.axis2.jaxws.message.util.MessageUtils;
 import org.apache.axis2.jaxws.message.util.Reader2Writer;
+import org.apache.axis2.jaxws.util.JavaUtils;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
@@ -287,8 +287,8 @@
 			omElement = null;
 			if (log.isDebugEnabled()) {
 				// The following stack trace consumes indicates where the message is consumed
-				log.debug("The follow stack trace indicates where the Block is consumed");
-				log.debug(MessageUtils.stackToString(new RuntimeException()));
+				log.debug("Message Block Monitor: Action=Consumed");
+				log.debug(JavaUtils.stackToString());
 			}
 		} else {
 			consumed = false;
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/util/MessageUtils.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/util/MessageUtils.java	(revision 476045)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/message/util/MessageUtils.java	(working copy)
@@ -326,18 +326,4 @@
             }
         }
     }
-    
-    /**
-     * Get a string containing the stack of the specified exception
-     * @param e Throwable
-     * @return String
-     */
-    public static String stackToString(Throwable e){
-        java.io.StringWriter sw= new java.io.StringWriter(); 
-        java.io.BufferedWriter bw = new java.io.BufferedWriter(sw);
-        java.io.PrintWriter pw= new java.io.PrintWriter(bw); 
-        e.printStackTrace(pw);
-        pw.close();
-        return sw.getBuffer().toString();
-      }
 }
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/MethodMarshallerUtils.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/MethodMarshallerUtils.java	(revision 476045)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/alt/MethodMarshallerUtils.java	(working copy)
@@ -55,9 +55,9 @@
 import org.apache.axis2.jaxws.message.XMLFaultReason;
 import org.apache.axis2.jaxws.message.databinding.JAXBBlockContext;
 import org.apache.axis2.jaxws.message.factory.JAXBBlockFactory;
-import org.apache.axis2.jaxws.message.util.MessageUtils;
 import org.apache.axis2.jaxws.registry.FactoryRegistry;
 import org.apache.axis2.jaxws.util.ClassUtils;
+import org.apache.axis2.jaxws.util.JavaUtils;
 import org.apache.axis2.jaxws.util.XMLRootElementUtil;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -429,8 +429,7 @@
         // Get the root cause of the throwable object
         if (log.isDebugEnabled()) {
             log.debug("Marshal Throwable =" + throwable.getClass().getName());
-            log.debug("  message=" + throwable.getMessage());
-            log.debug("  stack=" + MessageUtils.stackToString(throwable));
+            log.debug("  exception=" + throwable.toString());
         }
         Throwable t = ClassUtils.getRootCause(throwable);
 
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/MethodMarshallerImpl.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/MethodMarshallerImpl.java	(revision 476045)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/marshaller/impl/MethodMarshallerImpl.java	(working copy)
@@ -61,9 +61,9 @@
 import org.apache.axis2.jaxws.message.databinding.JAXBUtils;
 import org.apache.axis2.jaxws.message.factory.JAXBBlockFactory;
 import org.apache.axis2.jaxws.message.factory.MessageFactory;
-import org.apache.axis2.jaxws.message.util.MessageUtils;
 import org.apache.axis2.jaxws.registry.FactoryRegistry;
 import org.apache.axis2.jaxws.util.ClassUtils;
+import org.apache.axis2.jaxws.util.JavaUtils;
 import org.apache.axis2.jaxws.util.XMLRootElementUtil;
 import org.apache.axis2.jaxws.wrapper.JAXBWrapperTool;
 import org.apache.axis2.jaxws.wrapper.impl.JAXBWrapperException;
@@ -192,8 +192,7 @@
 			//  Get the root cause of the throwable object
             if (log.isDebugEnabled()) {
                 log.debug("Marshal Throwable =" + throwable.getClass().getName());
-                log.debug("  message=" + throwable.getMessage());
-                log.debug("  stack=" + MessageUtils.stackToString(throwable));
+                log.debug("  message=" + throwable.toString());
             }
 			XMLFault xmlfault = null;
 			
Index: C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/util/JavaUtils.java
===================================================================
--- C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/util/JavaUtils.java	(revision 476039)
+++ C:/workspace/live/axis2new/modules/jaxws/src/org/apache/axis2/jaxws/util/JavaUtils.java	(working copy)
@@ -17,6 +17,9 @@
 
 package org.apache.axis2.jaxws.util;
 
+import java.io.BufferedWriter;
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.util.ArrayList;
@@ -166,4 +169,22 @@
 
         return new String(sb);
     }
+
+    /**
+     * Get a string containing the stack of the specified exception
+     * @return String
+     */
+    public static String stackToString(){
+        Exception e = new RuntimeException();
+        java.io.StringWriter sw= new java.io.StringWriter(); 
+        java.io.BufferedWriter bw = new java.io.BufferedWriter(sw);
+        java.io.PrintWriter pw= new java.io.PrintWriter(bw); 
+        e.printStackTrace(pw);
+        pw.close();
+        String text = sw.getBuffer().toString();
+        // Jump past the "RuntimeException"
+        text = text.substring(text.indexOf("at"));
+        text = replace(text, "at ", "DEBUG_FRAME = ");
+        return text;
+      }
 }
