Index: test/org/apache/axis2/schema/XSD2JavaTest.java
===================================================================
--- test/org/apache/axis2/schema/XSD2JavaTest.java	(revision 0)
+++ test/org/apache/axis2/schema/XSD2JavaTest.java	(revision 0)
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.axis2.schema;
+
+import java.io.File;
+import org.junit.Test;
+
+public class XSD2JavaTest extends XMLSchemaTest {
+
+    @Test
+    public void testMain() throws Exception {
+        File file = null;
+
+        for (int i = 0; i <= 4; i++) {
+            file = new File(SampleSchemasDirectory + "sampleSchema" + i + ".xsd");
+           
+            if (file.exists()) {
+                XSD2Java.main(new String[] { SampleSchemasDirectory + "sampleSchema" + i + ".xsd",
+                        SampleSchemasDirectory+"XSD2JAVA" + i });
+                File temp=new File(SampleSchemasDirectory+"XSD2JAVA" + i );
+                //checks whether XSD2Java executed successfully
+                assertTrue(temp.exists());
+            }
+        }
+
+    }
+}
Index: test/org/apache/axis2/schema/SchemaCompilerTest.java
===================================================================
--- test/org/apache/axis2/schema/SchemaCompilerTest.java	(revision 0)
+++ test/org/apache/axis2/schema/SchemaCompilerTest.java	(revision 0)
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.axis2.schema;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import javax.xml.namespace.QName;
+
+import org.apache.ws.commons.schema.XmlSchema;
+import org.apache.ws.commons.schema.XmlSchemaCollection;
+import org.junit.Test;
+import org.w3c.dom.Document;
+
+public class SchemaCompilerTest extends XMLSchemaTest{
+    
+    private ArrayList<XmlSchema> schemas;
+    private SchemaCompiler schemaCompiler;
+    private HashMap<QName,String> processedElementMap;
+    
+
+    @Override
+    protected void setUp() throws Exception {
+        schemas=new ArrayList<XmlSchema>();
+        loadSampleSchemaFile(schemas);
+        schemaCompiler=new SchemaCompiler(null);
+    }
+
+    @Override
+    protected void tearDown() throws Exception {
+        schemas=null;
+        super.tearDown();
+    }
+    
+    @Test
+    public void testCompileSchema() throws Exception{
+        Map map=schemaCompiler.getProcessedModelMap();
+        schemaCompiler.compile(schemas);
+        processedElementMap=schemaCompiler.getProcessedElementMap();
+        map=schemaCompiler.getProcessedModelMap();
+        Iterator iterator=map.values().iterator();
+        int i=0;
+        while (iterator.hasNext()) {
+            i++;
+            Document document= (Document) iterator.next();
+            ByteArrayOutputStream  stream=new ByteArrayOutputStream();
+            XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+            XmlSchema schema = schemaCol.read(document, null);
+            schema.write(stream);
+            assertSimilarXML(stream.toString(), readXMLfromSchemaFile(customDirectoryLocation+"generated"+i+".xsd"));
+            
+        }
+        
+    }
+    
+   
+    
+    
+    
+
+}
Index: test/org/apache/axis2/schema/ExtensionUtilityTest.java
===================================================================
--- test/org/apache/axis2/schema/ExtensionUtilityTest.java	(revision 0)
+++ test/org/apache/axis2/schema/ExtensionUtilityTest.java	(revision 0)
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+
+package org.apache.axis2.schema;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import javax.xml.namespace.QName;
+
+import org.apache.axis2.description.AxisService;
+import org.apache.axis2.util.CommandLineOption;
+import org.apache.axis2.wsdl.codegen.CodeGenConfiguration;
+import org.apache.axis2.wsdl.databinding.TypeMapper;
+import org.apache.ws.commons.schema.XmlSchema;
+import org.apache.ws.commons.schema.XmlSchemaCollection;
+import org.w3c.dom.Document;
+
+
+
+public class ExtensionUtilityTest extends XMLSchemaTest {
+    
+    public void testInvoke() throws Exception{
+        ArrayList<XmlSchema> schemas=new ArrayList<XmlSchema>();
+        loadSampleSchemaFile(schemas);
+        AxisService service=new AxisService();
+        service.addSchema(schemas);
+        CodeGenConfiguration configuration=new CodeGenConfiguration(new HashMap<String, CommandLineOption>());
+        configuration.addAxisService(service);
+        ExtensionUtility.invoke(configuration);
+        TypeMapper mapper=configuration.getTypeMapper();
+        assertEquals(mapper.getParameterName(new QName("//www.w3schools.com", "order")), "order");
+     //   assertEquals(mapper., actual)
+        ByteArrayOutputStream stream =new ByteArrayOutputStream();
+        Document document=(Document)mapper.getTypeMappingObject(new QName("ColorType"));
+        XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+        XmlSchema schema = schemaCol.read(document, null);
+        schema.write(stream);
+        //writeToFile(customDirectoryLocation+"generated.xsd", stream.toString());
+        assertSimilarXML(stream.toString(), readXMLfromSchemaFile(customDirectoryLocation+"generated.xsd"));
+ 
+        
+    }
+
+}
Index: test/org/apache/axis2/schema/XMLSchemaTest.java
===================================================================
--- test/org/apache/axis2/schema/XMLSchemaTest.java	(revision 0)
+++ test/org/apache/axis2/schema/XMLSchemaTest.java	(revision 0)
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.axis2.schema;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.InputStream;
+import java.util.ArrayList;
+
+import javax.xml.transform.stream.StreamSource;
+
+import junit.framework.TestCase;
+
+import org.apache.axis2.util.XMLPrettyPrinter;
+import org.apache.ws.commons.schema.XmlSchema;
+import org.apache.ws.commons.schema.XmlSchemaCollection;
+import org.custommonkey.xmlunit.Diff;
+
+public abstract class XMLSchemaTest extends TestCase {
+
+    public final String XMLSchemaNameSpace = "xmlns:xs=\"http://www.w3.org/2001/XMLSchema\"";
+
+    public final String CustomSchemaLocation = "test-resources"
+            + File.separator + "schemas" + File.separator + "custom_schemas"
+            + File.separator + "note.xsd";
+    
+    public final String customDirectoryLocation = "test-resources"
+            + File.separator + "schemas" + File.separator + "custom_schemas"
+            + File.separator;
+
+    public final String SampleSchemasDirectory = "test-resources"
+            + File.separator + "schemas" + File.separator + "custom_schemas"
+            + File.separator;
+
+    public final String MappingFileLocation = "test-resources" + File.separator
+            + "schemas" + File.separator + "mapping_files" + File.separator
+            + "mapping1.txt";
+
+    public void assertSimilarXML(String XML1, String XML2) throws Exception {
+        Diff myDiff = new Diff(XML1, XML2);
+        assertTrue("XML similar " + myDiff.toString(), myDiff.similar());
+
+    }
+
+    public void assertIdenticalXML(String XML1, String XML2) throws Exception {
+        Diff myDiff = new Diff(XML1, XML2);
+        assertTrue("XML similar " + myDiff.toString(), myDiff.identical());
+
+    }
+
+    public void loadSampleSchemaFile(ArrayList<XmlSchema> schemas) throws Exception{
+        XmlSchemaCollection xmlSchemaCollection = new XmlSchemaCollection();
+        File file = null;
+        int i = 1;
+       
+            file = new File(SampleSchemasDirectory + "sampleSchema" + i
+                    + ".xsd");
+            while (file.exists()) {
+                InputStream is = new FileInputStream(file);
+                XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+                XmlSchema schema = schemaCol.read(new StreamSource(is), null);
+                schemas.add(schema);
+                i++;
+                file = new File(SampleSchemasDirectory + "sampleSchema" + i
+                        + ".xsd");
+            }
+       
+    }
+    
+    public XmlSchema loadSingleSchemaFile(int i) throws Exception{
+        File file = new File(SampleSchemasDirectory + "sampleSchema" + i
+                + ".xsd");
+        InputStream is = new FileInputStream(file);
+        XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+        XmlSchema schema = schemaCol.read(new StreamSource(is), null);
+        return schema;
+    }
+     
+    
+    public String readFile(String fileName) throws Exception {
+        File file = new File(fileName);
+        char[] buffer = null;
+        BufferedReader bufferedReader = new BufferedReader(new FileReader(file));
+        buffer = new char[(int) file.length()];
+        int i = 0;
+        int c = bufferedReader.read();
+        while (c != -1) {
+            buffer[i++] = (char) c;
+            c = bufferedReader.read(); 
+        }
+        return new String(buffer);
+    }
+
+    public String readXMLfromSchemaFile(String path) throws Exception {
+        InputStream is = new FileInputStream(path);
+        XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+        XmlSchema schema = schemaCol.read(new StreamSource(is), null);
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        schema.write(stream);
+        is.close();
+        return stream.toString();
+    }
+    
+   
+    public String readWSDLFromFile(String path) throws Exception {
+        File file=new File(path);
+        XMLPrettyPrinter.prettify(file);    //this is used to correct unnecessary formatting in the file
+        return readFile(path);
+    }
+    
+    public void writeToFile(String path,String data) throws Exception{
+        FileWriter fileWriter=new FileWriter(new File(path));
+        fileWriter.write(data);
+        fileWriter.flush();
+        fileWriter.close();        
+    }
+    
+    public String schemaToString(XmlSchema schema){
+        ByteArrayOutputStream stream=new ByteArrayOutputStream();
+        schema.write(stream);
+        return stream.toString();
+    }
+    
+    public XmlSchema loadSingleSchemaFile(String path) throws Exception{
+        File file = new File(path);
+        InputStream is = new FileInputStream(file);
+        XmlSchemaCollection schemaCol = new XmlSchemaCollection();
+        XmlSchema schema = schemaCol.read(new StreamSource(is), null);
+        return schema;
+    }
+    
+    
+
+}
Index: test-resources/schemas/custom_schemas/generated10.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated10.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated10.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated11.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated11.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated11.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated12.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated12.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated12.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated13.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated13.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated13.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/note.xsd
===================================================================
--- test-resources/schemas/custom_schemas/note.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/note.xsd	(revision 0)
@@ -0,0 +1,37 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<?xml version="1.0"?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+targetNamespace="http://www.w3schools.com"
+xmlns="http://www.w3schools.com"
+elementFormDefault="qualified">
+
+<xs:element name="note">
+  <xs:complexType>
+    <xs:sequence>
+      <xs:element name="to" type="xs:string"/>
+      <xs:element name="from" type="xs:string"/>
+      <xs:element name="heading" type="xs:string"/>
+      <xs:element name="body" type="xs:string"/>
+    </xs:sequence>
+  </xs:complexType>
+</xs:element>
+
+</xs:schema>
Index: test-resources/schemas/custom_schemas/generated1.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated1.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated1.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated2.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated2.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated2.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated3.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated3.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated3.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated4.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated4.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated4.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/sampleSchema1.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema1.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema1.xsd	(revision 0)
@@ -0,0 +1,38 @@
+<?xml version="1.0"?>
+
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+targetNamespace="http://www.w3schools.com"
+xmlns="http://www.w3schools.com"
+elementFormDefault="qualified">
+
+<xs:element name="order">
+  <xs:complexType>
+    <xs:sequence>
+      <xs:element name="to" type="xs:string"/>
+      <xs:element name="from" type="xs:string"/>
+      <xs:element name="heading" type="xs:string"/>
+      <xs:element name="body" type="xs:string"/>
+    </xs:sequence>
+  </xs:complexType>
+</xs:element>
+
+</xs:schema>
Index: test-resources/schemas/custom_schemas/generated5.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated5.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated5.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated6.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated6.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated6.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/sampleSchema2.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema2.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema2.xsd	(revision 0)
@@ -0,0 +1,104 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<!-- mails.xsd -->
+
+<xsd:schema xmlns:xsd="http://www.example.org/2001/XMLSchema">
+
+ <xsd:element name="mails" type="mailsType" />
+
+ <xsd:complexType name="mailsType">
+  <xsd:sequence minOccurs="0" maxOccurs="unbounded">
+   <xsd:element name="mail" type="mailType" />
+  </xsd:sequence>
+ </xsd:complexType>
+
+ <xsd:complexType name="mailType">
+  <xsd:sequence>
+   <xsd:element name="envelope" type="envelopeType" />
+   <xsd:element name="body" type="bodyType" />
+   <xsd:element name="attachment" type="attachmentType"
+		minOccurs="0" maxOccurs="unbounded" />
+  </xsd:sequence>
+  <xsd:attribute use="required" name="id" type="xsd:integer" />
+ </xsd:complexType>
+
+ <xsd:element name="header">
+  <xsd:complexType>
+   <xsd:simpleContent>
+    <xsd:extension base="xsd:string">
+     <xsd:attribute ref="name" use="required" />
+    </xsd:extension>
+   </xsd:simpleContent>
+  </xsd:complexType>
+ </xsd:element>
+
+ <xsd:element name="Date" type="xsd:dateTime" />
+
+ <xsd:complexType name="envelopeType">
+  <xsd:sequence>
+   <xsd:element name="From" type="xsd:string" />
+   <xsd:element name="To" type="xsd:string" />
+   <xsd:element ref="Date" />
+   <xsd:element name="Subject" type="xsd:string" />
+   <xsd:element ref="header" minOccurs="0" maxOccurs="unbounded" />
+  </xsd:sequence>
+  <xsd:attribute name="From" type="xsd:string" use="required" />
+ </xsd:complexType>
+
+ <xsd:simpleType name="bodyType">
+  <xsd:restriction base="xsd:string" />
+ </xsd:simpleType>
+
+ <xsd:complexType name="attachmentType">
+  <xsd:group ref="attachmentContent" />
+  <xsd:attribute ref="name" use="required" />
+ </xsd:complexType>
+
+ <xsd:group name="attachmentContent">
+  <xsd:sequence>
+   <xsd:element name="mimetype">
+    <xsd:complexType>
+     <xsd:attributeGroup ref="mimeTypeAttributes" />
+    </xsd:complexType>
+   </xsd:element>
+   <xsd:element name="content" type="xsd:string" minOccurs="0" />
+  </xsd:sequence>
+ </xsd:group>
+
+ <xsd:attribute name="name" type="xsd:string" />
+
+ <xsd:attributeGroup name="mimeTypeAttributes">
+  <xsd:attribute name="type" type="mimeTopLevelType" use="required" />
+  <xsd:attribute name="subtype" type="xsd:string" use="required" />
+ </xsd:attributeGroup>
+
+ <xsd:simpleType name="mimeTopLevelType">
+  <xsd:restriction base="xsd:string">
+   <xsd:enumeration value="text" />
+   <xsd:enumeration value="multipart" />
+   <xsd:enumeration value="application" />
+   <xsd:enumeration value="message" />
+   <xsd:enumeration value="image" />
+   <xsd:enumeration value="audio" />
+   <xsd:enumeration value="video" />
+  </xsd:restriction>
+ </xsd:simpleType>
+
+</xsd:schema>
Index: test-resources/schemas/custom_schemas/sampleSchema3.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema3.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema3.xsd	(revision 0)
@@ -0,0 +1,46 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<xsd:schema xmlns:xsd="http://www.example.com/2001/XMLSchema"
+            targetNamespace="urn:books"
+            xmlns:bks="urn:books">
+
+  <xsd:element name="books" type="bks:BooksForm"/>
+
+  <xsd:complexType name="BooksForm">
+    <xsd:sequence>
+      <xsd:element name="book" 
+                  type="bks:BookForm" 
+                  minOccurs="0" 
+                  maxOccurs="unbounded"/>
+      </xsd:sequence>
+  </xsd:complexType>
+
+  <xsd:complexType name="BookForm">
+    <xsd:sequence>
+      <xsd:element name="author"   type="xsd:string"/>
+      <xsd:element name="title"    type="xsd:string"/>
+      <xsd:element name="genre"    type="xsd:string"/>
+      <xsd:element name="price"    type="xsd:float" />
+      <xsd:element name="pub_date" type="xsd:date" />
+      <xsd:element name="review"   type="xsd:string"/>
+    </xsd:sequence>
+    <xsd:attribute name="id"   type="xsd:string"/>
+  </xsd:complexType>
+</xsd:schema>
Index: test-resources/schemas/custom_schemas/generated7.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated7.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated7.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/generated8.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated8.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated8.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/sampleSchema4.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema4.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema4.xsd	(revision 0)
@@ -0,0 +1,54 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<xs:schema xmlns:xs="http://www.w3.org/2000/XMLSchema"
+	xmlns:ns="http://axisversion.sample" xmlns:ns1="http://org.apache.axis2/xsd"
+	attributeFormDefault="qualified" elementFormDefault="qualified"
+	targetNamespace="http://axisversion.sample">
+	<xs:element name="VersionException">
+		<xs:complexType>
+			<xs:sequence>
+				<xs:element minOccurs="0" name="VersionException"
+					nillable="true" type="ns:Exception" />
+			</xs:sequence>
+		</xs:complexType>
+	</xs:element>
+	<xs:complexType name="Exception">
+		<xs:sequence>
+			<xs:element minOccurs="0" name="Message" nillable="true"
+				type="xs:string" />
+		</xs:sequence>
+	</xs:complexType>
+	<xs:element name="getVersion">
+		<xs:complexType>
+			<xs:sequence />
+		</xs:complexType>
+	</xs:element>
+	<xs:element name="getVersionResponse">
+		<xs:complexType>
+			<xs:sequence>
+				<xs:element minOccurs="0" name="return" nillable="true"
+					type="xs:string" />
+			</xs:sequence>
+		</xs:complexType>
+	</xs:element>
+</xs:schema>
+
+
Index: test-resources/schemas/custom_schemas/generated9.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated9.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated9.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
Index: test-resources/schemas/custom_schemas/sampleSchema5.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema5.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema5.xsd	(revision 0)
@@ -0,0 +1,31 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+            xmlns:ord="http://www.w3schools.com">
+  <xs:import namespace="http://www.w3schools.com"
+              schemaLocation="test-resources/schemas/custom_schemas/sampleSchema1.xsd"/>
+  <xs:element name="envelope" type="EnvelopeType"/>
+  <xs:complexType name="EnvelopeType">
+    <xs:sequence>
+      <xs:element ref="ord:order" maxOccurs="unbounded"/>
+    </xs:sequence>
+  </xs:complexType>
+</xs:schema>
Index: test-resources/schemas/custom_schemas/sampleSchema6.xsd
===================================================================
--- test-resources/schemas/custom_schemas/sampleSchema6.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/sampleSchema6.xsd	(revision 0)
@@ -0,0 +1,65 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
+  <xs:element name="items" type="ItemsType"/>
+  <xs:complexType name="ItemsType">
+    <xs:choice minOccurs="0" maxOccurs="unbounded">
+      <xs:element name="shirt" type="ProductType"/>
+      <xs:element name="hat" type="ProductType"/>
+      <xs:element name="umbrella" type="ProductType"/>
+    </xs:choice>
+  </xs:complexType>
+  <!-- Element only content -->
+  <xs:complexType name="ProductType">
+    <xs:sequence>
+      <xs:element name="number" type="xs:integer"/>
+      <xs:element name="name" type="xs:string"/>
+      <xs:choice minOccurs="0" maxOccurs="unbounded">
+        <xs:element name="size" type="SizeType"/>
+        <xs:element name="color" type="ColorType"/>
+        <xs:element name="description" type="DescriptionType"/>
+      </xs:choice>
+    </xs:sequence>
+    <xs:attribute  name="effDate" type="xs:date"
+                    default="1900-01-01"/>
+    <xs:anyAttribute namespace="##other" processContents="lax"/>
+  </xs:complexType>
+  <!-- Simple content -->
+  <xs:complexType name="SizeType">
+    <xs:simpleContent>
+      <xs:extension base="xs:integer">
+        <xs:attribute name="system" type="xs:token"/>
+      </xs:extension>
+    </xs:simpleContent>
+  </xs:complexType>
+  <!-- Empty content -->
+  <xs:complexType name="ColorType">
+    <xs:attribute name="value" type="xs:string"/>
+  </xs:complexType>
+  <!-- Mixed content -->
+  <xs:complexType name="DescriptionType" mixed="true">
+    <xs:sequence>
+      <xs:any namespace="http://www.w3.org/1999/xhtml"
+      minOccurs="0" maxOccurs="unbounded"
+      processContents="skip"/>
+    </xs:sequence>
+  </xs:complexType>
+</xs:schema>
Index: test-resources/schemas/custom_schemas/generated.xsd
===================================================================
--- test-resources/schemas/custom_schemas/generated.xsd	(revision 0)
+++ test-resources/schemas/custom_schemas/generated.xsd	(revision 0)
@@ -0,0 +1,21 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements. See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership. The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License. You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+
+<schema xmlns="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified"/>
