Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/ClientConfigurationFactory.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/ClientConfigurationFactory.java	(revision 448235)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/ClientConfigurationFactory.java	(working copy)
@@ -29,18 +29,16 @@
  */
 public class ClientConfigurationFactory {
     
-    private static ClientConfigurationFactory instance = new ClientConfigurationFactory();
-    
     private ConfigurationContext configContext = null;
     
     protected ClientConfigurationFactory(){
     }
 
     /**
-     * Returns the singleton ClientConfigurationFactory object.
+     * Returns a ClientConfigurationFactory object.
      */
     public static ClientConfigurationFactory newInstance() {
-        return instance;
+        return new ClientConfigurationFactory();
     }
 
     /**



