Index: src/org/apache/axis2/description/AxisService.java
===================================================================
--- src/org/apache/axis2/description/AxisService.java	(revision 1299986)
+++ src/org/apache/axis2/description/AxisService.java	(working copy)
@@ -47,6 +47,7 @@
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.neethi.Policy;
+import org.apache.woden.wsdl20.Description;
 import org.apache.ws.commons.schema.XmlSchema;
 import org.apache.ws.commons.schema.XmlSchemaElement;
 import org.apache.ws.commons.schema.XmlSchemaExternal;
@@ -1155,7 +1156,24 @@
 		}
 
 	}
+	
+    public void printUserWSDL2(OutputStream out, String wsdlName, String ip) throws AxisFault {
+        Description description = null;
+        // first find the correct wsdl definition
+        Parameter wsdlParameter = getParameter(WSDLConstants.WSDL_20_DESCRIPTION);
+        if (wsdlParameter != null) {
+            description = (Description) wsdlParameter.getValue();
+            try {
+                org.apache.woden.WSDLFactory factory = org.apache.woden.WSDLFactory.newInstance();
+                org.apache.woden.WSDLWriter writer = factory.newWSDLWriter();
+                writer.writeWSDL(description.toElement(), out);
+            } catch (org.apache.woden.WSDLException e) {
+                throw AxisFault.makeFault(e);
+            }
+        }
 
+}
+
 	/**
 	 * find the defintion object for given name
 	 * 
@@ -1747,6 +1765,11 @@
 	 * @throws AxisFault
 	 */
 	public void printWSDL2(OutputStream out, String requestIP) throws AxisFault {
+	 // If we're looking for pre-existing WSDL, use that.
+        if (isUseUserWSDL()) {
+            printUserWSDL2(out, null, requestIP);
+            return;
+        }
 		AxisService2WSDL20 axisService2WSDL2 = new AxisService2WSDL20(this);
 		try {
 			if (requestIP != null) {
