Index: xdocs/tools/1_1/CodegenToolReference.html
===================================================================
--- xdocs/tools/1_1/CodegenToolReference.html	(revision 547210)
+++ xdocs/tools/1_1/CodegenToolReference.html	(working copy)
@@ -245,6 +245,11 @@
         tool.</td>
 </tr>
 <tr>
+    <td>typesystemname</td>
+    <td>The typesystemname to use for XMLBeans data binding. Maps to the -tsn option of the command line
+        tool.</td>
+</tr>
+<tr>
     <td>portName</td>
     <td>The name of the port in the presence of multiple ports. Maps to -pn options of the command line
         tool.</td>
Index: xdocs/tools/1_1/maven-plugins/maven-wsdl2code-plugin.html
===================================================================
--- xdocs/tools/1_1/maven-plugins/maven-wsdl2code-plugin.html	(revision 547210)
+++ xdocs/tools/1_1/maven-plugins/maven-wsdl2code-plugin.html	(working copy)
@@ -173,6 +173,13 @@
       <td>both</td>
     </tr>
     <tr>
+      <td>typesystemname</td>
+      <td>${axis2.wsdl2code.typesystemname}</td>
+      <td>The typesystemname to use for XMLBeans data binding. By default,
+        the generated classes include a unique (but random) STS name.</td>
+      <td></td>
+    </tr>
+    <tr>
       <td>unpackClasses</td>
       <td>${axis2.wsdl2code.unpackClasses}</td>
       <td>Whether to unpack classes.</td>
Index: xdocs/tools/1_2/CodegenToolReference.html
===================================================================
--- xdocs/tools/1_2/CodegenToolReference.html	(revision 547210)
+++ xdocs/tools/1_2/CodegenToolReference.html	(working copy)
@@ -245,6 +245,11 @@
         tool.</td>
 </tr>
 <tr>
+    <td>typesystemname</td>
+    <td>The typesystemname to use for XMLBeans data binding. Maps to the -tsn option of the command line
+        tool.</td>
+</tr>
+<tr>
     <td>portName</td>
     <td>The name of the port in the presence of multiple ports. Maps to -pn options of the command line
         tool.</td>
Index: xdocs/tools/1_2/maven-plugins/maven-wsdl2code-plugin.html
===================================================================
--- xdocs/tools/1_2/maven-plugins/maven-wsdl2code-plugin.html	(revision 547210)
+++ xdocs/tools/1_2/maven-plugins/maven-wsdl2code-plugin.html	(working copy)
@@ -173,6 +173,13 @@
       <td>both</td>
     </tr>
     <tr>
+      <td>typesystemname</td>
+      <td>${axis2.wsdl2code.typesystemname}</td>
+      <td>The typesystemname to use for XMLBeans data binding. By default,
+        the generated classes include a unique (but random) STS name.</td>
+      <td></td>
+    </tr>
+    <tr>
       <td>unpackClasses</td>
       <td>${axis2.wsdl2code.unpackClasses}</td>
       <td>Whether to unpack classes.</td>
Index: modules/kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java
===================================================================
--- modules/kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java	(revision 547238)
+++ modules/kernel/src/org/apache/axis2/util/CommandLineOptionConstants.java	(working copy)
@@ -35,6 +35,7 @@
         String GENERATE_ALL_OPTION = "g";
         String PORT_NAME_OPTION = "pn";
         String SERVICE_NAME_OPTION = "sn";
+        String TYPESYSTEMNAME_OPTION = "tsn";
         String REPOSITORY_PATH_OPTION = "r";
         String NAME_SPACE_TO_PACKAGE_OPTION = "ns2p";
         String SERVER_SIDE_INTERFACE_OPTION = "ssi";
@@ -62,6 +63,7 @@
         String GENERATE_ALL_OPTION_LONG = "generate-all";
         String PORT_NAME_OPTION_LONG = "port-name";
         String SERVICE_NAME_OPTION_LONG = "service-name";
+        String TYPESYSTEMNAME_OPTION_LONG = "typesystemname";
         String INVALID_OPTION = "INVALID_OPTION";
         String EXTRA_OPTIONTYPE_PREFIX = "E";
         String REPOSITORY_PATH_OPTION_LONG = "repository-path";
Index: modules/codegen/src/org/apache/axis2/wsdl/i18n/resource.properties
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/i18n/resource.properties	(revision 547210)
+++ modules/codegen/src/org/apache/axis2/wsdl/i18n/resource.properties	(working copy)
@@ -38,6 +38,7 @@
 wsdl2code.arg22=\t  -f                      Flattens the generated files
 wsdl2code.arg23=\t  -uw                      Switch on un-wrapping.
 wsdl2code.arg24=\t  -xsdconfig <path to file> Use XMLBeans .xsdconfig file. Valid only with -d xmlbeans.
+wsdl2code.arg25=\t  -tsn <typesystemname>     Set XMLBeans typesystemname. Valid only with -d xmlbeans.
 #wsdl2code.arg19=\t  -b                       Generate Axis 1.x backword compatible code.
 
 ################## prop file loader #################################
Index: modules/codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java	(revision 547210)
+++ modules/codegen/src/org/apache/axis2/wsdl/codegen/CodegenConfigLoader.java	(working copy)
@@ -136,6 +136,11 @@
         config.setServiceName(
                 commandLineOption != null ? commandLineOption.getOptionValue() : null);
 
+        commandLineOption = loadOption(WSDL2JavaConstants.TYPESYSTEMNAME_OPTION,
+                WSDL2JavaConstants.TYPESYSTEMNAME_OPTION_LONG, optionMap);
+        config.setTypesystemname(
+                commandLineOption != null ? commandLineOption.getOptionValue() : null);
+
         commandLineOption = loadOption(WSDL2JavaConstants.REPOSITORY_PATH_OPTION,
                                        WSDL2JavaConstants.REPOSITORY_PATH_OPTION_LONG, optionMap);
         config.setRepositoryPath(
Index: modules/codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java	(revision 547210)
+++ modules/codegen/src/org/apache/axis2/wsdl/codegen/CodeGenConfiguration.java	(working copy)
@@ -101,6 +101,27 @@
         this.xsdConfigFile = xsdConfigFile;
     }
 
+    /** Value for XMLBeans typesystemname to use */
+    private String typesystemname = null;
+
+    /**
+     * Returns the typesystemname to use for XMLBeans data binding.
+     *
+     * @return Value for XMLBeans typesystemname to use
+     */
+    public String getTypesystemname() {
+        return typesystemname;
+    }
+
+    /**
+     * Sets the typesystemname to use for XMLBeans data binding.
+     *
+     * @param Value for XMLBeans typesystemname to use
+     */
+    public void setTypesystemname(String typesystemname) {
+        this.typesystemname = typesystemname;
+    }
+
     /** Says whether to flatten the files or put them in specific folders */
     private boolean flattenFiles = false;
 
Index: modules/codegen/src/org/apache/axis2/wsdl/WSDL2Code.java
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/WSDL2Code.java	(revision 547210)
+++ modules/codegen/src/org/apache/axis2/wsdl/WSDL2Code.java	(working copy)
@@ -59,6 +59,7 @@
         System.out.println(CodegenMessages.getMessage("wsdl2code.arg22"));
         System.out.println(CodegenMessages.getMessage("wsdl2code.arg23"));
         System.out.println(CodegenMessages.getMessage("wsdl2code.arg24"));
+        System.out.println(CodegenMessages.getMessage("wsdl2code.arg25"));
 
         System.exit(0);
     }
Index: modules/codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java
===================================================================
--- modules/codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java	(revision 547210)
+++ modules/codegen/src/org/apache/axis2/wsdl/util/WSDL2JavaOptionsValidator.java	(working copy)
@@ -55,6 +55,7 @@
                             (WSDL2JavaConstants.GENERATE_ALL_OPTION).equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.PORT_NAME_OPTION).equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.SERVICE_NAME_OPTION).equalsIgnoreCase(optionType) ||
+                            (WSDL2JavaConstants.TYPESYSTEMNAME_OPTION).equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.REPOSITORY_PATH_OPTION)
                                     .equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.NAME_SPACE_TO_PACKAGE_OPTION)
@@ -100,6 +101,8 @@
                                     .equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.SERVICE_NAME_OPTION_LONG)
                                     .equalsIgnoreCase(optionType) ||
+                            (WSDL2JavaConstants.TYPESYSTEMNAME_OPTION_LONG)
+                                    .equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.GENERATE_ALL_OPTION_LONG)
                                     .equalsIgnoreCase(optionType) ||
                             (WSDL2JavaConstants.NAME_SPACE_TO_PACKAGE_OPTION_LONG)
Index: modules/xmlbeans/src/org/apache/axis2/xmlbeans/CodeGenerationUtility.java
===================================================================
--- modules/xmlbeans/src/org/apache/axis2/xmlbeans/CodeGenerationUtility.java	(revision 547210)
+++ modules/xmlbeans/src/org/apache/axis2/xmlbeans/CodeGenerationUtility.java	(working copy)
@@ -172,10 +172,10 @@
 
 
             sts = XmlBeans.compileXmlBeans(
-                    //set the STS name to null. it makes the generated class
+                    // set the STS name; defaults to null, which makes the generated class
                     // include a unique (but random) STS name
+                    cgconfig.getTypesystemname(),
                     null,
-                    null,
                     convertToSchemaArray(topLevelSchemaList),
                     new Axis2BindingConfig(cgconfig.getUri2PackageNameMap(),
                                            cgconfig.getXsdConfigFile()),
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/bean/CodegenBean.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/bean/CodegenBean.java	(revision 547210)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/bean/CodegenBean.java	(working copy)
@@ -60,6 +60,7 @@
     private boolean isGenerateAll;
     private boolean isTestCase;
     private String serviceName;
+    private String typesystemname;
     private String portName;
     private String databindingName;
 
@@ -130,6 +131,14 @@
         this.serviceName = serviceName;
     }
 
+    public String getTypesystemname() {
+        return typesystemname;
+    }
+
+    public void setTypesystemname(String typesystemname) {
+        this.typesystemname = typesystemname;
+    }
+
     public String getPortName() {
         return portName;
     }
@@ -225,6 +234,11 @@
             optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION, new CommandLineOption(
                     CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION, getStringArray(serviceName)));
         }
+        //typesystemname
+        if (typesystemname != null) {
+            optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION, new CommandLineOption(
+                    CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION, getStringArray(typesystemname)));
+        }
         //server side interface  mapping
         if (isServerSideInterface){
             optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.SERVER_SIDE_INTERFACE_OPTION, new CommandLineOption(
Index: modules/tool/axis2-ant-plugin/src/main/java/org/apache/axis2/tool/ant/AntCodegenTask.java
===================================================================
--- modules/tool/axis2-ant-plugin/src/main/java/org/apache/axis2/tool/ant/AntCodegenTask.java	(revision 547210)
+++ modules/tool/axis2-ant-plugin/src/main/java/org/apache/axis2/tool/ant/AntCodegenTask.java	(working copy)
@@ -40,6 +40,7 @@
     private String databindingName = ConfigPropertyFileLoader.getDefaultDBFrameworkName();
     private String portName = null;
     private String serviceName = null;
+    private String typesystemname = null;
 
     private boolean asyncOnly = false;
     private boolean syncOnly = false;
@@ -275,6 +276,14 @@
                         CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION,
                         new String[] { serviceName }));
 
+        if (typesystemname != null) {
+            optionMap.put(
+                CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION,
+                new CommandLineOption(
+                        CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION,
+                        new String[] { typesystemname }));
+        }
+
         optionMap.put(
                 CommandLineOptionConstants.WSDL2JavaConstants.PORT_NAME_OPTION,
                 new CommandLineOption(
@@ -336,6 +345,10 @@
         this.serviceName = serviceName;
     }
 
+    public void setTypesystemname(String typesystemname) {
+        this.typesystemname = typesystemname;
+    }
+
     public void setGenerateAllClasses(boolean generateAllClasses) {
         this.generateAllClasses = generateAllClasses;
     }
Index: modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/eclipse/CodeGenWizard.java
===================================================================
--- modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/eclipse/CodeGenWizard.java	(revision 547210)
+++ modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/eclipse/CodeGenWizard.java	(working copy)
@@ -239,6 +239,7 @@
                                                                                                 optionsPage.isGenerateTestCase(),
                                                                                                 optionsPage.getGenerateAll(),
                                                                                                 optionsPage.getServiceName(),
+                                                                                                null, // TODO: add typesystemname option!
                                                                                                 optionsPage.getPortName(),
                                                                                                 optionsPage.getDatabinderName(),
                                                                                                 wsdlSelectionPage.getFileName(),
Index: modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/WSDL2JavaGenerator.java
===================================================================
--- modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/WSDL2JavaGenerator.java	(revision 547210)
+++ modules/tool/axis2-eclipse-codegen-plugin/src/main/java/org/apache/axis2/tool/codegen/WSDL2JavaGenerator.java	(working copy)
@@ -56,6 +56,7 @@
             		  boolean isTestCase,
             		  boolean isGenerateAll,
             		  String serviceName,
+            		  String typesystemname,
             		  String portName,
             		  String databindingName,
             		  String WSDLURI,
@@ -129,6 +130,11 @@
 	       optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION, new CommandLineOption(
 	    		   CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION, getStringArray(serviceName)));
        }
+       //typesystemname
+       if (typesystemname!= null){
+	       optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION, new CommandLineOption(
+	    		   CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION, getStringArray(typesystemname)));
+       }
        //ns2pkg mapping
        if (namespace2packageList!= null){
 	       optionMap.put(CommandLineOptionConstants.WSDL2JavaConstants.NAME_SPACE_TO_PACKAGE_OPTION, new CommandLineOption(
Index: modules/tool/axis2-wsdl2code-maven-plugin/src/main/java/org/apache/axis2/maven2/wsdl2code/WSDL2CodeMojo.java
===================================================================
--- modules/tool/axis2-wsdl2code-maven-plugin/src/main/java/org/apache/axis2/maven2/wsdl2code/WSDL2CodeMojo.java	(revision 547210)
+++ modules/tool/axis2-wsdl2code-maven-plugin/src/main/java/org/apache/axis2/maven2/wsdl2code/WSDL2CodeMojo.java	(working copy)
@@ -129,6 +129,14 @@
     private String serviceName;
 
     /**
+     * The typesystemname to use for XMLBeans data binding. By default, the generated classes
+     * include a unique (but random) STS name.
+     *
+     * @parameter expression="${axis2.wsdl2code.typesystemname}"
+     */
+    private String typesystemname;
+
+    /**
      * Mode, for which sources are being generated; either of "sync", "async" or "both".
      *
      * @parameter expression="${axis2.wsdl2code.syncMode}" default-value="both"
@@ -355,6 +363,14 @@
                         CommandLineOptionConstants.WSDL2JavaConstants.SERVICE_NAME_OPTION,
                         new String[] { serviceName }));
 
+        if (typesystemname != null) {
+          optionMap.put(
+              CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION,
+              new CommandLineOption(
+                      CommandLineOptionConstants.WSDL2JavaConstants.TYPESYSTEMNAME_OPTION,
+                      new String[] { typesystemname }));
+        }
+
         optionMap.put(
                 CommandLineOptionConstants.WSDL2JavaConstants.PORT_NAME_OPTION,
                 new CommandLineOption(
Index: modules/tool/axis2-wsdl2code-maven-plugin/src/site/apt/configuration.apt
===================================================================
--- modules/tool/axis2-wsdl2code-maven-plugin/src/site/apt/configuration.apt	(revision 547210)
+++ modules/tool/axis2-wsdl2code-maven-plugin/src/site/apt/configuration.apt	(working copy)
@@ -64,6 +64,10 @@
 |                             | both                                           | generated; either of "sync", "async", or       |
 |                             |                                                | "both" (default).                              |
 *-----------------------------+------------------------------------------------+------------------------------------------------+
+| typesystemname              | ${axis2.wsdl2code.typesystemname}              | The typesystemname to use for XMLBeans data    |
+|                             |                                                | binding. By default, the generated classes     |
+|                             |                                                | include a unique (but random) STS name.        |
+*-----------------------------+------------------------------------------------+------------------------------------------------+
 | unpackClasses               | ${axis2.wsdl2code.unpackClasses}               | Whether to unpack classes.                     |
 *-----------------------------+------------------------------------------------+------------------------------------------------|
 | wsdlFile                    | ${axis2.wsdl2code.wsdl}                        | Location of the WSDL file, which is read as    |
