Index: /home/hd2/work/axis2/trunk/java/modules/jaxws/src/org/apache/axis2/jaxws/message/databinding/JAXBUtils.java
===================================================================
--- /home/hd2/work/axis2/trunk/java/modules/jaxws/src/org/apache/axis2/jaxws/message/databinding/JAXBUtils.java	(revision 499659)
+++ /home/hd2/work/axis2/trunk/java/modules/jaxws/src/org/apache/axis2/jaxws/message/databinding/JAXBUtils.java	(working copy)
@@ -391,8 +391,7 @@
 	        //Catch Throwable as ClassLoader can throw an NoClassDefFoundError that
 	        //does not extend Exception. So we will absorb any Throwable exception here.
 	    } catch (Throwable e) {
-	        // Exception is thrown if class does not exist.
-	    	e.printStackTrace();
+	       log.info("ObjectFactory Class Not Found");
 	    }
 	    
         try {
@@ -403,8 +402,7 @@
             //Catch Throwable as ClassLoader can throw an NoClassDefFoundError that
 	        //does not extend Exception. So we will absorb any Throwable exception here.
         } catch (Throwable e) {
-            // Exception is thrown if class does not exist.
-        	e.printStackTrace();
+            log.info("package-info Class Not Found");
         }
 
 	    return false;
