Index: src/javax/xml/soap/SOAPConnection.java
===================================================================
--- src/javax/xml/soap/SOAPConnection.java	(revision 495207)
+++ src/javax/xml/soap/SOAPConnection.java	(working copy)
@@ -61,7 +61,14 @@
      */
     public abstract void close() throws SOAPException;
 
-    public SOAPMessage get(Object obj)
+    /**
+     * Gets a message from a specific endpoint and blocks until it receives,
+     * @param to - an Object that identifies where the request should be sent. Objects of type
+     * java.lang.String and java.net.URL must be supported.
+     * @return the SOAPMessage object that is the response to the get message request
+     * @throws SOAPException - if there is a SOAP error
+     */
+    public SOAPMessage get(Object to)
             throws SOAPException {
         throw new UnsupportedOperationException();
     }
Index: src/javax/xml/soap/SOAPFault.java
===================================================================
--- src/javax/xml/soap/SOAPFault.java	(revision 495207)
+++ src/javax/xml/soap/SOAPFault.java	(working copy)
@@ -209,7 +209,7 @@
      */
     public abstract Locale getFaultStringLocale();
 
-    public abstract void addFaultReasonText(java.lang.String s,
+    public abstract void addFaultReasonText(java.lang.String reasonText,
                         java.util.Locale locale)
                         throws SOAPException;
 
