Index: MTOMAssertion.java
===================================================================
--- MTOMAssertion.java	(revision 736025)
+++ MTOMAssertion.java	(working copy)
@@ -21,14 +21,33 @@
 
 import org.apache.neethi.Assertion;
 
+/**
+ * 
+ * This abstract class specifies the common features of a MTOM assertion. 
+ *
+ */
 public abstract class MTOMAssertion implements Assertion {
 
+	/** Specifies if the MTOM assertion is optional. The request can be MTOMised or non-MTOMised,</br>
+	 *  but the response will be MTOMised only if request is MTOMised. */
     protected boolean optional = false;
     
+    /**
+     * Checks if the MTOM assertion is optional. The request can be MTOMised or non-MTOMised,</br>
+	 * but the response will be MTOMised only if request is MTOMised.
+	 * 
+	 * @return <code>true</code> if the MTOM assertion is optional, otherwise returns <code>false</code>.
+     */
     public boolean isOptional() {
         return optional;
     }
     
+    /**
+     * Sets the <code>optional</code> parameter.  
+     * 
+     * @param isOptional sets if the MTOM assertion is optional or not. If set to <code>true</code> </br>
+     * then if the request is MTOMised then the response should be MTOMised, too.
+     */
     public void setOptional(boolean isOptional) {
         this.optional = isOptional;
     }
