Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/HTTPTransportUtils.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/HTTPTransportUtils.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/HTTPTransportUtils.java	(working copy)
@@ -250,6 +250,7 @@
                         builder =
                                 new StAXSOAPModelBuilder(xmlreader,
                                                          SOAP12Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+                        builder.enableFilter(msgContext);
                         envelope = (SOAPEnvelope) builder.getDocumentElement();
                     } else if (contentType.indexOf(SOAP11Constants.SOAP_11_CONTENT_TYPE) > -1) {
                         soap11 = true;
@@ -270,6 +271,7 @@
                                 SOAPFactory soapFactory = new SOAP11Factory();
 
                                 builder = new StAXOMBuilder(xmlreader);
+                                builder.enableFilter(msgContext);
                                 builder.setOMBuilderFactory(soapFactory);
                                 envelope = soapFactory.getDefaultEnvelope();
                                 envelope.getBody().addChild(builder.getDocumentElement());
@@ -277,6 +279,7 @@
                         } else {
                             builder = new StAXSOAPModelBuilder(
                                     xmlreader, SOAP11Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+                            builder.enableFilter(msgContext);
                             envelope = (SOAPEnvelope) builder.getDocumentElement();
                         }
                     }
@@ -292,6 +295,7 @@
                                        MessageContext.DEFAULT_CHAR_SET_ENCODING);
                 builder = new StAXSOAPModelBuilder(
                         xmlreader, SOAP11Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+                builder.enableFilter(msgContext);
                 envelope = (SOAPEnvelope) builder.getDocumentElement();
             }
 
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/http/util/RESTUtil.java	(working copy)
@@ -198,7 +198,9 @@
                     msgCtxt.setProperty(Constants.Configuration.CHARACTER_SET_ENCODING, charSetEnc);
                 }
 
-                OMNodeEx documentElement = (OMNodeEx)new StAXOMBuilder(xmlreader).getDocumentElement();
+                StAXOMBuilder builder = new StAXOMBuilder(xmlreader);
+                builder.enableFilter(msgCtxt);
+                OMNodeEx documentElement = (OMNodeEx) builder.getDocumentElement();
                 documentElement.setParent(null);
                 body.addChild(documentElement);
 
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/SimpleMailListener.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/SimpleMailListener.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/SimpleMailListener.java	(working copy)
@@ -304,6 +304,7 @@
             }
 
             StAXBuilder builder = new StAXSOAPModelBuilder(reader, soapNamespaceURI);
+            builder.enableFilter(msgContext);
             SOAPEnvelope envelope = (SOAPEnvelope) builder.getDocumentElement();
             msgContext.setEnvelope(envelope);
         }
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/server/MailSorter.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/server/MailSorter.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/mail/server/MailSorter.java	(working copy)
@@ -127,6 +127,7 @@
             }
 
             StAXBuilder builder = new StAXSOAPModelBuilder(reader, soapNamespaceURI);
+            builder.enableFilter(msgContext);
             SOAPEnvelope envelope = (SOAPEnvelope) builder.getDocumentElement();
 
             msgContext.setEnvelope(envelope);
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/tcp/TCPWorker.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/tcp/TCPWorker.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/tcp/TCPWorker.java	(working copy)
@@ -83,6 +83,7 @@
                 Reader in = new InputStreamReader(socket.getInputStream());
                 XMLStreamReader xmlreader = StAXUtils.createXMLStreamReader(in);
                 StAXBuilder builder = new StAXSOAPModelBuilder(xmlreader, null);
+                builder.enableFilter(msgContext);
                 SOAPEnvelope envelope = (SOAPEnvelope) builder.getDocumentElement();
 
                 msgContext.setEnvelope(envelope);
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/jms/JMSUtils.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/jms/JMSUtils.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/jms/JMSUtils.java	(working copy)
@@ -332,11 +332,13 @@
                     // it is SOAP 1.2
                     builder = new StAXSOAPModelBuilder(xmlreader,
                                                        SOAP12Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+                    builder.enableFilter(msgContext);
                     envelope = (SOAPEnvelope) builder.getDocumentElement();
                 } else if (contentType.indexOf(SOAP11Constants.SOAP_11_CONTENT_TYPE) > -1) {
                     // SOAP 1.1
                     builder =  new StAXSOAPModelBuilder(xmlreader,
                                                         SOAP11Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+                    builder.enableFilter(msgContext);
                     envelope = (SOAPEnvelope) builder.getDocumentElement();
                 }
             }
@@ -351,6 +353,7 @@
                                    MessageContext.DEFAULT_CHAR_SET_ENCODING);
             builder = new StAXSOAPModelBuilder(
                 xmlreader, SOAP11Constants.SOAP_ENVELOPE_NAMESPACE_URI);
+            builder.enableFilter(msgContext);
             envelope = (SOAPEnvelope) builder.getDocumentElement();
         }
 
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -74,6 +74,7 @@
             XMLStreamReader reader = StAXUtils.createXMLStreamReader(
                     new BufferedReader(new InputStreamReader(in)));
             StAXBuilder builder = new StAXSOAPModelBuilder(reader, null);
+            builder.enableFilter(msgCtx);
             SOAPEnvelope envelope = (SOAPEnvelope) builder.getDocumentElement();
 
             msgCtx.setEnvelope(envelope);
Index: D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/TransportUtils.java
===================================================================
--- D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/TransportUtils.java	(revision 449532)
+++ D:/soc/workspaces/axis2/axis2_enableFitler/modules/kernel/src/org/apache/axis2/transport/TransportUtils.java	(working copy)
@@ -103,6 +103,7 @@
                 SOAPFactory soapFactory = new SOAP11Factory();
 
                 builder = new StAXOMBuilder(xmlreader);
+                builder.enableFilter(msgContext);
                 builder.setOMBuilderFactory(soapFactory);
                 envelope = soapFactory.getDefaultEnvelope();
                 envelope.getBody().addChild(builder.getDocumentElement());
@@ -117,6 +118,7 @@
                         StAXUtils.createXMLStreamReader(inStream, charSetEnc);
 
                 builder = new StAXSOAPModelBuilder(xmlreader, soapNamespaceURI);
+                builder.enableFilter(msgContext);
                 envelope = (SOAPEnvelope) builder.getDocumentElement();
             }
 
@@ -265,10 +267,12 @@
             * Creates the MTOM specific MTOMStAXSOAPModelBuilder
             */             
             builder = new MTOMStAXSOAPModelBuilder(streamReader, attachments, soapEnvelopeNamespaceURI);
-        	 
+            builder.enableFilter(msgContext);
+
         } else if (attachments.getAttachmentSpecType().equals(MTOMConstants.SWA_TYPE)& null!=soapEnvelopeNamespaceURI) {
             builder = new StAXSOAPModelBuilder(streamReader,
                     soapEnvelopeNamespaceURI);
+            builder.enableFilter(msgContext);
         }
 
         return builder;
