Index: modules/adb/src/org/apache/axis2/databinding/typemapping/SimpleTypeMapper.java
===================================================================
--- modules/adb/src/org/apache/axis2/databinding/typemapping/SimpleTypeMapper.java	(revision 1213174)
+++ modules/adb/src/org/apache/axis2/databinding/typemapping/SimpleTypeMapper.java	(working copy)
@@ -79,10 +79,9 @@
     
     private static final String SQL_DATE_FORMAT = "yyyy-MM-dd";
 
-    public static Object getSimpleTypeObject(Class parameter, OMElement value) {
+    public static Object getSimpleTypeObject(Class parameter, String text) {
         String name = parameter.getName();
-        String text = value.getText();
-        
+
         if(name.equals(STRING)) {
             return text;
         } else  if (text == null || text.length() == 0) {
@@ -168,6 +167,10 @@
         }
     }
 
+    public static Object getSimpleTypeObject(Class parameter, OMElement value) {
+        return getSimpleTypeObject(parameter, value.getText());
+    }
+
     public static ArrayList getArrayList(OMElement element, String localName) {
         Iterator childitr = element.getChildrenWithName(new QName(localName));
         ArrayList list = new ArrayList();
