Index: modules/transport/http/src/org/apache/axis2/transport/http/AxisServlet.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/AxisServlet.java	(revision 1177716)
+++ modules/transport/http/src/org/apache/axis2/transport/http/AxisServlet.java	(working copy)
@@ -92,7 +92,7 @@
     public static final String SESSION_ID = "SessionId";
     
     private static final Set<String> metadataQueryParamNames = new HashSet<String>(
-            Arrays.asList("wsdl2", "wsdl", "xsd", "policy"));
+            Arrays.asList("wsdl2", "WSDL2", "wsdl", "WSDL", "xsd", "policy"));
     
     protected transient ConfigurationContext configContext;
     protected transient AxisConfiguration axisConfiguration;
Index: modules/transport/http/src/org/apache/axis2/transport/http/HTTPWorker.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/HTTPWorker.java	(revision 1177716)
+++ modules/transport/http/src/org/apache/axis2/transport/http/HTTPWorker.java	(working copy)
@@ -115,7 +115,7 @@
                     }
                 }
             }
-            if (uri.endsWith("?wsdl2")) {
+            if (uri.endsWith("?wsdl2") ||uri.endsWith("?WSDL2") ) {
                 String serviceName = uri.substring(uri.lastIndexOf("/") + 1, uri.length() - 6);
                 HashMap services = configurationContext.getAxisConfiguration().getServices();
                 AxisService service = (AxisService) services.get(serviceName);
@@ -131,7 +131,7 @@
                     return;
                 }
             }
-            if (uri.endsWith("?wsdl")) {
+            if (uri.endsWith("?wsdl") || uri.endsWith("?WSDL")) {
                 /**
                  * service name can be hierarchical (axis2/services/foo/1.0.0/Version?wsdl) or
                  * normal (axis2/services/Version?wsdl).
@@ -250,14 +250,28 @@
                     }
                 }
             }
-            if (uri.indexOf("?wsdl2=") > 0) {
-                String serviceName =
-                        uri.substring(uri.lastIndexOf("/") + 1, uri.lastIndexOf("?wsdl2="));
+            if (HttpUtils.getIndexOf(uri , "?wsdl2=") > 0 ) {
+                String serviceName;
+                boolean isLowerCase = ((uri.indexOf("?wsdl2=") > 0));
+                if(isLowerCase){
+                    serviceName=
+                            uri.substring(uri.lastIndexOf("/") + 1, uri.lastIndexOf("?wsdl2="));
+                }else{
+                    serviceName =
+                            uri.substring(uri.lastIndexOf("/") + 1, uri.lastIndexOf("?WSDL2="));
+                }
                 if (processInternalWSDL(uri, configurationContext, serviceName, response, getHost(request))) return;
             }
-            if (uri.indexOf("?wsdl=") > 0) {
-                String serviceName =
+            if (HttpUtils.getIndexOf(uri , "?wsdl=") > 0) {
+                String serviceName;
+                boolean isLowerCase = (uri.indexOf("?wsdl=") > 0);
+                if(isLowerCase){
+                    serviceName =
                         uri.substring(uri.lastIndexOf("/") + 1, uri.lastIndexOf("?wsdl="));
+                }else{
+                    serviceName =
+                        uri.substring(uri.lastIndexOf("/") + 1, uri.lastIndexOf("?WSDL="));
+                }
                 if (processInternalWSDL(uri, configurationContext, serviceName, response, getHost(request))) return;
             }
 
Index: modules/transport/http/src/org/apache/axis2/transport/http/server/HttpUtils.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/server/HttpUtils.java	(revision 1177716)
+++ modules/transport/http/src/org/apache/axis2/transport/http/server/HttpUtils.java	(working copy)
@@ -19,17 +19,9 @@
 
 package org.apache.axis2.transport.http.server;
 
-import org.apache.axis2.description.Parameter;
-import org.apache.axis2.engine.AxisConfiguration;
-import org.apache.axis2.transport.TransportListener;
 import org.apache.axis2.transport.http.HTTPConstants;
 import org.apache.http.Header;
 
-import java.net.InetAddress;
-import java.net.NetworkInterface;
-import java.net.SocketException;
-import java.util.Enumeration;
-
 public class HttpUtils {
 
     private HttpUtils() {
@@ -47,4 +39,20 @@
         }
     }
 
+    /**
+     *
+     * @param uri  - a String
+     * @param strPart - specified substring to be checked in uppercase and lowercase
+     * @return  if the specified String occurs as a substring within first String argument in fully lowercase or fully uppercase , then
+     * the index of the first character of the first such substring is returned; if it does not occur as a substring, -1 is returned.
+     */
+    public static int getIndexOf(String uri , String strPart){
+        String lowerCase = strPart.toLowerCase();
+        int index = uri.indexOf(lowerCase);
+        if(index < 0){
+            String upperCase = lowerCase.toUpperCase();
+            index = uri.indexOf(upperCase);
+        }
+        return index;
+    }
 }
Index: modules/transport/http/src/org/apache/axis2/transport/http/ListingAgent.java
===================================================================
--- modules/transport/http/src/org/apache/axis2/transport/http/ListingAgent.java	(revision 1177716)
+++ modules/transport/http/src/org/apache/axis2/transport/http/ListingAgent.java	(working copy)
@@ -26,6 +26,7 @@
 import org.apache.axis2.description.AxisService;
 import org.apache.axis2.description.PolicyInclude;
 import org.apache.axis2.description.Parameter;
+import org.apache.axis2.transport.http.server.HttpUtils;
 import org.apache.axis2.util.ExternalPolicySerializer;
 import org.apache.axis2.util.IOUtils;
 import org.apache.axis2.util.JavaUtils;
@@ -67,7 +68,7 @@
 
         String query = httpServletRequest.getQueryString();
         if (query != null) {
-            if (query.indexOf("wsdl2") > 0 || query.indexOf("wsdl") > 0 ||
+            if (HttpUtils.getIndexOf(query, "wsdl2") > 0 || HttpUtils.getIndexOf(query , "wsdl") > 0 ||
                 query.indexOf("xsd") > 0 || query.indexOf("policy") > 0) {
                 processListService(httpServletRequest, httpServletResponse);
             } else {
@@ -167,8 +168,8 @@
         String serviceName = extractServiceName(url);
         HashMap services = configContext.getAxisConfiguration().getServices();
         String query = req.getQueryString();
-        int wsdl2 = query.indexOf("wsdl2");
-        int wsdl = query.indexOf("wsdl");
+        int wsdl2 = HttpUtils.getIndexOf(query, "wsdl2");
+        int wsdl = HttpUtils.getIndexOf(query, "wsdl");
         int xsd = query.indexOf("xsd");
         int policy = query.indexOf("policy");
 
@@ -320,6 +321,9 @@
         res.setContentType("text/xml");
         String ip = extractHost(url);
         String wsdlName = req.getParameter("wsdl");
+        if(wsdlName==null) {
+            wsdlName = req.getParameter("WSDL");
+        }
 
         if (wsdlName != null && wsdlName.length()>0) {
             axisService.printUserWSDL(out, wsdlName, ip);
@@ -339,6 +343,9 @@
         res.setContentType("text/xml");
         String ip = extractHost(url);
         String wsdlName = req.getParameter("wsdl2");
+        if(wsdlName==null) {
+            wsdlName = req.getParameter("WSDL2");
+        }
 
         int ret = axisService.printWSDL2(res.getOutputStream(), ip, wsdlName);
         if (ret == 0) {
