Index: MIMEBuilder.java
===================================================================
--- MIMEBuilder.java	(revision 713004)
+++ MIMEBuilder.java	(working copy)
@@ -58,6 +58,11 @@
         // by subsequent builders(eg:MTOMBuilder) if needed..
         msgContext.setDoingSwA(true);
         
+        //TODO bgdki 07.11.2008 Set as a property that we have a MIME parts into the request.
+        // Should check if it is possible to have message with MIME parts and a binary "by value"
+        // at the same time. This seems to me as a legal case that needs a special attention!
+        msgContext.setProperty("mimeBinary", Boolean.TRUE);        
+        
         ContentType ct;
         try {
             ct = new ContentType(contentType);
