Index: src/org/apache/axis2/transport/local/LocalResponder.java
===================================================================
--- src/org/apache/axis2/transport/local/LocalResponder.java	(revision 1060336)
+++ src/org/apache/axis2/transport/local/LocalResponder.java	(working copy)
@@ -20,19 +20,29 @@
 
 package org.apache.axis2.transport.local;
 
+import org.apache.axiom.om.OMXMLParserWrapper;
+import org.apache.axiom.soap.SOAPEnvelope;
 import org.apache.axis2.AxisFault;
+import org.apache.axis2.Constants;
+import org.apache.axis2.addressing.AddressingConstants;
 import org.apache.axis2.addressing.EndpointReference;
+import org.apache.axis2.addressing.RelatesTo;
+import org.apache.axis2.builder.BuilderUtil;
 import org.apache.axis2.context.ConfigurationContext;
 import org.apache.axis2.context.MessageContext;
 import org.apache.axis2.description.TransportOutDescription;
+import org.apache.axis2.description.WSDL2Constants;
+import org.apache.axis2.engine.AxisEngine;
 import org.apache.axis2.handlers.AbstractHandler;
 import org.apache.axis2.transport.TransportSender;
 import org.apache.axis2.transport.TransportUtils;
+import org.apache.axis2.wsdl.WSDLConstants;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
-import java.io.ByteArrayOutputStream;
-import java.io.OutputStream;
+import javax.xml.stream.XMLStreamException;
+import java.io.*;
+import java.util.Map;
 
 /**
  * LocalResponder
@@ -40,7 +50,6 @@
 public class LocalResponder extends AbstractHandler implements TransportSender {
     protected static final Log log = LogFactory.getLog(LocalResponder.class);
     
-    
     //  fixed for Executing LocalTransport in MulthThread. 
     private OutputStream out;
 
@@ -87,13 +96,22 @@
                     TransportUtils.writeMessage(msgContext, out);
                 }
             } else {
-                out = (OutputStream) msgContext.getProperty(MessageContext.TRANSPORT_OUT);
+                MessageContext proxyInMessageContext = msgContext.
+                        getOperationContext().getMessageContext(WSDL2Constants.MESSAGE_LABEL_IN);
+                MessageContext initialMessageContext = (MessageContext)proxyInMessageContext.
+                        getProperty(LocalTransportReceiver.IN_MESSAGE_CONTEXT);
 
-                if (out != null) {
-                    TransportUtils.writeMessage(msgContext, out);
+                if (initialMessageContext != null) {
+                    handleResponse(msgContext, initialMessageContext);
                 } else {
-                    throw new AxisFault(
-                            "Both the TO and Property MessageContext.TRANSPORT_OUT is Null, No where to send");
+                    out = (OutputStream) msgContext.getProperty(MessageContext.TRANSPORT_OUT);
+
+                    if (out != null) {
+                        TransportUtils.writeMessage(msgContext, out);
+                    } else {
+                        throw new AxisFault(
+                                "Both the TO and Property MessageContext.TRANSPORT_OUT is Null, No where to send");
+                    }
                 }
             }
         } catch (AxisFault axisFault) {
@@ -106,4 +124,69 @@
         
         return InvocationResponse.CONTINUE;
     }
+
+    /**
+     * Retrieves the properties from the proxyOutMessageContext and sets the values to the
+     * inMessageContext.
+     *
+     * @param proxyOutMessageContext the active message context
+     * @param initialMessageContext the initial message context, which was stored as a property
+     * in the proxyOutMessageContext
+     * @throws AxisFault
+     */
+    private void handleResponse(MessageContext proxyOutMessageContext, MessageContext initialMessageContext) throws AxisFault {
+        MessageContext inMessageContext = initialMessageContext.getOperationContext().
+                getMessageContext(WSDL2Constants.MESSAGE_LABEL_IN);
+
+        // setting the properties
+        Map<String,Object> initialPropertyMap =  initialMessageContext.getProperties();
+        initialMessageContext.setProperties(initialPropertyMap);
+
+        inMessageContext.setEnvelope(getEnvelope(proxyOutMessageContext));
+        inMessageContext.setAxisServiceGroup(initialMessageContext.getAxisServiceGroup());
+        inMessageContext.setAxisService(initialMessageContext.getAxisService());
+        inMessageContext.setAxisOperation(initialMessageContext.getAxisOperation());
+        inMessageContext.setAxisMessage(initialMessageContext.getAxisOperation().getMessage(
+                WSDLConstants.MESSAGE_LABEL_OUT_VALUE));
+        inMessageContext.setIncomingTransportName(Constants.TRANSPORT_LOCAL);
+        inMessageContext.setServiceContext(initialMessageContext.getServiceContext());
+
+        // set properties on response
+        inMessageContext.setServerSide(true);
+        inMessageContext.setProperty(MessageContext.TRANSPORT_OUT,
+                initialMessageContext.getProperty(MessageContext.TRANSPORT_OUT));
+        inMessageContext.setProperty(Constants.OUT_TRANSPORT_INFO,
+                initialMessageContext.getProperty(Constants.OUT_TRANSPORT_INFO));
+        inMessageContext.setTransportIn(initialMessageContext.getTransportIn());
+        inMessageContext.setTransportOut(initialMessageContext.getTransportOut());
+
+        // copy the message type property that is used by the out message to the response message
+        inMessageContext.setProperty(Constants.Configuration.MESSAGE_TYPE,
+                initialMessageContext.getProperty(Constants.Configuration.MESSAGE_TYPE));
+
+        if (initialMessageContext.getMessageID() != null) {
+            inMessageContext.setRelationships(
+                    new RelatesTo[]{new RelatesTo(initialMessageContext.getMessageID())});
+        }
+
+        inMessageContext.setReplyTo(initialMessageContext.getReplyTo());
+        inMessageContext.setFaultTo(initialMessageContext.getFaultTo());
+
+        AxisEngine.receive(inMessageContext);
+    }
+
+    private SOAPEnvelope getEnvelope(MessageContext messageContext) throws AxisFault {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        TransportUtils.writeMessage(messageContext, out);
+
+        ByteArrayInputStream bs = new ByteArrayInputStream(out.toByteArray());
+        InputStreamReader streamReader = new InputStreamReader(bs);
+        OMXMLParserWrapper builder;
+        try {
+            builder = BuilderUtil.getBuilder(streamReader);
+        } catch (XMLStreamException e) {
+            throw AxisFault.makeFault(e);
+        }
+        return (SOAPEnvelope) builder.getDocumentElement();
+    }
 }
Index: src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 1060336)
+++ src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -43,6 +43,7 @@
     public static ConfigurationContext CONFIG_CONTEXT;
     private ConfigurationContext confContext;
     private MessageContext inMessageContext;
+    public static final String IN_MESSAGE_CONTEXT = "IN_MESSAGE_CONTEXT";
 
     public LocalTransportReceiver(ConfigurationContext configContext) {
         confContext = configContext;
@@ -79,6 +80,10 @@
                                OutputStream response)
             throws AxisFault {
         MessageContext msgCtx = confContext.createMessageContext();
+
+        // Set the in-message context as a property to the  current message context.
+        msgCtx.setProperty(IN_MESSAGE_CONTEXT, inMessageContext);
+
         if (inMessageContext != null) {
             msgCtx.setProperty(HTTPConstants.MC_HTTP_SERVLETREQUEST,
                                inMessageContext.getProperty(HTTPConstants.MC_HTTP_SERVLETREQUEST));
