Index: modules/kernel/src/org/apache/axis2/context/ConfigurationContext.java
===================================================================
--- modules/kernel/src/org/apache/axis2/context/ConfigurationContext.java	(revision 1050781)
+++ modules/kernel/src/org/apache/axis2/context/ConfigurationContext.java	(working copy)
@@ -251,18 +251,7 @@
                     serviceGroupContext =
                             getServiceGroupContextFromSoapSessionTable(serviceGroupContextId,
                                                                        messageContext);
-                    if (serviceGroupContext == null) {
 
-                        // TODO: Adding this code so that requests to services deployed in soapsession scope will work
-                        // TODO: However, soapsession functionality is still broken
-                        serviceGroupContext =
-                                new ServiceGroupContext(this,
-                                                        axisService.getAxisServiceGroup());
-                        serviceGroupContext.setId(serviceGroupContextId);
-                        addServiceGroupContextIntoSoapSessionTable(serviceGroupContext);
-//                        throw new AxisFault("Unable to find corresponding context" +
-//                                            " for the serviceGroupId: " + serviceGroupContextId);
-                    }
                 } else {
                     AxisServiceGroup axisServiceGroup = axisService.getAxisServiceGroup();
                     serviceGroupContext = createServiceGroupContext(axisServiceGroup);
