Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/input/README.txt
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/input/README.txt	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/input/README.txt	(working copy)
@@ -1,4 +0,0 @@
-This package will represent the wrapper around a collection of WSMService objects. The wrapper 
-class will possibly contain some type of information that signifies how the WSMService objects 
-were created. This will allow the user to determine if the data was created from wsdl or class 
-file annotations.
\ No newline at end of file
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/output/README.txt
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/output/README.txt	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/output/README.txt	(working copy)
@@ -1,5 +0,0 @@
-This package will be used to represent the data that is output as a result of MetaDataQuery 
-functionality. The output should be similar to the input, a collection of WSMService objects. 
-The WSMService objects will represent the merged data from both wsdl and annotations, if both 
-were present. The user will then write generators, similar to the WSM convention, that make use 
-of the data in the appropriate fashion.
\ No newline at end of file
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/README.txt
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/README.txt	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/README.txt	(working copy)
@@ -1,6 +0,0 @@
-This package will contain functionality to validate, process, and merge information provided 
-by annotation metadata and wsdl. The idea will be that the MDQ works off of collections of 
-WSMService objects as defined by the WSM project. Users will need to supply their own
-functionality that converts varying data into WSMService objects. For instance, users may 
-need a WSDL to WSMSerivce converter. Ideally, the MDQ will have knowledge of a single data 
-structure which will be a WSMService.
\ No newline at end of file
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/annotations/collector/README.txt
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/annotations/collector/README.txt	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/mdq/annotations/collector/README.txt	(working copy)
@@ -1,4 +0,0 @@
-This package will allow for independence of annotation collection techniques. There will be an 
-interface, AnnotationCollector, that defines a method which will return an array of WSMService 
-objects. These objects will be created by the collector after it has located all of the 
-annotations in the relevant class files.
\ No newline at end of file
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointDescription.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointDescription.java	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointDescription.java	(working copy)
@@ -51,13 +51,15 @@
 
  */
 /*
- * TODO: EndpointDescription should be created AxisService objects and not directly from WSDL
+ * TODO: EndpointDescription should be created via AxisService objects and not directly from WSDL
  * IMPORTANT NOTE: Axis2 currently only supports 1 service and 1 port under that service.  When that is
  * fixed, that will probably have an impact on this class.  In particular, I think this should be created 
  * somehow from an AxisService/AxisPort combination, and not directly from the WSDL.
  */
 public class EndpointDescription {
+    private ServiceDescription parentServiceDescription;
     private QName portQName;
+    private EndpointInterfaceDescription endpointInterfaceDescription;
     // TODO: This needs to be a collection of handler descriptions; use JAX-WS Appendix B Handler Chain Configuration File Schema as a starting point
     private ArrayList<String> handlerList = new ArrayList<String>();
     
@@ -69,15 +71,21 @@
      * @param wsdlPort The WSDL Port tag for this EndpointDescription.
      * @param definition The WSDL Definition target namespace used to create the port QName
      */
-    EndpointDescription(Port wsdlPort, Definition definition) {
+    EndpointDescription(Port wsdlPort, Definition definition, ServiceDescription parent) {
+        parentServiceDescription = parent;
         String localPart = wsdlPort.getName();
         String namespace = definition.getTargetNamespace();
         portQName = new QName(namespace, localPart);
+        endpointInterfaceDescription = new EndpointInterfaceDescription(this);
     }
     public QName getPortQName() {
         return portQName;
     }
     
+    public ServiceDescription getServiceDescription() {
+        return parentServiceDescription;
+    }
+    
     /**
      * Returns a live list describing the handlers on this port.
      * TODO: This is currently returning List<String>, but it should return a HandlerDescritpion
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/OperationDescription.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/OperationDescription.java	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/OperationDescription.java	(working copy)
@@ -18,6 +18,8 @@
 
 package org.apache.axis2.jaxws.description;
 
+import org.apache.axis2.description.AxisOperation;
+
 /**
  * 
  */
@@ -56,5 +58,19 @@
 
  */
 public class OperationDescription {
+    private EndpointInterfaceDescription parentEndpointInterfaceDescription;
+    private AxisOperation axisOperation;
+    
+    OperationDescription(AxisOperation operation, EndpointInterfaceDescription parent) {
+        parentEndpointInterfaceDescription = parent;
+        axisOperation = operation;       
+    }
 
+    public EndpointInterfaceDescription getEndpointInterfaceDescription() {
+        return parentEndpointInterfaceDescription;
+    }
+    
+    public AxisOperation getAxisOperation() {
+        return axisOperation;
+    }
 }
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/ServiceDescription.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/ServiceDescription.java	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/ServiceDescription.java	(working copy)
@@ -78,7 +78,7 @@
 TBD
 
 Properties available to JAXWS runtime:
-getJAXInterface(QName port) Needed by HandlerResolver
+getEndpointDescription(QName port) Needed by HandlerResolver
 TBD
 
  */
@@ -104,9 +104,9 @@
      * @param serviceQName  The name of the service in the WSDL.  This can not be null since a 
      *   javax.xml.ws.Service can not be created with a null service QName.
      * @param serviceClass  The JAX-WS service class.  This could be an instance of
-     *   javax.xml.ws.Service or a generated service subclass thereof.
+     *   javax.xml.ws.Service or a generated service subclass thereof.  This will not be null.
      */
-    public ServiceDescription(URL wsdlURL, QName serviceQName, Class serviceClass) {
+    ServiceDescription(URL wsdlURL, QName serviceQName, Class serviceClass) {
         if (serviceQName == null) {
         	// TODO NLS
             throw ExceptionFactory.makeWebServiceException("Invalid Service QName; cannot be null");
@@ -222,7 +222,7 @@
             Iterator portIterator = ports.values().iterator();
             while (portIterator.hasNext()) {
                 Port wsdlPort = (Port) portIterator.next();
-                EndpointDescription endpointDescription = new EndpointDescription(wsdlPort, definition);
+                EndpointDescription endpointDescription = new EndpointDescription(wsdlPort, definition, this);
                 QName portQName = endpointDescription.getPortQName();
                 endpointDescriptions.put(portQName, endpointDescription); 
             }
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointInterfaceDescription.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointInterfaceDescription.java	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/EndpointInterfaceDescription.java	(working copy)
@@ -18,6 +18,18 @@
 
 package org.apache.axis2.jaxws.description;
 
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.Iterator;
+
+import javax.wsdl.Binding;
+import javax.wsdl.Port;
+import javax.wsdl.PortType;
+import javax.xml.namespace.QName;
+
+import org.apache.axis2.description.AxisOperation;
+import org.apache.axis2.description.AxisService;
+
 /**
  * 
  */
@@ -51,5 +63,21 @@
 
  */
 public class EndpointInterfaceDescription {
+    EndpointDescription parentEndpointDescription;
+    Hashtable<QName, OperationDescription> operationDescriptions = new Hashtable<QName, OperationDescription>();
+    
+    
+    EndpointInterfaceDescription(EndpointDescription parent) {
+        parentEndpointDescription = parent;
+        
+        AxisService axisService = parentEndpointDescription.getServiceDescription().getAxisService();
+        ArrayList publishedOperations = axisService.getPublishedOperations();
+        Iterator operationsIterator = publishedOperations.iterator();
+        while (operationsIterator.hasNext()) {
+            AxisOperation operation = (AxisOperation) operationsIterator.next();
+            operationDescriptions.put(operation.getName(), new OperationDescription(operation, this));
+        }
+        
+    }
 
 }
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/DescriptionFactory.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/DescriptionFactory.java	(revision 0)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/description/DescriptionFactory.java	(revision 0)
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2004,2005 The Apache Software Foundation.
+ * Copyright 2006 International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.axis2.jaxws.description;
+
+import java.net.URL;
+
+import javax.xml.namespace.QName;
+
+/**
+ * Creates the JAX-WS metadata descritpion hierachy from some combinations of
+ * WSDL, Java classes with annotations, and (in the future) deployment descriptors.
+ */
+public class DescriptionFactory {
+    /**
+     * A DescrptionFactory can not be instantiated; all methods are static.
+     */
+    private DescriptionFactory() {
+    }
+    
+    public static ServiceDescription createServiceDescription(URL wsdlURL, QName serviceQName, Class serviceClass) {
+        return new ServiceDescription(wsdlURL, serviceQName, serviceClass);
+    }
+    
+    public static ServiceDescription updateEndpointInterface(ServiceDescription serviceDescription, Class seiClass) {
+        // TODO: Implement this method
+        return serviceDescription;
+    }
+
+}
Index: C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java
===================================================================
--- C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java	(revision 425742)
+++ C:/blddir/eclipse/axis2/axis2/modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java	(working copy)
@@ -48,6 +48,7 @@
 import org.apache.axis2.jaxws.client.factory.ProxyHandlerFactory;
 import org.apache.axis2.jaxws.client.proxy.BaseProxyHandler;
 import org.apache.axis2.jaxws.client.proxy.ProxyDescriptor;
+import org.apache.axis2.jaxws.description.DescriptionFactory;
 import org.apache.axis2.jaxws.description.ServiceDescription;
 import org.apache.axis2.jaxws.handler.PortData;
 import org.apache.axis2.jaxws.handler.PortInfoImpl;
@@ -76,7 +77,7 @@
     		throw ExceptionFactory.makeWebServiceException("Invalid Service QName, Service Name cannot be null or empty");
     	}
 
-        serviceDescription = new ServiceDescription(url, serviceQname, clazz);
+        serviceDescription = DescriptionFactory.createServiceDescription(url, serviceQname, clazz);
         if (isValidWSDLLocation()) {
             if(!isServiceDefined(serviceQname)){
                 throw new WebServiceException("Service " + serviceQname + " not defined in WSDL");
