--- axis2.sh.orig	2007-04-05 09:15:47.023914100 -0700
+++ axis2.sh	2007-04-05 09:16:32.992958300 -0700
@@ -28,8 +28,9 @@
 # -----------------------------------------------------------------------------
 
 # Get the context and from that find the location of setenv.sh
-. `dirname $0`/setenv.sh
+dir=`dirname "$0"`
+. "${dir}"/setenv.sh
 
 
-$JAVA_HOME/bin/java -classpath "$AXIS2_CLASSPATH" \
+"$JAVA_HOME"/bin/java -classpath "$AXIS2_CLASSPATH" \
 -Daxis2.xml="$AXIS2_HOME/conf/axis2.xml" -Daxis2.repo="$AXIS2_HOME/repository" $*
\ No newline at end of file
