Index: CommandLineOptionConstants.java
===================================================================
--- CommandLineOptionConstants.java	(revision 506926)
+++ CommandLineOptionConstants.java	(working copy)
@@ -46,8 +46,8 @@
         String SUPPRESS_PREFIXES_OPTION = "sp";
         String SOURCE_FOLDER_NAME_OPTION = "S";
         String RESOURCE_FOLDER_OPTION = "R";
+        String XSDCONFIG_OPTION = "xc";
 
-
         //long option constants
         String OUTPUT_LOCATION_OPTION_LONG = "output";
         String SERVER_SIDE_CODE_OPTION_LONG = "server-side";
@@ -75,6 +75,7 @@
         String SUPPRESS_PREFIXES_OPTION_LONG = "suppress-prefixes";
         String SOURCE_FOLDER_NAME_OPTION_LONG = "source-folder";
         String RESOURCE_FOLDER_OPTION_LONG = "resource-folder";
+        String XSDCONFIG_OPTION_LONG = "xsdconfig";
 
         String WSDL_VERSION_2 = "2.0";
         String WSDL_VERSION_2_OPTIONAL = "2";
