Index: src/org/apache/axis2/mtompolicy/MTOMOutHandler.java
===================================================================
--- src/org/apache/axis2/mtompolicy/MTOMOutHandler.java	(revision 767915)
+++ src/org/apache/axis2/mtompolicy/MTOMOutHandler.java	(working copy)
@@ -62,6 +62,7 @@
                 if (isOptional) {
                     msgCtx.setProperty(Constants.Configuration.ENABLE_MTOM,
                             Constants.VALUE_OPTIONAL);
+                    msgCtx.setDoingMTOM(Boolean.TRUE.booleanValue());
                 } else {
                     msgCtx.setProperty(Constants.Configuration.ENABLE_MTOM, Constants.VALUE_TRUE);
                 }
