Index: src/main/java/org/apache/axis2/tool/service/control/Controller.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/control/Controller.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/control/Controller.java	(working copy)
@@ -201,8 +201,9 @@
             throw new ProcessException(e);
         } finally {
             deleteDir(tempFileFolder);
-             if (isServiceCreated)
+             if (isServiceCreated){
                 serviceFile.delete();
+             }
         }
 
     }
Index: src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLOptionsPage.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLOptionsPage.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLOptionsPage.java	(working copy)
@@ -52,7 +52,7 @@
     private Button searchDeclaredMethodsCheckBox;
     private Table table;
     
-    private boolean dirty = false;
+//    private boolean dirty = false;
     
     public WSDLOptionsPage(){
         super("page6");
@@ -274,7 +274,7 @@
     }
     
     private void updateDirtyStatus(boolean status){
-        dirty = status;
+//        dirty = status;
         if (table.isVisible()){
             table.setEnabled(!status);
         }
Index: src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceArchiveWizard.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceArchiveWizard.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceArchiveWizard.java	(working copy)
@@ -47,7 +47,6 @@
     private ServiceArchiveOutputLocationPage serviceArchiveOutputLocationPage;
     private LibraryAddingPage libPage;
 
-    private boolean updateServiceGenerationStatus;
     private String classFileLocation;
     private String wsdlFileGenerationStatus;
     
@@ -148,8 +147,9 @@
         {
            protected void execute(IProgressMonitor monitor)
            throws CoreException , InvocationTargetException, InterruptedException{
-              if (monitor == null)
+              if (monitor == null){
                  monitor = new NullProgressMonitor();
+              }
 
               /*
                * "7" is the total amount of steps, see below monitor.worked(amount)
Index: src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceXMLFileSelectionPage.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceXMLFileSelectionPage.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/eclipse/ui/ServiceXMLFileSelectionPage.java	(working copy)
@@ -50,7 +50,6 @@
     
     
     private boolean skipNextPage=true;
-    private boolean pageComplete;
     
     public ServiceXMLFileSelectionPage(){
         super("page2");
@@ -186,8 +185,9 @@
     
  
     public void updateRecommendation(String message){
-        if (recommendationTextLable!=null)
-        recommendationTextLable.setText(message);
+		if (recommendationTextLable != null) {
+			recommendationTextLable.setText(message);
+		}
     }
     
    
Index: src/main/java/org/apache/axis2/tool/service/eclipse/ui/LibraryAddingPage.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/eclipse/ui/LibraryAddingPage.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/eclipse/ui/LibraryAddingPage.java	(working copy)
@@ -38,7 +38,6 @@
 
 public class LibraryAddingPage extends AbstractServiceWizardPage {
     
-    private Label libraryNameLabel;
     private Text libraryNameText;
     private Button browseButton;
     private Button addButton;
Index: src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLFileSelectionPage.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLFileSelectionPage.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/eclipse/ui/WSDLFileSelectionPage.java	(working copy)
@@ -38,9 +38,6 @@
 
 public class WSDLFileSelectionPage extends AbstractServiceWizardPage {
     
-    //  set the default to false. This step cannot be skipped
-    private boolean skipNextPage = false;
-    
     private Text wsdlTextBox;
     private Label selectionLabel;
     private Button browseButton;
Index: src/main/java/org/apache/axis2/tool/service/bean/LibrarySelectionBean.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/bean/LibrarySelectionBean.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/bean/LibrarySelectionBean.java	(working copy)
@@ -27,10 +27,10 @@
     
 
     /**
-     * @return Returns the fileList.
+     * @return Returns a clone of the fileList.
      */
     public String[] getFileList() {
-        return fileList;
+        return fileList.clone();
     }
     /**
      * @param fileList The fileList to set.
Index: src/main/java/org/apache/axis2/tool/service/swing/ui/WizardPane.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/swing/ui/WizardPane.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/swing/ui/WizardPane.java	(working copy)
@@ -79,10 +79,11 @@
         fc.setFileSelectionMode(JFileChooser.FILES_AND_DIRECTORIES);
         fc.addChoosableFileFilter(new FileFilter() {
             public boolean accept(File f) {
-                if (f.getName().endsWith(extension) || f.isDirectory())
+                if (f.getName().endsWith(extension) || f.isDirectory()){
                     return true;
-                else
+                }else{
                     return false;
+                }
             }
 
             public String getDescription() {
Index: src/main/java/org/apache/axis2/tool/service/swing/ui/MainWindow.java
===================================================================
--- src/main/java/org/apache/axis2/tool/service/swing/ui/MainWindow.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/service/swing/ui/MainWindow.java	(working copy)
@@ -105,8 +105,9 @@
                 bHeight);
         this.cancelButton.addActionListener(new ActionListener() {
             public void actionPerformed(ActionEvent e) {
-                if (confirmExit())
+                if (confirmExit()){
                     System.exit(0);
+                }
             }
         });
 
Index: src/main/java/org/apache/axis2/tool/core/ServiceFileCreator.java
===================================================================
--- src/main/java/org/apache/axis2/tool/core/ServiceFileCreator.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/core/ServiceFileCreator.java	(working copy)
@@ -49,10 +49,6 @@
         writeFile(getServiceModel(serviceName,implementationClassName,methodList),serviceXmlWriter,fileName);
 
         return new File(currentUserDir + File.separator + fileName);
-
-
-
-
     }
 
     private Document getServiceModel(String serviceName,String className,ArrayList methods){
Index: src/main/java/org/apache/axis2/tool/core/FileCopier.java
===================================================================
--- src/main/java/org/apache/axis2/tool/core/FileCopier.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/core/FileCopier.java	(working copy)
@@ -36,9 +36,9 @@
 
     public void copyFiles(File sourceFile,File destinationDirectory,String filter){
 
-        if (sourceFile.isFile())
+        if (sourceFile.isFile()){
             this.setFile(sourceFile);
-        else {
+        }else {
         	 FileSet fileset = new FileSet();
 	            fileset.setDir(sourceFile);
 	            if (filter!=null){
Index: src/main/java/org/apache/axis2/tool/core/ClassFileHandler.java
===================================================================
--- src/main/java/org/apache/axis2/tool/core/ClassFileHandler.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/core/ClassFileHandler.java	(working copy)
@@ -24,7 +24,9 @@
 import java.lang.reflect.Method;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.security.AccessController;
 import java.util.ArrayList;
+import java.security.PrivilegedAction;
 
 public class ClassFileHandler {
 
@@ -41,10 +43,15 @@
     public ArrayList getMethodNamesFromClass(String classFileName,String location) throws IOException, ClassNotFoundException{
         ArrayList returnList = new ArrayList();
         File fileEndpoint = new File(location);
-        if (!fileEndpoint.exists())
+        if (!fileEndpoint.exists()){
             throw new IOException("the location is invalid");
-        URL[] urlList = {fileEndpoint.toURL()};
-        URLClassLoader clazzLoader = new URLClassLoader(urlList);
+        }
+        final URL[] urlList = {fileEndpoint.toURL()};
+        URLClassLoader clazzLoader = AccessController.doPrivileged(new PrivilegedAction<URLClassLoader>() {
+			public URLClassLoader run() {
+				return new URLClassLoader(urlList);
+			}
+		});
         Class clazz = clazzLoader.loadClass(classFileName);
         Method[] methods = clazz.getDeclaredMethods();
 
Index: src/main/java/org/apache/axis2/tool/core/JarFileWriter.java
===================================================================
--- src/main/java/org/apache/axis2/tool/core/JarFileWriter.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/core/JarFileWriter.java	(working copy)
@@ -41,8 +41,9 @@
         if (!outputFolder.exists()){
             outputFolder.mkdir(); //create the output path
         }else{
-            if (!outputFolder.isDirectory())
+            if (!outputFolder.isDirectory()){
                 return;
+            }
         }
 
         File targetFile = new File(outputFolder,outputFileName);
Index: src/main/java/org/apache/axis2/tool/util/ServicePluginUtils.java
===================================================================
--- src/main/java/org/apache/axis2/tool/util/ServicePluginUtils.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/util/ServicePluginUtils.java	(working copy)
@@ -1,6 +1,7 @@
 package org.apache.axis2.tool.util;
 
 import java.io.File;
+import java.io.IOException;
 
 import javax.xml.transform.Source;
 import javax.xml.transform.stream.StreamSource;
@@ -10,8 +11,13 @@
 
 import org.apache.axis2.tool.service.eclipse.plugin.ServiceArchiver;
 import org.apache.axis2.tool.util.Constants.ServiceConstants;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.xml.sax.SAXException;
 
 public class ServicePluginUtils {
+	private static final Log logger = LogFactory.getLog(ServicePluginUtils.class);
+	
 	/**
 	 * Validates the given xml file against the axis2 services schema. 
 	 * @return return true if the xml is valid
@@ -19,20 +25,23 @@
 	public static boolean isServicesXMLValid(String servicesXmlPath){
         SchemaFactory factory = 
             SchemaFactory.newInstance(ServiceConstants.XML_SCHEMA);
-        
-        try {
-        	String resourcePath=addAnotherNodeToPath(
-        			ServiceConstants.RESOURCE_FOLDER, ServiceConstants.SERVICES_XSD_SCHEMA_NAME);
-            Schema schema = factory.newSchema(
-            		 ServiceArchiver.getDefault().getBundle().getEntry(resourcePath));
-            Validator validator = schema.newValidator();
-            Source source = new StreamSource(new File(servicesXmlPath));
-            validator.validate(source);
-            return true;
-        }
-        catch (Exception ex) {
-            return false;
-        }  
+       
+        	try {
+				String resourcePath=addAnotherNodeToPath(
+						ServiceConstants.RESOURCE_FOLDER, ServiceConstants.SERVICES_XSD_SCHEMA_NAME);
+				Schema schema = factory.newSchema(
+						 ServiceArchiver.getDefault().getBundle().getEntry(resourcePath));
+				Validator validator = schema.newValidator();
+				Source source = new StreamSource(new File(servicesXmlPath));
+				validator.validate(source);
+				return true;
+			} catch (SAXException e) {
+				logger.debug("Schema addition failed",e);
+				return false;
+			} catch (IOException e) {
+				logger.debug("Schema validation failed",e);
+				return false;
+			}
 	}
 	
 	public static String addAnotherNodeToPath(String currentPath, String newNode) {
Index: src/main/java/org/apache/axis2/tool/util/Constants.java
===================================================================
--- src/main/java/org/apache/axis2/tool/util/Constants.java	(revision 1057160)
+++ src/main/java/org/apache/axis2/tool/util/Constants.java	(working copy)
@@ -20,7 +20,7 @@
 package org.apache.axis2.tool.util;
 
 public class Constants {
-    public class UIConstants{
+    public static class UIConstants{
         public static final int LABEL_WIDTH=100;
         public static final int RADIO_BUTTON_WIDTH=200;
         public static final int TEXT_BOX_WIDTH=250;
@@ -31,7 +31,7 @@
     
     }
     
-    public class ServiceConstants{
+    public static class ServiceConstants{
     	public static final String RESOURCE_FOLDER="resources";						
     	public static final String SERVICES_XSD_SCHEMA_NAME="services.xsd";				
     	public static final String XML_SCHEMA="http://www.w3.org/2001/XMLSchema";	
