diff -urw ../origsrc/jms/JMSConnectionFactory.java org/apache/axis2/transport/jms/JMSConnectionFactory.java
--- ../origsrc/jms/JMSConnectionFactory.java	2008-12-03 16:11:38.562500000 +0100
+++ org/apache/axis2/transport/jms/JMSConnectionFactory.java	2008-12-03 18:35:51.515625000 +0100
@@ -225,6 +225,20 @@
 	}
 
     /**
+     * Set the actual factory username
+     */
+	public void setUser(String user) {
+		this.user = user;
+	}   
+    
+    /**
+     * Set the actual factory password
+     */
+	public void setPass(String pass) {
+		this.pass = pass;
+	}
+
+    /**
      * Add a listen destination on this connection factory on behalf of the given service
      *
      * @param destinationJndi destination JNDI name
@@ -240,7 +254,7 @@
 
             Connection con = null;
             try {
-            	if ((jndiUser == null) || (jndiPass == null)){
+            	if ((user == null) || (pass == null)){
             		// Use the OS username and credentials
                     con = conFactory.createConnection();            		
             	} else{
@@ -402,7 +416,7 @@
                     "Connection factory must be 'connected' before listening");
         } else {
             try {
-            	if ((jndiUser == null) || (jndiPass == null)){
+            	if ((user == null) || (pass == null)){
             		// User the OS username and credentials
                     connection = conFactory.createConnection();            		
             	} else{
diff -urw ../origsrc/jms/JMSListener.java org/apache/axis2/transport/jms/JMSListener.java
--- ../origsrc/jms/JMSListener.java	2008-12-03 16:11:38.578125000 +0100
+++ org/apache/axis2/transport/jms/JMSListener.java	2008-12-03 18:12:01.937500000 +0100
@@ -251,6 +251,10 @@
                     jmsConFactory.setJndiUser((String) p.getValue());
                 } else if (JMSConstants.CONFAC_JNDI_NAME_PASS.equals(p.getName())) {
                     jmsConFactory.setJndiPass((String) p.getValue());
+                } else if (JMSConstants.CONFAC_USER.equals(p.getName())) {
+                    jmsConFactory.setUser((String) p.getValue());
+                } else if (JMSConstants.CONFAC_PASS.equals(p.getName())) {
+                    jmsConFactory.setPass((String) p.getValue());
                 } else if (JMSConstants.DEST_PARAM.equals(p.getName())) {
                     StringTokenizer st =
                             new StringTokenizer((String) p.getValue(), " ,");
