Index: modules/core/src/org/apache/axis2/util/Utils.java
===================================================================
--- modules/core/src/org/apache/axis2/util/Utils.java	(revision 421863)
+++ modules/core/src/org/apache/axis2/util/Utils.java	(working copy)
@@ -80,7 +80,6 @@
         }
 
         newmsgCtx.setFaultTo(oldOptions.getTo());
-        newmsgCtx.setFrom(oldOptions.getTo());
         newmsgCtx.addRelatesTo(
                 new RelatesTo(
                         oldOptions.getMessageId(),
