Index: org/apache/axis2/transport/TransportUtils.java
===================================================================
--- org/apache/axis2/transport/TransportUtils.java	(revision 1022713)
+++ org/apache/axis2/transport/TransportUtils.java	(working copy)
@@ -525,14 +525,15 @@
            }
 
        	Attachments attachments = msgContext.getAttachmentMap();
-       	LifecycleManager lcm = (LifecycleManager)msgContext.getRootContext().getAxisConfiguration().getParameterValue(DeploymentConstants.ATTACHMENTS_LIFECYCLE_MANAGER);
+
            if (attachments != null) {
                // Get the list of Content IDs for the attachments...but does not try to pull the stream for new attachments.
                // (Pulling the stream for new attachments will probably fail...the stream is probably closed)
                List keys = attachments.getContentIDList();
-               if (keys != null) {
+               if (keys != null && keys.size() > 0) {
                	String key = null;
                	File file = null;
+               	   LifecycleManager lcm = (LifecycleManager)msgContext.getRootContext().getAxisConfiguration().getParameterValue(DeploymentConstants.ATTACHMENTS_LIFECYCLE_MANAGER);               		
                	DataSource dataSource = null;
                    for (int i = 0; i < keys.size(); i++) {
                        try {
Index: org/apache/axis2/transport/http/SOAPMessageFormatter.java
===================================================================
--- org/apache/axis2/transport/http/SOAPMessageFormatter.java	(revision 1022713)
+++ org/apache/axis2/transport/http/SOAPMessageFormatter.java	(working copy)
@@ -55,14 +55,15 @@
         }
         OMElement element = msgCtxt.getEnvelope();
         
-        int optimizedThreshold = Utils.getMtomThreshold(msgCtxt);       
-        if(optimizedThreshold > 0){
-        	if(log.isDebugEnabled()){
-        		log.debug("Setting MTOM optimized Threshold Value on OMOutputFormat");
-        	}
-        	format.setOptimizedThreshold(optimizedThreshold);
-        }	        
-
+        if (msgCtxt.isDoingMTOM()) {        	
+            int optimizedThreshold = Utils.getMtomThreshold(msgCtxt);       
+            if(optimizedThreshold > 0){
+            	if(log.isDebugEnabled()){
+            		log.debug("Setting MTOM optimized Threshold Value on OMOutputFormat");
+            	}
+            	format.setOptimizedThreshold(optimizedThreshold);
+            }        	
+        }
         try {
             if (!(format.isOptimized()) && format.isDoingSWA()) {
                 // Write the SOAPBody to an output stream
