Index: modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java	(revision 82195)
+++ modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java	(working copy)
@@ -72,6 +72,9 @@
         XmlSchemaSequence sequence;
 
         for (Method jMethod : declaredMethods) {
+            if (jMethod.isBridge()) {
+                continue;
+            }
             WebMethodAnnotation methodAnnon = JSR181Helper.INSTANCE.getWebMethodAnnotation(jMethod);
             if (methodAnnon != null) {
                 if (methodAnnon.isExclude()) {
Index: modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java	(revision 82195)
+++ modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java	(working copy)
@@ -279,6 +279,9 @@
         XmlSchemaSequence sequence = null;
 
         for (Method jMethod : declaredMethods) {
+            if (jMethod.isBridge()) {
+                continue;
+            }
             WebMethodAnnotation methodAnnon = JSR181Helper.INSTANCE.getWebMethodAnnotation(jMethod);
             if (methodAnnon != null) {
                 if (methodAnnon.isExclude()) {
Index: modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java
===================================================================
--- modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java	(revision 82195)
+++ modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java	(working copy)
@@ -94,6 +94,9 @@
                 Method[] methods = ImplClass.getMethods();
 
                 for (Method method1 : methods) {
+                    if (method1.isBridge()) {
+                        continue;
+                    }
                     if (method1.getName().equals(methodName)) {
                         method = method1;
                         op.addParameter("myMethod", method);
