### Eclipse Workspace Patch 1.0
#P axis2-trunk
Index: kernel/src/org/apache/axis2/deployment/DeploymentEngine.java
===================================================================
--- kernel/src/org/apache/axis2/deployment/DeploymentEngine.java	(revision 883984)
+++ kernel/src/org/apache/axis2/deployment/DeploymentEngine.java	(working copy)
@@ -1090,6 +1090,7 @@
         try {
             in = fileListUrl.openStream();
         } catch (IOException e) {
+        	log.info(e.getMessage() + " -  as per axis2.repository.url, the URL is "+fileListUrl+" that will be used relative to "+new File(".").getAbsolutePath());
             return fileList;
         }
         BufferedReader input = null;
