Index: modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java	(revision 1045019)
+++ modules/kernel/src/org/apache/axis2/description/java2wsdl/DocLitBareSchemaGenerator.java	(working copy)
@@ -72,6 +72,9 @@
         XmlSchemaSequence sequence;
 
         for (Method jMethod : declaredMethods) {
+            if (jMethod.isBridge()) {
+                continue;
+            }
             WebMethodAnnotation methodAnnon = JSR181Helper.INSTANCE.getWebMethodAnnotation(jMethod);
             if (methodAnnon != null) {
                 if (methodAnnon.isExclude()) {
Index: modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java	(revision 1045019)
+++ modules/kernel/src/org/apache/axis2/description/java2wsdl/DefaultSchemaGenerator.java	(working copy)
@@ -279,6 +279,9 @@
         XmlSchemaSequence sequence = null;
 
         for (Method jMethod : declaredMethods) {
+            if (jMethod.isBridge()) {
+                continue;
+            }
             WebMethodAnnotation methodAnnon = JSR181Helper.INSTANCE.getWebMethodAnnotation(jMethod);
             if (methodAnnon != null) {
                 if (methodAnnon.isExclude()) {
Index: modules/kernel/src/org/apache/axis2/description/java2wsdl/bytecode/MethodTable.java
===================================================================
--- modules/kernel/src/org/apache/axis2/description/java2wsdl/bytecode/MethodTable.java	(revision 1045019)
+++ modules/kernel/src/org/apache/axis2/description/java2wsdl/bytecode/MethodTable.java	(working copy)
@@ -43,6 +43,9 @@
         Method [] methods = cls.getMethods();
         for (int i = 0; i < methods.length; i++) {
             Method method = methods[i];
+            if (method.isBridge()) {
+                continue;
+            }
             nameToMethodMap.put(method.getName(), method);
         }
     }
Index: modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOnlyMessageReceiver.java
===================================================================
--- modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOnlyMessageReceiver.java	(revision 1045019)
+++ modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOnlyMessageReceiver.java	(working copy)
@@ -54,6 +54,9 @@
             String methodName = op.getName().getLocalPart();
             Method[] methods = ImplClass.getMethods();
             for (int i = 0; i < methods.length; i++) {
+                if (methods[i].isBridge()) {
+                    continue;
+                }
                 if (methods[i].getName().equals(methodName)) {
                     method = methods[i];
                     break;
Index: modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOutAsyncMessageReceiver.java
===================================================================
--- modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOutAsyncMessageReceiver.java	(revision 1045019)
+++ modules/adb/src/org/apache/axis2/rpc/receivers/RPCInOutAsyncMessageReceiver.java	(working copy)
@@ -79,6 +79,9 @@
             String methodName = op.getName().getLocalPart();
             Method[] methods = ImplClass.getMethods();
             for (int i = 0; i < methods.length; i++) {
+                if (methods[i].isBridge()) {
+                    continue;
+                }
                 if (methods[i].getName().equals(methodName)) {
                     method = methods[i];
                     break;
Index: modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java
===================================================================
--- modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java	(revision 1045019)
+++ modules/adb/src/org/apache/axis2/rpc/receivers/RPCMessageReceiver.java	(working copy)
@@ -93,6 +93,9 @@
                 Method[] methods = ImplClass.getMethods();
 
                 for (Method method1 : methods) {
+                    if (method1.isBridge()) {
+                        continue;
+                    }
                     if (method1.getName().equals(methodName)) {
                         method = method1;
                         op.addParameter("myMethod", method);
