Index: modules/metadata/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java
===================================================================
--- modules/metadata/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java	(revision 3542)
+++ modules/metadata/src/org/apache/axis2/jaxws/description/builder/WebServiceContextAnnot.java	(revision 3544)
@@ -23,6 +23,9 @@
 import java.lang.annotation.Annotation;
 import java.security.Principal;
 
+import javax.xml.ws.EndpointReference;
+import org.w3c.dom.Element;
+
 public class WebServiceContextAnnot implements javax.xml.ws.WebServiceContext {
 
     private MessageContext messageContext;
@@ -90,7 +93,14 @@
         this.userPrincipal = userPrincipal;
     }
 
+    public <T extends EndpointReference> T getEndpointReference(Class<T> arg0, Element... arg1) {
+        throw new UnsupportedOperationException();
+    }
 
+    public EndpointReference getEndpointReference(Element... arg0) {
+        throw new UnsupportedOperationException();
+    }
+
     //hmm, should we really do this
     public Class<Annotation> annotationType() {
         return Annotation.class;
Index: modules/jaxws/src/org/apache/axis2/jaxws/context/WebServiceContextImpl.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/context/WebServiceContextImpl.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/context/WebServiceContextImpl.java	(revision 3544)
@@ -22,10 +22,13 @@
 import org.apache.commons.logging.LogFactory;
 
 import javax.servlet.http.HttpServletRequest;
+import javax.xml.ws.EndpointReference;
 import javax.xml.ws.WebServiceContext;
 import javax.xml.ws.handler.MessageContext;
 import java.security.Principal;
 
+import org.w3c.dom.Element;
+
 public class WebServiceContextImpl implements WebServiceContext {
 
     private static final Log log = LogFactory.getLog(WebServiceContext.class);
@@ -91,4 +94,11 @@
         this.soapMessageContext = soapMessageContext;
     }
 
+    public <T extends EndpointReference> T getEndpointReference(Class<T> clazz, Element... referenceParameters) {
+        throw new UnsupportedOperationException();
+    }
+
+    public EndpointReference getEndpointReference(Element... referenceParameters) {
+        throw new UnsupportedOperationException();
+    }
 }
Index: modules/jaxws/src/org/apache/axis2/jaxws/spi/Provider.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/spi/Provider.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/spi/Provider.java	(revision 3544)
@@ -19,12 +19,19 @@
 package org.apache.axis2.jaxws.spi;
 
 import javax.xml.namespace.QName;
+import javax.xml.transform.Source;
 import javax.xml.ws.Endpoint;
+import javax.xml.ws.EndpointReference;
 import javax.xml.ws.spi.ServiceDelegate;
+import javax.xml.ws.WebServiceFeature;
+import javax.xml.ws.wsaddressing.W3CEndpointReference;
 
 import org.apache.axis2.jaxws.server.endpoint.EndpointImpl;
 
+import org.w3c.dom.Element;
+
 import java.net.URL;
+import java.util.List;
 
 public class Provider extends javax.xml.ws.spi.Provider {
 
@@ -44,4 +51,24 @@
     public ServiceDelegate createServiceDelegate(URL url, QName qname, Class clazz) {
         return new org.apache.axis2.jaxws.spi.ServiceDelegate(url, qname, clazz);
     }
+
+    @Override
+    public W3CEndpointReference createW3CEndpointReference(String address,
+            QName serviceName,
+            QName portName,
+            List<Element> metadata,
+            String wsdlDocumentLocation,
+            List<Element> referenceParameters) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T getPort(EndpointReference jaxwsEPR, Class<T> sei, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public EndpointReference readEndpointReference(Source eprInfoset) {
+        throw new UnsupportedOperationException();
+    }
 }
Index: modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/spi/ServiceDelegate.java	(revision 3544)
@@ -21,6 +21,7 @@
 
 import javax.xml.ws.handler.HandlerResolver;
 import org.apache.axis2.client.ServiceClient;
+import org.apache.axis2.context.ConfigurationContext;
 import org.apache.axis2.java.security.AccessController;
 import org.apache.axis2.jaxws.binding.BindingImpl;
 import org.apache.axis2.jaxws.ExceptionFactory;
@@ -47,9 +48,11 @@
 import javax.xml.soap.SOAPMessage;
 import javax.xml.transform.Source;
 import javax.xml.ws.Dispatch;
+import javax.xml.ws.EndpointReference;
 import javax.xml.ws.Service;
 import javax.xml.ws.Service.Mode;
 import javax.xml.ws.WebServiceException;
+import javax.xml.ws.WebServiceFeature;
 import javax.xml.ws.handler.HandlerResolver;
 import javax.xml.ws.http.HTTPBinding;
 import javax.xml.ws.soap.SOAPBinding;
@@ -201,6 +204,29 @@
         return dispatch;
     }
 
+
+    @Override
+    public <T> Dispatch<T> createDispatch(EndpointReference jaxwsEPR, Class<T> type, Mode mode, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Dispatch<Object> createDispatch(EndpointReference jaxwsEPR, JAXBContext context, Mode mode, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+
+    @Override
+    public <T> Dispatch<T> createDispatch(QName portName, Class<T> type, Mode mode, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Dispatch<Object> createDispatch(QName portName, JAXBContext context, Mode mode, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+
     /*
      * (non-Javadoc)
      * @see javax.xml.ws.spi.ServiceDelegate#getPort(java.lang.Class)
@@ -265,6 +291,21 @@
         return sei.cast(proxyClass);
     }
 
+    @Override
+    public <T> T getPort(Class<T> sei, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T getPort(EndpointReference jaxwsEPR, Class<T> sei, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T getPort(QName portName, Class<T> sei, WebServiceFeature... features) {
+        throw new UnsupportedOperationException();
+    }
+
     /*
     * (non-Javadoc)
     * @see javax.xml.ws.spi.ServiceDelegate#getExecutor()
Index: modules/jaxws/src/org/apache/axis2/jaxws/BindingProvider.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/BindingProvider.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/BindingProvider.java	(revision 3544)
@@ -22,7 +22,10 @@
 import java.util.Map;
 
 import javax.xml.ws.Binding;
+import javax.xml.ws.EndpointReference;
 import javax.xml.ws.handler.HandlerResolver;
+import javax.xml.ws.wsaddressing.W3CEndpointReference;
+import org.apache.axis2.AxisFault;
 import org.apache.axis2.jaxws.binding.BindingUtils;
 import org.apache.axis2.jaxws.binding.SOAPBinding;
 import org.apache.axis2.jaxws.client.PropertyValidator;
@@ -225,4 +228,16 @@
             }
         }
     }
+
+    public EndpointReference getEndpointReference() {
+        throw new UnsupportedOperationException();
+    }
+
+    public <T extends EndpointReference> T getEndpointReference(Class<T> clazz) {
+        throw new UnsupportedOperationException();
+    }
+
+    public org.apache.axis2.addressing.EndpointReference getAxis2EndpointReference(String addressingNamespace) throws AxisFault {
+        throw new UnsupportedOperationException();
+    }
 }
Index: modules/jaxws/src/org/apache/axis2/jaxws/server/endpoint/EndpointImpl.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/server/endpoint/EndpointImpl.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/server/endpoint/EndpointImpl.java	(revision 3544)
@@ -31,8 +31,12 @@
 import org.apache.axis2.transport.http.server.SimpleHttpServer;
 import org.apache.axis2.transport.http.server.WorkerFactory;
 
+import org.w3c.dom.Element;
+
 import javax.xml.transform.Source;
 import javax.xml.ws.Binding;
+import javax.xml.ws.EndpointReference;
+import javax.xml.ws.wsaddressing.W3CEndpointReference;
 
 import java.io.IOException;
 import java.util.List;
@@ -202,4 +206,12 @@
             e.printStackTrace();
         }
     }
+
+    public <T extends EndpointReference> T getEndpointReference(Class<T> clazz, Element... referenceParameters) {
+        throw new UnsupportedOperationException();
+    }
+
+    public EndpointReference getEndpointReference(Element... referenceParameters) {
+        throw new UnsupportedOperationException();
+    }
 }
Index: modules/jaxws/src/org/apache/axis2/jaxws/binding/BindingImpl.java
===================================================================
--- modules/jaxws/src/org/apache/axis2/jaxws/binding/BindingImpl.java	(revision 3542)
+++ modules/jaxws/src/org/apache/axis2/jaxws/binding/BindingImpl.java	(revision 3544)
@@ -81,4 +81,7 @@
         }
     }
 
+    public String getBindingID() {
+        throw new UnsupportedOperationException();
+    }
 }
Index: modules/jaxws-api/src/javax/xml/ws/EndpointReference.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/EndpointReference.java	(revision 0)
+++ modules/jaxws-api/src/javax/xml/ws/EndpointReference.java	(revision 3544)
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *      
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package javax.xml.ws;
+
+import java.io.ByteArrayOutputStream;
+
+import javax.xml.transform.Result;
+import javax.xml.transform.Source;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.ws.spi.Provider;
+
+public abstract class EndpointReference {
+    protected EndpointReference() {
+    }
+    
+    public static EndpointReference readFrom(Source eprInfoset) {
+        return Provider.provider().readEndpointReference(eprInfoset);
+    }
+    
+    public abstract void writeTo(Result result);
+    
+    public <T> T getPort(Class<T> serviceEndpointInterface, WebServiceFeature... features) {
+        return Provider.provider().getPort(this, serviceEndpointInterface, features);
+    }
+    
+    @Override
+    public String toString() {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        writeTo(new StreamResult(baos));
+        
+        return baos.toString();
+    }
+}
+
Index: modules/jaxws-api/src/javax/xml/ws/WebServiceFeature.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/WebServiceFeature.java	(revision 0)
+++ modules/jaxws-api/src/javax/xml/ws/WebServiceFeature.java	(revision 3544)
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *      
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package javax.xml.ws;
+
+public abstract class WebServiceFeature {
+    protected boolean enabled;
+    
+    protected WebServiceFeature() {
+    }
+    
+    public abstract String getID();
+    
+    public boolean isEnabled() {
+        return enabled;
+    }
+}
+
Index: modules/jaxws-api/src/javax/xml/ws/spi/Provider.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/spi/Provider.java	(revision 3542)
+++ modules/jaxws-api/src/javax/xml/ws/spi/Provider.java	(revision 3544)
@@ -19,10 +19,18 @@
 
 package javax.xml.ws.spi;
 
+import java.net.URL;
+import java.util.List;
+
 import javax.xml.namespace.QName;
+import javax.xml.transform.Source;
 import javax.xml.ws.Endpoint;
-import java.net.URL;
+import javax.xml.ws.EndpointReference;
+import javax.xml.ws.WebServiceFeature;
+import javax.xml.ws.wsaddressing.W3CEndpointReference;
 
+import org.w3c.dom.Element;
+
 public abstract class Provider {
 
     protected Provider() {
@@ -38,6 +46,20 @@
 
     public abstract Endpoint createAndPublishEndpoint(String s, Object obj);
 
+    //TODO
+    public abstract EndpointReference readEndpointReference(Source eprInfoset);
+    
+    //TODO
+    public abstract <T> T getPort(EndpointReference endpointReference, Class<T> serviceEndpointInterface, WebServiceFeature... features);
+    
+    //TODO
+    public abstract W3CEndpointReference createW3CEndpointReference(String address,
+            QName serviceName,
+            QName portName,
+            List<Element> metadata,
+            String wsdlDocumentLocation,
+            List<Element> referenceParameters);
+    
     public static final String JAXWSPROVIDER_PROPERTY = "javax.xml.ws.spi.Provider";
     private static final String DEFAULT_JAXWSPROVIDER = "org.apache.axis2.jaxws.spi.Provider";
 }
Index: modules/jaxws-api/src/javax/xml/ws/spi/ServiceDelegate.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/spi/ServiceDelegate.java	(revision 3542)
+++ modules/jaxws-api/src/javax/xml/ws/spi/ServiceDelegate.java	(revision 3544)
@@ -21,8 +21,10 @@
 import javax.xml.bind.JAXBContext;
 import javax.xml.namespace.QName;
 import javax.xml.ws.Dispatch;
+import javax.xml.ws.EndpointReference;
+import javax.xml.ws.Service;
+import javax.xml.ws.WebServiceFeature;
 import javax.xml.ws.handler.HandlerResolver;
-import java.net.URI;
 import java.net.URL;
 import java.util.Iterator;
 import java.util.concurrent.Executor;
@@ -36,15 +38,36 @@
 
     public abstract <T> T getPort(Class<T> class1);
 
+    //TODO
+    public abstract <T> T getPort(QName portName, Class<T> serviceEndpointInterface, WebServiceFeature... features);
+    
+    //TODO
+    public abstract <T> T getPort(EndpointReference endpointReference, Class<T> serviceEndpointInterface, WebServiceFeature... features);
+    
+    //TODO
+    public abstract <T> T getPort(Class<T> serviceEndpointInterface, WebServiceFeature... features);
+    
     public abstract void addPort(QName qname, String bindingId, String s);
 
-    public abstract <T>Dispatch<T> createDispatch(QName qname, Class<T> class1, javax.xml.ws.Service.Mode mode);
+    public abstract <T>Dispatch<T> createDispatch(QName qname, Class<T> class1, Service.Mode mode);
 
-    public abstract Dispatch<java.lang.Object> createDispatch(QName qname, JAXBContext jaxbcontext, javax.xml.ws.Service.Mode mode);
+    public abstract Dispatch<Object> createDispatch(QName qname, JAXBContext jaxbcontext, javax.xml.ws.Service.Mode mode);
 
+    //TODO
+    public abstract <T> Dispatch<T> createDispatch(QName portName, Class<T> type, Service.Mode mode, WebServiceFeature... features);
+    
+    //TODO
+    public abstract <T> Dispatch<T> createDispatch(EndpointReference endpointReference, Class<T> type, Service.Mode mode, WebServiceFeature... features);
+    
+    //TODO
+    public abstract Dispatch<Object> createDispatch(QName portName, JAXBContext context, Service.Mode mode, WebServiceFeature... features);
+    
+    //TODO
+    public abstract Dispatch<Object> createDispatch(EndpointReference endpointReference, JAXBContext context, Service.Mode mode, WebServiceFeature... features);
+    
     public abstract QName getServiceName();
 
-    public abstract Iterator<javax.xml.namespace.QName> getPorts();
+    public abstract Iterator<QName> getPorts();
 
     public abstract URL getWSDLDocumentLocation();
 
@@ -56,3 +79,4 @@
 
     public abstract void setExecutor(Executor executor);
 }
+
Index: modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReferenceBuilder.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReferenceBuilder.java	(revision 0)
+++ modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReferenceBuilder.java	(revision 3544)
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *      
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package javax.xml.ws.wsaddressing;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.xml.namespace.QName;
+import javax.xml.ws.spi.Provider;
+
+import org.w3c.dom.Element;
+
+public final class W3CEndpointReferenceBuilder {
+    private String address;
+    private QName serviceName;
+    private QName endpointName;
+    private String wsdlDocumentLocation;
+    private List<Element> referenceParameters;
+    private List<Element> metadataElements;
+    
+    public W3CEndpointReferenceBuilder() {
+    }
+    
+    public W3CEndpointReferenceBuilder address(String address) {
+        this.address = address;
+        return this;
+    }
+    
+    public W3CEndpointReferenceBuilder serviceName(QName serviceName) {
+        this.serviceName = serviceName;
+        return this;
+    }
+    
+    public W3CEndpointReferenceBuilder endpointName(QName endpointName) {
+        //TODO NLS enable
+        if (this.serviceName == null) {
+            throw new IllegalStateException("The endpoint qname cannot be set before the service qname.");
+        }
+        
+        this.endpointName = endpointName;
+        return this;
+    }
+    
+    public W3CEndpointReferenceBuilder wsdlDocumentLocation(String wsdlDocumentLocation) {
+        this.wsdlDocumentLocation = wsdlDocumentLocation;
+        return this;
+    }
+    
+    public W3CEndpointReferenceBuilder referenceParameter(Element referenceParameter) {
+        //TODO NLS enable
+        if (referenceParameter == null) {
+            throw new IllegalArgumentException("A reference parameter cannot be null.");
+        }
+        
+        if (this.referenceParameters == null) {
+            this.referenceParameters = new ArrayList<Element>();
+        }
+        
+        this.referenceParameters.add(referenceParameter);
+        return this;
+    }
+    
+    public W3CEndpointReferenceBuilder metadata(Element metadataElement) {
+        //TODO NLS enable
+        if (metadataElement == null) {
+            throw new IllegalArgumentException("A metadata element cannot be null.");
+        }
+        
+        if (this.metadataElements == null) {
+            this.metadataElements = new ArrayList<Element>();
+        }
+        
+        this.metadataElements.add(metadataElement);
+        return this;
+    }
+    
+    public W3CEndpointReference build() {
+        return Provider.provider().createW3CEndpointReference(address,
+                serviceName,
+                endpointName,
+                metadataElements,
+                wsdlDocumentLocation,
+                referenceParameters);
+    }
+
+    @Override
+    public String toString() {
+        // TODO Auto-generated method stub
+        return super.toString();
+    }
+}
+
Index: modules/jaxws-api/src/javax/xml/ws/wsaddressing/package-info.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/wsaddressing/package-info.java	(revision 0)
+++ modules/jaxws-api/src/javax/xml/ws/wsaddressing/package-info.java	(revision 3544)
@@ -0,0 +1,21 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *      
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+@javax.xml.bind.annotation.XmlSchema(namespace = "http://www.w3.org/2005/08/addressing", elementFormDefault = javax.xml.bind.annotation.XmlNsForm.QUALIFIED)
+package javax.xml.ws.wsaddressing;
+
Index: modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReference.java
===================================================================
--- modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReference.java	(revision 0)
+++ modules/jaxws-api/src/javax/xml/ws/wsaddressing/W3CEndpointReference.java	(revision 3544)
@@ -0,0 +1,252 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *      
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package javax.xml.ws.wsaddressing;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlAnyAttribute;
+import javax.xml.bind.annotation.XmlAnyElement;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlSchemaType;
+import javax.xml.bind.annotation.XmlTransient;
+import javax.xml.bind.annotation.XmlType;
+import javax.xml.bind.annotation.XmlValue;
+import javax.xml.namespace.QName;
+import javax.xml.transform.Result;
+import javax.xml.transform.Source;
+import javax.xml.ws.EndpointReference;
+import javax.xml.ws.WebServiceException;
+
+/**
+ * <p>Java class for EndpointReferenceType complex type.
+ * 
+ * <p>The following schema fragment specifies the expected content contained within this class.
+ * 
+ * <pre>
+ * &lt;complexType name="EndpointReferenceType">
+ *   &lt;complexContent>
+ *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType">
+ *       &lt;sequence>
+ *         &lt;element name="Address" type="{http://www.w3.org/2005/08/addressing}AttributedURIType"/>
+ *         &lt;element ref="{http://www.w3.org/2005/08/addressing}ReferenceParameters" minOccurs="0"/>
+ *         &lt;element ref="{http://www.w3.org/2005/08/addressing}Metadata" minOccurs="0"/>
+ *         &lt;any/>
+ *       &lt;/sequence>
+ *     &lt;/restriction>
+ *   &lt;/complexContent>
+ * &lt;/complexType>
+ * </pre>
+ * 
+ * 
+ */
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(name = "EndpointReferenceType", propOrder = {
+    "address",
+    "referenceParameters",
+    "metadata",
+    "any"
+})
+public final class W3CEndpointReference extends EndpointReference {
+    @XmlTransient
+    protected static volatile JAXBContext jaxbContext;
+    @XmlTransient
+    protected static final String NS = "http://www.w3.org/2005/08/addressing";
+    @XmlTransient
+    protected static final QName NAME = new QName(NS, "EndpointReference", "wsa");
+    
+    @XmlElement(name = "Address", required = true)
+    protected AttributedURIType address;
+    @XmlElement(name = "ReferenceParameters")
+    protected ReferenceParametersType referenceParameters;
+    @XmlElement(name = "Metadata")
+    protected MetadataType metadata;
+    @XmlAnyElement(lax = true)
+    protected List<Object> any;
+    @XmlAnyAttribute
+    private Map<QName, String> otherAttributes = new HashMap<QName, String>();
+
+    protected W3CEndpointReference() {
+    }
+    
+    public W3CEndpointReference(Source eprInfoset) {
+        super();
+        
+        try {
+            JAXBContext jaxbContext = getJAXBContext();
+            Unmarshaller um = jaxbContext.createUnmarshaller();
+            JAXBElement<W3CEndpointReference> element =
+                um.unmarshal(eprInfoset, W3CEndpointReference.class);
+            W3CEndpointReference w3cEPR = element.getValue();
+            
+            address = w3cEPR.address;
+            referenceParameters = w3cEPR.referenceParameters;
+            metadata = w3cEPR.metadata;
+            any = w3cEPR.any;
+            otherAttributes.putAll(w3cEPR.otherAttributes);
+        }
+        catch (Exception e) {
+            //TODO NLS enable.
+            throw new WebServiceException("Unable to create W3C endpoint reference.", e);
+        }
+    }
+    
+    @Override
+    public void writeTo(Result result) {
+        if (result == null) {
+            //TODO NLS enable
+            throw new IllegalArgumentException("Null is not allowed.");
+        }
+        
+        try {
+            JAXBContext jaxbContext = getJAXBContext();
+            Marshaller m = jaxbContext.createMarshaller();
+            m.setProperty(Marshaller.JAXB_FRAGMENT, Boolean.TRUE);
+            JAXBElement<W3CEndpointReference> element =
+                new JAXBElement<W3CEndpointReference>(NAME, W3CEndpointReference.class, this);
+            m.marshal(element, result);
+        }
+        catch (Exception e) {
+            //TODO NLS enable
+            throw new WebServiceException("writeTo failure.", e);
+        }
+    }
+
+    private JAXBContext getJAXBContext() throws JAXBException {
+        //This is an implementation of double-checked locking.
+        //It works because jaxbContext is volatile.
+        if (jaxbContext == null) {
+            synchronized (W3CEndpointReference.class) {
+                if (jaxbContext == null)
+                    jaxbContext = JAXBContext.newInstance(W3CEndpointReference.class);
+            }
+        }
+        
+        return jaxbContext;
+    }
+
+    /**
+     * <p>Java class for AttributedURIType complex type.
+     * 
+     * <p>The following schema fragment specifies the expected content contained within this class.
+     * 
+     * <pre>
+     * &lt;complexType name="AttributedURIType">
+     *   &lt;simpleContent>
+     *     &lt;extension base="&lt;http://www.w3.org/2001/XMLSchema>anyURI">
+     *     &lt;/extension>
+     *   &lt;/simpleContent>
+     * &lt;/complexType>
+     * </pre>
+     * 
+     * 
+     */
+    @XmlAccessorType(XmlAccessType.FIELD)
+    @XmlType(name = "AttributedURIType", propOrder = {
+        "value"
+    })
+    private static class AttributedURIType {
+
+        @XmlValue
+        @XmlSchemaType(name = "anyURI")
+        protected String value;
+        @XmlAnyAttribute
+        private Map<QName, String> otherAttributes = new HashMap<QName, String>();
+        
+        public AttributedURIType() {
+        }
+    }
+
+    /**
+     * <p>Java class for ReferenceParametersType complex type.
+     * 
+     * <p>The following schema fragment specifies the expected content contained within this class.
+     * 
+     * <pre>
+     * &lt;complexType name="ReferenceParametersType">
+     *   &lt;complexContent>
+     *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType">
+     *       &lt;sequence>
+     *         &lt;any/>
+     *       &lt;/sequence>
+     *     &lt;/restriction>
+     *   &lt;/complexContent>
+     * &lt;/complexType>
+     * </pre>
+     * 
+     * 
+     */
+    @XmlAccessorType(XmlAccessType.FIELD)
+    @XmlType(name = "ReferenceParametersType", propOrder = {
+        "any"
+    })
+    private static class ReferenceParametersType {
+
+        @XmlAnyElement(lax = true)
+        protected List<Object> any;
+        @XmlAnyAttribute
+        private Map<QName, String> otherAttributes = new HashMap<QName, String>();
+        
+        public ReferenceParametersType() {
+        }
+    }
+
+    /**
+     * <p>Java class for MetadataType complex type.
+     * 
+     * <p>The following schema fragment specifies the expected content contained within this class.
+     * 
+     * <pre>
+     * &lt;complexType name="MetadataType">
+     *   &lt;complexContent>
+     *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType">
+     *       &lt;sequence>
+     *         &lt;any/>
+     *       &lt;/sequence>
+     *     &lt;/restriction>
+     *   &lt;/complexContent>
+     * &lt;/complexType>
+     * </pre>
+     * 
+     * 
+     */
+    @XmlAccessorType(XmlAccessType.FIELD)
+    @XmlType(name = "MetadataType", propOrder = {
+        "any"
+    })
+    private static class MetadataType {
+
+        @XmlAnyElement(lax = true)
+        protected List<Object> any;
+        @XmlAnyAttribute
+        private Map<QName, String> otherAttributes = new HashMap<QName, String>();
+        
+        public MetadataType() {
+        }
+    }
+}
+
