Index: modules/jaxbri/src/org/apache/axis2/jaxbri/CodeGenerationUtility.java
===================================================================
--- modules/jaxbri/src/org/apache/axis2/jaxbri/CodeGenerationUtility.java	(revision 677175)
+++ modules/jaxbri/src/org/apache/axis2/jaxbri/CodeGenerationUtility.java	(revision )
@@ -201,7 +201,22 @@
                 });
 
                 sc.parseSchema((InputSource) schemaToInputSourceMap.get(key));
+                sc.getOptions().addGrammar((InputSource) schemaToInputSourceMap.get(key));
 
+                for (Object property : properties.keySet()){
+                    String propertyName = (String) property;
+                    if (propertyName.startsWith("X")) {
+                        String[] args = null;
+                        String propertyValue = (String) properties.get(property);
+                        if (propertyValue != null && propertyValue.trim().length() > 0) {
+                            args = new String[]{"-" + propertyName, propertyValue};
+                        } else {
+                            args = new String[]{"-" + propertyName};
+                        }
+                        sc.getOptions().parseArguments(args);
+                    }
+                }
+
                 // Bind the XML
                 S2JJAXBModel jaxbModel = sc.bind();
 
