Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/ideaplugin/frames/Axi2PluginPage.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/ideaplugin/frames/Axi2PluginPage.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/ideaplugin/frames/Axi2PluginPage.java	(working copy)
@@ -3,6 +3,7 @@
 import com.intellij.openapi.project.Project;
 import org.apache.axis2.tools.idea.Java2CodeFrame;
 import org.apache.axis2.tools.java2wsdl.Java2WSDLFrame;
+import org.apache.axis2.tools.wizardframe.CodegenFrame;
 
 
 import javax.swing.*;
@@ -35,14 +36,12 @@
 public class Axi2PluginPage extends JFrame implements ActionListener {
     ButtonGroup cbg;
     JRadioButton service;
-    JRadioButton wsdlGen;
-    JRadioButton sourceGen;
+    JRadioButton codeGen;
     JButton butOK;
     JButton butCancle;
     JPanel imglbl;
     Project project;
-    Java2CodeFrame winj2c;
-    Java2WSDLFrame winj2w;
+    CodegenFrame codegenFrame;
 
     private int defaultCloseOperation;
 
@@ -61,22 +60,18 @@
 
         /*Create a service archive Radio Button  */
 
-        service = new JRadioButton("Create a service archive", true);
+        service = new JRadioButton("Axis2 Aervice Archive", true);
         service.setToolTipText("Hepls package classes, libs and WSDLs to create a archive that can be deployed in Axis2");
         cbg.add(service);
         getContentPane().add(service);
 
         /*Create a Code code generation Radio Button */
 
-        wsdlGen = new JRadioButton("Generate a WSDL from a java source file", false);
-        wsdlGen.setToolTipText("you can generate a WSDL from a java source file ");
-        cbg.add(wsdlGen);
-        getContentPane().add(wsdlGen);
+        codeGen = new JRadioButton("Axis2 Code Generator", false);
+        codeGen.setToolTipText("you can generate a WSDL from a java source file or java code from a WSDL ");
+        cbg.add(codeGen);
+        getContentPane().add(codeGen);
 
-        sourceGen =new JRadioButton("Generate java source code from a WSDL file",true);
-        sourceGen.setToolTipText("you can generate java code from a WSDL");
-        cbg.add(sourceGen);
-        getContentPane().add(sourceGen);
 
 
         butOK = new JButton("OK");
@@ -134,24 +129,10 @@
         } else if (obj == butOK) {
             this.hide();
             setVisible(false);
-            if (wsdlGen.isSelected()) {                
-
-                winj2c = new Java2CodeFrame();
-                winj2c.setResizable(false);
-                winj2c.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
-                winj2c.setProject(project);
-                winj2c.pack();
-                winj2c.show();
-                
-            } else if (sourceGen.isSelected() ) {
-
-                winj2w = new Java2WSDLFrame();
-                winj2w.setResizable(false);
-                winj2w.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
-                winj2w.setProject(project);
-                winj2w.pack();
-                winj2w.show();
-
+            if (codeGen.isSelected()) {
+                codegenFrame = new CodegenFrame();
+                codegenFrame.setVisible(true);
+                codegenFrame .setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);                
             }else{
                 ServiceArciveFrame window = new ServiceArciveFrame();
                 window.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
@@ -204,20 +185,16 @@
         if (c.isVisible()) {
             c.setBounds(insets.left + 24, insets.top + 136, 296, 24);
         }
-         c = parent.getComponent(2);
+        c = parent.getComponent(2);
         if (c.isVisible()) {
-            c.setBounds(insets.left + 24, insets.top + 168, 296, 24);
+            c.setBounds(insets.left + 130, insets.top + 210, 80, 24);
         }
         c = parent.getComponent(3);
         if (c.isVisible()) {
-            c.setBounds(insets.left + 130, insets.top + 210, 80, 24);
+            c.setBounds(insets.left + 215, insets.top + 210, 80, 24);
         }
         c = parent.getComponent(4);
         if (c.isVisible()) {
-            c.setBounds(insets.left + 215, insets.top + 210, 80, 24);
-        }
-        c = parent.getComponent(5);
-        if (c.isVisible()) {
             c.setBounds(insets.left, insets.top, 320, 80);
         }
     }
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/ImagePanel.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/ImagePanel.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/ImagePanel.java	(working copy)
@@ -1,108 +0,0 @@
-package org.apache.axis2.tools.idea;
-
-import javax.swing.*;
-import java.awt.*;
-
-/*
-* Copyright 2004,2005 The Apache Software Foundation.
-*
-* Licensed under the Apache License, Version 2.0 (the "License");
-* you may not use this file except in compliance with the License.
-* You may obtain a copy of the License at
-*
-*      http://www.apache.org/licenses/LICENSE-2.0
-*
-* Unless required by applicable law or agreed to in writing, software
-* distributed under the License is distributed on an "AS IS" BASIS,
-* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-* See the License for the specific language governing permissions and
-* limitations under the License.
-*
-*
-*/
-
-/**
- * Author : Deepal Jayasinghe
- * Date: Jul 20, 2005
- * Time: 9:33:14 PM
- */
-public class ImagePanel extends JPanel {
-    JLabel lblImage;
-    JLabel labTop;
-    JLabel lblBottom;
-    ImageIcon apachelogo;
-
-    public ImagePanel() {
-        ImageLayout customLayout = new ImageLayout();
-        setLayout(customLayout);
-
-        java.net.URL resource = ImagePanel.class.getResource("/icons/asf-feather.png");
-        apachelogo = new ImageIcon(resource);
-
-        lblImage = new JLabel(apachelogo);
-        add(lblImage);
-        lblImage.setBackground(Color.white);
-
-        labTop = new JLabel();
-        labTop.setBackground(Color.white);
-        add(labTop);
-        labTop.setFont(new Font("Helvetica", Font.BOLD, 12));
-
-        lblBottom = new JLabel();
-        add(lblBottom);
-        lblBottom.setBackground(Color.white);
-        lblBottom.setFont(new Font("Helvetica", Font.PLAIN, 10));
-
-        setSize(getPreferredSize());
-        this.setBackground(Color.white);
-    }
-
-    public void setCaptions(String lbl1, String lbl2) {
-        labTop.setText(lbl1);
-        lblBottom.setText(lbl2);
-    }
-}
-
-class ImageLayout implements LayoutManager {
-
-    public ImageLayout() {
-    }
-
-    public void addLayoutComponent(String name, Component comp) {
-    }
-
-    public void removeLayoutComponent(Component comp) {
-    }
-
-    public Dimension preferredLayoutSize(Container parent) {
-        Dimension dim = new Dimension(0, 0);
-
-        Insets insets = parent.getInsets();
-        dim.width = 535 + insets.left + insets.right;
-        dim.height = 77 + insets.top + insets.bottom;
-
-        return dim;
-    }
-
-    public Dimension minimumLayoutSize(Container parent) {
-        return new Dimension(0, 0);
-    }
-
-    public void layoutContainer(Container parent) {
-        Insets insets = parent.getInsets();
-
-        Component c;
-        c = parent.getComponent(0);
-        if (c.isVisible()) {
-            c.setBounds(insets.left + 368, insets.top, 168, 80);
-        }
-        c = parent.getComponent(1);
-        if (c.isVisible()) {
-            c.setBounds(insets.left, insets.top, 368, 40);
-        }
-        c = parent.getComponent(2);
-        if (c.isVisible()) {
-            c.setBounds(insets.left, insets.top + 40, 368, 40);
-        }
-    }
-}
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/Java2CodeFrame.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/Java2CodeFrame.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/idea/Java2CodeFrame.java	(working copy)
@@ -41,7 +41,7 @@
  * Time: 2:26:15 PM
  */
 public class Java2CodeFrame extends JFrame {
-    ImagePanel panel_3;
+   // ImagePanel panel_3;
     FirstPanel plMiddle;
     BottomPanel lblBottom;
     //    SecondPanel secondPanel;
@@ -64,11 +64,11 @@
 
         codegenBean = new CodegenBean();
 
-        panel_3 = new ImagePanel();
+      /*  panel_3 = new ImagePanel();
         panel_3.setCaptions("  WSDL selection page"
-                , "  Welcome to the Axis2 code generation wizard. Select the WSDL file");
+                , "  Welcome to the Axis2 code generation wizard. Select the WSDL file");*/
 
-        getContentPane().add(panel_3);
+       // getContentPane().add(panel_3);
 
         plMiddle = new FirstPanel(codegenBean);
         getContentPane().add(plMiddle);
@@ -259,8 +259,8 @@
                     panleID--;
                     break;
                 }
-                panel_3.setCaptions("  Custom Options"
-                        , "  Set the options for the code generation");
+              /*  panel_3.setCaptions("  Custom Options"
+                        , "  Set the options for the code generation");*/
 
                 this.secondPanel.setVisible(true);
                 if(this.secondPanel.codegenBean == null)
@@ -273,8 +273,8 @@
             }
             case 2: {
 
-                panel_3.setCaptions("  Set the output location for the generated code"
-                        , "  set the output project for the generated code");
+               /* panel_3.setCaptions("  Set the output location for the generated code"
+                        , "  set the output project for the generated code");*/
                 this.secondPanel.setVisible(false);
                 this.plMiddle.setVisible(false);
                 this.secondPanel.fillBean();
@@ -310,8 +310,8 @@
         panleID--;
         switch (panleID) {
             case 0: {
-                panel_3.setCaptions("  WSDL selection page"
-                        , "  Welcome to the Axis2 code generation wizard. Select the WSDL file");
+               /* panel_3.setCaptions("  WSDL selection page"
+                        , "  Welcome to the Axis2 code generation wizard. Select the WSDL file");*/
                 this.secondPanel.setVisible(false);
                 this.plMiddle.setVisible(true);
                 this.outputpane.setVisible(false);
@@ -319,8 +319,8 @@
                 break;
             }
             case 1: {
-                panel_3.setCaptions("  Custom  Options"
-                        , "  Set the options for the code generation");
+              /*  panel_3.setCaptions("  Custom  Options"
+                        , "  Set the options for the code generation");*/
 
                 this.secondPanel.setVisible(true);
                 this.plMiddle.setVisible(false);
@@ -377,23 +377,23 @@
         Insets insets = parent.getInsets();
 
         Component c;
-        c = parent.getComponent(0);
+       /* c = parent.getComponent(0);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top, 600, 80);
-        }
-        c = parent.getComponent(1);
+        }*/
+        c = parent.getComponent(0);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 80, 600, 480);
         }
-        c = parent.getComponent(3);
+        c = parent.getComponent(2);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 80, 600, 480);
         }
-        c = parent.getComponent(4);
+        c = parent.getComponent(3);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 80, 600, 480);
         }
-        c = parent.getComponent(2);
+        c = parent.getComponent(1);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 550, 600, 50);
         }
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/CodegenFrame.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/CodegenFrame.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/CodegenFrame.java	(working copy)
@@ -3,6 +3,7 @@
 import org.apache.axis2.tools.bean.WsdlgenBean;
 import org.apache.axis2.tools.bean.CodegenBean;
 import org.apache.axis2.tools.idea.FirstPanel;
+import org.apache.axis2.tools.idea.ChooserPanel;
 import org.apache.axis2.tools.component.Utilities;
 import org.apache.axis2.tools.component.WizardPanel;
 
@@ -34,10 +35,10 @@
 
         wsdlgenBean= new WsdlgenBean();
         codegenBean = new CodegenBean();
-       /*
+
         panel = new ChooserPanel(getWizardComponents());
         getWizardComponents().addWizardPanel(PANEL_CHOOSER , panel);
-
+      /*
         panel = new FirstPanel(getWizardComponents(),codegenBean);
         getWizardComponents().addWizardPanel(PANEL_FIRST_A , panel);
 
@@ -56,7 +57,7 @@
         panel = new J2WLocationWizardPanel(getWizardComponents(),wsdlgenBean);
         getWizardComponents().addWizardPanel(PANEL_LAST_B , panel);*/
 
-        setSize(550, 200);
+        setSize(650, 600);
         Utilities.centerComponentOnScreen(this);
 
     }
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/WizardFrame.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/WizardFrame.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/wizardframe/WizardFrame.java	(working copy)
@@ -1,9 +1,6 @@
 package org.apache.axis2.tools.wizardframe;
 
-import org.apache.axis2.tools.component.CancelAction;
-import org.apache.axis2.tools.component.WizardComponents;
-import org.apache.axis2.tools.component.FinishAction;
-import org.apache.axis2.tools.component.DefaultWizardComponents;
+import org.apache.axis2.tools.component.*;
 
 import javax.swing.*;
 import java.awt.*;
@@ -16,6 +13,15 @@
  * wizardFrame class
  */
 public class WizardFrame extends JFrame {
+     /**
+     * variable
+     */
+
+    private JLabel panelImageLabel;
+    private JLabel panelTopTitleLabel;
+    private JLabel panelBottomTitleLabel;
+    private JLabel errorLabel;
+    private JPanel errorPanel;
     private WizardComponents wizardComponents;
 
     public WizardFrame() {
@@ -27,25 +33,25 @@
         wizardComponents = new DefaultWizardComponents();
 
         this.getContentPane().setLayout(new GridBagLayout());
-       /* this.getContentPane().add(new ImagePanel(wizardComponents)
+        this.getContentPane().add(createTitlePanel()
                 , new GridBagConstraints(0, 0, 1, 1, 1.0, 0.0
-                , GridBagConstraints.NORTH, GridBagConstraints.BOTH
+                , GridBagConstraints.NORTH, GridBagConstraints.HORIZONTAL
                 , new Insets(0, 0, 0, 0), 0, 0));
 
         this.getContentPane().add(new JSeparator()
                 , new GridBagConstraints(0, 1, 1, 1, 1.0, 0.0
                 , GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL,
-                new Insets(1, 1, 1, 1), 0, 0));
+                new Insets(0, 0, 0, 0), 0, 0));
 
         this.getContentPane().add(wizardComponents.getWizardPanelsContainer()
                 , new GridBagConstraints(0, 2, 1, 1, 1.0, 1.0
-                , GridBagConstraints.PAGE_START , GridBagConstraints.BOTH
+                , GridBagConstraints.NORTHWEST , GridBagConstraints.NONE
                 , new Insets(10, 0, 0, 0), 0, 0));
 
-        this.getContentPane().add(new ProgressBarPanel(wizardComponents)
+       /* this.getContentPane().add(new ProgressBarPanel(wizardComponents)
                 , new GridBagConstraints(0, 3, 1, 1, 1.0, 0.0
                 , GridBagConstraints.CENTER, GridBagConstraints.NONE
-                , new Insets(10, 0, 0, 0), 0, 0));
+                , new Insets(10, 0, 0, 0), 0, 0));  */
 
         this.getContentPane().add(new JSeparator()
                 , new GridBagConstraints(0, 4, 1, 1, 1.0, 0.0
@@ -55,14 +61,18 @@
         this.getContentPane().add(createButtonPanel(),
                 new GridBagConstraints(0, 5, 1, 1, 1.0, 0.0
                         ,GridBagConstraints.EAST, GridBagConstraints.NONE,
-                        new Insets(20, 20, 20, 20), 0, 0));*/
+                        new Insets(20, 20, 20, 20), 0, 0));
 
         ImageIcon  img=new ImageIcon("icons/icon.png");
         this.setIconImage(img.getImage());
 
         wizardComponents.addPropertyChangeListener(new PropertyChangeListener() {
             public void propertyChange(PropertyChangeEvent event) {
-
+                setPanelTopTitle(((WizardPanel)event.getNewValue()).getPanelTopTitle());
+                setPanelTitleImage(((WizardPanel)event.getNewValue()).getPanelImage());
+                setPanelBottomTitle(((WizardPanel)event.getNewValue()).getPanelBottomTitle());
+                setErrorVisible(false
+                        ,((WizardPanel)event.getNewValue()).getError());
             }
         });
         wizardComponents.setFinishAction(createFinishAction());
@@ -82,7 +92,77 @@
         wizardComponents.updateComponents();
         super.show();
     }
+      // set error message
+    protected void setErrorVisible(boolean flag ,String  error){
+        errorLabel.setText(error);
+        errorPanel.setVisible(flag);
 
+    }
+    protected void setBottomVisible(boolean flag){
+        panelBottomTitleLabel.setVisible(flag);
+    }
+    //Set Title Panel
+    protected void setPanelTopTitle(String title) {
+        panelTopTitleLabel.setText(title);
+    }
+
+    protected void setPanelBottomTitle(String title) {
+        panelBottomTitleLabel.setText(title);
+    }
+    // set title image
+    protected void setPanelTitleImage(ImageIcon image){
+        panelImageLabel.setIcon(image );
+    }
+
+     protected JPanel createTitlePanel() {
+
+        JPanel panel = new JPanel(new GridBagLayout() );
+
+        panel.setBackground(Color.white );
+
+        panelTopTitleLabel = new JLabel();
+
+        panelBottomTitleLabel=new JLabel();
+
+        panelImageLabel= new JLabel();
+
+        errorLabel = new JLabel();
+        errorLabel.setVisible(false);
+        errorLabel.setBackground(Color.blue);
+
+        errorPanel =new JPanel();
+        errorPanel .setLayout(new GridBagLayout());
+        errorPanel .setBorder(BorderFactory.createLineBorder(Color.black));
+
+        errorPanel.add(errorLabel,
+                new GridBagConstraints(0, 0, 1, 1, 0.0, 0.0
+                , GridBagConstraints.CENTER, GridBagConstraints.BOTH
+                , new Insets(10, 10, 0, 10), 0, 0));
+
+        panel.add(panelTopTitleLabel
+                , new GridBagConstraints(0, 0, 1, 1, 0.5, 0.0
+                , GridBagConstraints.LINE_START , GridBagConstraints.BOTH
+                , new Insets(10, 20, 0, 0), 0, 0));
+
+        panel.add(panelBottomTitleLabel
+                , new GridBagConstraints(0, 1, 1, 1, 0.5, 0.0
+                , GridBagConstraints.LINE_START, GridBagConstraints.BOTH
+                , new Insets(10,20, 0, 0), 0, 0));
+
+        panel.add(panelImageLabel
+                , new GridBagConstraints(1, 1, 1, 1, 0.0, 0.0
+                , GridBagConstraints.CENTER, GridBagConstraints.BOTH
+                , new Insets(0, 0, 0, 0), 0, 0));
+
+
+        panel.add(errorPanel
+                , new GridBagConstraints(0, 1, 1, 1, 0.0, 0.0
+                , GridBagConstraints.SOUTH, GridBagConstraints.BOTH
+                , new Insets(20, 0, 0, 0), 0, 0));
+
+        return panel;
+
+    }
     protected JPanel createButtonPanel() {
         JPanel panel = new JPanel(new GridLayout());
         panel.add(wizardComponents.getBackButton());
Index: modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/java2wsdl/Java2WSDLFrame.java
===================================================================
--- modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/java2wsdl/Java2WSDLFrame.java	(revision 551504)
+++ modules/tool/axis2-idea-plugin/src/main/java/org/apache/axis2/tools/java2wsdl/Java2WSDLFrame.java	(working copy)
@@ -1,6 +1,5 @@
 package org.apache.axis2.tools.java2wsdl;
 
-import org.apache.axis2.tools.idea.*;
 import org.apache.axis2.tools.bean.WsdlgenBean;
 import com.intellij.openapi.project.Project;
 import javax.swing.*;
@@ -10,7 +9,7 @@
 public class Java2WSDLFrame extends JFrame {
 
     // first panel
-    ImagePanel imgPanel;
+  //  ImagePanel imgPanel;
     MiddlePanel plMiddle;
     BottomPanel plBottom;
     OptionPanel opPanel;
@@ -35,10 +34,10 @@
         wsdlgenBean=new WsdlgenBean();
         //add image panel
 
-        imgPanel = new ImagePanel();
+       /* imgPanel = new ImagePanel();
         imgPanel.setCaptions("  Java source/classpath selection"
                 , "  Welcome to the Axis2 Java source code generation wizard.");
-        getContentPane().add(imgPanel);
+        getContentPane().add(imgPanel);*/
 
         //add bottom panel
 
@@ -122,9 +121,9 @@
 
         switch (panleID) {
             case 1:{
-                this.imgPanel .setCaptions(" Java source/classpath selection" ,
+                /*this.imgPanel .setCaptions(" Java source/classpath selection" ,
                         "  Select the classes and the libraries.");
-
+*/
                 this.plMiddle .setVisible(true);
 
                 this.opPanel.setVisible(false);
@@ -136,10 +135,10 @@
                 break;
             }
             case 2: {
-                this.imgPanel .setCaptions(" Java to WSDL Options " ,
+              /*  this.imgPanel .setCaptions(" Java to WSDL Options " ,
                         "  Set the Options for the generator.");
                 this.plMiddle .setVisible(false);
-
+*/
                 this.opPanel.setVisible(true);
 
                 this.outPanel .setVisible(false);
@@ -149,10 +148,10 @@
                 break;
             }
             case 3: {
-                this.imgPanel .setCaptions(" WSDL file Output location " ,
+               /* this.imgPanel .setCaptions(" WSDL file Output location " ,
                         "  Select the location for the generated WSDL.");
                 this.plMiddle .setVisible(false);
-
+*/
                 this.opPanel.setVisible(false);
 
                 this.outPanel .setVisible(true);
@@ -169,9 +168,9 @@
         switch (panleID) {
             case 1: {
 
-                this.imgPanel .setCaptions(" Java source/classpath selection" ,
+               /* this.imgPanel .setCaptions(" Java source/classpath selection" ,
                         "  Select the classes and the libraries.");
-
+*/
                 this.opPanel.setVisible(false);
 
                 this.plMiddle.setVisible(true);
@@ -185,9 +184,9 @@
 
             case 2: {
 
-                this.imgPanel .setCaptions(" Java to WSDL Options " ,
+              /*  this.imgPanel .setCaptions(" Java to WSDL Options " ,
                         "  Set the Options for the generator.");
-
+*/
                 this.plMiddle .setVisible(false);
 
                 this.outPanel.setVisible(false);
@@ -201,9 +200,9 @@
 
             case 3: {
 
-                this.imgPanel .setCaptions(" WSDL file Output location " ,
+               /* this.imgPanel .setCaptions(" WSDL file Output location " ,
                         "  Select the location for the generated WSDL.");
-
+*/
                 this.plMiddle .setVisible(false);
 
                 this.outPanel.setVisible(true);
@@ -249,9 +248,13 @@
         Insets insets = parent.getInsets();
 
         Component c;
-        c = parent.getComponent(0);
+       /* c = parent.getComponent(0);
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top, 550, 80);
+        }*/
+        c = parent.getComponent(1);
+        if (c.isVisible()) {
+            c.setBounds(insets.left, insets.top + 80, 500, 450);
         }
         c = parent.getComponent(2);
         if (c.isVisible()) {
@@ -261,12 +264,8 @@
         if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 80, 500, 450);
         }
-        c = parent.getComponent(4);
+        c = parent.getComponent(0);
         if (c.isVisible()) {
-            c.setBounds(insets.left, insets.top + 80, 500, 450);
-        }
-        c = parent.getComponent(1);
-        if (c.isVisible()) {
             c.setBounds(insets.left, insets.top + 525, 500, 50);
         }
 
