Index: src/org/apache/axis2/policy/builders/MTOM10AssertionBuilder.java
===================================================================
--- src/org/apache/axis2/policy/builders/MTOM10AssertionBuilder.java	(revision 703931)
+++ src/org/apache/axis2/policy/builders/MTOM10AssertionBuilder.java	(working copy)
@@ -43,7 +43,7 @@
  */
 public class MTOM10AssertionBuilder implements AssertionBuilder {
 
-    private static Log log = LogFactory.getLog(MTOM10AssertionBuilder.class);
+	public static final Log log = LogFactory.getLog(MTOM10AssertionBuilder.class);
 
     public Assertion build(OMElement element, AssertionBuilderFactory factory)
             throws IllegalArgumentException {
Index: src/org/apache/axis2/policy/builders/MTOM11AssertionBuilder.java
===================================================================
--- src/org/apache/axis2/policy/builders/MTOM11AssertionBuilder.java	(revision 703931)
+++ src/org/apache/axis2/policy/builders/MTOM11AssertionBuilder.java	(working copy)
@@ -16,7 +16,7 @@
 
 public class MTOM11AssertionBuilder implements AssertionBuilder{
     
-    private static Log log = LogFactory.getLog(MTOM10AssertionBuilder.class);
+    public static final Log log = LogFactory.getLog(MTOM10AssertionBuilder.class);
 
     public Assertion build(OMElement element, AssertionBuilderFactory factory)
             throws IllegalArgumentException {
