Index: modules/transport/local/src/org/apache/axis2/transport/local/LocalResponder.java
===================================================================
--- modules/transport/local/src/org/apache/axis2/transport/local/LocalResponder.java	(revision 1060336)
+++ modules/transport/local/src/org/apache/axis2/transport/local/LocalResponder.java	(working copy)
@@ -20,19 +20,29 @@
 
 package org.apache.axis2.transport.local;
 
+import org.apache.axiom.om.OMXMLParserWrapper;
+import org.apache.axiom.soap.SOAPEnvelope;
 import org.apache.axis2.AxisFault;
+import org.apache.axis2.Constants;
+import org.apache.axis2.addressing.AddressingConstants;
 import org.apache.axis2.addressing.EndpointReference;
+import org.apache.axis2.addressing.RelatesTo;
+import org.apache.axis2.builder.BuilderUtil;
 import org.apache.axis2.context.ConfigurationContext;
 import org.apache.axis2.context.MessageContext;
 import org.apache.axis2.description.TransportOutDescription;
+import org.apache.axis2.description.WSDL2Constants;
+import org.apache.axis2.engine.AxisEngine;
 import org.apache.axis2.handlers.AbstractHandler;
 import org.apache.axis2.transport.TransportSender;
 import org.apache.axis2.transport.TransportUtils;
+import org.apache.axis2.wsdl.WSDLConstants;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
-import java.io.ByteArrayOutputStream;
-import java.io.OutputStream;
+import javax.xml.stream.XMLStreamException;
+import java.io.*;
+import java.util.Map;
 
 /**
  * LocalResponder
@@ -40,7 +50,6 @@
 public class LocalResponder extends AbstractHandler implements TransportSender {
     protected static final Log log = LogFactory.getLog(LocalResponder.class);
     
-    
     //  fixed for Executing LocalTransport in MulthThread. 
     private OutputStream out;
 
@@ -61,39 +70,52 @@
     /**
      * Method invoke
      *
-     * @param msgContext the active MessageContext
+     * @param proxyOutMessageContext the active MessageContext
      * @throws AxisFault
      */
-    public InvocationResponse invoke(MessageContext msgContext) throws AxisFault {
+    public InvocationResponse invoke(MessageContext proxyOutMessageContext) throws AxisFault {
 
         // Check for the REST behaviour, if you desire rest beahaviour
         // put a <parameter name="doREST" value="true"/> at the axis2.xml
-        msgContext.setDoingMTOM(TransportUtils.doWriteMTOM(msgContext));
-        msgContext.setDoingSwA(TransportUtils.doWriteSwA(msgContext));
+        proxyOutMessageContext.setDoingMTOM(TransportUtils.doWriteMTOM(proxyOutMessageContext));
+        proxyOutMessageContext.setDoingSwA(TransportUtils.doWriteSwA(proxyOutMessageContext));
 
         EndpointReference epr = null;
 
-        if (msgContext.getTo() != null && !msgContext.getTo().hasAnonymousAddress()) {
-            epr = msgContext.getTo();
+        if (proxyOutMessageContext.getTo() != null && !proxyOutMessageContext.getTo().hasAnonymousAddress()) {
+            epr = proxyOutMessageContext.getTo();
         }
 
         try {
             if (log.isDebugEnabled()) {
-                log.debug("Response - " + msgContext.getEnvelope().toString());
+                log.debug("Response - " + proxyOutMessageContext.getEnvelope().toString());
             }
 
             if (epr != null) {
                 if (!epr.hasNoneAddress()) {
-                    TransportUtils.writeMessage(msgContext, out);
+                    TransportUtils.writeMessage(proxyOutMessageContext, out);
                 }
             } else {
-                out = (OutputStream) msgContext.getProperty(MessageContext.TRANSPORT_OUT);
+                // check whether the property is set.
+                // use the operation context and get the in message context
+                MessageContext proxyInMessageContext = proxyOutMessageContext.
+                        getOperationContext().getMessageContext(WSDL2Constants.MESSAGE_LABEL_IN);
+                // get the message context
+                MessageContext initialMessageContext = (MessageContext)proxyInMessageContext.
+                        getProperty("IN_MESSAGE_CONTEXT");
 
-                if (out != null) {
-                    TransportUtils.writeMessage(msgContext, out);
+                if (initialMessageContext != null) {
+                    handleResponse(proxyOutMessageContext, initialMessageContext);
+
                 } else {
-                    throw new AxisFault(
-                            "Both the TO and Property MessageContext.TRANSPORT_OUT is Null, No where to send");
+                    out = (OutputStream) proxyOutMessageContext.getProperty(MessageContext.TRANSPORT_OUT);
+
+                    if (out != null) {
+                        TransportUtils.writeMessage(proxyOutMessageContext, out);
+                    } else {
+                        throw new AxisFault(
+                                "Both the TO and Property MessageContext.TRANSPORT_OUT is Null, No where to send");
+                    }
                 }
             }
         } catch (AxisFault axisFault) {
@@ -102,8 +124,78 @@
             log.error("Error sending response", axisFault);
         }
 
-        TransportUtils.setResponseWritten(msgContext, true);
+        TransportUtils.setResponseWritten(proxyOutMessageContext, true);
         
         return InvocationResponse.CONTINUE;
     }
+
+    /**
+     * Retrieves the properties from the proxyOutMessageContext and sets the values to the
+     * inMessageContext.
+     *
+     * @param proxyOutMessageContext the active message context
+     * @param initialMessageContext the initial message context, which was stored as a property
+     * in the proxyOutMessageContext
+     * @throws AxisFault
+     */
+    private void handleResponse(MessageContext proxyOutMessageContext, MessageContext initialMessageContext) throws AxisFault {
+        MessageContext inMessageContext = initialMessageContext.getOperationContext().
+                getMessageContext(WSDL2Constants.MESSAGE_LABEL_IN);
+
+        // setting the properties
+        Map<String,Object> initialPropertyMap =  initialMessageContext.getProperties();
+        initialMessageContext.setProperties(initialPropertyMap);
+
+        inMessageContext.setEnvelope(getEnvelope(proxyOutMessageContext));
+        inMessageContext.setAxisServiceGroup(initialMessageContext.getAxisServiceGroup());
+        inMessageContext.setAxisService(initialMessageContext.getAxisService());
+        inMessageContext.setAxisOperation(initialMessageContext.getAxisOperation());
+        inMessageContext.setAxisMessage(initialMessageContext.getAxisOperation().getMessage(
+                WSDLConstants.MESSAGE_LABEL_OUT_VALUE));
+        inMessageContext.setIncomingTransportName(Constants.TRANSPORT_LOCAL);
+        inMessageContext.setServiceContext(initialMessageContext.getServiceContext());
+
+        // set properties on response
+        inMessageContext.setServerSide(true);
+        //inMessageContext.setProperty("synapse.isresponse", Boolean.TRUE);
+        inMessageContext.setProperty(MessageContext.TRANSPORT_OUT,
+                initialMessageContext.getProperty(MessageContext.TRANSPORT_OUT));
+        inMessageContext.setProperty(Constants.OUT_TRANSPORT_INFO,
+                initialMessageContext.getProperty(Constants.OUT_TRANSPORT_INFO));
+        inMessageContext.setTransportIn(initialMessageContext.getTransportIn());
+        inMessageContext.setTransportOut(initialMessageContext.getTransportOut());
+
+        // copy the message type property that is used by the out message to the response message
+        inMessageContext.setProperty(Constants.Configuration.MESSAGE_TYPE,
+                initialMessageContext.getProperty(Constants.Configuration.MESSAGE_TYPE));
+
+        if (initialMessageContext.getMessageID() != null) {
+            inMessageContext.setRelationships(
+                    new RelatesTo[]{new RelatesTo(initialMessageContext.getMessageID())});
+        }
+
+        inMessageContext.setReplyTo(initialMessageContext.getReplyTo());
+        inMessageContext.setFaultTo(initialMessageContext.getFaultTo());
+
+        if (initialMessageContext.isPropertyTrue("IGNORE_SC_ACCEPTED")) {
+            inMessageContext.setProperty("FORCE_SC_ACCEPTED", Constants.VALUE_TRUE);
+        }
+
+        AxisEngine.receive(inMessageContext);
+    }
+
+    private SOAPEnvelope getEnvelope(MessageContext messageContext) throws AxisFault {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        TransportUtils.writeMessage(messageContext, out);
+
+        ByteArrayInputStream bs = new ByteArrayInputStream(out.toByteArray());
+        InputStreamReader streamReader = new InputStreamReader(bs);
+        OMXMLParserWrapper builder;
+        try {
+            builder = BuilderUtil.getBuilder(streamReader);
+        } catch (XMLStreamException e) {
+            throw AxisFault.makeFault(e);
+        }
+        return (SOAPEnvelope) builder.getDocumentElement();
+    }
 }
Index: modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 1060336)
+++ modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -79,6 +79,10 @@
                                OutputStream response)
             throws AxisFault {
         MessageContext msgCtx = confContext.createMessageContext();
+
+        // Set the in-message context as a property to the  current message context.
+        msgCtx.setProperty("IN_MESSAGE_CONTEXT", inMessageContext);
+
         if (inMessageContext != null) {
             msgCtx.setProperty(HTTPConstants.MC_HTTP_SERVLETREQUEST,
                                inMessageContext.getProperty(HTTPConstants.MC_HTTP_SERVLETREQUEST));
