Index: quickstartaxiom/src/samples/quickstart/clients/AXIOMClient.java
===================================================================
--- quickstartaxiom/src/samples/quickstart/clients/AXIOMClient.java	(Revision 537941)
+++ quickstartaxiom/src/samples/quickstart/clients/AXIOMClient.java	(Arbeitskopie)
@@ -57,7 +57,7 @@
             sender.setOptions(options);
 
             sender.fireAndForget(updatePayload);
-            System.err.println("done");
+            System.err.println("price updated");
             Thread.sleep(3000);
             OMElement result = sender.sendReceive(getPricePayload);
 
Index: quickstartjibx/src/samples/quickstart/clients/JiBXClient.java
===================================================================
--- quickstartjibx/src/samples/quickstart/clients/JiBXClient.java	(Revision 537941)
+++ quickstartjibx/src/samples/quickstart/clients/JiBXClient.java	(Arbeitskopie)
@@ -11,6 +11,7 @@
 
             getPrice(stub);
             update(stub);
+            getPrice(stub);
 
         } catch(Exception e){
             e.printStackTrace();
@@ -21,8 +22,8 @@
     /* fire and forget */
     public static void update(StockQuoteServiceStub stub){
         try{
-            stub.update("ABC", new Double(42.35));
-            System.err.println("done");
+            stub.update("CDE", new Double(42.35));
+            System.err.println("price updated");
         } catch(Exception e){
             e.printStackTrace();
             System.err.println("\n\n\n");
@@ -32,7 +33,7 @@
     /* two way call/receive */
     public static void getPrice(StockQuoteServiceStub stub){
         try{
-            System.err.println(stub.getPrice("ABC"));
+            System.err.println(stub.getPrice("CDE"));
         } catch(Exception e){
             e.printStackTrace();
             System.err.println("\n\n\n");
Index: quickstartjibx/build.xml
===================================================================
--- quickstartjibx/build.xml	(Revision 537941)
+++ quickstartjibx/build.xml	(Arbeitskopie)
@@ -19,11 +19,13 @@
         </fileset>
     </path>
 
-    <target name="init">
-        <delete dir="${build.dir}"/>
-        <mkdir dir="${build.dir}"/>
-    </target>
-
+    <target name="init">
+        <delete dir="${build.dir}"/>
+        <mkdir dir="${build.dir}"/>
+        <mkdir dir="${build.dir}/service"/>
+        <mkdir dir="${build.dir}/client"/>
+    </target>
+	
     <target name="generate.service" depends="init">
         <taskdef name="wsdl2java"
                  classname="org.apache.axis2.tool.ant.AntCodegenTask"
Index: quickstartxmlbeans/src/samples/quickstart/clients/XMLBEANSClient.java
===================================================================
--- quickstartxmlbeans/src/samples/quickstart/clients/XMLBEANSClient.java	(Revision 537941)
+++ quickstartxmlbeans/src/samples/quickstart/clients/XMLBEANSClient.java	(Arbeitskopie)
@@ -15,6 +15,7 @@
 
             getPrice(stub);
             update(stub);
+            getPrice(stub);
 
         } catch(Exception e){
             e.printStackTrace();
@@ -27,11 +28,11 @@
         try{
             UpdateDocument reqDoc = UpdateDocument.Factory.newInstance();
             UpdateDocument.Update req = reqDoc.addNewUpdate();
-            req.setSymbol ("ABC");
+            req.setSymbol ("BCD");
             req.setPrice (42.32);
 
             stub.update(reqDoc);
-            System.err.println("done");
+            System.err.println("price updated");
         } catch(Exception e){
             e.printStackTrace();
             System.err.println("\n\n\n");
@@ -43,7 +44,7 @@
         try{
             GetPriceDocument reqDoc = GetPriceDocument.Factory.newInstance();
             GetPriceDocument.GetPrice req = reqDoc.addNewGetPrice();
-            req.setSymbol("ABC");
+            req.setSymbol("BCD");
 
             GetPriceResponseDocument res =
                 stub.getPrice(reqDoc);
Index: quickstartxmlbeans/build.xml
===================================================================
--- quickstartxmlbeans/build.xml	(Revision 537941)
+++ quickstartxmlbeans/build.xml	(Arbeitskopie)
@@ -19,10 +19,11 @@
         </fileset>
     </path>
 
-    <target name="init">
-        <mkdir dir="${build.dir}"/>
-        <mkdir dir="${build.dir}/service"/>
-    </target>
+    <target name="init">
+        <mkdir dir="${build.dir}"/>
+        <mkdir dir="${build.dir}/service"/>
+        <mkdir dir="${build.dir}/client"/>
+    </target>
 
     <target name="generate.service" depends="init">
         <taskdef name="wsdl2java"
Index: quickstartadb/src/samples/quickstart/clients/ADBClient.java
===================================================================
--- quickstartadb/src/samples/quickstart/clients/ADBClient.java	(Revision 537941)
+++ quickstartadb/src/samples/quickstart/clients/ADBClient.java	(Arbeitskopie)
@@ -10,6 +10,7 @@
 
             getPrice(stub);
             update(stub);
+            getPrice(stub);
 
         } catch(Exception e){
             e.printStackTrace();
@@ -25,7 +26,7 @@
             req.setPrice (42.35);
 
             stub.update(req);
-            System.err.println("done");
+            System.err.println("price updated");
         } catch(Exception e){
             e.printStackTrace();
             System.err.println("\n\n\n");
Index: quickstartadb/build.xml
===================================================================
--- quickstartadb/build.xml	(Revision 537941)
+++ quickstartadb/build.xml	(Arbeitskopie)
@@ -19,10 +19,11 @@
         </fileset>
     </path>
 
-    <target name="init">
-        <mkdir dir="${build.dir}"/>
-        <mkdir dir="${build.dir}/service"/>
-    </target>
+    <target name="init">
+        <mkdir dir="${build.dir}"/>
+        <mkdir dir="${build.dir}/service"/>
+        <mkdir dir="${build.dir}/client"/>
+    </target>
 
     <target name="generate.service" depends="init">
         <taskdef name="wsdl2java"
