Index: modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java
===================================================================
--- modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(revision 1103616)
+++ modules/transport/local/src/org/apache/axis2/transport/local/LocalTransportReceiver.java	(working copy)
@@ -43,7 +43,7 @@
 import java.io.InputStreamReader;
 import java.io.OutputStream;
 
-public class LocalTransportReceiver implements TransportListener {
+public class LocalTransportReceiver {
     protected static final Log log = LogFactory.getLog(LocalTransportReceiver.class);
 
     public static ConfigurationContext CONFIG_CONTEXT;
@@ -184,27 +184,4 @@
         }
     }
 
-    public void init(ConfigurationContext axisConf, TransportInDescription transprtIn) throws AxisFault {
-    }
-
-    public void start() throws AxisFault {
-    }
-
-    public void stop() throws AxisFault {
-    }
-
-    public EndpointReference getEPRForService(String serviceName, String ip) throws AxisFault {
-        return null;
-    }
-
-    public EndpointReference[] getEPRsForService(String serviceName, String ip) throws AxisFault {
-        return new EndpointReference[0];
-    }
-
-    public SessionContext getSessionContext(MessageContext messageContext) {
-        return null;
-    }
-
-    public void destroy() {
-    }
 }
