Index: src/main/java/org/apache/axis2/transport/mail/MailTransportSender.java
===================================================================
--- src/main/java/org/apache/axis2/transport/mail/MailTransportSender.java	(revision 121460)
+++ src/main/java/org/apache/axis2/transport/mail/MailTransportSender.java	(working copy)
@@ -32,6 +32,7 @@
 import org.apache.axis2.transport.MessageFormatter;
 import org.apache.axiom.om.OMOutputFormat;
 import org.apache.axiom.om.util.CommonUtils;
+import org.wso2.securevault.SecretResolver;
 
 import javax.mail.*;
 import javax.mail.internet.*;
@@ -109,6 +110,11 @@
 
         smtpUsername = (String) props.get(MailConstants.MAIL_SMTP_USERNAME);
         smtpPassword = (String) props.get(MailConstants.MAIL_SMTP_PASSWORD);
+        
+        SecretResolver secretResolver = cfgCtx.getAxisConfiguration().getSecretResolver();
+        if(secretResolver != null && secretResolver.isTokenProtected(MailConstants.MAIL_SMTP_PASSWORD)){
+            smtpPassword = secretResolver.resolve(MailConstants.MAIL_SMTP_PASSWORD);
+        }
 
         if (smtpUsername != null && smtpPassword != null) {
             session = Session.getInstance(props, new Authenticator() {
